---
title: "R Notebook"
output: html_notebook
editor_options: 
  chunk_output_type: console
---

#tutorial
{
This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute
code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by 
placing your cursor inside it and pressing *Ctrl+Shift+Enter*. 

Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by 
pressing *Ctrl+Alt+I*.

When you save the notebook, an HTML file containing the code and output will be 
saved alongside it (click the *Preview* button or press *Ctrl+Shift+K* to 
preview the HTML file).

The preview shows you a rendered HTML copy of the contents of the editor. 
Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, 
the output of the chunk when it was last run in the editor is displayed.
}


#loading libraries, setting up workspace and directories
```{r}

#loading libraries and defining settings
{
  #analysis libraries
  library(DESeq2)
  library(Rtsne)
  library(umap)
  library(biomaRt)
  library(fgsea)
  
  #for parallelisation
  library(BiocParallel)
  library(foreach)
  library(doMC)
  
  #graphics libraries
  #library(PCAtools)
  library(ggplot2)
  #library(ggpmisc)
  library(reshape2)
  library(ggrepel)
  library(patchwork)
  library(cowplot)
  #library(scales)
  library(data.tree)
  library(plotly)
  library(viridis)
  library(ggnewscale)
  library(ggExtra)
  #library(grid)
  #library(heatmap.plus)
  library(NMF)
  #library(ggridges)
  library(RColorBrewer)
  library(gridExtra)
  
  #other libraries
  library(stringr)
  library(reshape2)
  
  #setting up parallelisation parameters for BiocParallel
  numCores <- parallel::detectCores()
  #utilise ~3/4 of the available cores/threads
  register(MulticoreParam(ceiling(numCores*(3/4))))
  registerDoMC(ceiling(numCores*(3/4)))
  
  
}

#setting up directories
{
  rm(list = ls(all.names = T))
  orig_loc <- getwd()
  
  setwd("/home/tareens/Workspace/Tregs/")
  #wrkspc <- "/home/tareens/Workspace/Tregs/"
  
  #locations of inputs
  loc_counts <- paste0(getwd(), "/05_counts/")
  loc_diffexprs_sngltrnscrpt <- paste0(getwd(), "/08_single_transcript/", 
                                       "Differential Expression Results/")
  loc_gsea_gmt <- paste0(getwd(), "/09_GSEA/GMTs/")
  
  #locations of outputs
  loc_diffexprs <- paste0(getwd(), "/06_differential_expression/")
  loc_diffexprs_merged <- paste0(getwd(), 
                                 "/06b_differential_expression_merged/")
  loc_plots <- paste0(getwd(), "/06b_differential_expression_merged/",
                      "Plots/")
  loc_tissue_specific_exprs <- paste0(getwd(), "/08_single_transcript/", 
                                      "Tissue Specific Expression/")
  loc_gsea <- paste0(getwd(), "/09_GSEA/")
  
  #location of functions and scripts
  loc_scripts <- paste0(getwd(), "/00b_R/")
  
  # #location of FANTOM5 data
  # loc_FANTOM <- paste0(getwd(), "/00d_other_data/" , "E-MTAB-3579 - ", 
  #                      "RNA-Seq CAGE (Cap Analysis of Gene Expression) ", 
  #                      "analysis of mice tissue in RIKEN FANTOM5 project/")
  
  #loading saved environment
  #load("/home/tareens/Workspace/Tregs/.RData")
  
}

#sourcing functions and scripts
{
  #sourcing the unified differential expression function
  source(paste0(loc_scripts, "DiffExprsDESeqAll.R"))
  
  #sourcing multiplot function
  source(paste0(loc_scripts, "multiplot.r"))
  
  #sourcing custom user-made heatmap.3 script; 
  # sourced from: https://www.biostars.org/p/18211/
  #source(paste0(loc_scripts, "heatmap.3.R"))
  
  #sourcing the plotting functions of ggbiplot as the library is no longer 
  #compatible with R 3.5.2+
  source(paste0(loc_scripts, "ggbiplot.r"))
  source(paste0(loc_scripts, "ggscreeplot.r"))
  
}

```


#importing and formatting counts matrix
```{r}

#importing and formatting counts matrix
{
  counts_matrix <- read.table(file = paste(loc_counts, "counts.txt", sep = ""),
                              sep = "\t",
                              stringsAsFactors = F,
                              header = T)
  rownames(counts_matrix) <- counts_matrix[,1]
  counts_matrix <- counts_matrix[,-1]
  
  meta_data <- counts_matrix [,1:5]
  counts_matrix <- counts_matrix[,-(1:5)]
    
  ids <- c(1:ncol(counts_matrix))
  counts_matrix[,ids] <- as.numeric( unlist( counts_matrix[,ids] ) )
  
  # #removing contaminated sample Pancreas_T_3
  # counts_matrix <- counts_matrix[,!grepl(pattern = "Pancreas_T_3", 
  #                                        x = colnames(counts_matrix), 
  #                                        ignore.case = T)]
  
  counts_matrix <- as.matrix(counts_matrix)
  
}

```


#checking received tissues against expected tissues
```{r}

#vector of expected tissues
{
  expected_tissues <- c("Blood", "Intraepithelial Layer", "Kidney", 
                      "Lamina Propria Lymphocyte", "Liver", "Lung", "Lymph Node", 
                      "Pancreas", "Peyers Patches", "Spleen")
}

#vector of sample conditions and their labels
condition_labels <- unique(gsub("_([^_]*)$", "", colnames(counts_matrix)))
names(condition_labels) <- rep(expected_tissues, each = 2)
print(condition_labels)


#stopping script if received data does not contain expected tissues defined 
#above
{

  if(!all(tolower(unique(gsub("_.*| ", "", colnames(counts_matrix)))) %in% 
     tolower(unique(gsub("_.*| ", "", expected_tissues)))) & 
     !all(tolower(unique(gsub("_.*| ", "", expected_tissues))) %in% 
     tolower(unique(gsub("_.*| ", "", colnames(counts_matrix))))))
  {
    stop(paste0("Some of the expected tissues are not in the data, or are ", 
                "possibly under a different name"))
  }
  
}

```


#setting up data annotations for analysis
```{r}

#constructing data annotations for differential expression analysis
{
  annotations <- as.data.frame(t(as.data.frame(strsplit(colnames(counts_matrix),
                                                        "_"))))
  rownames(annotations) <- colnames(counts_matrix)
  annotations <- annotations[,-3]
  colnames(annotations) <- c("tissue", "cellType")
  
  annotations$tissue <- as.factor(annotations$tissue)
  annotations$cellType <- as.factor(annotations$cellType)
  
  #check that the rownames and colnames are in the same order (redundant via 
  #above code) as its required by DESeq2
  print(paste0("Checking if row names of annotations are in the same order as ", 
               "column names of the counts matrix: ", 
               unique(rownames(annotations) == colnames(counts_matrix))))
  #unique(rownames(annotations) == colnames(counts_matrix))
  
}
```


#DESeq2 differential expression analysis
```{r}
#DESeq2 analysis using unified function
{
  
  #initialising DESeq object
  print("initialising DESeq object")
  DESeq_object <- DESeqDataSetFromMatrix(countData = counts_matrix,
                                         colData = annotations,
                                         design = ~ 0)
  # DESeq_object <- DESeqDataSetFromMatrix(countData =
  #                                          subset(counts_matrix,
  #                                                 rownames(counts_matrix) %in%
  #                                                   rownames(
  #                                                     filtered_tcounts_0.1)),
  #                                        colData = annotations,
  #                                        design = ~ 0)
  
  #adding meta_data to the dataset
  print("adding meta_data to the dataset")
  mcols(DESeq_object) <- DataFrame(mcols(DESeq_object),  meta_data)
  # mcols(DESeq_object) <- DataFrame(mcols(DESeq_object),
  #                                  subset(meta_data, rownames(meta_data) %in%
  #                                           rownames(filtered_tcounts_0.1)))
  
  #prefiltering zero count genes
  print("prefiltering zero count genes")
  keep <- rowSums(counts(DESeq_object)) >0
  DESeq_object <- DESeq_object[keep,]
  
  #estimating size factors to extract normalised count data
  print("estimating size factors to extract normalised count data")
  DESeq_object <- estimateSizeFactors(DESeq_object)
  # sizeFactors_55421 <- sizeFactors(DESeq_object)
  # sizeFactors(DESeq_object) <- sizeFactors_55421
  
  #extracting and writing normalised data
  print("extracting and writing normalised data")
  norm_counts <- counts(DESeq_object, normalized = T)
  write.table(x = norm_counts,
              file = paste0(loc_counts, "normalised_counts.txt"),
              sep = "\t", quote = F)
  
  #creating the (custom) model matrix to extract contrasts that we need
  #print("creating the (custom) model matrix to extract contrasts that we need")
  #model_matrix <- model.matrix(~0+tissue+tissue:cellType, colData(DESeq_object))
  
  #creating and formatting the contrast groups
  print("creating and formatting the contrast groups")
  #contrast groups for tissue vs tissue and Treg vs Treg contrasts
  #   i- Blood
  #  ii- Spleen, Lymph Node
  # iii- Intraepithelial Layer, Lamina Propria Lymphocyte, Peyer's Patches
  #  iv- Kidney, Lung, Liver, Pancreas
  
  contrast_groups <- list(c("Blood"), 
                          c("Spleen", "Lymph Node", "Peyer's Patches"), 
                          c("Intraepithelial Layer", 
                            "Lamina Propria Lymphocyte"), 
                          c("Kidney", "Lung", "Liver", "Pancreas"),
                          c("cellTypeT"),
                          c("cellTypeNT"))
  
  #ensuring that names in contrast_groups list are all lower letter and do not 
  # contain spaces or punctuation
  for(i in seq(1, length(contrast_groups)))
  {
    contrast_groups[[i]] <- tolower(unique(
      gsub("_.*|[[:space:]]|[[:punct:]]", "", contrast_groups[[i]])
      ))
  }
  rm(i)
  
  #creating the vector of formulae to use in the unified function
  print("creating the vector of formulae to use in the unified function")
  formulae <- c("~0+tissue:cellType")
  #design(DESeq_object) <- ~0+tissue:cellType
  
  #creating vector of stratification for use with certain formulae to separate 
  #cellTypeT and cellTypeNTs
  stratification <- c("cellType")
  
  #keep this commented -- only execute when you need to update the DESeq object
  #for the webtool
  # design(DESeq_object) <- formula(formulae[1])
  # DESeq_object <- DESeq(object = DESeq_object,
  #                       test = "Wald",
  #                       betaPrior = F,
  #                       minReplicatesForReplace = Inf,
  #                       parallel = T)
  # saveRDS(object = DESeq_object, file = "DESeq_Object.rds")
  
  #initialising a list to store all the results
  print("initialising a list to store all the results")
  merged_results_list <- list()
  
  source(paste0(loc_scripts, "DiffExprsDESeqAll.R"))
  
  #applying the unified differential expression function on the formulae
  print("applying the unified differential expression function on the formulae")
  for(i in seq(1, length(formulae)))
  {
    writeLines(paste0("\nPerforming the analysis for the formula: ", formulae[i]))
    
    merged_results_list[[length(merged_results_list)+1]] <- 
      DiffExprsDESeq(DESeq_object = DESeq_object, 
                     formula = formulae[i], 
                     contrast_groups = contrast_groups, 
                     stratification = stratification[i])
    
    #naming the result based on the formula
    names(merged_results_list)[length(merged_results_list)] <- formulae[i]
    
  }
  rm(i)
  
  #printing number of significant genes per formula per comparison
  # writeLines(paste0("\nShowing the number of significant genes (adj.p.value<", 
  #                   "0.05) from the contrast results\n"))
  # for(i in seq(1, length(merged_results_list))){
  #   writeLines(paste0("Formula: ", names(merged_results_list)[i]))
  #   
  #   if(length(merged_results_list[[i]])<1){
  #     writeLines("\n\n")
  #     next
  #   }
  #   
  #   for(j in seq(1, length(merged_results_list[[i]]))){
  #     writeLines(paste0("Permutation: ", names(merged_results_list[[i]])[j]))
  #     
  #     for(k in seq(1, length(merged_results_list[[i]][[j]]))){
  #       writeLines(paste0(nrow(
  #         subset(merged_results_list[[i]][[j]][[k]], 
  #                merged_results_list[[i]][[j]][[k]]$padj<0.05)),
  #         " genes in comparison: ",
  #         names(merged_results_list[[i]][[j]])[k]))
  #     }
  #     writeLines("\n")
  #   }
  #   writeLines("\n")
  # }
  # rm(i,j,k)
  
  writeLines("\nDifferential expression analyses completed")
  
  writeLines("The results are structured as below")
  xtree <- FromListSimple(merged_results_list, nodeName = "Root", )
  xtree
  
}

```


#writing out the differential expression files
```{r}

#reading norm counts
norm_counts <- read.delim(file = paste0(loc_counts, "normalised_counts.txt"),
                          sep = "\t",
                          stringsAsFactors = F)

#filtering norm_counts using transcripts per cell filtering
#creating list of genes having >x transcripts per cell 
#(as per Oliver, each sample has about 2000 cells; norm_count/2000 > 100)
## change to FALSE if you do not want this calculated
if(TRUE){
  
  #filtered_tcounts <- as.data.frame((norm_counts/2000)>0.1)
  #filtered_tcounts <- filtered_tcounts[rowSums(filtered_tcounts) > 0L,]
  
  #creating a vector of rows to keep
  selection <- logical(length = nrow(norm_counts))
  #creating groups to check the transcript per cell in
  groups <- unique(gsub(pattern = "_[0-9]+$", 
                        replacement = "", 
                        x = colnames(norm_counts), 
                        ignore.case = T))
  groups <- groups[grepl(pattern = "_T", x = groups, ignore.case = T)]
  #looping through each gene and checking if it has at least 0.01 transcript per
  # cell in the tissue_cellTypeT groups
  for(i in seq(1, length(selection))){
    
    for(j in groups){
      
      if(sum((norm_counts[i, grepl(pattern = j, 
                                       x = colnames(norm_counts), 
                                       ignore.case = T)]/2000)>0.01) == 
         sum(grepl(pattern = j, x = colnames(norm_counts), ignore.case = T))){
        selection[i] <- TRUE
        break
      }
    }
  }
  
  filtered_tcounts <- norm_counts[selection,]
  
  norm_counts <- filtered_tcounts
  
}
#for the webtool, rerun the above code while making changes to the filter (0.01 
# -> 0.1 -> 1) and then using the below code to export the filter object as RDS 
# for use in the webtool
{
  # #filtered_tcounts_0.01 <- filtered_tcounts
  # #filtered_tcounts_0.1 <- filtered_tcounts
  # #filtered_tcounts_1 <- filtered_tcounts
  # #filtered_tcounts_10 <- filtered_tcounts
  # #filtered_tcounts_100 <- filtered_tcounts
  # 
  # filtered_genes_lists <- list(
  #   gsub(pattern = "\\.[0-9]+$", 
  #        replacement = "", 
  #        x = rownames(filtered_tcounts_0.01), 
  #        ignore.case = T), 
  #   gsub(pattern = "\\.[0-9]+$", 
  #        replacement = "", 
  #        x = rownames(filtered_tcounts_0.1), 
  #        ignore.case = T), 
  #   gsub(pattern = "\\.[0-9]+$", 
  #        replacement = "", 
  #        x = rownames(filtered_tcounts_1), 
  #        ignore.case = T)#, 
  #   # gsub(pattern = "\\.[0-9]+$", 
  #   #      replacement = "", 
  #   #      x = rownames(filtered_tcounts_10), 
  #   #      ignore.case = T) 
  #   )
  # names(filtered_genes_lists) <- c("Filt0.01", "Filt0.1", "Filt1")#, "Filt10")
  # saveRDS(object = filtered_genes_lists, file = "Filtered_Gene_Lists.rds")
}

#calculating mean log2 exp based on tissue and cell type
{
  # mean_norm_counts <- data.frame(matrix(nrow = nrow(norm_counts), ncol = 0))
  mean_norm_exp <- data.frame(matrix(nrow = nrow(norm_counts), ncol = 0))
  
  #condition_labels <- unique(gsub("_([^_]*)$", "", colnames(norm_counts)))
  
  #iterate thorugh condition labels and calculate means for Tregs
  for(i in seq(1, length(condition_labels))){
    
    columns <- norm_counts[, grepl(pattern = paste0(
      condition_labels[i], "_[0-9]{1}$"), 
      x = colnames(norm_counts), 
      ignore.case = T), 
      drop = F]
    
    #tested the logic via the following code
    # temp <- matrix(c(1:3,NaN,4,Inf,5:6, 7:10),4,3)
    # rowMeans(temp)
    # temp2 <- temp * is.finite(temp)
    # rowMeans(temp2)
    
    log2_cols <- as.matrix(log2(columns))
    #this produces NaNs for Infs which are then ignored via na.rm = T
    log2_cols <- log2_cols * is.finite(log2_cols)
    # commenting the above line to retain infinites so that I can use facet_grid
    # commenting it gives false results as log2 of a c(1,0,4) row becomes -Inf, 
    # reverting to older code
    
    if(ncol(columns)>1){
      # log2_cols <- as.matrix(log2(columns))
      # #this produces NaNs for Infs which are then ignored
      # log2_cols <- log2_cols * is.finite(log2_cols)
      # print(head(log2_cols[!is.finite(log2_cols[,1]) | 
      #                        !is.finite(log2_cols)[,2] | 
      #                        !is.finite(log2_cols)[,3],]))
      mean_norm_exp <- cbind(mean_norm_exp, 
                             rowMeans(log2_cols, na.rm = T)
                             )
    }
    else mean_norm_exp <- cbind(mean_norm_exp, log2_cols)
    # else mean_norm_exp <- cbind(mean_norm_exp,
    #                             log2(columns))
    
    colnames(mean_norm_exp)[ncol(mean_norm_exp)] <-
        paste0(gsub(pattern = "_[0-9]{1}$", 
                    replacement = "", 
                    x = condition_labels[i], 
                    ignore.case = T))
  }
  #removing iteration related variables
  rm(i, columns)
  
  #replacing undetected (+/-log2(0)) mean_norm_exp with 0
  #mean_norm_exp[mean_norm_exp == "Inf" | mean_norm_exp == "-Inf"] <- 0
  
  #formatting the row names to Gene.stable.ID for merging with other results
  # mean_norm_counts$Gene.stable.ID <- gsub(pattern = "\\..*",
  #                                             replacement = "",
  #                                             x = rownames(norm_counts))
}

#calculating real log2FC and renaming DESeq2 log2FC to 'DeseqBeta'
{
  writeLines(paste0("\nCalculating the real log2 fold change and adding to the",
                    " results"))
  
  for(i in seq(1, length(merged_results_list))){
    
    for(j in seq(1, length(merged_results_list[[i]]))){
    #for(j in seq(1, 1)){
      
      cellType <- gsub(pattern = "^.*cellType", 
                       replacement = "", 
                       x = names(merged_results_list[[i]][j]), 
                       ignore.case = T)
      
      for(k in seq(1, length(merged_results_list[[i]][[j]]))){
        
        names <- unlist(strsplit(x = names(merged_results_list[[i]][[j]][k]), 
                                 split = "__Vs__", 
                                 fixed = T))
        
        numerator <- unlist(strsplit(x = names[1], split = "_", fixed = T))
        denominator <- unlist(strsplit(x = names[2], split = "_", fixed = T))
        
        numerator <- paste(numerator, "_", cellType, sep = "")
        denominator <- paste(denominator, "_", cellType, sep = "")
        
        {
        if(length(numerator) < 2){
          # numerator_temp <- as.data.frame(
          #   norm_counts[, grepl(pattern = numerator, 
          #                       x = colnames(norm_counts), 
          #                       ignore.case = T), drop = F])
          numerator_temp <- as.data.frame(
            mean_norm_exp[, grepl(pattern = numerator, 
                                     x = colnames(mean_norm_exp), 
                                     ignore.case = T), drop = F])
          # numerator_temp$means <- rowMeans(numerator_temp)
          # numerator_temp <- numerator_temp[,!grepl(pattern = "_[0-9]+$", 
          #                                          x = colnames(numerator_temp), 
          #                                          ignore.case = T), drop = F]
          colnames(numerator_temp) <- numerator
          #numerator_temp$ids <- rownames(numerator_temp)
        }
        else{
          
          # numerator_temp <- as.data.frame(
          #   norm_counts[, grepl(pattern = numerator[1], 
          #                       x = colnames(norm_counts), 
          #                       ignore.case = T), drop = F])
          numerator_temp <- as.data.frame(
            mean_norm_exp[, grepl(pattern = numerator[1], 
                                     x = colnames(mean_norm_exp), 
                                     ignore.case = T), drop = F])
          # numerator_temp$means <- rowMeans(numerator_temp)
          # numerator_temp <- numerator_temp[,!grepl(pattern = "_[0-9]+$", 
          #                                          x = colnames(numerator_temp), 
          #                                          ignore.case = T), drop = F]
          colnames(numerator_temp) <- numerator[1]
          
          for(l in seq(2, length(numerator))){
            # temp <- as.data.frame(
            #   norm_counts[, grepl(pattern = numerator[l], 
            #                       x = colnames(norm_counts), 
            #                       ignore.case = T), drop = F])
            temp <- as.data.frame(
              mean_norm_exp[, grepl(pattern = numerator[l], 
                                       x = colnames(mean_norm_exp), 
                                       ignore.case = T), drop = F])
            
            # temp$means <- rowMeans(temp)
            # temp <- temp[,!grepl(pattern = "_[0-9]+$", 
            #                      x = colnames(temp), 
            #                      ignore.case = T), drop = F]
            
            numerator_temp <- cbind(numerator_temp, temp)
            colnames(numerator_temp)[l] <- numerator[l]
            rm(temp)
          }
          rm(l)
          #numerator_temp$ids <- rownames(numerator_temp)
        }
        }
        # numerator_temp <- 
        #   numerator_temp[match(rownames(merged_results_list[[i]][[j]][[k]]), 
        #                        rownames(numerator_temp)),]
        
        {
        if(length(denominator) < 2){
          # denominator_temp <- as.data.frame(
          #   norm_counts[, grepl(pattern = denominator, 
          #                       x = colnames(norm_counts), 
          #                       ignore.case = T), drop = F])
          denominator_temp <- as.data.frame(
            mean_norm_exp[, grepl(pattern = denominator, 
                                     x = colnames(mean_norm_exp), 
                                     ignore.case = T), drop = F])
          # denominator_temp$means <- rowMeans(denominator_temp)
          # denominator_temp <- denominator_temp[,!grepl(pattern = "_[0-9]+$", 
          #                                          x = colnames(denominator_temp), 
          #                                          ignore.case = T), drop = F]
          colnames(denominator_temp) <- denominator
          #denominator_temp$ids <- rownames(denominator_temp)
        }
        else{
          # denominator_temp <- as.data.frame(
          #   norm_counts[, grepl(pattern = denominator[1], 
          #                       x = colnames(norm_counts), 
          #                       ignore.case = T), drop = F])
          denominator_temp <- as.data.frame(
            mean_norm_exp[, grepl(pattern = denominator[1], 
                                     x = colnames(mean_norm_exp), 
                                     ignore.case = T), drop = F])
          # denominator_temp$means <- rowMeans(denominator_temp)
          # denominator_temp <- denominator_temp[,!grepl(pattern = "_[0-9]+$", 
          #                                          x = colnames(denominator_temp), 
          #                                          ignore.case = T), drop = F]
          colnames(denominator_temp) <- denominator[1]
          
          for(l in seq(2, length(denominator))){
            # temp <- as.data.frame(
            #   norm_counts[, grepl(pattern = denominator[l], 
            #                       x = colnames(norm_counts), 
            #                       ignore.case = T), drop = F])
            temp <- as.data.frame(
            mean_norm_exp[, grepl(pattern = denominator[l], 
                                     x = colnames(mean_norm_exp), 
                                     ignore.case = T), drop = F])
            
            # temp$means <- rowMeans(temp)
            # temp <- temp[,!grepl(pattern = "_[0-9]+$", 
            #                      x = colnames(temp), 
            #                      ignore.case = T), drop = F]
            
            denominator_temp <- cbind(denominator_temp, temp)
            colnames(denominator_temp)[l] <- denominator[l]
            rm(temp)
          }
          rm(l)
          #denominator_temp$ids <- rownames(denominator_temp)
        }
        }
        
        # Actuallog2FC <- data.frame(Actuallog2FC = 
        #                              log2(rowMeans(numerator_temp) / 
        #                                     rowMeans(denominator_temp))
        #                            )
        Actuallog2FC <- data.frame(Actuallog2FC = 
                                     rowMeans(numerator_temp, na.rm = T) - 
                                     rowMeans(denominator_temp, na.rm = T))
        
        #means matrix
        means_matrix <- cbind(rowMeans(numerator_temp, na.rm = T), 
                              rowMeans(denominator_temp, na.rm = T))
        
        #compensating for infinite mean due to undetected counts
        # is useful to avoid -Inf in the max.col() below
        ## deprecated as all Infs are now NaNs via calculation from above block
        #means_matrix[means_matrix == "Inf" | means_matrix == "-Inf"] <- Inf
        
        #calculating quantiles
        Quantile <- data.frame(Quantile = rowMins(abs(means_matrix), na.rm = T))#, 
                                #stringsAsFactors = F)
        #rowMins on row of NaNs produces Inf, replacing with NaN below
        Quantile[!is.finite(Quantile$Quantile),] <- NaN
        rownames(Quantile) <- rownames(means_matrix)
        #getting indices/rows for which quantiles would be retained (contain min of 0)
        #quantile_indices <- (Quantile$Quantile==0)
        quantile_indices <- which(Quantile$Quantile==0)
        
        Detected_Exp <- data.frame(DetectedExp = Quantile$Quantile)
        rownames(Detected_Exp) <- rownames(means_matrix)
        
        # mincolindex <- max.col(m = -means_matrix,
        #                        ties.method = "first")
        # the above produces NAs for rows with NaN, reverting to Infs for just this part
        ## reversion necessary as max.col() of any row having a single NaN is NA
        means_matrix_infs <- means_matrix
        means_matrix_infs[is.nan(means_matrix_infs)] <- Inf
        mincolindex <- max.col(m = -means_matrix_infs,
                               ties.method = "first")
        mincolindex[quantile_indices] <- 
          max.col(m = means_matrix_infs, 
                  ties.method = "first")[quantile_indices]
        rm(means_matrix_infs)
        
        distrib_num <- ecdf(rowMeans(numerator_temp, na.rm = T))
        distrib_denom <- ecdf(rowMeans(denominator_temp, na.rm = T))
        # distrib_num <- ecdf(log2(rowMeans(numerator_temp)))
        # distrib_denom <- ecdf(log2(rowMeans(denominator_temp)))
        
        for(m in seq(1, nrow(Quantile))){
          
          if(mincolindex[m] == 1){
            Quantile$Quantile[m] <- distrib_num(means_matrix[m,1])
            Detected_Exp$DetectedExp[m] <- (means_matrix[m,1])
            # Quantile$Quantile[m] <- distrib_num(log2(means_matrix[m,1]))
            # Detected_Exp$DetectedExp[m] <- log2(means_matrix[m,1])
            #distrib <- ecdf(rowMeans(numerator_temp))
          }
          else{
            Quantile$Quantile[m] <- distrib_denom(means_matrix[m,2])
            Detected_Exp$DetectedExp[m] <- -(means_matrix[m,2])
            # Quantile$Quantile[m] <- distrib_denom(log2(means_matrix[m,2]))
            # Detected_Exp$DetectedExp[m] <- -log2(means_matrix[m,2])
            #distrib <- ecdf(rowMeans(denominator_temp))
          }
          #Quantile$Quantile[m] <- distrib(Quantile$Quantile[m])
        }
        
        #Quantile$Quantile <- Quantile$Quantile*100
        #Quantile$Quantile[!quantile_indices] <- NA
        #Quantile[-quantile_indices, 1] <- NA
        #Detected_Counts$DetectedExp[!quantile_indices] <- NA
        #Detected_Counts[-quantile_indices, 1, drop = F] <- NA
        #temp <- ecdf(mean_tissue_counts[,i])
        #percentiles[,i] <- temp(percentiles[,i])
        
        Actuallog2FC <- merge(x = Actuallog2FC,
                              y = Detected_Exp,
                              by = "row.names",
                              #by = 0,
                              all = F)
        rownames(Actuallog2FC) <- Actuallog2FC$Row.names
        Actuallog2FC$Row.names <- NULL
        
        Actuallog2FC <- merge(x = Actuallog2FC,
                              y = Quantile,
                              by = "row.names",
                              #by = 0,
                              all = F)
        rownames(Actuallog2FC) <- Actuallog2FC$Row.names
        Actuallog2FC$Row.names <- NULL
        
        merged_results_list[[i]][[j]][[k]] <- 
          merge(x = as.data.frame(merged_results_list[[i]][[j]][[k]]), 
                y = Actuallog2FC, 
                by = "row.names", 
                #by = 0, 
                all = F)
        
        # merged_results_list[[i]][[j]][[k]] <-
        #   merge(x = merged_results_list[[i]][[j]][[k]],
        #         y = Quantile,
        #         #by = "row.names",
        #         by = 0,
        #         all = F)
        #merged_results_list[[i]][[j]][[k]]$Quantile <- Quantile$Quantile
        
        merged_results_list[[i]][[j]][[k]] <- 
          merged_results_list[[i]][[j]][[k]][
            order(merged_results_list[[i]][[j]][[k]]$padj, decreasing = F),]
        
        row.names(merged_results_list[[i]][[j]][[k]]) <- 
          merged_results_list[[i]][[j]][[k]]$Row.names
        
        merged_results_list[[i]][[j]][[k]]$Row.names <- NULL
        
        # merged_results_list[[i]][[j]][[k]]$Actuallog2FC <-
        #   log2(rowMeans(numerator_temp)/rowMeans(denominator_temp))
        # 
        # colnames(merged_results_list[[i]][[j]][[k]])[
        #   ncol(merged_results_list[[i]][[j]][[k]])] <- "Actuallog2FC"
        
        # temp_data_struct <- merged_results_list[[i]][[j]][[k]]
        # temp_data_struct$ids <- rownames(temp_data_struct)
        
      }
      
    }
    
  }
}
rm(i,j,k,m)

#writing all the results tables to directories in loc_diffexprs
{
  writeLines("\nCreating directories and writing file output")

  #writing merged_results_list as an Rdata object to preserve data structure
  saveRDS(object = merged_results_list,
          file = paste0(loc_diffexprs, "merged_results_list.rds"),
          compress = T)

  for(i in seq(1, length(merged_results_list))){

    if(length(merged_results_list[[i]])<1) next

    for(j in seq(1, length(merged_results_list[[i]]))){

      #creating directories
      path <- paste0(loc_diffexprs,
                     names(merged_results_list)[i], "/",
                     names(merged_results_list[[i]])[j], "/")
      dir.create(path = path, recursive = T)

      for(k in seq(1, length(merged_results_list[[i]][[j]]))){

        write.table(x = merged_results_list[[i]][[j]][[k]],
                    file = paste0(path,
                                  names(merged_results_list[[i]][[j]])[k],
                                  ".txt"),
                    quote = F,
                    sep = "\t",
                    row.names = T,
                    col.names = T)
      }
      path <- NULL
    }
  }
  #removing iteration related variables
  rm(i, j, k, path)

  writeLines("\nOutput writing completed")
}


#mean expression on log10 scale for use in webtool ## needs to be log2
## uncomment when need to update RDS for the tool
{
  # # mean_norm_counts <- data.frame(matrix(nrow = nrow(norm_counts), ncol = 0))
  # # mean_norm_exp_log10 <- data.frame(matrix(nrow = nrow(norm_counts), ncol = 0))
  # mean_norm_exp_log2 <- data.frame(matrix(nrow = nrow(norm_counts), ncol = 0))
  # 
  # #iterate thorugh condition labels and calculate means for Tregs
  # for(i in seq(1, length(condition_labels))){
  # 
  #   columns <- norm_counts[, grepl(pattern = paste0(
  #     condition_labels[i], "_[0-9]{1}$"),
  #     x = colnames(norm_counts),
  #     ignore.case = T),
  #     drop = F]
  # 
  #   #tested the logic via the following code
  #   # temp <- matrix(c(1:3,NaN,4,Inf,5:6, 7:10),4,3)
  #   # rowMeans(temp)
  #   # temp2 <- temp * is.finite(temp)
  #   # rowMeans(temp2)
  # 
  #   #log10_cols <- as.matrix(log10(columns))
  #   log2_cols <- as.matrix(log2(columns))
  #   #this produces NaNs for Infs which are then ignored via na.rm = T
  #   #log10_cols <- log10_cols * is.finite(log10_cols)
  #   log2_cols <- log2_cols * is.finite(log2_cols)
  # 
  #   if(ncol(columns)>1){
  #     # # log2_cols <- as.matrix(log2(columns))
  #     # # #this produces NaNs for Infs which are then ignored
  #     # # log2_cols <- log2_cols * is.finite(log2_cols)
  #     # # print(head(log2_cols[!is.finite(log2_cols[,1]) |
  #     # #                        !is.finite(log2_cols)[,2] |
  #     # #                        !is.finite(log2_cols)[,3],]))
  #     # mean_norm_exp_log10 <- cbind(mean_norm_exp_log10,
  #     #                        rowMeans(log10_cols, na.rm = T)
  #     #                        )
  #     mean_norm_exp_log2 <- cbind(mean_norm_exp_log2,
  #                            rowMeans(log2_cols, na.rm = T)
  #                            )
  #   }
  #   #else mean_norm_exp_log10 <- cbind(mean_norm_exp_log10, log10_cols)
  #   else mean_norm_exp_log2 <- cbind(mean_norm_exp_log2, log2_cols)
  #   # else mean_norm_exp <- cbind(mean_norm_exp,
  #   #                             log2(columns))
  # 
  #   # colnames(mean_norm_exp_log10)[ncol(mean_norm_exp_log10)] <-
  #   #     paste0(gsub(pattern = "_[0-9]{1}$",
  #   #                 replacement = "",
  #   #                 x = condition_labels[i],
  #   #                 ignore.case = T))
  #   colnames(mean_norm_exp_log2)[ncol(mean_norm_exp_log2)] <-
  #       paste0(gsub(pattern = "_[0-9]{1}$",
  #                   replacement = "",
  #                   x = condition_labels[i],
  #                   ignore.case = T))
  # }
  # #removing iteration related variables
  # rm(i, columns)
  # 
  # #replacing undetected (+/-log2(0)) mean_norm_exp with 0
  # #mean_norm_exp[mean_norm_exp == "Inf" | mean_norm_exp == "-Inf"] <- 0
  # 
  # #formatting the row names to Gene.stable.ID for merging with other results
  # # mean_norm_counts$Gene.stable.ID <- gsub(pattern = "\\..*",
  # #                                             replacement = "",
  # #                                             x = rownames(norm_counts))
  # 
  # #saveRDS(object = mean_norm_exp_log10, file = "Mean_Normalised_Counts.rds")
  # saveRDS(object = mean_norm_exp_log2, file = "Mean_Normalised_Counts.rds")
}


```


#creating and writing wide output table of results
# run the pre-writing code from the previous block first (writing out diff...)
# to get the filtered genes and the mean norm exp
```{r}
{
  
  #reading in DESeq2 outputs
  {
    #reading norm counts
    #use filtered counts from previous code block
    
    # norm_counts <- read.delim(file = paste0(loc_counts, "normalised_counts.txt"),
    #                           sep = "\t",
    #                           stringsAsFactors = F)
    #norm_counts <- norm_counts[rowSums(as.matrix(norm_counts))>0,]
    
    #reading the DESeq2 analysis output table
    merged_results_list <- readRDS(file = paste0(loc_diffexprs, 
                                          "merged_results_list.rds"))
    
    # contrast_groups <- list(c("Blood"), 
    #                         c("Spleen", "Lymph Node"), 
    #                         c("Intraepithelial Layer", 
    #                           "Lamina Propria Lymphocyte", "Peyer's Patches"), 
    #                         c("Kidney", "Lung", "Liver", "Pancreas"),
    #                         c("cellTypeT"),
    #                         c("cellTypeNT"))
    
    #ensuring that names in contrast_groups list are all lower letter and do not 
    # contain spaces or punctuation
    # for(i in seq(1, length(contrast_groups)))
    # {
    #   contrast_groups[[i]] <- tolower(unique(
    #     gsub("_.*|[[:space:]]|[[:punct:]]", "", contrast_groups[[i]])
    #     ))
    # }
    # rm(i)
    
  }
  
  #creating list of genes having >x transcripts per cell 
  #(as per Oliver, each sample has about 2000 cells; norm_count/2000 > 100)
  ## moved to writing out the differential expression files code block
  {
    
    # #filtered_tcounts <- as.data.frame((norm_counts/2000)>0.1)
    # #filtered_tcounts <- filtered_tcounts[rowSums(filtered_tcounts) > 0L,]
    # 
    # #creating a vector of rows to keep
    # selection <- logical(length = nrow(norm_counts))
    # #creating groups to check the transcript per cell in
    # groups <- unique(gsub(pattern = "_[0-9]+$", 
    #                       replacement = "", 
    #                       x = colnames(norm_counts), 
    #                       ignore.case = T))
    # groups <- groups[grepl(pattern = "_T", x = groups, ignore.case = T)]
    # #looping through each gene and checking if it has at least 0.1 transcript per
    # # cell in the tissue_cellTypeT groups
    # for(i in seq(1, length(selection))){
    #   
    #   for(j in groups){
    #     
    #     if(rowSums((norm_counts[i, grepl(pattern = j, 
    #                                      x = colnames(norm_counts), 
    #                                      ignore.case = T)]/2000)>0.01) == 
    #        sum(grepl(pattern = j, x = colnames(norm_counts), ignore.case = T))){
    #       selection[i] <- TRUE
    #       break
    #     }
    #   }
    # }
    # 
    # filtered_tcounts <- norm_counts[selection,]
    
  }
  
  #beginning the output table via merging
  output_table <- data.frame(matrix(nrow = nrow(norm_counts), ncol = 0))
  output_table$Gene.stable.ID <- gsub(pattern = "\\..*", 
                                   replacement = "", 
                                   x = rownames(norm_counts))
  
  ## uncomment code in here to retreive information from biomart and write it 
  #   out
  {
    # #retreiving info from biomart
    # {
    #   ensembl <- useMart(biomart = "ensembl", dataset = "mmusculus_gene_ensembl")
    #   filters <- listFilters(ensembl) #ensembl_gene_id
    #   attributes <- listAttributes(ensembl) #mgi_symbol, external_synonym and
    #   #mgi_description
    #   MGI_info <- getBM(attributes = c('mgi_symbol', 'external_synonym',
    #                                    'entrezgene_accession',
    #                                    'mgi_description', 'ensembl_gene_id'),
    #                     filters = 'ensembl_gene_id',
    #                     values = output_table$Gene.stable.ID,
    #                     mart = ensembl)
    # 
    #   MGI_info <- MGI_info[order(MGI_info$ensembl_gene_id),]
    # }
    # #for the complete list from counts matrix
    # {
    #   ensembl <- useMart(biomart = "ensembl", dataset = "mmusculus_gene_ensembl")
    #   #filters <- listFilters(ensembl) #ensembl_gene_id
    #   #attributes <- listAttributes(ensembl) #mgi_symbol, external_synonym and
    #   #mgi_description
    #   complete_gene_info <- getBM(attributes = c('mgi_symbol', 'external_synonym',
    #                                    'mgi_description', 'ensembl_gene_id', 
    #                                    'entrezgene_id'),
    #                     filters = 'ensembl_gene_id',
    #                     values = gsub(pattern = "\\.[0-9]+$", 
    #                                   replacement = "", 
    #                                   x = rownames(counts_matrix), 
    #                                   ignore.case = T),
    #                     mart = ensembl)
    # 
    #   complete_gene_info <- complete_gene_info[order(complete_gene_info$ensembl_gene_id),]
    # }
    # 
    # #iteratively collapsing rows in the data frame
    # {
    #   restruct_MGI_info <- MGI_info[1,]
    #   #curr_row <- as.list(MGI_info[1,])
    #   for (i in seq(2, nrow(MGI_info), 1)){
    #     if (restruct_MGI_info[nrow(restruct_MGI_info),5] !=
    #         MGI_info$ensembl_gene_id[i]){
    # 
    #       restruct_MGI_info[nrow(restruct_MGI_info),1] <- paste(unique(
    #         unlist(strsplit(x = restruct_MGI_info[nrow(restruct_MGI_info),1],
    #                         split = ";"))),
    #         collapse = "; ")
    # 
    #       restruct_MGI_info[nrow(restruct_MGI_info),2] <- paste(unique(
    #         unlist(strsplit(x = restruct_MGI_info[nrow(restruct_MGI_info),2],
    #                         split = ";"))),
    #         collapse = "; ")
    # 
    #       restruct_MGI_info[nrow(restruct_MGI_info),3] <- paste(unique(
    #         unlist(strsplit(x = restruct_MGI_info[nrow(restruct_MGI_info),3],
    #                         split = ";"))),
    #         collapse = "; ")
    #       
    #       restruct_MGI_info[nrow(restruct_MGI_info),4] <- paste(unique(
    #         unlist(strsplit(x = restruct_MGI_info[nrow(restruct_MGI_info),4],
    #                         split = ";"))),
    #         collapse = "; ")
    # 
    #       restruct_MGI_info <- rbind(restruct_MGI_info, MGI_info[i,])
    #     }
    #     else {
    #       restruct_MGI_info[nrow(restruct_MGI_info),] <- c(
    #         paste(restruct_MGI_info[nrow(restruct_MGI_info),1],
    #               MGI_info[i,1],
    #               sep = ";"),
    #         paste(restruct_MGI_info[nrow(restruct_MGI_info),2],
    #               MGI_info[i,2],
    #               sep = ";"),
    #         paste(restruct_MGI_info[nrow(restruct_MGI_info),3],
    #               MGI_info[i,3],
    #               sep = ";"),
    #         paste(restruct_MGI_info[nrow(restruct_MGI_info),4],
    #               MGI_info[i,4],
    #               sep = ";"),
    #         restruct_MGI_info[nrow(restruct_MGI_info),5]
    #         )
    #     }
    #   }
    #   restruct_MGI_info[nrow(restruct_MGI_info),1] <- paste(unique(
    #     unlist(strsplit(x = restruct_MGI_info[nrow(restruct_MGI_info),1],
    #                     split = ";"))),
    #     collapse = "; ")
    # 
    #   restruct_MGI_info[nrow(restruct_MGI_info),2] <- paste(unique(
    #     unlist(strsplit(x = restruct_MGI_info[nrow(restruct_MGI_info),2],
    #                     split = ";"))),
    #     collapse = "; ")
    # 
    #   restruct_MGI_info[nrow(restruct_MGI_info),3] <- paste(unique(
    #     unlist(strsplit(x = restruct_MGI_info[nrow(restruct_MGI_info),3],
    #                     split = ";"))),
    #     collapse = "; ")
    #   
    #   restruct_MGI_info[nrow(restruct_MGI_info),4] <- paste(unique(
    #     unlist(strsplit(x = restruct_MGI_info[nrow(restruct_MGI_info),4],
    #                     split = ";"))),
    #     collapse = "; ")
    #   rm(i)
    # 
    #   restruct_MGI_info$mgi_symbol <- gsub(pattern = "; $",
    #                            replacement = "",
    #                            x = paste(restruct_MGI_info$mgi_symbol,
    #                                      restruct_MGI_info$entrezgene_accession,
    #                                      restruct_MGI_info$external_synonym,
    #                              sep = "; "),
    #                            ignore.case = T)
    #   restruct_MGI_info$external_synonym <- NULL
    #   restruct_MGI_info$entrezgene_accession <- NULL
    #   
    #   for(i in seq(1, nrow(restruct_MGI_info))){
    #     restruct_MGI_info[i,1] <- paste(unique(
    #       unlist(strsplit(x = restruct_MGI_info[i,1], split = "; "))), 
    #       collapse = "; ")
    #   }
    #   rm(i)
    #   
    #   restruct_MGI_info$mgi_symbol <- gsub(
    #     pattern = "(^[[:space:]]*;[[:space:]]*|[[:space:]]*;[[:space:]]*$)", 
    #     replacement = "", 
    #     x = restruct_MGI_info$mgi_symbol, 
    #     ignore.case = T)
    #   
    # }
    # 
    # #writing out the MGI information
    # {
    #   write.table(x = restruct_MGI_info,
    #               file = paste0(loc_diffexprs_merged, "combined_MGI_info.txt"),
    #               quote = F,
    #               sep = "\t",
    #               row.names = F,
    #               col.names = T)
    # }
  }
  
  
  #importing combined MGI information so as not to query biomart and redo all
  {
    restruct_MGI_info <- read.delim(file = paste0(loc_diffexprs_merged, 
                                      "combined_MGI_info.txt"), 
                        header = T, 
                        sep = "\t", 
                        stringsAsFactors = F)
    
    colnames(restruct_MGI_info) <- c("MGI.symbol",
                                     "MGI.description",
                                     "ensembl_gene_id")
  }
  
  # MGI_sym <- read.delim(file = 
  #                         paste0(loc_diffexprs_merged, "ENSMUSG_MGIsym.txt"), 
  #                       sep = "\t", stringsAsFactors = F)
  # 
  # restruct_MGI_info <- merge(x = restruct_MGI_info, y = MGI_sym, 
  #                            by.x = "ensembl_gene_id", by.y = "Gene.stable.ID", 
  #                            all.x = T, all.y = F)
  
  output_table <- merge(x = output_table, y = restruct_MGI_info, 
                        by.x = "Gene.stable.ID", by.y = "ensembl_gene_id", 
                        all.x = T, all.y = F)
  
  
  #creating custom header rows for final output
  bottom_header_row <- c("Gene Information", 
                         gsub(pattern = FALSE, replacement = "", 
                              x = vector(length = ncol(output_table)-1))
                         )
  
  #adding normalised expression to the output table
  temporary_var <- as.data.frame(norm_counts)
  temporary_var$Gene.stable.ID <- gsub(pattern = "\\..*",
                                       replacement = "",
                                       x = rownames(temporary_var))

  output_table <- merge(x = output_table, y = temporary_var,
                        by = "Gene.stable.ID", all = T)
  rm(temporary_var)
  
  #continuing to construct the bottom  custom header row
  bottom_header_row <- c(bottom_header_row, "Normalised Expression Counts", 
                         gsub(pattern = FALSE, replacement = "", 
                              x = vector(length = 
                                           ncol(output_table)-1-length(
                                             bottom_header_row)))
                         )
  
  #calculating mean counts based on tissue and cell type
  ## moved to preceding code block for calculation of actual log2 fold change
  {
    # mean_norm_counts <- data.frame(matrix(nrow = nrow(norm_counts), ncol = 0))
    # 
    # #iterate thorugh condition labels and calculate means for Tregs
    # for(i in seq(1, length(condition_labels))){
    #   
    #   columns <- norm_counts[, grepl(pattern = paste0(
    #     condition_labels[i], "_[0-9]{1}$"), 
    #     x = colnames(norm_counts), 
    #     ignore.case = T), 
    #     drop = F]
    #   
    #   if(ncol(columns)>1) mean_norm_counts <- cbind(mean_norm_counts, 
    #                                                   rowMeans(
    #                                                     as.matrix(columns))
    #                                                   )
    #   else mean_norm_counts <- cbind(mean_norm_counts, columns)
    #   
    #   colnames(mean_norm_counts)[ncol(mean_norm_counts)] <-
    #       paste0(gsub(pattern = "_[0-9]{1}$", 
    #                   replacement = "", 
    #                   x = condition_labels[i], 
    #                   ignore.case = T))
    # }
    # #removing iteration related variables
    # rm(i, columns)
    # 
    # #formatting the row names to Gene.stable.ID for merging with other results
    # mean_norm_counts$Gene.stable.ID <- gsub(pattern = "\\..*",
    #                                             replacement = "",
    #                                             x = rownames(norm_counts))
  }
  
  #merging gene ids, symbols, description and mean expression per tissue per 
  #cell type
  # output_table <- merge(x = output_table, y = mean_norm_counts, 
  #                    by = "Gene.stable.ID", all = T)
  mean_norm_exp$Gene.stable.ID <- gsub(pattern = "\\.[0-9]+$", 
                                       replacement = "", 
                                       x = rownames(mean_norm_exp), 
                                       ignore.case = T)
  output_table <- merge(x = output_table, y = mean_norm_exp, 
                     by = "Gene.stable.ID", all = T)
  mean_norm_exp$Gene.stable.ID <- NULL
  
  #continuing to construct the bottom  custom header row
  bottom_header_row <- c(bottom_header_row, "Mean log2 Expression", 
                         gsub(pattern = FALSE, replacement = "", 
                              x = vector(length = 
                                           ncol(output_table)-1-length(
                                             bottom_header_row)))
                         )
  
  #creating new header rows to accommodate differential expression details
  mid_header_row <- gsub(pattern = FALSE, replacement = "", 
                         x = vector(length = length(bottom_header_row)))
  top_header_row <- gsub(pattern = FALSE, replacement = "", 
                         x = vector(length = length(bottom_header_row)))
  
  #diff.exprs results start from the below column -- is used later
  diff_exprs_col_no <- ncol(output_table)+1
  
  #now adding differential expression results
  for(i in seq_along(merged_results_list)){
    
    if(length(merged_results_list[[i]])<1) next
    
    for(j in seq_along(merged_results_list[[i]])){
      
      if(length(merged_results_list[[i]][[j]])<1) next
      
      for(k in seq_along(merged_results_list[[i]][[j]])){
        
        # temporary_var <- as.data.frame(
        #   merged_results_list[[i]][[j]][[k]][,c(7,2,5,6)])
        temporary_var <- as.data.frame(
          merged_results_list[[i]][[j]][[k]][,c(7,2,5,6,8,9)])
        #temporary_var$Significant <- temporary_var$padj<0.05
        temporary_var$Gene.stable.ID <- gsub(pattern = "\\..*",
                                             replacement = "",
                                             x = rownames(temporary_var))

        output_table <- merge(x = output_table, y = temporary_var,
                              by = "Gene.stable.ID", all.x = T, all.y = F)

        #continuing to construct the bottom  custom header row
        bottom_header_row <- c(bottom_header_row,
                               names(merged_results_list[[i]][[j]])[k],
                               gsub(pattern = FALSE, replacement = "",
                                    x = vector(length =
                                                 ncol(output_table)-1-length(
                                                   bottom_header_row)))
                               )
        
        #print(paste0("k: ", k , ", add length: ", ncol(output_table)-1-length(bottom_header_row)))
      }
      #continuing to construct the mid  custom header row
      mid_header_row <- c(mid_header_row,
                          names(merged_results_list[[i]])[j],
                          gsub(pattern = FALSE, replacement = "",
                               x = vector(length =
                                            ncol(output_table)-1-length(
                                              mid_header_row)))
                          )
      
      #print(paste0("j: ", j , ", add length: ", ncol(output_table)-1-length(mid_header_row)))
    }
    #continuing to construct the mid  custom header row
    top_header_row <- c(top_header_row,
                        paste0("Model: ", names(merged_results_list)[i]),
                        gsub(pattern = FALSE, replacement = "",
                             x = vector(length =
                                          ncol(output_table)-1-length(
                                            top_header_row)))
                        )
    
    #print(paste0("i: ", i , ", add length: ", ncol(output_table)-1-length(top_header_row)))
  }
  rm(i,j,k)
  
  
  #removing results for non-finite actual log2 fold change
  for(i in seq(64, ncol(output_table),6)){
    output_table[!is.finite(output_table[,i]), c(i, i+1, i+2, i+3)] <- NA
  }
  
  #removing detected expression and quantile for finite actual log2 fold change
  for(i in seq(64, ncol(output_table),6)){
    output_table[is.finite(output_table[,i]), c(i+4, i+5)] <- NA
  }
  
  #removing quantiles for max detected expression of 0
  for(i in seq(68, ncol(output_table),6)){
    output_table[(output_table[,i] == 0 |
                    !is.finite(output_table[,i])) &
                   !is.na(output_table[,i]), i+1] <- NA
    # output_table[output_table[,i] == 0 & !is.na(output_table[,i]), i] <-
    #   "Undetected in both groups"
    output_table[(output_table[,i] == 0 |
                    !is.finite(output_table[,i])) &
                   !is.na(output_table[,i]), i] <- NA
  }
  
  #removing trailing dots and x/y in the column names that arise from merging
  colnames(output_table) <- gsub(pattern = "\\.+[xy]$", 
                                 replacement = "", 
                                 x = colnames(output_table))
  
  #converting 'log2FoldChange' to 'Log2 Fold Change' and 'padj' to 'Adjusted 
  #p.value <0.05'
  colnames(output_table) <- gsub(pattern = "log2FoldChange", 
                                 replacement = "DESeq2 Beta", 
                                 x = colnames(output_table))
  colnames(output_table) <- gsub(pattern = "padj", 
                                 replacement = "Adjusted p.value", 
                                 x = colnames(output_table))
  colnames(output_table) <- gsub(pattern = "pvalue", 
                                 replacement = "p.value", 
                                 x = colnames(output_table))
  colnames(output_table) <- gsub(pattern = "Actuallog2FC", 
                                 replacement = "log2 Fold Change", 
                                 x = colnames(output_table))
  # colnames(output_table) <- gsub(pattern = "DetectedCounts",
  #                                replacement = "Mean Counts Detected",
  #                                x = colnames(output_table))
  colnames(output_table) <- gsub(pattern = "DetectedExp",
                                 replacement = "One-Side Mean log2 Expression",
                                 x = colnames(output_table))
  
  #filtering the output table if the filtering exists
  if(exists("filtered_tcounts")){
    output_table <- subset(output_table, output_table$Gene.stable.ID %in% 
                             gsub(pattern = "\\.[0-9]+$", 
                                  replacement = "", 
                                  x = rownames(filtered_tcounts), 
                                  ignore.case = T))
  }
  
  #cleaning up naming in the headers
  mid_header_row <- gsub(pattern = "_NT", 
                         replacement = " Tconv", 
                         x = gsub(pattern = "_T", 
                                  replacement = " Treg", 
                                  x = gsub(pattern = "tissue.cellType", 
                                           replacement = "tissue_", 
                                           x = mid_header_row, 
                                           ignore.case = T), 
                                  ignore.case = T), 
                         ignore.case = T)
  
  bottom_header_row <- 
    gsub(pattern = "(, ,)", 
         replacement = "", 
         x = gsub(pattern = "peyerspatches", 
                  replacement = "PP", 
                  x = gsub(pattern = "intraepitheliallayer", 
                           replacement = "IEL", 
                           x = gsub(pattern = "laminaproprialymphocyte", 
                                    replacement = "LPL", 
                                    x = gsub(pattern = "lymphnode", 
                                             replacement = "LN", 
                                             x = gsub(pattern = "_", 
                                                      replacement = ", ", 
                                                      x = bottom_header_row, 
                                                      ignore.case = T), 
                                             ignore.case = T), 
                                    ignore.case = T), 
                           ignore.case = T), 
                  ignore.case = T), 
         ignore.case = T)
  
  bottom_header_row <- gsub(pattern = "blood", 
                         replacement = "Blood", 
                         x = gsub(pattern = "kidney", 
                                  replacement = "Kidney", 
                                  x = gsub(pattern = "lung", 
                                           replacement = "Lung", 
                                           x = gsub(pattern = "liver", 
                                                    replacement = "Liver", 
                                                    x = gsub(pattern = "pancreas", 
                                                             replacement = "Pancreas", 
                                                             x = gsub(pattern = "spleen", 
                                                                      replacement = "Spleen", 
                                                                      x = bottom_header_row, 
                                                                      ignore.case = F), 
                                                             ignore.case = F), 
                                                    ignore.case = F), 
                                           ignore.case = F), 
                                  ignore.case = F), 
                         ignore.case = F)
  
  colnames(output_table) <- gsub(pattern = "_NT", 
                                 replacement = "_Tconv", 
                                 x = gsub(pattern = "_T", 
                                          replacement = "_Treg", 
                                          x = gsub(pattern = "\\.[0-9]+$", 
                                                   replacement = "", 
                                                   x = colnames(output_table), 
                                                   ignore.case = T), 
                                          ignore.case = T), 
                                 ignore.case = T)
  
  colnames(output_table) <- 
    gsub(pattern = "peyerspatches", 
         replacement = "PP", 
         x = gsub(pattern = "intraepitheliallayer", 
                  replacement = "IEL", 
                  x = gsub(pattern = "laminaproprialymphocyte", 
                           replacement = "LPL", 
                           x = gsub(pattern = "lymphnode", 
                                    replacement = "LN", 
                                    x = colnames(output_table), 
                                    ignore.case = T), 
                           ignore.case = T), 
                  ignore.case = T), 
         ignore.case = T)
  
  
  column_header_row <- colnames(output_table)
  
  column_header_row <- gsub(pattern = "_", 
                            replacement = " ", 
                            x = column_header_row, 
                            ignore.case = T)
  
  column_header_row[1:3] <- gsub(pattern = "\\.", 
                                 replacement = " ", 
                                 x = column_header_row[1:3], 
                                 ignore.case = T)
  
  #replacing blank information with NA
  output_table[output_table == ""] <- NA
  
  #merging custom header rows atop the output table to be exported as a tab 
  #separated file
  output_table <- rbind(top_header_row, 
                        mid_header_row, 
                        bottom_header_row, 
                        column_header_row, 
                        output_table)
  
  #lastly, removing all genes which have NA is all differential expression 
  #results
  #in this case, column 64 onwards
  # output_table <- output_table[rowSums(is.na(
  #   output_table[,diff_exprs_col_no:ncol(output_table)])) != ncol(
  #     output_table[,diff_exprs_col_no:ncol(output_table)]),]
  
  ##modify code after this comment if genes need to be sorted in a particular 
  #order
  
  #writing out the file
  write.table(x = output_table,
              file = paste0(loc_diffexprs_merged, "output_table.txt"),
              quote = F, sep = "\t", row.names = F, col.names = F)
  
  
  #the below code was used to calculate and add percentiles to the end of table
  ## no longer needed as this is not what Carlos has meant to be done
  {
    # #creating vector for row selection
    # to_keep <- output_table[5:nrow(output_table),64:ncol(output_table)]
    # to_keep <- to_keep[,grepl(pattern = "Actual", 
    #                           x = colnames(to_keep), 
    #                           ignore.case = T)]
    # to_keep <- is.finite(data.matrix(to_keep))
    # #to_keep <- rowSums(to_keep) > 0L
    # to_keep <- rowSums(to_keep) >= ncol(to_keep)
    # to_keep <- c(TRUE, TRUE, TRUE, TRUE, to_keep)
    # #making sure that the top four header rows are retained
    # #to_keep <- c(TRUE, TRUE, TRUE, TRUE, to_keep)
    # #names(to_keep)[1:4] <- c('1', '2', '3', '4')
    # 
    # #writing out only actual log fold change calculable rows
    # # write.table(x = output_table[to_keep,],
    # #             file = paste0(loc_diffexprs_merged, "output_table.txt"),
    # #             quote = F, sep = "\t", row.names = F, col.names = F)
    # 
    # #writing out only till counts for actual log fold change incalculable rows
    # # also calculating their quantile expression
    # output_table_excluded <- output_table[
    #   c(TRUE, TRUE, TRUE, TRUE, !to_keep[5:length(to_keep)]), 
    #   c(1, 2, 3, grep(pattern = "_T(conv|reg)", 
    #                   x = colnames(output_table), 
    #                   ignore.case = T))]
    # 
    # mean_tissue_counts <- (output_table[5:nrow(output_table), 
    #                              grep(pattern = "_T(conv|reg)$", 
    #                                   x = colnames(output_table), 
    #                                   ignore.case = T)])
    # 
    # percentiles <- output_table_excluded[5:nrow(output_table_excluded), 
    #                                      grepl(pattern = "_T(conv|reg)$", 
    #                                            x = colnames(output_table), 
    #                                            ignore.case = T)]
    # 
    # for(i in seq(1, ncol(percentiles))){
    #   temp <- ecdf(mean_tissue_counts[,i])
    #   percentiles[,i] <- temp(percentiles[,i])
    #   #percentiles <- percentiles*100
    # }
    # rm(i, temp)
    # percentiles <- percentiles*100
    # 
    # percentiles <- rbind(vector(mode = "character", length = ncol(percentiles)), 
    #                      vector(mode = "character", length = ncol(percentiles)), 
    #                      c("Percentiles in respective tissue expression", 
    #                        vector(mode = "character", length = ncol(percentiles)-1)), 
    #                      gsub(pattern = "_", 
    #                           replacement = " ", 
    #                           x = colnames(percentiles), 
    #                           ignore.case = T), 
    #                      percentiles)
    # 
    # # write.table(x = cbind(output_table_excluded, percentiles),
    # #   file = paste0(loc_diffexprs_merged, "output_table_excluded.txt"),
    # #   quote = F, sep = "\t", row.names = F, col.names = F)
    # 
    # 
    # percentiles$Gene.stable.ID <- output_table_excluded$Gene.stable.ID
    # 
    # output_table_combined <- merge(x = output_table[5:nrow(output_table),], 
    #                                y = percentiles[5:nrow(percentiles),], 
    #                                by = "Gene.stable.ID", all = T)
    # 
    # headers <- cbind(output_table[1:4,], percentiles[1:4, 1:ncol(percentiles)-1])
    # colnames(headers) <- colnames(output_table_combined)
    # output_table_combined <- rbind(headers, 
    #                                output_table_combined)
    # 
    # output_table_excluded_genes <- is.na(rowSums(
    #   data.matrix(output_table_combined[,112:ncol(output_table_combined)])
    #   ))
    # 
    # output_table_combined[c(FALSE, FALSE, FALSE, FALSE, 
    #                         output_table_excluded_genes[
    #                           5:length(output_table_excluded_genes)]), 
    #                       112:ncol(output_table_combined)] <- ""
    # 
    # # output_table_combined[c(FALSE, FALSE, FALSE, FALSE,
    # #                         !output_table_excluded_genes[
    # #                           5:length(output_table_excluded_genes)]),
    # #                       64:111] <- ""
    # # output_table_combined[c(FALSE, FALSE, FALSE, FALSE,
    # #                       !output_table_excluded_genes[
    # #                         5:length(output_table_excluded_genes)]),
    # #                       seq(64, 111, 4)] <- ""
    # # #seq(64, 111, 4)
    # 
    # #writing out the file
    # write.table(x = output_table_combined,
    #             file = paste0(loc_diffexprs_merged, "output_table_combined.txt"),
    #             quote = F, sep = "\t", row.names = F, col.names = F)
  }
  
  
}

```


#performing PCA and MDS on normalised count data
```{r}

#reading norm counts
{
  #read normalised counts from the '#writing out the differential expression 
  # files' code block to get them filtered by min transcripts per cell
  # norm_counts <- read.delim(file = paste0(loc_counts, "normalised_counts.txt"), 
  #                           sep = "\t", 
  #                           stringsAsFactors = F)
  
  contrast_groups <- list(c("Blood"), 
                          c("Spleen", "Lymph Node", "Peyer's Patches"), 
                          c("Intraepithelial Layer", 
                            "Lamina Propria Lymphocyte"), 
                          c("Kidney", "Lung", "Liver", "Pancreas"),
                          c("cellTypeT"),
                          c("cellTypeNT"))
    
  for(i in seq(1, length(contrast_groups))){
    contrast_groups[[i]] <- tolower(unique(
      gsub("_.*|[[:space:]]|[[:punct:]]", "", contrast_groups[[i]])
      ))
  }
}

#calculating log2 of normalised counts and removing genes with missing data for 
# the PCA, t-SNE and UMAP plots below
norm_exp_log2 <- data.matrix(log2(norm_counts))
norm_exp_log2[!is.finite(norm_exp_log2)] <- NA
norm_exp_log2 <- as.data.frame(norm_exp_log2[complete.cases(norm_exp_log2),])

#stripped_norm_counts <- norm_counts[rowSums(norm_counts)>0,]

#creating PCA plots via ggplot2
{
  
  #samples_PCA <- prcomp(t(norm_counts[rowSums(norm_counts)>0,]))
  #summary(samples_PCA)
  samples_PCA <- prcomp(t(norm_exp_log2))
  # samples_PCA <- data.matrix(mean_norm_exp)
  # samples_PCA[!is.finite(samples_PCA)] <- NA
  # samples_PCA <- samples_PCA[complete.cases(samples_PCA),]
  # samples_PCA <- prcomp(t(samples_PCA[complete.cases(samples_PCA),]))
  #samples_PCA <- prcomp(t(mean_norm_exp[complete.cases(mean_norm_exp),]))
  
  #plotting variance of PCs
  variance_proportion <- ((samples_PCA$sdev^2) / (sum(samples_PCA$sdev^2)))*100
  dev.new()
  barplot(variance_proportion, cex.names=1,
          xlab=paste("Principal component (PC), 1-",
                     length(samples_PCA$sdev)), 
          ylab="Proportion of variation (%)",
          main="Scree plot", ylim=c(0,100))
  #dev.off()
  
  ##PCA Plots
  {
    #plot(samples_PCA$x[,1], samples_PCA$x[,2]) --- individual colours
    samples_PCA_data <- as.data.frame(samples_PCA$x)
    #samples_PCA_data$group <- gsub("_[0-9]{1}$", "", colnames(norm_counts))
    #samples_PCA_data$group <- gsub("_[0-9]{1}$", "", colnames(mean_norm_exp))
    samples_PCA_data$group <- gsub("_[0-9]{1}$", "", colnames(norm_exp_log2))
    
    # #PCA loadings
    # PCA_loadings <- samples_PCA$rotation
    # 
    # #PCA coordinates?
    # PCA_coordinates <- samples_PCA$x
    # 
    # PCA_plot <- ggplot(samples_PCA_data, aes(x=PC1, y=PC2, color=group))
    # PCA_plot <- PCA_plot + geom_point() + 
    #   geom_text(aes(label=rownames(samples_PCA_data)),
    #             size=2.5, hjust=0.5, vjust=1.5, show.legend = F) + 
    #   theme_light()
    # dev.new()
    # PCA_plot
    # #dev.off()
    
    #plot PCA -- data points grouped by tissue groups
    tissue_group <- samples_PCA_data$group
    
    tissue_group <- gsub(pattern = "_(NT|T)", 
                       replacement = "", 
                       x = tissue_group, 
                       ignore.case = T)
    
    for(i in seq_along(tissue_group)){
      if (tolower(tissue_group[i]) == "blood"){
        tissue_group[i] <- "Blood"
      }
      else if (tolower(tissue_group[i]) %in% c("spleen", "lymphnode", 
                                               "peyerspatches")){
        tissue_group[i] <- "Lymphoid"
      }
      else if (tolower(tissue_group[i]) %in% 
          c("intraepitheliallayer", "laminaproprialymphocyte")){
        tissue_group[i] <- paste0("Gut-associated")
      }
      else if (tolower(tissue_group[i]) %in% 
          c("kidney", "lung", "liver", "pancreas")){
        tissue_group[i] <- "Non-lymphoid"
      }
      else if (tolower(tissue_group[i]) == "celltypet"){
        tissue_group[i] <- "Treg"
      }
      else if (tolower(tissue_group[i]) == "celltypent"){
        tissue_group[i] <- "Tconv"
      }
      
    }
    rm(i)
    
    samples_PCA_data$tissue_group <- as.factor(tissue_group)
    
    # PCA_plot <- ggplot(samples_PCA_data, 
    #                    aes(x=PC1, y=PC2, 
    #                        color= gsub(pattern = ", ",
    #                                  replacement = ", \n",
    #                                  x = tissue_group))
    #                    ) + 
    #   geom_point() + 
    #   geom_text(aes_(label=rownames(samples_PCA_data)),
    #             size=4, hjust=0.5, vjust=1.5, show.legend = F) + 
    #   #geom_jitter(width = 2, height = 2) + 
    #   labs(colour = "Tissue Groups") + 
    #   theme_light()
    
    #master plot
    {
      # PCA_plot <- ggplot(samples_PCA_data, 
      #                    aes(x=PC1, y=PC2, 
      #                        #color=tissue_group)
      #                        color= gsub(pattern = ", ",
      #                                    replacement = ", \n",
      #                                    x = tissue_group))
      #                  ) + 
      #   geom_point(aes_(shape = as.factor(ifelse(grepl(pattern = "_NT", 
      #                                 x = rownames(samples_PCA_data), 
      #                                 ignore.case = T), "CD4", "T-reg")))
      #              ) + 
      #   geom_text_repel(aes_(label= 
      #                          qdap::multigsub(pattern = 
      #                                            c("InterEpithelialLayer", 
      #                                              "LaminaPropriaLymphocyte", 
      #                                              "PeyersPatches", "LymphNode"), 
      #                                          replacement = c("IEL", "LPL", "PP",
      #                                                          "LN"), 
      #                                          order.pattern = F, 
      #                                          text.var = gsub(pattern = "_.*", 
      #                                                          replacement = "", 
      #                                                          x = rownames(
      #                                                            samples_PCA_data), 
      #                                                          ignore.case = T)
      #                                            )
      #                        ), 
      #                   size = 4, segment.alpha = 0.4,
      #                   force = 1, direction = "both", show.legend = F) + 
      #   scale_shape_manual(values = c(4, 20)) + 
      #   labs(colour = "Tissue Groups", 
      #        shape = "Cell Types",
      #        caption = paste0("IEL: Intraepithelial layer; LPL: ", 
      #                         "Lamina propria lymphocytes; PP: Peyer's patches; ",
      #                         "LN: Lymph node"),
      #        title = "PCA plot of 40 samples"
      #        ) + 
      #   theme_light() + 
      #   theme(legend.key.size = unit(2.8, 'lines'), 
      #         plot.caption = element_text(hjust = 0, face= "italic"),
      #         panel.grid.major = element_blank(), panel.grid.minor = element_blank()
      #         )
      # 
      # PCA_plot
    }
    
    
    #only CD4 and Treg style
    {
      Type <- as.factor(ifelse(grepl(pattern = "_NT", 
                                      x = rownames(samples_PCA_data), 
                                      ignore.case = T), "Tconv", "Treg"))
      
      Label <- qdap::multigsub(pattern = 
                                   c("IntraEpithelialLayer", 
                                     "LaminaPropriaLymphocyte", 
                                     "PeyersPatches", "LymphNode"), 
                                 replacement = c("IEL", "LPL", "PP",
                                                 "LN"), 
                                 order.pattern = F, 
                                 text.var = gsub(pattern = "_.*", 
                                                 replacement = "", 
                                                 x = rownames(
                                                   samples_PCA_data), 
                                                 ignore.case = T)
                                   )
      
      PCA_plot1 <- ggplot(samples_PCA_data, 
                         aes(#Label = Label, Type = Type, 
                              x=PC1, 
                              y=PC2)
                       ) + 
        geom_point(aes_(shape = as.factor(ifelse(grepl(pattern = "_NT", 
                                      x = rownames(samples_PCA_data), 
                                      ignore.case = T), "Tconv", "Treg"))), 
                   size = 4) + 
        geom_text_repel(aes_(label= 
                               qdap::multigsub(pattern = 
                                                 c("IntraEpithelialLayer", 
                                                   "LaminaPropriaLymphocyte", 
                                                   "PeyersPatches", 
                                                   "LymphNode"), 
                                               replacement = c("IEL", "LPL", 
                                                               "PP", "LN"), 
                                               order.pattern = F, 
                                               text.var = gsub(pattern = "_.*", 
                                                               replacement = "", 
                                                               x = rownames(
                                                                 samples_PCA_data), 
                                                               ignore.case = T)
                                                 )
                             ), 
                        #size = 4, segment.alpha = 0.4, #for offline images
                        size = 5, segment.alpha = 0.4, #for webtool rds files
                        force = 1, direction = "both", show.legend = F) + 
        scale_shape_manual(values = c(4, 20)) + 
        labs(shape = "Cell Types", 
             #title = paste0("PCA plot of ", nrow(samples_PCA_data) , " samples")
             title = paste0("All ", nrow(samples_PCA_data) , " samples")
             ) + 
        #theme_light(base_size = 13.5) + #for offline images
        theme_light(base_size = 18) + #for webtool rds files
        theme(legend.key.size = unit(0.15, 'lines'), 
              plot.caption = element_text(hjust = 0, face= "italic"),
              panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank()
              )
      
      #PCA_plot1
      
      # #for ggplotly in the webtool
      # PCA_plot1 <- ggplotly(PCA_plot1,
      #                       tooltip = c("Label", "Type", "PC1", "PC2")) %>%
      #   toWebGL()
    }
    
    #only Treg plot
    {
      subset_data <- samples_PCA_data[grepl(pattern = "_T", 
                                            x = rownames(samples_PCA_data), 
                                            ignore.case = F),]
      
      Type <- as.factor(ifelse(grepl(pattern = "_NT", 
                                      x = rownames(subset_data), 
                                      ignore.case = F), "Tconv", "Treg"))
      
      Label <- qdap::multigsub(pattern = 
                                     c("IntraEpithelialLayer", 
                                       "LaminaPropriaLymphocyte", 
                                       "PeyersPatches", "LymphNode"), 
                                   replacement = c("IEL", "LPL", "PP",
                                                   "LN"), 
                                   order.pattern = F, 
                                   text.var = gsub(pattern = "_.*", 
                                                   replacement = "", 
                                                   x = rownames(
                                                     subset_data), 
                                                   ignore.case = T)
                                     )
      
      #manually assigning class colours
      {
        tissue_group_col <- tissue_group
        names(tissue_group_col) <- tissue_group_col
        
        tissue_group_col[grepl(pattern = "Blood",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#D81B60"
        tissue_group_col[grepl(pattern = "(Spleen|PP|LN|^Lymphoid)",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#FFC107"
        tissue_group_col[grepl(pattern = "(IEL|LPL|Gut)",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#1E88E5"
        tissue_group_col[grepl(pattern = "(Kidney|Lung|Liver|Pancreas|Non)",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#004D40"
      }
      
      
      PCA_plot2 <- ggplot(subset_data, 
                          aes(#Type = Type,
                              #Label = Label, 
                              x = PC1, y = PC2, 
                              #color=tissue_group)
                              color = gsub(pattern = ", ",
                                          replacement = ", \n",
                                          x = tissue_group)
                              )
                       ) + 
        geom_point(aes_(shape = as.factor(ifelse(grepl(pattern = "_NT", 
                                      x = rownames(subset_data), 
                                      ignore.case = F), "Tconv", "Treg"))), 
                   size = 4) + 
        geom_text_repel(aes_(label=
                               qdap::multigsub(pattern =
                                                 c("IntraEpithelialLayer",
                                                   "LaminaPropriaLymphocyte",
                                                   "PeyersPatches",
                                                   "LymphNode"),
                                               replacement = c("IEL", "LPL",
                                                               "PP", "LN"),
                                               order.pattern = F,
                                               text.var = gsub(pattern = "_.*",
                                                               replacement = "",
                                                               x = rownames(
                                                                 subset_data),
                                                               ignore.case = T)
                                                 )
                             ),
                        #size = 7, segment.alpha = 0.4, #for offline images
                        size = 5, segment.alpha = 0.4, #for webtool rds files
                        force = 1, direction = "both", show.legend = F) +
        scale_shape_manual(values = c(20, 4)) + 
        scale_colour_manual(values = tissue_group_col) + 
        labs(colour = "Tissue", 
             shape = "Cell Types",
             title = "Treg samples"
             ) + 
        #theme_light(base_size = 13.5) + #for offline images
        theme_light(base_size = 18) + #for webtool rds files
        theme(legend.key.size = unit(0.15, 'lines'), 
              plot.caption = element_text(hjust = 0, face= "italic"),
              panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank()
              ) + 
        scale_x_continuous(limits = c(min(samples_PCA_data$PC1), 
                                      max(samples_PCA_data$PC1))) + 
        scale_y_continuous(limits = c(min(samples_PCA_data$PC2), 
                                      max(samples_PCA_data$PC2))) + 
        guides(shape = F)
      
      #PCA_plot2
      
      # #for ggplotly in the webtool -- modify the weird brackets in the legend
      # myplot <- ggplotly(PCA_plot2, tooltip = c("Label", "Type", "PC1", "PC2")) %>% 
      #   toWebGL()
      # 
      # for (i in 1:length(myplot$x$data)){
      #   if (!is.null(myplot$x$data[[i]]$name)){
      #     myplot$x$data[[i]]$name =  gsub("\\(|\\)","",str_split(myplot$x$data[[i]]$name,",")[[1]][2])
      #   }
      # }
      # 
      # myplot
      
    }
    
    #only CD4 plot
    {
      subset_data <- samples_PCA_data[grepl(pattern = "_NT", 
                                            x = rownames(samples_PCA_data), 
                                            ignore.case = F),]
      
      Type <- as.factor(ifelse(grepl(pattern = "_NT", 
                                      x = rownames(subset_data), 
                                      ignore.case = F), "Tconv", "Treg"))
      
      Label <- qdap::multigsub(pattern = 
                                   c("IntraEpithelialLayer", 
                                     "LaminaPropriaLymphocyte", 
                                     "PeyersPatches", "LymphNode"), 
                                 replacement = c("IEL", "LPL", "PP",
                                                 "LN"), 
                                 order.pattern = F, 
                                 text.var = gsub(pattern = "_.*", 
                                                 replacement = "", 
                                                 x = rownames(
                                                   subset_data), 
                                                 ignore.case = T)
                                   )
      
      #manually assigning class colours
      {
        tissue_group_col <- tissue_group
        names(tissue_group_col) <- tissue_group_col
        
        tissue_group_col[grepl(pattern = "Blood",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#D81B60"
        tissue_group_col[grepl(pattern = "(Spleen|PP|LN|^Lymphoid)",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#FFC107"
        tissue_group_col[grepl(pattern = "(IEL|LPL|Gut)",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#1E88E5"
        tissue_group_col[grepl(pattern = "(Kidney|Lung|Liver|Pancreas|Non)",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#004D40"
      }
      
      PCA_plot3 <- ggplot(subset_data, 
                          aes(#Type = Type,
                              #Label = Label, 
                              x = PC1, y = PC2, 
                              #color=tissue_group)
                              color= gsub(pattern = ", ",
                                          replacement = ", \n",
                                          x = tissue_group))
                       ) + 
        geom_point(aes_(shape = as.factor(ifelse(grepl(pattern = "_NT", 
                                      x = rownames(subset_data), 
                                      ignore.case = F), "Tconv", "Treg"))), 
                   size = 4) + 
        geom_text_repel(aes_(label=
                               qdap::multigsub(pattern =
                                                 c("IntraEpithelialLayer",
                                                   "LaminaPropriaLymphocyte",
                                                   "PeyersPatches",
                                                   "LymphNode"),
                                               replacement = c("IEL", "LPL",
                                                               "PP", "LN"),
                                               order.pattern = F,
                                               text.var = gsub(pattern = "_.*",
                                                               replacement = "",
                                                               x = rownames(
                                                                 subset_data),
                                                               ignore.case = T)
                                                 )
                             ),
                        #size = 4, segment.alpha = 0.4, #for offline images
                        size = 5, segment.alpha = 0.4, #for webtool rds files
                        force = 1, direction = "both", show.legend = F) +
        scale_shape_manual(values = c(4, 20)) + 
        scale_colour_manual(values = tissue_group_col) + 
        labs(colour = "Tissue Groups", 
             shape = "Cell Types",
             title = "Tconv samples"
             ) + 
        #theme_light(base_size = 13.5) + #for offline images
        theme_light(base_size = 18) + #for webtool rds files
        theme(legend.key.size = unit(0.15, 'lines'), 
              plot.caption = element_text(hjust = 0, face= "italic"),
              panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank()
              ) + 
        scale_x_continuous(limits = c(min(samples_PCA_data$PC1), 
                                      max(samples_PCA_data$PC1))) + 
        scale_y_continuous(limits = c(min(samples_PCA_data$PC2), 
                                      max(samples_PCA_data$PC2))) + 
        guides(shape = F)
      
      #PCA_plot3
      
      # #for ggplotly in the webtool -- modify the weird brackets in the legend
      # myplot <- ggplotly(PCA_plot3, tooltip = c("Label", "Type", "PC1", "PC2")) %>%
      #   toWebGL()
      # 
      # for (i in 1:length(myplot$x$data)){
      #   if (!is.null(myplot$x$data[[i]]$name)){
      #     myplot$x$data[[i]]$name =  gsub("\\(|\\)","",str_split(myplot$x$data[[i]]$name,",")[[1]][2])
      #   }
      # }
      # 
      # myplot
      
    }
    
    #multiplot(PCA_plot1, PCA_plot2, PCA_plot3, cols = 2)
    
    PCA_plot1 + PCA_plot2 + PCA_plot3 + 
      plot_layout(ncol = 2, nrow = 2, byrow = T)
    gc()
    
  }
  
  
  #unneeded
  {
    
  # ##PCA Plots using centroid of tissue_T_# samples
  # {
  #   #plot(samples_PCA$x[,1], samples_PCA$x[,2]) --- individual colours
  #   
  #   T_centroid_PCA_data <- as.data.frame(matrix(data = 0, 
  #                                               nrow = 0, 
  #                                               ncol = ncol(samples_PCA$x)))
  #   colnames(T_centroid_PCA_data) <- colnames(samples_PCA$x)
  #   i <- 1
  #   while(i < nrow(samples_PCA$x)) {
  #     
  #     if((i-1)%%4 == 0){
  #       #print("In simple copy")
  #       T_centroid_PCA_data[nrow(T_centroid_PCA_data)+1,] <- 
  #         (samples_PCA$x[i,])
  #       
  #       rownames(T_centroid_PCA_data)[nrow(T_centroid_PCA_data)] <- 
  #         gsub(pattern = "_[0-9]{1}$", 
  #              replacement = "", 
  #              x = rownames(samples_PCA$x)[i])
  #       
  #       i <- i+1
  #     }
  #     else{
  #       #print("In rowMeans")
  #       T_centroid_PCA_data[nrow(T_centroid_PCA_data)+1,] <- 
  #         rowMeans(t(samples_PCA$x[i:(i+2),]))
  #       
  #       rownames(T_centroid_PCA_data)[nrow(T_centroid_PCA_data)] <- 
  #         gsub(pattern = "_[0-9]{1}$", 
  #              replacement = "", 
  #              x = rownames(samples_PCA$x)[i])
  #       
  #       i <- i+3
  #     }
  #     
  #   }
  #   rm(i)
  #   
  #   T_centroid_PCA_data$group <- unique(gsub(pattern = "_[0-9]{1}$", 
  #                                            replacement = "", 
  #                                            x = rownames(samples_PCA$x)))
  #   
  #   NT_coordinates <- T_centroid_PCA_data[grepl(
  #     pattern = "_NT$", x = rownames(T_centroid_PCA_data)),]
  #   T_coordinates <- T_centroid_PCA_data[grepl(
  #     pattern = "_T$", x = rownames(T_centroid_PCA_data)),]
  #   
  #   PCA_plot <- ggplot(T_centroid_PCA_data, aes(x=PC1, y=PC2, color=group))
  #   PCA_plot <- PCA_plot + geom_point() + 
  #     geom_text(aes(label=rownames(T_centroid_PCA_data)),
  #               size=3, hjust=0.5, vjust=1.5, show.legend = F) + 
  #     geom_segment(data = NT_coordinates, aes(x = NT_coordinates[,1], y = NT_coordinates[,2], 
  #                      xend = T_coordinates[,1], yend = T_coordinates[,2])) + 
  #     theme_light()
  #   
  #   dev.new()
  #   PCA_plot
  #   #dev.off()
  #   
  # }
  # 
  # ##PCA Plots after removing the three samples
  # {
  #   #plot(samples_PCA$x[,1], samples_PCA$x[,2]) --- individual colours
  #   samples_PCA_data <- as.data.frame(samples_PCA$x)[!grepl(
  #     pattern = "Pancreas_T_3|Liver_T_3|PeyersPatches_T_1", 
  #     x = rownames(samples_PCA$x), 
  #     ignore.case = T),]
  #   samples_PCA_data$group <- gsub("_[0-9]{1}$", "", rownames(samples_PCA_data))
  #   
  #   # #PCA loadings
  #   # PCA_loadings <- samples_PCA$rotation
  #   # 
  #   # #PCA coordinates?
  #   # PCA_coordinates <- samples_PCA$x
  #   
  #   #plot PCA -- data points grouped by tissue groups
  #   tissue_group <- samples_PCA_data$group
  #   
  #   tissue_group <- gsub(pattern = "_(NT|T)", 
  #                      replacement = "", 
  #                      x = tissue_group, 
  #                      ignore.case = T)
  #   
  #   for(i in seq_along(tissue_group)){
  #     if (tolower(tissue_group[i]) == "blood"){
  #       tissue_group[i] <- "Blood"
  #     }
  #     else if (tolower(tissue_group[i]) %in% c("spleen", "lymphnode", 
  #                                              "peyerspatches")){
  #       tissue_group[i] <- "Lymphoid tissue"
  #     }
  #     else if (tolower(tissue_group[i]) %in% 
  #         c("intraepitheliallayer", "laminaproprialymphocyte")){
  #       tissue_group[i] <- paste0("Gut-associated\ntissues")
  #     }
  #     else if (tolower(tissue_group[i]) %in% 
  #         c("kidney", "lung", "liver", "pancreas")){
  #       tissue_group[i] <- "Non-lymphoid\ntissues"
  #     }
  #     else if (tolower(tissue_group[i]) == "celltypet"){
  #       tissue_group[i] <- "Regulatory T-cells"
  #     }
  #     else if (tolower(tissue_group[i]) == "celltypent"){
  #       tissue_group[i] <- "CD4 T-cells"
  #     }
  #     
  #   }
  #   
  #   samples_PCA_data$tissue_group <- as.factor(tissue_group)
  #   
  #   # PCA_plot <- ggplot(samples_PCA_data, 
  #   #                    aes(x=PC1, y=PC2, color=tissue_group)
  #   #                    ) + 
  #   #   geom_point() + 
  #   #   geom_text(aes_(label=rownames(samples_PCA_data)),
  #   #             size=4, hjust=0.5, vjust=1.5, show.legend = F) + 
  #   #   labs(colour = "Tissue Groups") + 
  #   #   theme_light()
  #   
  #   
  #   #master plot
  #   {
  #     # PCA_plot <- ggplot(samples_PCA_data, 
  #     #                    aes(x=PC1, y=PC2, 
  #     #                        #color=tissue_group)
  #     #                        color= gsub(pattern = ", ",
  #     #                                    replacement = ", \n",
  #     #                                    x = tissue_group))
  #     #                  ) + 
  #     #   geom_point(aes_(shape = as.factor(ifelse(grepl(pattern = "_NT", 
  #     #                                 x = rownames(samples_PCA_data), 
  #     #                                 ignore.case = T), "CD4", "T-reg")))
  #     #              ) + 
  #     #   geom_text_repel(aes_(label= 
  #     #                          qdap::multigsub(pattern = 
  #     #                                            c("InterEpithelialLayer", 
  #     #                                              "LaminaPropriaLymphocyte", 
  #     #                                              "PeyersPatches", "LymphNode"), 
  #     #                                          replacement = c("IEL", "LPL", "PP",
  #     #                                                          "LN"), 
  #     #                                          order.pattern = F, 
  #     #                                          text.var = gsub(pattern = "_.*", 
  #     #                                                          replacement = "", 
  #     #                                                          x = rownames(
  #     #                                                            samples_PCA_data), 
  #     #                                                          ignore.case = T)
  #     #                                            )
  #     #                        ), 
  #     #                   size = 4, segment.alpha = 0.4,
  #     #                   force = 1, direction = "both", show.legend = F) + 
  #     #   scale_shape_manual(values = c(4, 20)) + 
  #     #   labs(colour = "Tissue Groups", 
  #     #        shape = "Cell Types",
  #     #        caption = paste0("IEL: Intraepithelial layer; LPL: ", 
  #     #                         "Lamina propria lymphocytes; PP: Peyer's patches; ",
  #     #                         "LN: Lymph node"),
  #     #        title = "PCA plot of 37 samples"
  #     #        ) + 
  #     #   theme_light() + 
  #     #   theme(legend.key.size = unit(2.8, 'lines'), 
  #     #         plot.caption = element_text(hjust = 0, face= "italic"),
  #     #         panel.grid.major = element_blank(), panel.grid.minor = element_blank()
  #     #         )
  #     # 
  #     # PCA_plot
  #   }
  #   
  #   
  #   #only CD4 and Treg style
  #   {
  #     PCA_plot1 <- ggplot(samples_PCA_data, 
  #                        aes(x=PC1, y=PC2)
  #                      ) + 
  #       geom_point(aes_(shape = as.factor(ifelse(grepl(pattern = "_NT", 
  #                                     x = rownames(samples_PCA_data), 
  #                                     ignore.case = T), "CD4", "T-reg")))
  #                  ) + 
  #       geom_text_repel(aes_(label= 
  #                              qdap::multigsub(pattern = 
  #                                                c("InterEpithelialLayer", 
  #                                                  "LaminaPropriaLymphocyte", 
  #                                                  "PeyersPatches", "LymphNode"), 
  #                                              replacement = c("IEL", "LPL", "PP",
  #                                                              "LN"), 
  #                                              order.pattern = F, 
  #                                              text.var = gsub(pattern = "_.*", 
  #                                                              replacement = "", 
  #                                                              x = rownames(
  #                                                                samples_PCA_data), 
  #                                                              ignore.case = T)
  #                                                )
  #                            ), 
  #                       size = 4, segment.alpha = 0.4,
  #                       force = 1, direction = "both", show.legend = F) + 
  #       scale_shape_manual(values = c(4, 20)) + 
  #       labs(shape = "Cell Types", 
  #            title = paste0("PCA plot of ", nrow(samples_PCA_data), " samples")
  #            ) + 
  #       theme_light() + 
  #       theme(legend.key.size = unit(1, 'lines'), 
  #             plot.caption = element_text(hjust = 0, face= "italic"),
  #             panel.grid.major = element_blank(), panel.grid.minor = element_blank()
  #             )
  #     
  #     PCA_plot1
  #   }
  #   
  #   #only T-reg plot
  #   {
  #     subset_data <- samples_PCA_data[grepl(pattern = "_T", 
  #                                           x = rownames(samples_PCA_data), 
  #                                           ignore.case = F),]
  #     PCA_plot2 <- ggplot(subset_data, 
  #                         aes(x=PC1, y=PC2, 
  #                             #color=tissue_group)
  #                             color= gsub(pattern = ", ",
  #                                         replacement = ", \n",
  #                                         x = tissue_group))
  #                      ) + 
  #       geom_point(aes_(shape = as.factor(ifelse(grepl(pattern = "_NT", 
  #                                     x = rownames(subset_data), 
  #                                     ignore.case = F), "CD4", "T-reg")))
  #                  ) + 
  #       geom_text_repel(aes_(label= 
  #                              qdap::multigsub(pattern = 
  #                                                c("InterEpithelialLayer", 
  #                                                  "LaminaPropriaLymphocyte", 
  #                                                  "PeyersPatches", "LymphNode"), 
  #                                              replacement = c("IEL", "LPL", "PP",
  #                                                              "LN"), 
  #                                              order.pattern = F, 
  #                                              text.var = gsub(pattern = "_.*", 
  #                                                              replacement = "", 
  #                                                              x = rownames(
  #                                                                subset_data), 
  #                                                              ignore.case = T)
  #                                                )
  #                            ), 
  #                       size = 4, segment.alpha = 0.4,
  #                       force = 1, direction = "both", show.legend = F) + 
  #       scale_shape_manual(values = c(20, 4)) + 
  #       labs(colour = "Tissue Groups", 
  #            shape = "Cell Types",
  #            title = "PCA plot of T-reg samples"
  #            ) + 
  #       theme_light() + 
  #       theme(legend.key.size = unit(1.5, 'lines'), 
  #             plot.caption = element_text(hjust = 0, face= "italic"),
  #             panel.grid.major = element_blank(), panel.grid.minor = element_blank()
  #             ) + 
  #       scale_x_continuous(limits = c(min(samples_PCA_data$PC1), 
  #                                     max(samples_PCA_data$PC1))) + 
  #       scale_y_continuous(limits = c(min(samples_PCA_data$PC2), 
  #                                     max(samples_PCA_data$PC2))) + 
  #       guides(shape = F)
  #     
  #     PCA_plot2
  #   }
  #   
  #   #only CD4 plot
  #   {
  #     subset_data <- samples_PCA_data[grepl(pattern = "_NT", 
  #                                           x = rownames(samples_PCA_data), 
  #                                           ignore.case = F),]
  #     PCA_plot3 <- ggplot(subset_data, 
  #                         aes(x=PC1, y=PC2, 
  #                             #color=tissue_group)
  #                             color= gsub(pattern = ", ",
  #                                         replacement = ", \n",
  #                                         x = tissue_group))
  #                      ) + 
  #       geom_point(aes_(shape = as.factor(ifelse(grepl(pattern = "_NT", 
  #                                     x = rownames(subset_data), 
  #                                     ignore.case = F), "CD4", "T-reg")))
  #                  ) + 
  #       geom_text_repel(aes_(label= 
  #                              qdap::multigsub(pattern = 
  #                                                c("InterEpithelialLayer", 
  #                                                  "LaminaPropriaLymphocyte", 
  #                                                  "PeyersPatches", "LymphNode"), 
  #                                              replacement = c("IEL", "LPL", "PP",
  #                                                              "LN"), 
  #                                              order.pattern = F, 
  #                                              text.var = gsub(pattern = "_.*", 
  #                                                              replacement = "", 
  #                                                              x = rownames(
  #                                                                subset_data), 
  #                                                              ignore.case = T)
  #                                                )
  #                            ), 
  #                       size = 4, segment.alpha = 0.4,
  #                       force = 1, direction = "both", show.legend = F) + 
  #       scale_shape_manual(values = c(4, 20)) + 
  #       labs(colour = "Tissue Groups", 
  #            shape = "Cell Types",
  #            title = "PCA plot of CD4 samples"
  #            ) + 
  #       theme_light() + 
  #       theme(legend.key.size = unit(1.5, 'lines'), 
  #             plot.caption = element_text(hjust = 0, face= "italic"),
  #             panel.grid.major = element_blank(), panel.grid.minor = element_blank()
  #             ) + 
  #       scale_x_continuous(limits = c(min(samples_PCA_data$PC1), 
  #                                     max(samples_PCA_data$PC1))) + 
  #       scale_y_continuous(limits = c(min(samples_PCA_data$PC2), 
  #                                     max(samples_PCA_data$PC2))) + 
  #       guides(shape = F)
  #     
  #     PCA_plot3
  #   }
  #   
  #   #multiplot(PCA_plot1, PCA_plot2, PCA_plot3, cols = 2)
  #   
  #   PCA_plot1 + PCA_plot2 + PCA_plot3 + 
  #     plot_layout(ncol = 2, nrow = 2, byrow = T)
  #   
  #   
  # }
  
  }
  
}

#performing PCA on (mean) normalised count data
{
  
  #samples_PCA <- prcomp(t(norm_counts[rowSums(norm_counts)>0,]))
  #summary(samples_PCA)
  #samples_PCA <- prcomp(t(norm_exp_log2))
  samples_PCA <- data.matrix(mean_norm_exp)
  samples_PCA[!is.finite(samples_PCA)] <- NA
  samples_PCA <- samples_PCA[complete.cases(samples_PCA),]
  samples_PCA <- prcomp(t(samples_PCA[complete.cases(samples_PCA),]))
  #samples_PCA <- prcomp(t(mean_norm_exp[complete.cases(mean_norm_exp),]))
  
  #plotting variance of PCs
  variance_proportion <- ((samples_PCA$sdev^2) / (sum(samples_PCA$sdev^2)))*100
  dev.new()
  barplot(variance_proportion, cex.names=1,
          xlab=paste("Principal component (PC), 1-",
                     length(samples_PCA$sdev)), 
          ylab="Proportion of variation (%)",
          main="Scree plot", ylim=c(0,100))
  #dev.off()
  
  ##PCA Plots
  {
    #plot(samples_PCA$x[,1], samples_PCA$x[,2]) --- individual colours
    samples_PCA_data <- as.data.frame(samples_PCA$x)
    #samples_PCA_data$group <- gsub("_[0-9]{1}$", "", colnames(norm_counts))
    samples_PCA_data$group <- gsub("_[0-9]{1}$", "", colnames(mean_norm_exp))
    #samples_PCA_data$group <- gsub("_[0-9]{1}$", "", colnames(norm_exp_log2))
    
    #plot PCA -- data points grouped by tissue groups
    tissue_group <- samples_PCA_data$group
    
    tissue_group <- gsub(pattern = "_(NT|T)", 
                       replacement = "", 
                       x = tissue_group, 
                       ignore.case = T)
    
    for(i in seq_along(tissue_group)){
      if (tolower(tissue_group[i]) == "blood"){
        tissue_group[i] <- "Blood"
      }
      else if (tolower(tissue_group[i]) %in% c("spleen", "lymphnode", 
                                               "peyerspatches")){
        tissue_group[i] <- "Lymphoid"
      }
      else if (tolower(tissue_group[i]) %in% 
          c("intraepitheliallayer", "laminaproprialymphocyte")){
        tissue_group[i] <- paste0("Gut-associated")
      }
      else if (tolower(tissue_group[i]) %in% 
          c("kidney", "lung", "liver", "pancreas")){
        tissue_group[i] <- "Non-lymphoid"
      }
      else if (tolower(tissue_group[i]) == "celltypet"){
        tissue_group[i] <- "Treg"
      }
      else if (tolower(tissue_group[i]) == "celltypent"){
        tissue_group[i] <- "Tconv"
      }
      
    }
    rm(i)
    
    samples_PCA_data$tissue_group <- as.factor(tissue_group)
    
    #only CD4 and Treg style
    {
      Type <- as.factor(ifelse(grepl(pattern = "_NT", 
                                      x = rownames(samples_PCA_data), 
                                      ignore.case = T), "Tconv", "Treg"))
      
      Label <- qdap::multigsub(pattern = 
                                   c("IntraEpithelialLayer", 
                                     "LaminaPropriaLymphocyte", 
                                     "PeyersPatches", "LymphNode"), 
                                 replacement = c("IEL", "LPL", "PP",
                                                 "LN"), 
                                 order.pattern = F, 
                                 text.var = gsub(pattern = "_.*", 
                                                 replacement = "", 
                                                 x = rownames(
                                                   samples_PCA_data), 
                                                 ignore.case = T)
                                   )
      
      PCA_plot1 <- ggplot(samples_PCA_data, 
                         aes(#Label = Label, Type = Type, 
                              x=PC1, 
                              y=PC2)
                       ) + 
        geom_point(aes_(shape = as.factor(ifelse(grepl(pattern = "_NT", 
                                      x = rownames(samples_PCA_data), 
                                      ignore.case = T), "Tconv", "Treg"))), 
                   size = 4) + 
        geom_text_repel(aes_(label= 
                               qdap::multigsub(pattern = 
                                                 c("IntraEpithelialLayer", 
                                                   "LaminaPropriaLymphocyte", 
                                                   "PeyersPatches", 
                                                   "LymphNode"), 
                                               replacement = c("IEL", "LPL", 
                                                               "PP", "LN"), 
                                               order.pattern = F, 
                                               text.var = gsub(pattern = "_.*", 
                                                               replacement = "", 
                                                               x = rownames(
                                                                 samples_PCA_data), 
                                                               ignore.case = T)
                                                 )
                             ), 
                        #size = 4, segment.alpha = 0.4, #for offline images
                        size = 5, segment.alpha = 0.4, #for webtool rds files
                        force = 1, direction = "both", show.legend = F) + 
        scale_shape_manual(values = c(4, 20)) + 
        labs(shape = "Cell Types", 
             #title = paste0("PCA plot of ", nrow(samples_PCA_data) , " samples")
             title = paste0("All ", nrow(samples_PCA_data) , " samples")
             ) + 
        #theme_light(base_size = 13.5) + #for offline images
        theme_light(base_size = 18) + #for webtool rds files
        theme(legend.key.size = unit(0.15, 'lines'), 
              plot.caption = element_text(hjust = 0, face= "italic"),
              panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank()
              )
      
      #PCA_plot1
      
      # #for ggplotly in the webtool
      # PCA_plot1 <- ggplotly(PCA_plot1,
      #                       tooltip = c("Label", "Type", "PC1", "PC2")) %>%
      #   toWebGL()
    }
    
    #only Treg plot
    {
      subset_data <- samples_PCA_data[grepl(pattern = "_T", 
                                            x = rownames(samples_PCA_data), 
                                            ignore.case = F),]
      
      Type <- as.factor(ifelse(grepl(pattern = "_NT", 
                                      x = rownames(subset_data), 
                                      ignore.case = F), "Tconv", "Treg"))
      
      Label <- qdap::multigsub(pattern = 
                                     c("IntraEpithelialLayer", 
                                       "LaminaPropriaLymphocyte", 
                                       "PeyersPatches", "LymphNode"), 
                                   replacement = c("IEL", "LPL", "PP",
                                                   "LN"), 
                                   order.pattern = F, 
                                   text.var = gsub(pattern = "_.*", 
                                                   replacement = "", 
                                                   x = rownames(
                                                     subset_data), 
                                                   ignore.case = T)
                                     )
      
      #manually assigning class colours
      {
        tissue_group_col <- tissue_group
        names(tissue_group_col) <- tissue_group_col
        
        tissue_group_col[grepl(pattern = "Blood",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#D81B60"
        tissue_group_col[grepl(pattern = "(Spleen|PP|LN|^Lymphoid)",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#FFC107"
        tissue_group_col[grepl(pattern = "(IEL|LPL|Gut)",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#1E88E5"
        tissue_group_col[grepl(pattern = "(Kidney|Lung|Liver|Pancreas|Non)",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#004D40"
      }
      
      
      PCA_plot2 <- ggplot(subset_data, 
                          aes(#Type = Type,
                              #Label = Label, 
                              x = PC1, y = PC2, 
                              #color=tissue_group)
                              color = gsub(pattern = ", ",
                                          replacement = ", \n",
                                          x = tissue_group)
                              )
                       ) + 
        geom_point(aes_(shape = as.factor(ifelse(grepl(pattern = "_NT", 
                                      x = rownames(subset_data), 
                                      ignore.case = F), "Tconv", "Treg"))), 
                   size = 4) + 
        geom_text_repel(aes_(label=
                               qdap::multigsub(pattern =
                                                 c("IntraEpithelialLayer",
                                                   "LaminaPropriaLymphocyte",
                                                   "PeyersPatches",
                                                   "LymphNode"),
                                               replacement = c("IEL", "LPL",
                                                               "PP", "LN"),
                                               order.pattern = F,
                                               text.var = gsub(pattern = "_.*",
                                                               replacement = "",
                                                               x = rownames(
                                                                 subset_data),
                                                               ignore.case = T)
                                                 )
                             ),
                        #size = 7, segment.alpha = 0.4, #for offline images
                        size = 5, segment.alpha = 0.4, #for webtool rds files
                        force = 1, direction = "both", show.legend = F) +
        scale_shape_manual(values = c(20, 4)) + 
        scale_colour_manual(values = tissue_group_col) + 
        labs(colour = "Tissue", 
             shape = "Cell Types",
             title = "Treg samples"
             ) + 
        #theme_light(base_size = 13.5) + #for offline images
        theme_light(base_size = 18) + #for webtool rds files
        theme(legend.key.size = unit(0.15, 'lines'), 
              plot.caption = element_text(hjust = 0, face= "italic"),
              panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank()
              ) + 
        scale_x_continuous(limits = c(min(samples_PCA_data$PC1), 
                                      max(samples_PCA_data$PC1))) + 
        scale_y_continuous(limits = c(min(samples_PCA_data$PC2), 
                                      max(samples_PCA_data$PC2))) + 
        guides(shape = F)
      
      #PCA_plot2
      
      # #for ggplotly in the webtool -- modify the weird brackets in the legend
      # myplot <- ggplotly(PCA_plot2, tooltip = c("Label", "Type", "PC1", "PC2")) %>% 
      #   toWebGL()
      # 
      # for (i in 1:length(myplot$x$data)){
      #   if (!is.null(myplot$x$data[[i]]$name)){
      #     myplot$x$data[[i]]$name =  gsub("\\(|\\)","",str_split(myplot$x$data[[i]]$name,",")[[1]][2])
      #   }
      # }
      # 
      # myplot
      
    }
    
    #only CD4 plot
    {
      subset_data <- samples_PCA_data[grepl(pattern = "_NT", 
                                            x = rownames(samples_PCA_data), 
                                            ignore.case = F),]
      
      Type <- as.factor(ifelse(grepl(pattern = "_NT", 
                                      x = rownames(subset_data), 
                                      ignore.case = F), "Tconv", "Treg"))
      
      Label <- qdap::multigsub(pattern = 
                                   c("IntraEpithelialLayer", 
                                     "LaminaPropriaLymphocyte", 
                                     "PeyersPatches", "LymphNode"), 
                                 replacement = c("IEL", "LPL", "PP",
                                                 "LN"), 
                                 order.pattern = F, 
                                 text.var = gsub(pattern = "_.*", 
                                                 replacement = "", 
                                                 x = rownames(
                                                   subset_data), 
                                                 ignore.case = T)
                                   )
      
      #manually assigning class colours
      {
        tissue_group_col <- tissue_group
        names(tissue_group_col) <- tissue_group_col
        
        tissue_group_col[grepl(pattern = "Blood",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#D81B60"
        tissue_group_col[grepl(pattern = "(Spleen|PP|LN|^Lymphoid)",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#FFC107"
        tissue_group_col[grepl(pattern = "(IEL|LPL|Gut)",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#1E88E5"
        tissue_group_col[grepl(pattern = "(Kidney|Lung|Liver|Pancreas|Non)",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#004D40"
      }
      
      PCA_plot3 <- ggplot(subset_data, 
                          aes(#Type = Type,
                              #Label = Label, 
                              x = PC1, y = PC2, 
                              #color=tissue_group)
                              color= gsub(pattern = ", ",
                                          replacement = ", \n",
                                          x = tissue_group))
                       ) + 
        geom_point(aes_(shape = as.factor(ifelse(grepl(pattern = "_NT", 
                                      x = rownames(subset_data), 
                                      ignore.case = F), "Tconv", "Treg"))), 
                   size = 4) + 
        geom_text_repel(aes_(label=
                               qdap::multigsub(pattern =
                                                 c("IntraEpithelialLayer",
                                                   "LaminaPropriaLymphocyte",
                                                   "PeyersPatches",
                                                   "LymphNode"),
                                               replacement = c("IEL", "LPL",
                                                               "PP", "LN"),
                                               order.pattern = F,
                                               text.var = gsub(pattern = "_.*",
                                                               replacement = "",
                                                               x = rownames(
                                                                 subset_data),
                                                               ignore.case = T)
                                                 )
                             ),
                        #size = 4, segment.alpha = 0.4, #for offline images
                        size = 5, segment.alpha = 0.4, #for webtool rds files
                        force = 1, direction = "both", show.legend = F) +
        scale_shape_manual(values = c(4, 20)) + 
        scale_colour_manual(values = tissue_group_col) + 
        labs(colour = "Tissue Groups", 
             shape = "Cell Types",
             title = "Tconv samples"
             ) + 
        #theme_light(base_size = 13.5) + #for offline images
        theme_light(base_size = 18) + #for webtool rds files
        theme(legend.key.size = unit(0.15, 'lines'), 
              plot.caption = element_text(hjust = 0, face= "italic"),
              panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank()
              ) + 
        scale_x_continuous(limits = c(min(samples_PCA_data$PC1), 
                                      max(samples_PCA_data$PC1))) + 
        scale_y_continuous(limits = c(min(samples_PCA_data$PC2), 
                                      max(samples_PCA_data$PC2))) + 
        guides(shape = F)
      
      #PCA_plot3
      
      # #for ggplotly in the webtool -- modify the weird brackets in the legend
      # myplot <- ggplotly(PCA_plot3, tooltip = c("Label", "Type", "PC1", "PC2")) %>%
      #   toWebGL()
      # 
      # for (i in 1:length(myplot$x$data)){
      #   if (!is.null(myplot$x$data[[i]]$name)){
      #     myplot$x$data[[i]]$name =  gsub("\\(|\\)","",str_split(myplot$x$data[[i]]$name,",")[[1]][2])
      #   }
      # }
      # 
      # myplot
      
    }
    
    #multiplot(PCA_plot1, PCA_plot2, PCA_plot3, cols = 2)
    
    PCA_plot1 + PCA_plot2 + PCA_plot3 + 
      plot_layout(ncol = 2, nrow = 2, byrow = T)
    gc()
    
  }
}

#performing t-SNE via the Rtsne library
{
  data_tsne <- data.matrix(log2(norm_counts))
  data_tsne[!is.finite(data_tsne)] <- NA
  data_tsne <- data_tsne[complete.cases(data_tsne),]
  # data_tsne <- data_tsne[,grepl(pattern = "_T_", 
  #                               x = colnames(data_tsne), 
  #                               ignore.case = T)]
  data_tsne <- t(data_tsne)
  
  #data_tsne <- t(norm_counts)
  tissue_group <- rownames(data_tsne)
  tissue_group <- gsub(pattern = "_(NT|T)_[1|2|3]$", 
                       replacement = "", 
                       x = tissue_group, 
                       ignore.case = T)
    
  for(i in seq_along(tissue_group)){
    if (tolower(tissue_group[i]) == "blood"){
      tissue_group[i] <- "Blood"
    }
    else if (tolower(tissue_group[i]) %in% c("spleen", "lymphnode", 
                                             "peyerspatches")){
      tissue_group[i] <- "Lymphoid tissue"
    }
    else if (tolower(tissue_group[i]) %in% 
        c("intraepitheliallayer", "laminaproprialymphocyte")){
      tissue_group[i] <- paste0("Gut-associated")
    }
    else if (tolower(tissue_group[i]) %in% 
        c("kidney", "lung", "liver", "pancreas")){
      tissue_group[i] <- "Non-lymphoid"
    }
    else if (tolower(tissue_group[i]) == "celltypet"){
      tissue_group[i] <- "Treg"
    }
    else if (tolower(tissue_group[i]) == "celltypent"){
      tissue_group[i] <- "Tconv"
    }
    
  }
  
  #data_tsne[!is.finite(data_tsne)] <- NA
  
  tsne_results <- Rtsne(X = data_tsne, 
                        dims = 2, 
                        initial_dims = 50, 
                        pca = T, 
                        partial_pca = F, 
                        pca_center = T, 
                        pca_scale = F, 
                        normalize = T, 
                        check_duplicates = F, 
                        verbose = T, 
                        perplexity = 3, 
                        theta = 0, 
                        max_iter = 2000, 
                        num_threads = 9)
  
  #master plot
  {
    # tSNE_plot <- ggplot(as.data.frame(tsne_results$Y), 
    #                    aes(x=tsne_results$Y[,1], 
    #                        y=tsne_results$Y[,2], 
    #                        #color=tissue_group)
    #                        color= gsub(pattern = ", ",
    #                                    replacement = ", \n",
    #                                    x = tissue_group))
    #                    ) + 
    #   geom_point(aes_(shape = as.factor(ifelse(grepl(pattern = "_NT", 
    #                                   x = rownames(data_tsne), 
    #                                   ignore.case = T), "CD4", "T-reg")))
    #                ) + 
    #   geom_text_repel(aes_(label= 
    #                          qdap::multigsub(pattern = 
    #                                            c("InterEpithelialLayer", 
    #                                              "LaminaPropriaLymphocyte", 
    #                                              "PeyersPatches", "LymphNode"), 
    #                                          replacement = c("IEL", "LPL", "PP", 
    #                                                          "LN"), 
    #                                        order.pattern = F, 
    #                                        text.var = gsub(pattern = "_.*", 
    #                                                        replacement = "", 
    #                                                        x = rownames(
    #                                                          data_tsne), 
    #                                                        ignore.case = T))
    #                        ), 
    #                   size = 4, segment.alpha = 0.4,
    #                   force = 1, direction = "both", show.legend = F) + 
    #   scale_shape_manual(values = c(4, 20)) + 
    #   labs(x = "Dimension 1", 
    #        y = "Dimension 2", 
    #        colour = "Tissue Groups", 
    #        shape = "Cell Types", 
    #        caption = paste0("IEL: Intraepithelial layer; LPL: ", 
    #                         "Lamina propria lymphocytes; PP: Peyer's patches; ",
    #                         "LN: Lymph node"),
    #        title = "t-SNE plot of 40 samples"
    #        ) + 
    #   theme_light() + 
    #   theme(legend.key.size = unit(2.8, 'lines'), 
    #         plot.caption = element_text(hjust = 0, face= "italic"),
    #         panel.grid.major = element_blank(), panel.grid.minor = element_blank()
    #         )
    # 
    # tSNE_plot
  }
  
  #only CD4 and Treg style
  {
    tSNE_plot1 <- ggplot(as.data.frame(tsne_results$Y), 
                       aes_(x=tsne_results$Y[,1], 
                           y=tsne_results$Y[,2])
                           ) + 
      geom_point(aes_(shape = as.factor(ifelse(grepl(pattern = "_NT", 
                                    x = rownames(data_tsne), 
                                    ignore.case = T), "Tconv", "Treg"))), 
                   size = 4) + 
      geom_text_repel(aes_(label= 
                             qdap::multigsub(pattern = 
                                               c("IntraEpithelialLayer", 
                                                 "LaminaPropriaLymphocyte", 
                                                 "PeyersPatches", "LymphNode"), 
                                             replacement = c("IEL", "LPL", "PP",
                                                             "LN"), 
                                             order.pattern = F, 
                                             text.var = gsub(pattern = "_.*", 
                                                             replacement = "", 
                                                             x = rownames(
                                                               data_tsne), 
                                                             ignore.case = T)
                                               )
                           ), 
                      #size = 3.25, segment.alpha = 0.4,
                      size = 5, segment.alpha = 0.4,
                      force = 1, direction = "both", show.legend = F) + 
      scale_shape_manual(values = c(4, 20)) + 
      labs(x = "Dimension 1", 
           y = "Dimension 2", 
           shape = "Cell Types", 
           title = paste0("All ", nrow(data_tsne), " samples")
           ) + 
      #theme_light(base_size = 13) + 
      theme_light(base_size = 18) + 
      theme(legend.key.size = unit(0.15, 'lines'), 
            plot.caption = element_text(hjust = 0, face= "italic"),
            panel.grid.major = element_blank(), panel.grid.minor = element_blank()
            )
    
    #tSNE_plot1
  }
  
  #only Treg plot
  {
    subset_data <- data_tsne[grepl(pattern = "_T", 
                                          x = rownames(data_tsne), 
                                          ignore.case = F),]
    
    #manually assigning class colours
      {
        tissue_group_col <- tissue_group
        names(tissue_group_col) <- tissue_group_col
        
        tissue_group_col[grepl(pattern = "Blood",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#D81B60"
        tissue_group_col[grepl(pattern = "(Spleen|PP|LN|^Lymphoid)",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#FFC107"
        tissue_group_col[grepl(pattern = "(IEL|LPL|Gut)",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#1E88E5"
        tissue_group_col[grepl(pattern = "(Kidney|Lung|Liver|Pancreas|Non)",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#004D40"
      }
    
    tSNE_plot2 <- ggplot(as.data.frame(subset_data), 
                        aes_(x=tsne_results$Y[which(
                          rownames(data_tsne) %in% rownames(subset_data)),1], 
                           y=tsne_results$Y[which(
                             rownames(data_tsne) %in% rownames(subset_data)),2], 
                            color= gsub(pattern = ", ",
                                        replacement = ", \n",
                                        x = tissue_group[which(
                                          rownames(data_tsne) %in% 
                                            rownames(subset_data))]))
                     ) + 
      geom_point(aes_(shape = as.factor(ifelse(grepl(pattern = "_NT", 
                                    x = rownames(subset_data), 
                                    ignore.case = F), "Tconv", "Treg"))),
                 size = 4) + 
      geom_text_repel(aes_(label= 
                             qdap::multigsub(pattern = 
                                               c("IntraEpithelialLayer", 
                                                 "LaminaPropriaLymphocyte", 
                                                 "PeyersPatches", "LymphNode"), 
                                             replacement = c("IEL", "LPL", "PP",
                                                             "LN"), 
                                             order.pattern = F, 
                                             text.var = gsub(pattern = "_.*", 
                                                             replacement = "", 
                                                             x = rownames(
                                                               subset_data), 
                                                             ignore.case = T)
                                               )
                           ), 
                      #size = 3.25, segment.alpha = 0.4,
                      size = 5, segment.alpha = 0.4,
                      force = 1, direction = "both", show.legend = F) + 
      scale_shape_manual(values = c(20, 4)) + 
      scale_colour_manual(values = tissue_group_col) + 
      labs(x = "Dimension 1", 
           y = "Dimension 2", 
           colour = "Tissue Groups",
           title = "Treg samples"
           ) + 
      #theme_light(base_size = 13) + 
      theme_light(base_size = 18) + 
      theme(legend.key.size = unit(0.15, 'lines'), 
            plot.caption = element_text(hjust = 0, face= "italic"),
            panel.grid.major = element_blank(), panel.grid.minor = element_blank()
            ) + 
      scale_x_continuous(limits = c(min(tsne_results$Y[,1]), 
                                    max(tsne_results$Y[,1]))) + 
      scale_y_continuous(limits = c(min(tsne_results$Y[,2]), 
                                    max(tsne_results$Y[,2]))) + 
      guides(shape = F)
    
    #tSNE_plot2
  }
  
  #only CD4 plot
  {
    subset_data <- data_tsne[grepl(pattern = "_NT", 
                                          x = rownames(data_tsne), 
                                          ignore.case = F),]
    
    #manually assigning class colours
      {
        tissue_group_col <- tissue_group
        names(tissue_group_col) <- tissue_group_col
        
        tissue_group_col[grepl(pattern = "Blood",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#D81B60"
        tissue_group_col[grepl(pattern = "(Spleen|PP|LN|^Lymphoid)",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#FFC107"
        tissue_group_col[grepl(pattern = "(IEL|LPL|Gut)",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#1E88E5"
        tissue_group_col[grepl(pattern = "(Kidney|Lung|Liver|Pancreas|Non)",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#004D40"
      }
    
    tSNE_plot3 <- ggplot(as.data.frame(subset_data), 
                        aes_(x=tsne_results$Y[which(
                          rownames(data_tsne) %in% rownames(subset_data)),1], 
                           y=tsne_results$Y[which(
                             rownames(data_tsne) %in% rownames(subset_data)),2], 
                            color= gsub(pattern = ", ",
                                        replacement = ", \n",
                                        x = tissue_group[which(
                                          rownames(data_tsne) %in% 
                                            rownames(subset_data))]))
                     ) + 
      geom_point(aes_(shape = as.factor(ifelse(grepl(pattern = "_NT", 
                                    x = rownames(subset_data), 
                                    ignore.case = F), "Tconv", "Treg"))), 
                   size = 4) + 
      geom_text_repel(aes_(label= 
                             qdap::multigsub(pattern = 
                                               c("IntraEpithelialLayer", 
                                                 "LaminaPropriaLymphocyte", 
                                                 "PeyersPatches", "LymphNode"), 
                                             replacement = c("IEL", "LPL", "PP",
                                                             "LN"), 
                                             order.pattern = F, 
                                             text.var = gsub(pattern = "_.*", 
                                                             replacement = "", 
                                                             x = rownames(
                                                               subset_data), 
                                                             ignore.case = T)
                                               )
                           ), 
                      #size = 3.25, segment.alpha = 0.4,
                      size = 5, segment.alpha = 0.4,
                      force = 1, direction = "both", show.legend = F) + 
      scale_shape_manual(values = c(4, 20)) + 
      scale_colour_manual(values = tissue_group_col) + 
      labs(x = "Dimension 1", 
           y = "Dimension 2", 
           colour = "Tissue Groups", 
           title = "Tconv samples"
           ) + 
      #theme_light(base_size = 13) + 
      theme_light(base_size = 18) + 
      theme(legend.key.size = unit(0.15, 'lines'), 
            plot.caption = element_text(hjust = 0, face= "italic"),
            panel.grid.major = element_blank(), panel.grid.minor = element_blank()
            ) + 
      scale_x_continuous(limits = c(min(tsne_results$Y[,1]), 
                                    max(tsne_results$Y[,1]))) + 
      scale_y_continuous(limits = c(min(tsne_results$Y[,2]), 
                                    max(tsne_results$Y[,2]))) + 
      guides(shape = F)
    
    #tSNE_plot3
  }
  
  #multiplot(PCA_plot1, PCA_plot2, PCA_plot3, cols = 2)
  
  tSNE_plot1 + tSNE_plot2 + tSNE_plot3 + 
    plot_layout(ncol = 2, nrow = 2, byrow = T)
  gc()
  
}

#performing UMAP via the umap library
{
  data_umap <- data.matrix(log2(norm_counts))
  data_umap[!is.finite(data_umap)] <- NA
  data_umap <- data_umap[complete.cases(data_umap),]
  data_umap <- t(data_umap)
  #data_umap <- t(norm_counts)
  
  tissue_group <- rownames(t(norm_counts))
  tissue_group <- gsub(pattern = "_(NT|T)_[1|2|3]$", 
                       replacement = "", 
                       x = tissue_group, 
                       ignore.case = T)
  
  for(i in seq_along(tissue_group)){
    
    if (tolower(tissue_group[i]) == "blood"){
      tissue_group[i] <- "Blood"
    }
    else if (tolower(tissue_group[i]) %in% c("spleen", "lymphnode", 
                                             "peyerspatches")){
      tissue_group[i] <- "Lymphoid tissue"
    }
    else if (tolower(tissue_group[i]) %in% 
        c("intraepitheliallayer", "laminaproprialymphocyte")){
      tissue_group[i] <- paste0("Gut-associated")
    }
    else if (tolower(tissue_group[i]) %in% 
        c("kidney", "lung", "liver", "pancreas")){
      tissue_group[i] <- "Non-lymphoid"
    }
    else if (tolower(tissue_group[i]) == "celltypet"){
      tissue_group[i] <- "Treg"
    }
    else if (tolower(tissue_group[i]) == "celltypent"){
      tissue_group[i] <- "Tconv"
    }
    
  }
  
  #changing umap default settings
  custom_config <- umap.defaults
  custom_config$n_neighbors <- 9
  custom_config$n_epochs <- 200
  custom_config$verbose <- TRUE
  
  # umap_results <- umap(d = t(norm_counts), method = "naive", 
  #                      config = custom_config)
  umap_results <- umap(d = data_umap, method = "naive", 
                       config = custom_config)
  
  #master plot
  {
    # umap_plot <- ggplot(as.data.frame(umap_results$layout), 
    #                    aes(x=umap_results$layout[,1], 
    #                        y=umap_results$layout[,2], 
    #                        color= gsub(pattern = ", ",
    #                                    replacement = ", \n",
    #                                    x = tissue_group))
    #                    ) + 
    #   geom_point(aes_(shape = as.factor(ifelse(grepl(pattern = "_NT", 
    #                                   x = rownames(umap_results$layout), 
    #                                   ignore.case = T), "CD4", "T-reg")))
    #                ) + 
    #   geom_text_repel(aes_(label= 
    #                          qdap::multigsub(pattern = 
    #                                            c("InterEpithelialLayer", 
    #                                              "LaminaPropriaLymphocyte", 
    #                                              "PeyersPatches", "LymphNode"), 
    #                                          replacement = c("IEL", "LPL", "PP", 
    #                                                          "LN"), 
    #                                          order.pattern = F, 
    #                                          text.var = gsub(pattern = "_.*", 
    #                                                          replacement = "", 
    #                                                          x = rownames(
    #                                                            umap_results$layout), 
    #                                                          ignore.case = T))
    #                        ), 
    #                   size = 4, segment.alpha = 0.4,
    #                   force = 1, direction = "both", show.legend = F) + 
    #   scale_shape_manual(values = c(4, 20)) + 
    #   labs(x = "Dimension 1", 
    #        y = "Dimension 2", 
    #        colour = "Tissue Groups", 
    #        shape = "Cell Types", 
    #        caption = paste0("IEL: Intraepithelial layer; LPL: ", 
    #                         "Lamina propria lymphocytes; PP: Peyer's patches; ",
    #                         "LN: Lymph node"),
    #        title = "UMAP plot of 40 samples"
    #        ) + 
    #   theme_light() + 
    #   theme(legend.key.size = unit(2.8, 'lines'), 
    #         plot.caption = element_text(hjust = 0, face= "italic"),
    #         panel.grid.major = element_blank(), panel.grid.minor = element_blank()
    #         )
    # 
    # umap_plot
  }
  
  
  #only CD4 and Treg style
  {
    umap_plot1 <- ggplot(as.data.frame(umap_results$layout), 
                       aes_(x=umap_results$layout[,1], 
                           y=umap_results$layout[,2])
                           ) + 
      geom_point(aes_(shape = as.factor(ifelse(grepl(pattern = "_NT", 
                                    x = rownames(umap_results$layout), 
                                    ignore.case = T), "Tconv", "Treg"))), 
                   size = 4) + 
      geom_text_repel(aes_(label= 
                             qdap::multigsub(pattern = 
                                               c("IntraEpithelialLayer", 
                                                 "LaminaPropriaLymphocyte", 
                                                 "PeyersPatches", "LymphNode"), 
                                             replacement = c("IEL", "LPL", "PP",
                                                             "LN"), 
                                             order.pattern = F, 
                                             text.var = gsub(pattern = "_.*", 
                                                             replacement = "", 
                                                             x = rownames(
                                                               umap_results$layout), 
                                                             ignore.case = T)
                                               )
                           ), 
                      #size = 3.25, segment.alpha = 0.4,
                      size = 5, segment.alpha = 0.4,
                      force = 1, direction = "both", show.legend = F) + 
      scale_shape_manual(values = c(4, 20)) + 
      labs(x = "Dimension 1", 
           y = "Dimension 2", 
           shape = "Cell Types", 
           title = paste0("All ", nrow(data_umap), " samples")
           ) + 
      #theme_light(base_size = 13) + 
      theme_light(base_size = 18) + 
      theme(legend.key.size = unit(0.15, 'lines'), 
            plot.caption = element_text(hjust = 0, face= "italic"),
            panel.grid.major = element_blank(), panel.grid.minor = element_blank()
            )
    
    #umap_plot1
  }
  
  #only T-reg plot
  {
    subset_data <- data_umap[grepl(pattern = "_T", 
                                          x = rownames(data_umap), 
                                          ignore.case = F),]
    
    #manually assigning class colours
      {
        tissue_group_col <- tissue_group
        names(tissue_group_col) <- tissue_group_col
        
        tissue_group_col[grepl(pattern = "Blood",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#D81B60"
        tissue_group_col[grepl(pattern = "(Spleen|PP|LN|^Lymphoid)",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#FFC107"
        tissue_group_col[grepl(pattern = "(IEL|LPL|Gut)",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#1E88E5"
        tissue_group_col[grepl(pattern = "(Kidney|Lung|Liver|Pancreas|Non)",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#004D40"
      }
    
    umap_plot2 <- ggplot(as.data.frame(subset_data), 
                        aes_(x=umap_results$layout[which(
                          rownames(data_umap) %in% rownames(subset_data)),1], 
                           y=umap_results$layout[which(
                             rownames(data_umap) %in% rownames(subset_data)),2], 
                            color= gsub(pattern = ", ",
                                        replacement = ", \n",
                                        x = tissue_group[which(
                                          rownames(data_umap) %in% 
                                            rownames(subset_data))]))
                     ) + 
      geom_point(aes_(shape = as.factor(ifelse(grepl(pattern = "_NT", 
                                    x = rownames(subset_data), 
                                    ignore.case = F), "Tconv", "Treg"))), 
                   size = 4) + 
      geom_text_repel(aes_(label= 
                             qdap::multigsub(pattern = 
                                               c("IntraEpithelialLayer", 
                                                 "LaminaPropriaLymphocyte", 
                                                 "PeyersPatches", "LymphNode"), 
                                             replacement = c("IEL", "LPL", "PP",
                                                             "LN"), 
                                             order.pattern = F, 
                                             text.var = gsub(pattern = "_.*", 
                                                             replacement = "", 
                                                             x = rownames(
                                                               subset_data), 
                                                             ignore.case = T)
                                               )
                           ), 
                      #size = 3.25, segment.alpha = 0.4,
                      size = 5, segment.alpha = 0.4,
                      force = 1, direction = "both", show.legend = F) + 
      scale_shape_manual(values = c(20, 4)) + 
      scale_colour_manual(values = tissue_group_col) +
      labs(x = "Dimension 1", 
           y = "Dimension 2", 
           colour = "Tissue Groups",
           title = "Treg samples"
           ) + 
      #theme_light(base_size = 13) + 
      theme_light(base_size = 18) + 
      theme(legend.key.size = unit(0.15, 'lines'), 
            plot.caption = element_text(hjust = 0, face= "italic"),
            panel.grid.major = element_blank(), panel.grid.minor = element_blank()
            ) + 
      scale_x_continuous(limits = c(min(umap_results$layout[,1]), 
                                    max(umap_results$layout[,1]))) + 
      scale_y_continuous(limits = c(min(umap_results$layout[,2]), 
                                    max(umap_results$layout[,2]))) + 
      guides(shape = F)
    
    #umap_plot2
  }
  
  #only CD4 plot
  {
    subset_data <- data_umap[grepl(pattern = "_NT", 
                                          x = rownames(data_umap), 
                                          ignore.case = F),]
    
    #manually assigning class colours
      {
        tissue_group_col <- tissue_group
        names(tissue_group_col) <- tissue_group_col
        
        tissue_group_col[grepl(pattern = "Blood",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#D81B60"
        tissue_group_col[grepl(pattern = "(Spleen|PP|LN|^Lymphoid)",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#FFC107"
        tissue_group_col[grepl(pattern = "(IEL|LPL|Gut)",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#1E88E5"
        tissue_group_col[grepl(pattern = "(Kidney|Lung|Liver|Pancreas|Non)",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#004D40"
      }
    
    umap_plot3 <- ggplot(as.data.frame(subset_data), 
                        aes_(x=umap_results$layout[which(
                          rownames(data_umap) %in% rownames(subset_data)),1], 
                           y=umap_results$layout[which(
                             rownames(data_umap) %in% rownames(subset_data)),2], 
                            color= gsub(pattern = ", ",
                                        replacement = ", \n",
                                        x = tissue_group[which(
                                          rownames(data_umap) %in% 
                                            rownames(subset_data))]))
                     ) + 
      geom_point(aes_(shape = as.factor(ifelse(grepl(pattern = "_NT", 
                                    x = rownames(subset_data), 
                                    ignore.case = F), "Tconv", "Treg"))), 
                   size = 4) + 
      geom_text_repel(aes_(label= 
                             qdap::multigsub(pattern = 
                                               c("IntraEpithelialLayer", 
                                                 "LaminaPropriaLymphocyte", 
                                                 "PeyersPatches", "LymphNode"), 
                                             replacement = c("IEL", "LPL", "PP",
                                                             "LN"), 
                                             order.pattern = F, 
                                             text.var = gsub(pattern = "_.*", 
                                                             replacement = "", 
                                                             x = rownames(
                                                               subset_data), 
                                                             ignore.case = T)
                                               )
                           ), 
                      #size = 3.25, segment.alpha = 0.4,
                      size = 5, segment.alpha = 0.4,
                      force = 1, direction = "both", show.legend = F) + 
      scale_shape_manual(values = c(4, 20)) + 
      scale_colour_manual(values = tissue_group_col) + 
      labs(x = "Dimension 1", 
           y = "Dimension 2", 
           colour = "Tissue Groups", 
           title = "Tconv samples"
           ) + 
      #theme_light(base_size = 13) + 
      theme_light(base_size = 18) + 
      theme(legend.key.size = unit(0.15, 'lines'), 
            plot.caption = element_text(hjust = 0, face= "italic"),
            panel.grid.major = element_blank(), panel.grid.minor = element_blank()
            ) + 
      scale_x_continuous(limits = c(min(umap_results$layout[,1]), 
                                    max(umap_results$layout[,1]))) + 
      scale_y_continuous(limits = c(min(umap_results$layout[,2]), 
                                    max(umap_results$layout[,2]))) + 
      guides(shape = F)
    
    #umap_plot3
  }
  
  #multiplot(PCA_plot1, PCA_plot2, PCA_plot3, cols = 2)
  
  umap_plot1 + umap_plot2 + umap_plot3 + 
    plot_layout(ncol = 2, nrow = 2, byrow = T)
  gc()
  
}


#perform t-SNE on mean log2 transformed data
{
  #reading in DESeq2 outputs
  {
    #read normalised counts from the '#writing out the differential expression 
    # files' code block to get them filtered by min transcripts per cell
    #reading norm counts
    norm_counts <- read.delim(file = paste0(loc_counts, "normalised_counts.txt"),
                              sep = "\t",
                              stringsAsFactors = F)
    #norm_counts <- norm_counts[rowSums(as.matrix(norm_counts))>0,]
    
    #reading the DESeq2 analysis output table
    merged_results_list <- readRDS(file = paste0(loc_diffexprs, 
                                          "merged_results_list.rds"))
    
    contrast_groups <- list(c("Blood"), 
                            c("Spleen", "Lymph Node", "Peyer's Patches"), 
                            c("Intraepithelial Layer", 
                              "Lamina Propria Lymphocyte"), 
                            c("Kidney", "Lung", "Liver", "Pancreas"),
                            c("cellTypeT"),
                            c("cellTypeNT"))
    
    #ensuring that names in contrast_groups list are all lower letter and do not 
    # contain spaces or punctuation
    for(i in seq(1, length(contrast_groups)))
    {
      contrast_groups[[i]] <- tolower(unique(
        gsub("_.*|[[:space:]]|[[:punct:]]", "", contrast_groups[[i]])
        ))
    }
    rm(i)
    names(contrast_groups) <- c("Blood", 
                                "Lymphoid", 
                                "Gut-Associated", 
                                "Non-Lymphoid", 
                                "Treg", 
                                "Tconv")
    
    # # output_table <- read.delim(file = paste0(loc_diffexprs_merged,
    # #                                          "output_table_combined.txt"),
    # #                            sep = "\t", stringsAsFactors = F)
    # output_table <- read.delim(file = paste0(loc_diffexprs_merged,
    #                                          "output_table.txt"),
    #                            sep = "\t", stringsAsFactors = F)
    # 
    # colnames(output_table) <- output_table[3,]
    # output_table <- output_table[-(1:3),]
    # 
    # mean_norm_counts <- output_table[, c(1, 44:63)]
    # rownames(mean_norm_counts) <- mean_norm_counts$`Gene stable ID`
    # mean_norm_counts$`Gene stable ID` <- NULL
    # # rownames(mean_norm_counts) <- gsub(pattern = "\\.[0-9]+$",
    # #                                    replacement = "",
    # #                                    x = rownames(mean_norm_counts),
    # #                                    ignore.case = T)
    # #mean_norm_counts$`Gene stable ID` <- output_table$`Gene stable ID`
    # #colnames(temp) <- temp[3,]
    # #temp <- temp[4:nrow(temp),]
    
  }
  
  data_tsne <- t(mean_norm_counts)
  
  tissue_group <- rownames(data_tsne)
  tissue_group <- gsub(pattern = " (Treg|Tconv)$", 
                       replacement = "", 
                       x = tissue_group, 
                       ignore.case = T)
    
  for(i in seq_along(tissue_group)){
    if (tolower(tissue_group[i]) == "blood"){
      tissue_group[i] <- "Blood"
    }
    else if (tolower(tissue_group[i]) %in% c("spleen", "ln", "pp")){
      tissue_group[i] <- "Lymphoid tissue"
    }
    else if (tolower(tissue_group[i]) %in% c("iel", "lpl")){
      tissue_group[i] <- paste0("Gut-associated\ntissues")
    }
    else if (tolower(tissue_group[i]) %in% 
        c("kidney", "lung", "liver", "pancreas")){
      tissue_group[i] <- "Non-lymphoid\ntissues"
    }
    
  }
  
  data_tsne <- as.data.frame(data_tsne)
  
  tsne_results <- Rtsne(X = data_tsne, 
                        dims = 2, 
                        initial_dims = 50, 
                        pca = T, 
                        partial_pca = F, 
                        pca_center = T, 
                        pca_scale = F, 
                        normalize = T, 
                        check_duplicates = F, 
                        verbose = T, 
                        perplexity = 3, 
                        theta = 0, 
                        max_iter = 2000, 
                        num_threads = 9)
  
  #only CD4 and Treg style
  {
    # tSNE_plot1 <- ggplot(as.data.frame(tsne_results$Y), 
    #                    aes_(x=tsne_results$Y[,1], 
    #                        y=tsne_results$Y[,2])
    #                        ) + 
    #   geom_point(aes_(shape = as.factor(ifelse(grepl(pattern = "_NT", 
    #                                 x = rownames(data_tsne), 
    #                                 ignore.case = T), "Tconv", "Treg")))
    #              ) + 
    #   geom_text_repel(aes_(label= 
    #                          qdap::multigsub(pattern = 
    #                                            c("IntraEpithelialLayer", 
    #                                              "LaminaPropriaLymphocyte", 
    #                                              "PeyersPatches", "LymphNode"), 
    #                                          replacement = c("IEL", "LPL", "PP",
    #                                                          "LN"), 
    #                                          order.pattern = F, 
    #                                          text.var = gsub(pattern = "_.*", 
    #                                                          replacement = "", 
    #                                                          x = rownames(
    #                                                            data_tsne), 
    #                                                          ignore.case = T)
    #                                            )
    #                        ), 
    #                   #size = 3.25, segment.alpha = 0.4,
    #                   size = 5, segment.alpha = 0.4,
    #                   force = 1, direction = "both", show.legend = F) + 
    #   scale_shape_manual(values = c(4, 20)) + 
    #   labs(x = "Dimension 1", 
    #        y = "Dimension 2", 
    #        shape = "Cell Types", 
    #        title = paste0("All ", nrow(data_tsne), " samples")
    #        ) + 
    #   #theme_light(base_size = 13) + 
    #   theme_light(base_size = 18) + 
    #   theme(legend.key.size = unit(0.15, 'lines'), 
    #         plot.caption = element_text(hjust = 0, face= "italic"),
    #         panel.grid.major = element_blank(), panel.grid.minor = element_blank()
    #         )
    # 
    # #tSNE_plot1
  }
  
  #only Treg plot
  {
    subset_data <- data_tsne[grepl(pattern = " Treg", 
                                          x = rownames(data_tsne), 
                                          ignore.case = F),]
    tSNE_plot2 <- ggplot(as.data.frame(subset_data), 
                        aes_(x=tsne_results$Y[which(
                          rownames(data_tsne) %in% rownames(subset_data)),1], 
                           y=tsne_results$Y[which(
                             rownames(data_tsne) %in% rownames(subset_data)),2], 
                            color= gsub(pattern = ", ",
                                        replacement = ", \n",
                                        x = tissue_group[which(
                                          rownames(data_tsne) %in% 
                                            rownames(subset_data))]))
                     ) + 
      geom_point(aes_(shape = as.factor(ifelse(grepl(pattern = "_NT", 
                                    x = rownames(subset_data), 
                                    ignore.case = F), "Tconv", "Treg"))),
                 size = 4) + 
      geom_text_repel(aes_(label= 
                             qdap::multigsub(pattern = 
                                               c("IntraEpithelialLayer", 
                                                 "LaminaPropriaLymphocyte", 
                                                 "PeyersPatches", "LymphNode"), 
                                             replacement = c("IEL", "LPL", "PP",
                                                             "LN"), 
                                             order.pattern = F, 
                                             text.var = gsub(pattern = "_.*", 
                                                             replacement = "", 
                                                             x = rownames(
                                                               subset_data), 
                                                             ignore.case = T)
                                               )
                           ), 
                      #size = 3.25, segment.alpha = 0.4,
                      size = 7, segment.alpha = 0.4,
                      force = 1, direction = "both", show.legend = F) + 
      scale_shape_manual(values = c(20, 4)) + 
      labs(x = "Dimension 1", 
           y = "Dimension 2", 
           colour = "Tissue Groups",
           title = "Treg samples"
           ) + 
      #theme_light(base_size = 13) + 
      theme_light(base_size = 18) + 
      theme(legend.key.size = unit(0.15, 'lines'), 
            plot.caption = element_text(hjust = 0, face= "italic"),
            panel.grid.major = element_blank(), panel.grid.minor = element_blank()
            ) + 
      scale_x_continuous(limits = c(min(tsne_results$Y[,1]), 
                                    max(tsne_results$Y[,1]))) + 
      scale_y_continuous(limits = c(min(tsne_results$Y[,2]), 
                                    max(tsne_results$Y[,2]))) + 
      guides(shape = F)
    
    #tSNE_plot2
  }
  
  #only CD4 plot
  {
    # subset_data <- data_tsne[grepl(pattern = "_NT", 
    #                                       x = rownames(data_tsne), 
    #                                       ignore.case = F),]
    # tSNE_plot3 <- ggplot(as.data.frame(subset_data), 
    #                     aes_(x=tsne_results$Y[which(
    #                       rownames(data_tsne) %in% rownames(subset_data)),1], 
    #                        y=tsne_results$Y[which(
    #                          rownames(data_tsne) %in% rownames(subset_data)),2], 
    #                         color= gsub(pattern = ", ",
    #                                     replacement = ", \n",
    #                                     x = tissue_group[which(
    #                                       rownames(data_tsne) %in% 
    #                                         rownames(subset_data))]))
    #                  ) + 
    #   geom_point(aes_(shape = as.factor(ifelse(grepl(pattern = "_NT", 
    #                                 x = rownames(subset_data), 
    #                                 ignore.case = F), "Tconv", "Treg")))
    #              ) + 
    #   geom_text_repel(aes_(label= 
    #                          qdap::multigsub(pattern = 
    #                                            c("IntraEpithelialLayer", 
    #                                              "LaminaPropriaLymphocyte", 
    #                                              "PeyersPatches", "LymphNode"), 
    #                                          replacement = c("IEL", "LPL", "PP",
    #                                                          "LN"), 
    #                                          order.pattern = F, 
    #                                          text.var = gsub(pattern = "_.*", 
    #                                                          replacement = "", 
    #                                                          x = rownames(
    #                                                            subset_data), 
    #                                                          ignore.case = T)
    #                                            )
    #                        ), 
    #                   #size = 3.25, segment.alpha = 0.4,
    #                   size = 5, segment.alpha = 0.4,
    #                   force = 1, direction = "both", show.legend = F) + 
    #   scale_shape_manual(values = c(4, 20)) + 
    #   labs(x = "Dimension 1", 
    #        y = "Dimension 2", 
    #        colour = "Tissue Groups", 
    #        title = "Tconv samples"
    #        ) + 
    #   #theme_light(base_size = 13) + 
    #   theme_light(base_size = 18) + 
    #   theme(legend.key.size = unit(0.15, 'lines'), 
    #         plot.caption = element_text(hjust = 0, face= "italic"),
    #         panel.grid.major = element_blank(), panel.grid.minor = element_blank()
    #         ) + 
    #   scale_x_continuous(limits = c(min(tsne_results$Y[,1]), 
    #                                 max(tsne_results$Y[,1]))) + 
    #   scale_y_continuous(limits = c(min(tsne_results$Y[,2]), 
    #                                 max(tsne_results$Y[,2]))) + 
    #   guides(shape = F)
    # 
    # #tSNE_plot3
  }
  
  #multiplot(PCA_plot1, PCA_plot2, PCA_plot3, cols = 2)
  
  # tSNE_plot1 + tSNE_plot2 + tSNE_plot3 + 
  #   plot_layout(ncol = 2, nrow = 2, byrow = T)
  # gc()
  
}


#performing PCA after removing ENSMUSG00000097971.3 and/or ENSMUSG00000046008.8,
#ENSMUSG00000058579.5, ENSMUSG00000026818.5 and ENSMUSG00000023031.8
{
  # samples_PCA_removeGenes <-
  #   prcomp(t(norm_counts[(rowSums(norm_counts)>0 &
  #                           !grepl(pattern = "ENSMUSG00000097971\\.3|ENSMUSG00000029580\\.14|ENSMUSG00000046008\\.8|ENSMUSG00000058579\\.5|ENSMUSG00000026818\\.5|ENSMUSG00000023031\\.8|ENSMUSG00000106106\\.2|ENSMUSG00000035202\\.8|ENSMUSG00000052837\\.6|ENSMUSG00000019850\\.11|ENSMUSG00000057163\\.3|ENSMUSG00000011463\\.5|ENSMUSG00000054446\\.8|ENSMUSG00000031957\\.6|ENSMUSG00000098973\\.1|ENSMUSG00000064339\\.1|ENSMUSG00000076258\\.1|ENSMUSG00000049775\\.16|ENSMUSG00000023034\\.7|ENSMUSG00000024190\\.7|ENSMUSG00000026011\\.13",
  #                                  x = rownames(norm_counts),
  #                                  ignore.case = T)),]
  #            ))
  # 
  # # samples_PCA_removeGenes <- 
  # #   prcomp(t(norm_counts[(rowSums(norm_counts)>0 & 
  # #                           !grepl(pattern = "ENSMUSG00000097971\\.3", 
  # #                                  x = rownames(norm_counts), 
  # #                                  ignore.case = T)),
  # #                        !grepl(pattern = "pancreas", x = colnames(norm_counts), ignore.case = T)]
  # #            ))
  # 
  # #summary(samples_PCA)
  # 
  # #plotting variance of PCs
  # variance_proportion <- ((samples_PCA_removeGenes$sdev^2) / (sum(samples_PCA_removeGenes$sdev^2)))*100
  # dev.new()
  # barplot(variance_proportion, cex.names=1,
  #         xlab=paste("Principal component (PC), 1-",
  #                    length(samples_PCA_removeGenes$sdev)), 
  #         ylab="Proportion of variation (%)",
  #         main="Scree plot", ylim=c(0,100))
  # #dev.off()
  # 
  # ##PCA Plots
  # {
  #   #plot(samples_PCA$x[,1], samples_PCA$x[,2]) --- individual colours
  #   samples_PCA_data <- as.data.frame(samples_PCA_removeGenes$x)
  #   samples_PCA_data$group <- gsub("_[0-9]{1}$", "", colnames(norm_counts))
  #   # samples_PCA_data$group <- gsub("_[0-9]{1}$", "", colnames(norm_counts))[
  #   #   !grepl("pancreas", gsub("_[0-9]{1}$", "", colnames(norm_counts)), 
  #   #          ignore.case = T)]
  #   
  #   #PCA loadings
  #   PCA_loadings <- samples_PCA_removeGenes$rotation
  #   
  #   #PCA coordinates?
  #   PCA_coordinates <- samples_PCA_removeGenes$x
  #   
  #   # PCA_plot <- ggplot(samples_PCA_data, aes(x=PC1, y=PC2, color=group))
  #   # PCA_plot <- PCA_plot + geom_point() + 
  #   #   geom_text(aes(label=rownames(samples_PCA_data)),
  #   #             size=2.5, hjust=0.5, vjust=1.5)
  #   # dev.new()
  #   # PCA_plot
  #   #dev.off()
  #   
  #   contrast_groups <- list(c("Blood"), 
  #                         c("Spleen", "Lymph Node"), 
  #                         c("Intraepithelial Layer", 
  #                           "Lamina Propria Lymphocyte", "Peyer's Patches"), 
  #                         c("Kidney", "Lung", "Liver", "Pancreas"),
  #                         c("cellTypeT"),
  #                         c("cellTypeNT"))
  #   
  #   for(i in seq(1, length(contrast_groups))){
  #     contrast_groups[[i]] <- tolower(unique(
  #       gsub("_.*|[[:space:]]|[[:punct:]]", "", contrast_groups[[i]])
  #       ))
  #   }
  #   
  #   #plot PCA -- data points grouped by tissue groups
  #   tissue_group <- samples_PCA_data$group
  #   indexing <- data.frame(tissue = unlist(contrast_groups), 
  #                              index = rep(seq_along(contrast_groups), 
  #                                          lapply(contrast_groups, length)))
  #   
  #   for(i in seq_along(tissue_group)){
  #     
  #     # if(!length(indexing$index[which(indexing$tissue %in% tolower(
  #     #   gsub(pattern = "_(NT|T)$", 
  #     #        replacement = "", 
  #     #        x = tissue_group[i])))])>0) next
  #     
  #     tissue_group[i] <- indexing$index[which(indexing$tissue %in% tolower(
  #       gsub(pattern = "_(NT|T)$", 
  #            replacement = "", 
  #            x = tissue_group[i])))]
  #   }
  #   
  #   samples_PCA_data$tissue_group <- as.factor(tissue_group)
  #   
  #   PCA_plot <- ggplot(samples_PCA_data, 
  #                      aes(x=PC1, y=PC2, color=tissue_group)
  #                      ) + 
  #     geom_point() + 
  #     geom_text(aes_(label=rownames(samples_PCA_data)),
  #               size=3, hjust=0.5, vjust=1.5, show.legend = F,
  #               position=position_jitter(width=2,height=2)) + 
  #     #geom_jitter(width = 2, height = 2) + 
  #     labs(x = "Dimension 1", y = "Dimension 2", colour = "Tissue Groups") + 
  #     theme_light()
  #   
  #   dev.new()
  #   PCA_plot
  #   
  #   
  #   ##PCA Plots using centroid of tissue_T_# samples
  #   {
  #     #plot(samples_PCA$x[,1], samples_PCA$x[,2]) --- individual colours
  #     
  #     T_centroid_PCA_data <- as.data.frame(matrix(data = 0, 
  #                                                 nrow = 0, 
  #                                                 ncol = ncol(samples_PCA_removeGenes$x)))
  #     colnames(T_centroid_PCA_data) <- colnames(samples_PCA_removeGenes$x)
  #     i <- 1
  #     while(i < nrow(samples_PCA_removeGenes$x)) {
  #       
  #       if((i-1)%%4 == 0){
  #         #print("In simple copy")
  #         T_centroid_PCA_data[nrow(T_centroid_PCA_data)+1,] <- 
  #           (samples_PCA_removeGenes$x[i,])
  #         
  #         rownames(T_centroid_PCA_data)[nrow(T_centroid_PCA_data)] <- 
  #           gsub(pattern = "_[0-9]{1}$", 
  #                replacement = "", 
  #                x = rownames(samples_PCA_removeGenes$x)[i])
  #         
  #         i <- i+1
  #       }
  #       else{
  #         #print("In rowMeans")
  #         T_centroid_PCA_data[nrow(T_centroid_PCA_data)+1,] <- 
  #           rowMeans(t(samples_PCA_removeGenes$x[i:(i+2),]))
  #         
  #         rownames(T_centroid_PCA_data)[nrow(T_centroid_PCA_data)] <- 
  #           gsub(pattern = "_[0-9]{1}$", 
  #                replacement = "", 
  #                x = rownames(samples_PCA_removeGenes$x)[i])
  #         
  #         i <- i+3
  #       }
  #       
  #     }
  #     rm(i)
  #     
  #     T_centroid_PCA_data$group <- unique(gsub(pattern = "_[0-9]{1}$", 
  #                                              replacement = "", 
  #                                              x = rownames(samples_PCA_removeGenes$x)))
  #     
  #     NT_coordinates <- T_centroid_PCA_data[grepl(
  #       pattern = "_NT$", x = rownames(T_centroid_PCA_data)),]
  #     T_coordinates <- T_centroid_PCA_data[grepl(
  #       pattern = "_T$", x = rownames(T_centroid_PCA_data)),]
  #     
  #     PCA_plot <- ggplot(T_centroid_PCA_data, aes(x=PC1, y=PC2, color=group))
  #     PCA_plot <- PCA_plot + geom_point() + 
  #       geom_text(aes(label=rownames(T_centroid_PCA_data)),
  #                 size=3, hjust=0.5, vjust=1.5, show.legend = F) + 
  #       geom_segment(data = NT_coordinates, aes(x = NT_coordinates[,1], y = NT_coordinates[,2], 
  #                        xend = T_coordinates[,1], yend = T_coordinates[,2])) + 
  #       theme_light()
  #     
  #     dev.new()
  #     PCA_plot
  #     #dev.off()
  #   }
  #   
  #   
  # }
}

#performing classical multi-dimensional scaling (MDS)
{
  #classical (metric) MDS
  {
    #calculating variances
    variances <- cbind(norm_counts, rowVars(norm_counts))
    variances <- variances[order(variances[,ncol(variances)], decreasing = T),]
    
    #filtering out zero variance genes
    variances <- variances[variances[,ncol(variances)]>0,]
    
    #calculating euclidiean distances
    distances <- dist(x = t(
      variances[1:(nrow(variances)),-(ncol(variances))]), 
      method = "euclidean")
    
    ### R session creashes at the below command after maing out 32GB RAM and 32GB 
    #   Swap
    #k is the target number of dimensions
    fit <- cmdscale(d = distances, eig = T, k = 2)
    
    #below needs to be modified
    MDS_plot <- ggplot(data = as.data.frame(fit$points), 
                       aes(x=fit$points[,1], y=fit$points[,2], 
                           color=rownames(fit$points)))
    
    MDS_plot <- MDS_plot + geom_point() +
      geom_text(aes(label=rownames(fit$points)),
                size=3.5, hjust=0.5, vjust=1.5)
    
    MDS_plot
    
  }
  
}

#performing t-SNE on PCA via the Rtsne library without Pancreas_T_3, Liver_T_3 
#and PeyersPatches_T_1
{
  # data_tsne <- t(norm_counts)
  # data_tsne <- data_tsne[!grepl(
  #   pattern = "Pancreas_T_3|Liver_T_3|PeyersPatches_T_1", 
  #   x = rownames(data_tsne), 
  #   ignore.case = T),]
  # 
  # tissue_group <- rownames(data_tsne)
  # tissue_group <- gsub(pattern = "_(NT|T)_[1|2|3]$", 
  #                      replacement = "", 
  #                      x = tissue_group, 
  #                      ignore.case = T)
  # 
  # for(i in seq_along(tissue_group)){
  #   if (tolower(tissue_group[i]) == "blood"){
  #     tissue_group[i] <- "Blood"
  #   }
  #   else if (tolower(tissue_group[i]) %in% c("spleen", "lymphnode")){
  #     tissue_group[i] <- "Lymphoid tissue"
  #   }
  #   else if (tolower(tissue_group[i]) %in% 
  #       c("intraepitheliallayer", "laminaproprialymphocyte", "peyerspatches")){
  #     tissue_group[i] <- paste0("Gut-associated\ntissues")
  #   }
  #   else if (tolower(tissue_group[i]) %in% 
  #       c("kidney", "lung", "liver", "pancreas")){
  #     tissue_group[i] <- "Non-lymphoid\ntissues"
  #   }
  #   else if (tolower(tissue_group[i]) == "celltypet"){
  #     tissue_group[i] <- "Regulatory T-cells"
  #   }
  #   else if (tolower(tissue_group[i]) == "celltypent"){
  #     tissue_group[i] <- "CD4 T-cells"
  #   }
  #   
  # }
  # 
  # tsne_results <- Rtsne(X = data_tsne, 
  #                       dims = 2, 
  #                       initial_dims = 50, 
  #                       pca = T, 
  #                       partial_pca = F, 
  #                       pca_center = T, 
  #                       pca_scale = F, 
  #                       normalize = T, 
  #                       check_duplicates = F, 
  #                       verbose = T, 
  #                       perplexity = 3, 
  #                       theta = 0, 
  #                       max_iter = 2000, 
  #                       num_threads = 9)
  # 
  # #master plot
  # {
  #   tSNE_plot <- ggplot(as.data.frame(tsne_results$Y), 
  #                      aes(x=tsne_results$Y[,1], 
  #                          y=tsne_results$Y[,2], 
  #                          color= gsub(pattern = ", ",
  #                                      replacement = ", \n",
  #                                      x = tissue_group))
  #                      ) + 
  #     geom_point(aes_(shape = as.factor(ifelse(grepl(pattern = "NT", 
  #                                     x = rownames(data_tsne), 
  #                                     ignore.case = T), "CD4", "T-reg")))
  #                  ) + 
  #     geom_text_repel(aes_(label= 
  #                            qdap::multigsub(pattern = 
  #                                              c("InterEpithelialLayer", 
  #                                                "LaminaPropriaLymphocyte", 
  #                                                "PeyersPatches", "LymphNode"), 
  #                                            replacement = c("IEL", "LPL", "PP", 
  #                                                            "LN"), 
  #                                            order.pattern = F, 
  #                                            text.var = gsub(pattern = "_.*", 
  #                                                            replacement = "", 
  #                                                            x = rownames(
  #                                                              data_tsne), 
  #                                                            ignore.case = T))
  #                          ), 
  #                     size = 4, segment.alpha = 0.4,
  #                     force = 1, direction = "both", show.legend = F) + 
  #     scale_shape_manual(values = c(4, 20)) + 
  #     labs(x = "Dimension 1", 
  #          y = "Dimension 2", 
  #          colour = "Tissue Groups", 
  #          shape = "Cell Types", 
  #          caption = paste0("IEL: Intraepithelial layer; LPL: ", 
  #                           "Lamina propria lymphocytes; PP: Peyer's patches; ",
  #                           "LN: Lymph node"),
  #          title = "t-SNE plot of 37 samples"
  #          ) + 
  #     theme_light() + 
  #     theme(legend.key.size = unit(2.8, 'lines'), 
  #           plot.caption = element_text(hjust = 0, face= "italic"),
  #           panel.grid.major = element_blank(), panel.grid.minor = element_blank()
  #           )
  #   
  #   tSNE_plot
  # }
  # 
  # 
  # #only CD4 and Treg style
  # {
  #   tSNE_plot1 <- ggplot(as.data.frame(tsne_results$Y), 
  #                      aes(x=tsne_results$Y[,1], 
  #                          y=tsne_results$Y[,2])
  #                          ) + 
  #     geom_point(aes_(shape = as.factor(ifelse(grepl(pattern = "_NT", 
  #                                   x = rownames(data_tsne), 
  #                                   ignore.case = T), "CD4", "T-reg")))
  #                ) + 
  #     geom_text_repel(aes_(label= 
  #                            qdap::multigsub(pattern = 
  #                                              c("InterEpithelialLayer", 
  #                                                "LaminaPropriaLymphocyte", 
  #                                                "PeyersPatches", "LymphNode"), 
  #                                            replacement = c("IEL", "LPL", "PP",
  #                                                            "LN"), 
  #                                            order.pattern = F, 
  #                                            text.var = gsub(pattern = "_.*", 
  #                                                            replacement = "", 
  #                                                            x = rownames(
  #                                                              data_tsne), 
  #                                                            ignore.case = T)
  #                                              )
  #                          ), 
  #                     size = 4, segment.alpha = 0.4,
  #                     force = 1, direction = "both", show.legend = F) + 
  #     scale_shape_manual(values = c(4, 20)) + 
  #     labs(x = "Dimension 1", 
  #          y = "Dimension 2", 
  #          shape = "Cell Types", 
  #          title = "t-SNE plot of 37 samples"
  #          ) + 
  #     theme_light() + 
  #     theme(legend.key.size = unit(1, 'lines'), 
  #           plot.caption = element_text(hjust = 0, face= "italic"),
  #           panel.grid.major = element_blank(), panel.grid.minor = element_blank()
  #           )
  #   
  #   tSNE_plot1
  # }
  # 
  # #only T-reg plot
  # {
  #   subset_data <- data_tsne[grepl(pattern = "_T", 
  #                                         x = rownames(data_tsne), 
  #                                         ignore.case = F),]
  #   tSNE_plot2 <- ggplot(as.data.frame(subset_data), 
  #                       aes_(x=tsne_results$Y[which(
  #                         rownames(data_tsne) %in% rownames(subset_data)),1], 
  #                          y=tsne_results$Y[which(
  #                            rownames(data_tsne) %in% rownames(subset_data)),2], 
  #                           color= gsub(pattern = ", ",
  #                                       replacement = ", \n",
  #                                       x = tissue_group[which(
  #                                         rownames(data_tsne) %in% 
  #                                           rownames(subset_data))]))
  #                    ) + 
  #     geom_point(aes_(shape = as.factor(ifelse(grepl(pattern = "_NT", 
  #                                   x = rownames(subset_data), 
  #                                   ignore.case = F), "CD4", "T-reg")))
  #                ) + 
  #     geom_text_repel(aes_(label= 
  #                            qdap::multigsub(pattern = 
  #                                              c("InterEpithelialLayer", 
  #                                                "LaminaPropriaLymphocyte", 
  #                                                "PeyersPatches", "LymphNode"), 
  #                                            replacement = c("IEL", "LPL", "PP",
  #                                                            "LN"), 
  #                                            order.pattern = F, 
  #                                            text.var = gsub(pattern = "_.*", 
  #                                                            replacement = "", 
  #                                                            x = rownames(
  #                                                              subset_data), 
  #                                                            ignore.case = T)
  #                                              )
  #                          ), 
  #                     size = 4, segment.alpha = 0.4,
  #                     force = 1, direction = "both", show.legend = F) + 
  #     scale_shape_manual(values = c(20, 4)) + 
  #     labs(x = "Dimension 1", 
  #          y = "Dimension 2", 
  #          colour = "Tissue Groups",
  #          title = "t-SNE plot of T-reg samples"
  #          ) + 
  #     theme_light() + 
  #     theme(legend.key.size = unit(1.5, 'lines'), 
  #           plot.caption = element_text(hjust = 0, face= "italic"),
  #           panel.grid.major = element_blank(), panel.grid.minor = element_blank()
  #           ) + 
  #     scale_x_continuous(limits = c(min(tsne_results$Y[,1]), 
  #                                   max(tsne_results$Y[,1]))) + 
  #     scale_y_continuous(limits = c(min(tsne_results$Y[,2]), 
  #                                   max(tsne_results$Y[,2]))) + 
  #     guides(shape = F)
  #   
  #   tSNE_plot2
  # }
  # 
  # #only CD4 plot
  # {
  #   subset_data <- data_tsne[grepl(pattern = "_NT", 
  #                                         x = rownames(data_tsne), 
  #                                         ignore.case = F),]
  #   tSNE_plot3 <- ggplot(as.data.frame(subset_data), 
  #                       aes_(x=tsne_results$Y[which(
  #                         rownames(data_tsne) %in% rownames(subset_data)),1], 
  #                          y=tsne_results$Y[which(
  #                            rownames(data_tsne) %in% rownames(subset_data)),2], 
  #                           color= gsub(pattern = ", ",
  #                                       replacement = ", \n",
  #                                       x = tissue_group[which(
  #                                         rownames(data_tsne) %in% 
  #                                           rownames(subset_data))]))
  #                    ) + 
  #     geom_point(aes_(shape = as.factor(ifelse(grepl(pattern = "_NT", 
  #                                   x = rownames(subset_data), 
  #                                   ignore.case = F), "CD4", "T-reg")))
  #                ) + 
  #     geom_text_repel(aes_(label= 
  #                            qdap::multigsub(pattern = 
  #                                              c("InterEpithelialLayer", 
  #                                                "LaminaPropriaLymphocyte", 
  #                                                "PeyersPatches", "LymphNode"), 
  #                                            replacement = c("IEL", "LPL", "PP",
  #                                                            "LN"), 
  #                                            order.pattern = F, 
  #                                            text.var = gsub(pattern = "_.*", 
  #                                                            replacement = "", 
  #                                                            x = rownames(
  #                                                              subset_data), 
  #                                                            ignore.case = T)
  #                                              )
  #                          ), 
  #                     size = 4, segment.alpha = 0.4,
  #                     force = 1, direction = "both", show.legend = F) + 
  #     scale_shape_manual(values = c(4, 20)) + 
  #     labs(x = "Dimension 1", 
  #          y = "Dimension 2", 
  #          colour = "Tissue Groups", 
  #          title = "t-SNE plot of CD4 samples"
  #          ) + 
  #     theme_light() + 
  #     theme(legend.key.size = unit(1.5, 'lines'), 
  #           plot.caption = element_text(hjust = 0, face= "italic"),
  #           panel.grid.major = element_blank(), panel.grid.minor = element_blank()
  #           ) + 
  #     scale_x_continuous(limits = c(min(tsne_results$Y[,1]), 
  #                                   max(tsne_results$Y[,1]))) + 
  #     scale_y_continuous(limits = c(min(tsne_results$Y[,2]), 
  #                                   max(tsne_results$Y[,2]))) + 
  #     guides(shape = F)
  #   
  #   tSNE_plot3
  # }
  # 
  # #multiplot(PCA_plot1, PCA_plot2, PCA_plot3, cols = 2)
  # 
  # tSNE_plot1 + tSNE_plot2 + tSNE_plot3 + 
  #   plot_layout(ncol = 2, nrow = 2, byrow = T)
  # 
}

#performing UMAP via the umap library without Pancreas_T_3, Liver_T_3 and 
#PeyersPatches_T_1
{
  # data_umap <- t(norm_counts)
  # 
  # data_umap <- data_umap[!grepl(
  #   pattern = "Pancreas_T_3|Liver_T_3|PeyersPatches_T_1", 
  #   x = rownames(data_umap), 
  #   ignore.case = T),]
  # 
  # tissue_group <- rownames(data_umap)
  # tissue_group <- gsub(pattern = "_(NT|T)_[1|2|3]$", 
  #                      replacement = "", 
  #                      x = tissue_group, 
  #                      ignore.case = T)
  # 
  # for(i in seq_along(tissue_group)){
  #   if (tolower(tissue_group[i]) == "blood"){
  #     tissue_group[i] <- "Blood"
  #   }
  #   else if (tolower(tissue_group[i]) %in% c("spleen", "lymphnode")){
  #     tissue_group[i] <- "Lymphoid tissue"
  #   }
  #   else if (tolower(tissue_group[i]) %in% 
  #       c("intraepitheliallayer", "laminaproprialymphocyte", "peyerspatches")){
  #     tissue_group[i] <- paste0("Gut-associated\ntissues")
  #   }
  #   else if (tolower(tissue_group[i]) %in% 
  #       c("kidney", "lung", "liver", "pancreas")){
  #     tissue_group[i] <- "Non-lymphoid\ntissues"
  #   }
  #   else if (tolower(tissue_group[i]) == "celltypet"){
  #     tissue_group[i] <- "Regulatory T-cells"
  #   }
  #   else if (tolower(tissue_group[i]) == "celltypent"){
  #     tissue_group[i] <- "CD4 T-cells"
  #   }
  #   
  # }
  # 
  # #changing umap default settings
  # custom_config <- umap.defaults
  # custom_config$n_neighbors <- 9
  # custom_config$n_epochs <- 200
  # custom_config$verbose <- TRUE
  # 
  # umap_results <- umap(d = data_umap, method = "naive", 
  #                      config = custom_config)
  # 
  # #master plot
  # {
  #   umap_plot <- ggplot(as.data.frame(umap_results$layout), 
  #                      aes(x=umap_results$layout[,1], 
  #                          y=umap_results$layout[,2], 
  #                          color= gsub(pattern = ", ",
  #                                      replacement = ", \n",
  #                                      x = tissue_group))
  #                      ) + 
  #     geom_point(aes_(shape = as.factor(ifelse(grepl(pattern = "NT", 
  #                                     x = rownames(data_umap), 
  #                                     ignore.case = T), "CD4", "T-reg")))
  #                  ) + 
  #     geom_text_repel(aes_(label= 
  #                            qdap::multigsub(pattern = 
  #                                              c("InterEpithelialLayer", 
  #                                                "LaminaPropriaLymphocyte", 
  #                                                "PeyersPatches", "LymphNode"), 
  #                                            replacement = c("IEL", "LPL", "PP", 
  #                                                            "LN"), 
  #                                            order.pattern = F, 
  #                                            text.var = gsub(pattern = "_.*", 
  #                                                            replacement = "", 
  #                                                            x = rownames(
  #                                                              data_umap), 
  #                                                            ignore.case = T))
  #                          ), 
  #                     size = 4, segment.alpha = 0.4,
  #                     force = 1, direction = "both", show.legend = F) + 
  #     scale_shape_manual(values = c(4, 16)) + 
  #     labs(x = "Dimension 1", 
  #          y = "Dimension 2", 
  #          colour = "Tissue Groups", 
  #          shape = "Cell Types", 
  #          caption = paste0("IEL: Intraepithelial layer; LPL: ", 
  #                           "Lamina propria lymphocytes; PP: Peyer's patches; ",
  #                           "LN: Lymph node"),
  #          title = "UMAP plot of 37 samples"
  #          ) + 
  #     theme_light() + 
  #     theme(legend.key.size = unit(2.8, 'lines'), 
  #           plot.caption = element_text(hjust = 0, face= "italic"),
  #           panel.grid.major = element_blank(), panel.grid.minor = element_blank()
  #           )
  #   
  #   umap_plot
  # }
  # 
  # 
  # #only CD4 and Treg style
  # {
  #   umap_plot1 <- ggplot(as.data.frame(umap_results$layout), 
  #                      aes_(x=umap_results$layout[,1], 
  #                          y=umap_results$layout[,2])
  #                          ) + 
  #     geom_point(aes_(shape = as.factor(ifelse(grepl(pattern = "_NT", 
  #                                   x = rownames(umap_results$layout), 
  #                                   ignore.case = T), "CD4", "T-reg")))
  #                ) + 
  #     geom_text_repel(aes_(label= 
  #                            qdap::multigsub(pattern = 
  #                                              c("InterEpithelialLayer", 
  #                                                "LaminaPropriaLymphocyte", 
  #                                                "PeyersPatches", "LymphNode"), 
  #                                            replacement = c("IEL", "LPL", "PP",
  #                                                            "LN"), 
  #                                            order.pattern = F, 
  #                                            text.var = gsub(pattern = "_.*", 
  #                                                            replacement = "", 
  #                                                            x = rownames(
  #                                                              umap_results$layout), 
  #                                                            ignore.case = T)
  #                                              )
  #                          ), 
  #                     size = 4, segment.alpha = 0.4,
  #                     force = 1, direction = "both", show.legend = F) + 
  #     scale_shape_manual(values = c(4, 20)) + 
  #     labs(x = "Dimension 1", 
  #          y = "Dimension 2", 
  #          shape = "Cell Types", 
  #          title = "UMAP plot of 37 samples"
  #          ) + 
  #     theme_light() + 
  #     theme(legend.key.size = unit(1, 'lines'), 
  #           plot.caption = element_text(hjust = 0, face= "italic"),
  #           panel.grid.major = element_blank(), panel.grid.minor = element_blank()
  #           )
  #   
  #   umap_plot1
  # }
  # 
  # #only T-reg plot
  # {
  #   subset_data <- data_umap[grepl(pattern = "_T", 
  #                                         x = rownames(data_umap), 
  #                                         ignore.case = F),]
  #   umap_plot2 <- ggplot(as.data.frame(subset_data), 
  #                       aes_(x=umap_results$layout[which(
  #                         rownames(data_umap) %in% rownames(subset_data)),1], 
  #                          y=umap_results$layout[which(
  #                            rownames(data_umap) %in% rownames(subset_data)),2], 
  #                           color= gsub(pattern = ", ",
  #                                       replacement = ", \n",
  #                                       x = tissue_group[which(
  #                                         rownames(data_umap) %in% 
  #                                           rownames(subset_data))]))
  #                    ) + 
  #     geom_point(aes_(shape = as.factor(ifelse(grepl(pattern = "_NT", 
  #                                   x = rownames(subset_data), 
  #                                   ignore.case = F), "CD4", "T-reg")))
  #                ) + 
  #     geom_text_repel(aes_(label= 
  #                            qdap::multigsub(pattern = 
  #                                              c("InterEpithelialLayer", 
  #                                                "LaminaPropriaLymphocyte", 
  #                                                "PeyersPatches", "LymphNode"), 
  #                                            replacement = c("IEL", "LPL", "PP",
  #                                                            "LN"), 
  #                                            order.pattern = F, 
  #                                            text.var = gsub(pattern = "_.*", 
  #                                                            replacement = "", 
  #                                                            x = rownames(
  #                                                              subset_data), 
  #                                                            ignore.case = T)
  #                                              )
  #                          ), 
  #                     size = 4, segment.alpha = 0.4,
  #                     force = 1, direction = "both", show.legend = F) + 
  #     scale_shape_manual(values = c(20, 4)) + 
  #     labs(x = "Dimension 1", 
  #          y = "Dimension 2", 
  #          colour = "Tissue Groups",
  #          title = "UMAP plot of T-reg samples"
  #          ) + 
  #     theme_light() + 
  #     theme(legend.key.size = unit(1.5, 'lines'), 
  #           plot.caption = element_text(hjust = 0, face= "italic"),
  #           panel.grid.major = element_blank(), panel.grid.minor = element_blank()
  #           ) + 
  #     scale_x_continuous(limits = c(min(umap_results$layout[,1]), 
  #                                   max(umap_results$layout[,1]))) + 
  #     scale_y_continuous(limits = c(min(umap_results$layout[,2]), 
  #                                   max(umap_results$layout[,2]))) + 
  #     guides(shape = F)
  #   
  #   umap_plot2
  # }
  # 
  # #only CD4 plot
  # {
  #   subset_data <- data_umap[grepl(pattern = "_NT", 
  #                                         x = rownames(data_umap), 
  #                                         ignore.case = F),]
  #   umap_plot3 <- ggplot(as.data.frame(subset_data), 
  #                       aes_(x=umap_results$layout[which(
  #                         rownames(data_umap) %in% rownames(subset_data)),1], 
  #                          y=umap_results$layout[which(
  #                            rownames(data_umap) %in% rownames(subset_data)),2], 
  #                           color= gsub(pattern = ", ",
  #                                       replacement = ", \n",
  #                                       x = tissue_group[which(
  #                                         rownames(data_umap) %in% 
  #                                           rownames(subset_data))]))
  #                    ) + 
  #     geom_point(aes_(shape = as.factor(ifelse(grepl(pattern = "_NT", 
  #                                   x = rownames(subset_data), 
  #                                   ignore.case = F), "CD4", "T-reg")))
  #                ) + 
  #     geom_text_repel(aes_(label= 
  #                            qdap::multigsub(pattern = 
  #                                              c("InterEpithelialLayer", 
  #                                                "LaminaPropriaLymphocyte", 
  #                                                "PeyersPatches", "LymphNode"), 
  #                                            replacement = c("IEL", "LPL", "PP",
  #                                                            "LN"), 
  #                                            order.pattern = F, 
  #                                            text.var = gsub(pattern = "_.*", 
  #                                                            replacement = "", 
  #                                                            x = rownames(
  #                                                              subset_data), 
  #                                                            ignore.case = T)
  #                                              )
  #                          ), 
  #                     size = 4, segment.alpha = 0.4,
  #                     force = 1, direction = "both", show.legend = F) + 
  #     scale_shape_manual(values = c(4, 20)) + 
  #     labs(x = "Dimension 1", 
  #          y = "Dimension 2", 
  #          colour = "Tissue Groups", 
  #          title = "UMAP plot of CD4 samples"
  #          ) + 
  #     theme_light() + 
  #     theme(legend.key.size = unit(1.5, 'lines'), 
  #           plot.caption = element_text(hjust = 0, face= "italic"),
  #           panel.grid.major = element_blank(), panel.grid.minor = element_blank()
  #           ) + 
  #     scale_x_continuous(limits = c(min(umap_results$layout[,1]), 
  #                                   max(umap_results$layout[,1]))) + 
  #     scale_y_continuous(limits = c(min(umap_results$layout[,2]), 
  #                                   max(umap_results$layout[,2]))) + 
  #     guides(shape = F)
  #   
  #   umap_plot3
  # }
  # 
  # #multiplot(PCA_plot1, PCA_plot2, PCA_plot3, cols = 2)
  # 
  # umap_plot1 + umap_plot2 + umap_plot3 + 
  #   plot_layout(ncol = 2, nrow = 2, byrow = T)
  
}


```


#saving the volcano plots
```{r}

#reading in DESeq2 outputs
{
  #read normalised counts from the '#writing out the differential expression 
  # files' code block to get them filtered by min transcripts per cell
  #reading norm counts
  # norm_counts <- read.delim(file = paste0(loc_counts, "normalised_counts.txt"), 
  #                           sep = "\t", 
  #                           stringsAsFactors = F)
  #norm_counts <- norm_counts[rowSums(as.matrix(norm_counts))>0,]
  
  #reading the DESeq2 analysis output table
  merged_results_list <- readRDS(file = paste0(loc_diffexprs, 
                                        "merged_results_list.rds"))
  
  # contrast_groups <- list(c("Blood"), 
  #                         c("Spleen", "Lymph Node", "Peyer's Patches"), 
  #                         c("Intraepithelial Layer", 
  #                           "Lamina Propria Lymphocyte"), 
  #                         c("Kidney", "Lung", "Liver", "Pancreas"),
  #                         c("cellTypeT"),
  #                         c("cellTypeNT"))
  # 
  # #ensuring that names in contrast_groups list are all lower letter and do not 
  # # contain spaces or punctuation
  # for(i in seq(1, length(contrast_groups)))
  # {
  #   contrast_groups[[i]] <- tolower(unique(
  #     gsub("_.*|[[:space:]]|[[:punct:]]", "", contrast_groups[[i]])
  #     ))
  # }
  # rm(i)
  
  output_table <- read.delim(file = paste0(loc_diffexprs_merged, 
                                           "output_table.txt"), 
                             sep = "\t", stringsAsFactors = F)
  
  colnames(output_table) <- output_table[3,]
  output_table <- output_table[-(1:3),]
  
}


#creating list of genes having >x transcripts per cell 
#(as per Oliver, each sample has about 2000 cells; norm_count/2000 > 100)
## commented here and moved to the '#writing out the differential expression 
## files' code block
{
  # 
  # #filtered_tcounts <- as.data.frame((norm_counts/2000)>0.1)
  # #filtered_tcounts <- filtered_tcounts[rowSums(filtered_tcounts) > 0L,]
  # 
  # #creating a vector of rows to keep
  # selection <- logical(length = nrow(norm_counts))
  # #creating groups to check the transcript per cell in
  # groups <- unique(gsub(pattern = "_[0-9]+$", 
  #                       replacement = "", 
  #                       x = colnames(norm_counts), 
  #                       ignore.case = T))
  # groups <- groups[grepl(pattern = "_T", x = groups, ignore.case = T)]
  # #looping through each gene and checking if it has at least 0.1 transcript per
  # # cell in the tissue_cellTypeT groups
  # for(i in seq(1, length(selection))){
  #   
  #   for(j in groups){
  #     
  #     if(rowSums((norm_counts[i, grepl(pattern = j, 
  #                                      x = colnames(norm_counts), 
  #                                      ignore.case = T)]/2000)>1) == 
  #        sum(grepl(pattern = j, x = colnames(norm_counts), ignore.case = T))){
  #       selection[i] <- TRUE
  #       break
  #     }
  #   }
  # }
  # 
  # filtered_tcounts <- norm_counts[selection,]
  # 
  # #results
  # # filter    # of genes
  # # >100      18
  # # >50       35
  # # >10       336
  # # >5        805
  # # >1        4558
  # 
}

#additional genes of interest that need to be highlighted
{
  # # genes_of_int <- c("CD44", "CD62L", "CD69", "CD103", "ICOS", "KLRG1", "NK1.1", 
  # #                   "NRP1", "ST2", "PD1", "NK1")
  # # genes_of_int <- c("CD44", "SELL", "CD69", "ITGAE", "ICOS", "KLRG1", "KLRB1c", 
  # #                   "NRP1", "IL1RL1", "PDCD1", "NK1")
  # genes_of_int <- c("Ccr7", "Ccr8", "Ccr9", "CD44", "CD69", "Cxcr3", "Itgae", 
  #                   "Itgav", "Itgal", "Itgb1", "Itgb7", "Itgb8", "S1pr1", "Sell", 
  #                   "Ahr", "Areg", "Batf", "Bcl2", "Bcl6", "Ctla4", "Ebi3", "Fas", 
  #                   "Gata3", "Gpr43", "Gpr83", "Gzmb", "Havcr2", "Hif1a", "Icos", 
  #                   "Ikzf2", "Ikzf3", "Ikzf4", "Il10", "il1rl1", "il2ra", "Irf4", 
  #                   "Izumor1", "Klrg1", "Lag3", "Lgals1", "Ly6c1", "Maf", "Nfil3", 
  #                   "Nr4a1", "Nrp1", "Pdcd1", "Pparg", "Ptprj", "Relb", "Rora", 
  #                   "Rorc", "S100a4", "S1p1r", "Samsn1", "Socs1", "Socs2", 
  #                   "Tbx21", "Tigit", "Tnfrsf4", "Tnfrsf8", "Tnfrsf9", 
  #                   "ICOS", "KLRB1c", "NK1")
  # genes_of_int <- paste0("\\b(", paste(genes_of_int, collapse = "|"), ")\\b")
  # genes_of_int <- gsub(pattern = "\\.", replacement = "\\\\.", x = genes_of_int)
  
  #list of important genes given by oliver
  olivers_list <- read.delim(file = paste0(loc_diffexprs_merged, 
                                           "oliver_list.tab"), 
                             header = T, 
                             sep = "\t", 
                             stringsAsFactors = F)
  oliver_list_pattern <- paste0("\\b(", 
                                paste(olivers_list$Symbol, 
                                      collapse = "|"), 
                                ")\\b")
  oliver_list_pattern <- gsub(pattern = "\\.", 
                              replacement = "\\\\.", 
                              x = oliver_list_pattern)
  
  #manually constructing the list of genes of interest because the plots are 
  # really small in the paper so I can only label a handful of genes
  genes_of_int <- list(
    Treg = list(
      blood__Vs__spleen_lymphnode_peyerspatches = 
        c("Dexras1", "Clambda1", "Tubal3", "Ccr10", "Klf2", "Trpc1", "Nrgn", 
          "Trpc1", "Penk", "Il1r2", "Egr3", "Ccr8", "Tnfsf11", "Lag3", "Cesc2c", 
          "Ier5l", "Smpdl3b", "Gjb2", "Slc7a10", "Cc2d2a", "C3", "Il21", 
          "C1qc", "Cst7"), 
      blood__Vs__intraepitheliallayer_laminaproprialymphocyte = 
        c("Aldh1a2", "Pla2g7", "Scd1", "Mcub", "Ly6c1", "Ms4a4c", 
          "1110032F04Rik", "AB124611", "Itgb1", "Ms4a4b", "Atp1b3", 
          "Sell", "Ms4a1", "Ccr10", "Klf2", 
          "4930503L19Rik", "Il10", "Asb2", "Gzmb", "Hic1", "Ccr9", "Cd160", 
          "Itgae", "Ier5l", "Lilr4b", "Gzma", "Cables1", "Dapk2", 
          "Aldob", "Cldn7", "Rbm24", "Cps1"), 
      blood__Vs__kidney_lung_liver_pancreas = 
        c("Ighg2b", "Trgc4", "Gm128", "Gm45110", "Xkrx", "Cr2", "Ccr8", "Pdcd1", 
          "Bhlhe40", "Ier5l", "Areg", "Trf", "Atf3", "Perp", "Sirpa", "Acod1", 
          "Nr4a1", "Penk", "Nr4a2", "Cc2d2a"), 
      spleen_lymphnode_peyerspatches__Vs__intraepitheliallayer_laminaproprialymphocyte = 
        c("Aldh1a2", "Pla2g7", "Mcub", "Ms4a4c", "AB124611", "Ms4a4b", "Itgb1", 
          "Ramp1", "Gm42809", "Abca1", "Tktl1", "Wnt10a", "Sell", "Ccr9", 
          "4930503L19Rik", "Ppp1r10", "Gzmb", "Acot11", "Hic1", "Il10", "Gzma", 
          "Olfm1", "Rasd1", "Ppp1r1b", "Hnf4a", "Clca3b", "Ppp1r3d", "Lypd8", 
          "Dmbt1"), 
      spleen_lymphnode_peyerspatches__Vs__kidney_lung_liver_pancreas = 
        c("Ighg2b", "Ltbp3", "Anxa3", "Sema6b", "Slc7a10", "Trgc4", "Egr2", 
          "Olfm1", "Lmna", "Myadm", "Il1rl1", "Trp53inp2", "Fosb", "Trpc1", 
          "Dusp8", "Rasd1", "Esm1", "Mmp10", "Cyp4b1", "Ppp1r1b", "Ppp1r3d"), 
      intraepitheliallayer_laminaproprialymphocyte__Vs__kidney_lung_liver_pancreas = 
        c("Clca3b", "Ighg2b", "Rasgrf1", "Ahrr", "Gzma", "Asb2", "Gpr55", 
          "Itgae", "Abi3", "Gzmb", "Cd160", "4930503L19Rik", "Ccr9", "Itgb1", 
          "Ass1", "Atp1b3", "AB124611", "Mcub", "Ramp1", "Anza1", "Lyz2", 
          "Tktl1", "Pla2g7", "Aldh1a2", "Scd1")
      ), 
    Tconv = list(
      blood__Vs__spleen_lymphnode_peyerspatches = 
        c("Cdk20", "Ints6l", "Ppp1r21", "Bbc3", "Tnfrsf9", "Tnfsf11", 
          "Marcksl1", "Penk", "Egr2", "Egr1", "Bcl2a1d", "Pdcd1", "Cd160", 
          "Il2ra", "Irf4", "Egr3", "Asb2", "Nr4a1", "Fgl2", "Cxcr5"), 
      blood__Vs__intraepitheliallayer_laminaproprialymphocyte = 
        c("Gucy1a1", "1110032F04Rik", "Gm9889", "Klhl35", "Mical3", "Afp", 
          "Klf2", "Hdac10", "Cd27", "Ass1", "Atp1b3", "Itgae", "Cd160", 
          "Entpd1", "Cd38", "Frmd4b", "Gzmb", "Fgl2", "Asb2", "Bcl2a1d", 
          "B3galt5", "Trgv2", "Il2rb"), 
      blood__Vs__kidney_lung_liver_pancreas = 
        c("Tubal3", "Reep1", "Fxyd4", "Alas2", "Kdm5d", "A90007A09Rik", "Ksr2", 
          "Gucy1b1", "Kifc2", "5330438D12Rik", "Apbb3", "Il2rb", "Il2ra", 
          "Entpd1", "Fgl2", "Socs2", "Bhlhe40", "Ifng", "Mmp9", "Bcl2a1d", 
          "Ltb4r1", "Cxcl10", "Klrb1c"), 
      spleen_lymphnode_peyerspatches__Vs__intraepitheliallayer_laminaproprialymphocyte = 
        c("C3", "1110032F04Rik", "Gm32633", "Gm42809", "Afp", "AB124611", 
          "Cd27", "Itgae", "Ahr", "Entpd1", "Socs2", "Cpd", "Gzmb", "Smim3", 
          "Trpc1", "Olfm1", "Gab2", "Lypd8", "Cldn7", "Lilr4b", "Lacm2", 
          "Cables1"), 
      spleen_lymphnode_peyerspatches__Vs__kidney_lung_liver_pancreas = 
        c("C3", "Sptb", "Rnf24", "Iglc1", "5330438D12Rik", "H2-Q2", "Pak1", 
          "Fer1l5", "Socs2", "Il2rb", "Entpd1", "Ctsw", "Fgl2", "Il1rl1", 
          "Klrg1", "Mmp9", "Ltb4r1", "Pparg", "Olfm1", "Gab2", "Cercam", 
          "Col1a1"), 
      intraepitheliallayer_laminaproprialymphocyte__Vs__kidney_lung_liver_pancreas = 
        c("Ppp1r1b", "Lypd8", "Sema6b", "Nectin2", "Rnf24", "Pheta2", "Gzmb", 
          "Gm17764", "Abi3", "4930503L19Rik", "Itgae", "Ass1", "Itgb1", 
          "Osbpl5", "Sema4a", "Ctla2a", "Mcub", "Hrh4", "Hba-a1")
      )
    )
  
}


## the plan was to create a list of plots and save them as called but for some
## reason, the developers of R and ggplot thought to make something as simple as
## that neigh impossible, with little open documentation as to why, thereby 
## forcing people like me to waste hours upon hours of pain and suffering 
## while digging through tons of unrelated stuff to find out that this is by
## some hellish design. As such, the code below is now writing plots as created
## (non-rant reason: there is no direct or easy-to-mediocre way to save device
## information in the ggplot2 object, so when a plot is called from the list the
## aesthetics information is lost, leading to oversized plot objects and text 
## with no way to control it via ggsave or any other means)


#before running this, make sure the actual log2FC have been calculated via the 
# '#writing out the differential expression files' code block
#creating a list of volcano plots via ggplot2 scatter plots
{
  saved_plots <- list()
  
  for(i in seq(1, length(merged_results_list))){
    #print(paste0("i: ", i))
    
    saved_plots[[length(saved_plots)+1]] <- list()
    names(saved_plots)[length(saved_plots)] <- names(merged_results_list)[i]
    
    writeLines(paste0("\nProcessing results from the model: ", 
                      names(merged_results_list)[i]))
    
    for(j in seq(1, length(merged_results_list[[i]]))){
      #print(paste0("j: ", j))
      
      saved_plots[[i]][[length(saved_plots[[i]])+1]] <- list()
      names(saved_plots[[i]])[length(saved_plots[[i]])] <- 
        names(merged_results_list[[i]])[j]
      
      #print(names(merged_results_list[[i]])[j])
      
      #constructing the path and creating directories
      path <- paste0(loc_plots, 
                     "Volcano Plots/", 
                     names(merged_results_list)[i], "/", 
                     names(merged_results_list[[i]])[j], "/")
      dir.create(path = path, recursive = T)
      
      writeLines(paste0("Exporting plots for: ", 
                        names(merged_results_list[[i]])[j]))
      
      for(k in seq(1, length(merged_results_list[[i]][[j]]))){
        #print(paste0("k: ", k))
        
        print(paste0("Exporting plot: ", 
                     names(merged_results_list[[i]][[j]])[k]))
        
        # plot_data <- as.data.frame(
        #   merged_results_list[[i]][[j]][[k]])[
        #     complete.cases(merged_results_list[[i]][[j]][[k]]),]
        plot_data <- as.data.frame(merged_results_list[[i]][[j]][[k]])
        
        # plot_data <- as.data.frame(
        #   merged_results_list[[i]][[j]][[k]])[
        #     is.finite(merged_results_list[[i]][[j]][[k]]$Actuallog2FC),]
        
        #subsetting only data with calculated actual log2 fold change
        #plot_data <- subset(plot_data, is.finite(plot_data$Actuallog2FC))
        #retaining infinite actual log2 fold change to represent as column sub-
        # plots using facets_grid
        #plot_data <- subset(plot_data, !is.na(plot_data$Actuallog2FC))
        plot_data$Facet <- "Both"
        plot_data$Facet[!is.finite(plot_data$Actuallog2FC) & 
                          plot_data$log2FoldChange > 0] <- "Right"
        plot_data$Facet[!is.finite(plot_data$Actuallog2FC) & 
                        plot_data$log2FoldChange < 0] <- "Left"
        plot_data$Facet <- factor(x = plot_data$Facet, 
                                  levels = c("Left", "Both", "Right"))
        # plot_data$log2FoldChange[plot_data$Facet == "Left"] <- 0
        # plot_data$log2FoldChange[plot_data$Facet == "Right"] <- 0
        
        
        #filter the genes if filtered_tcounts exists
        if(exists("filtered_tcounts")){
          plot_data <- subset(plot_data, rownames(plot_data) %in% 
                                rownames(filtered_tcounts))
        }
        
        # #filtering actual log fold change excluded genes if object exists
        # if(exists(("output_table_excluded"))){
        #   plot_data <- subset(plot_data, 
        #                       !(gsub(pattern = "\\.[0-9]+$", 
        #                              replacement = "", 
        #                              x = rownames(plot_data), 
        #                              ignore.case = T) %in% 
        #                           output_table_excluded[
        #                             5:nrow(output_table_excluded),1]
        #                         )
        #                       )
        # }
        
        # #filtering actual log fold change excluded genes if object exists
        # if(exists(("output_table"))){
        #   plot_data <- subset(plot_data, 
        #                       !(gsub(pattern = "\\.[0-9]+$", 
        #                              replacement = "", 
        #                              x = rownames(plot_data), 
        #                              ignore.case = T) %in% 
        #                           output_table[
        #                             5:nrow(output_table),1]
        #                         )
        #                       )
        # }
        
        # plot_data$Gene.stable.ID <- gsub(pattern = "\\.[0-9]+$", 
        #                                  replacement = "", 
        #                                  x = rownames(plot_data))
        plot_data$`Gene stable ID` <- gsub(pattern = "\\.[0-9]+$", 
                                         replacement = "", 
                                         x = rownames(plot_data))
        
        # plot_data <- merge(x = plot_data, y = output_table[,1:2], 
        #                    by = "Gene.stable.ID", 
        #                    all.x = T, all.y = F, 
        #                    sort = F)
        plot_data <- merge(x = plot_data, y = output_table[,1:2], 
                           by = "Gene stable ID", 
                           all.x = T, all.y = F, 
                           sort = F)
        
        plot_data <- plot_data[order(plot_data$log2FoldChange, decreasing = T),]
        
        #subsetting plot_data using oliver's list and plotting DESeq and Actual log2FC
        {
          # plot_data_oliver <- plot_data[grepl(pattern = oliver_list_pattern, 
          #                                     x = plot_data$MGI.symbol, 
          #                                     ignore.case = T),]
          # png(filename = paste0("DESeq2_v_Actual_", 
          #                       names(merged_results_list[[i]][[j]])[k], 
          #                       ifelse(j == 1, "_Treg", "_Tconv"),
          #                       ".png"))
          # # plot(x = seq(1, nrow(plot_data_oliver)), 
          # #      y = plot_data_oliver$log2FoldChange, 
          # #      type = "b", 
          # #      col = "blue", 
          # #      # main = paste0(ifelse(j == 1, "Treg: ", "Tconv: "), 
          # #      #               names(merged_results_list[[i]][[j]])[k])
          # #      main = paste0(ifelse(j == 1, "Treg:\n", "Tconv:\n"), 
          # #                    paste0(unlist(str_split(
          # #                      string = names(merged_results_list[[i]][[j]])[k], 
          # #                      pattern = "__Vs__")), 
          # #                    collapse = "\nVs\n"))
          # #      )
          # # lines(x = seq(1, nrow(plot_data_oliver)), 
          # #       y = plot_data_oliver$Actuallog2FC, 
          # #       type = "l", 
          # #       col = "red")
          # 
          # plot(x = plot_data_oliver$log2FoldChange, 
          #      y = plot_data_oliver$Actuallog2FC, 
          #      xlim = range(c(plot_data_oliver$log2FoldChange, 
          #                     plot_data_oliver$Actuallog2FC)), 
          #      ylim = range(c(plot_data_oliver$log2FoldChange, 
          #                     plot_data_oliver$Actuallog2FC)), 
          #      type = "p", 
          #      col = "black", 
          #      main = paste0(ifelse(j == 1, "Treg:\n", "Tconv:\n"), 
          #                    paste0(unlist(str_split(
          #                      string = names(merged_results_list[[i]][[j]])[k],
          #                      pattern = "__Vs__")),
          #                    collapse = "\nVs\n")), 
          #      asp = 1
          #      )
          # 
          # dev.off()
        }
        
        #simple theshold based selection criteria
        plot_data$coloured <- "not_coloured"
        plot_data$coloured[which(
          plot_data$padj <= 0.01 & 
            abs(plot_data$log2FoldChange) >= 2)] <- "coloured"
        
        
        #creating a new column for labelling genes of interest
        plot_data$named <- "not_named"
        #creating a pattern for genes to be labelles
        curr_genes_of_int <- NULL
        # if(length(genes_of_int[[j]][[k]]) > 0){
        #   curr_genes_of_int <- paste0("\\b(", 
        #                             paste(genes_of_int[[j]][[k]], 
        #                                   collapse = "|"), 
        #                             ")\\b")
        #   curr_genes_of_int <- gsub(pattern = "\\.", 
        #                             replacement = "\\\\.", 
        #                             x = curr_genes_of_int)
        #   
        #   plot_data$named[grepl(pattern = curr_genes_of_int, 
        #                       x = plot_data$MGI.symbol, 
        #                       ignore.case = T) & 
        #                   plot_data$coloured == "coloured"] <- "named"
        #   }
        # plot_data$named[grepl(pattern = genes_of_int, 
        #                       x = plot_data$MGI.symbol, 
        #                       ignore.case = T) & 
        #                   plot_data$coloured == "coloured"] <- "named"
        
        #calculating euclidean distance for finding significant genes
        {
          # #coordinates <- cbind(plot_data$log2FoldChange, plot_data$padj)
          # centroid <- c(mean(plot_data$log2FoldChange), 
          #               mean(-log10(plot_data$padj)))
          # #weighting adj.p.value at half due to its higher dynamic range in 
          # #-log10
          # weights <- c(1, 0.5)
          # #subsetting relevant (adj.p < 0.05) data for coloured euclidean dist
          # plot_data1 <- subset(plot_data, plot_data$padj <= 0.01 & 
          #                        abs(plot_data$log2FoldChange) >=2)
          # #plot_data2 <- subset(plot_data, plot_data$padj > 0.05)
          # coordinates <- cbind(plot_data1$log2FoldChange, 
          #                      -log10(plot_data1$padj))
          # 
          # plot_data1$dist  <- foreach(l=1:nrow(coordinates),
          #                            .combine = 'c',
          #                            .inorder = TRUE
          #                            ) %dopar% {
          #                              sqrt(sum(((coordinates[l,]*weights) - 
          #                                          centroid)^2))
          #                            }
          # 
          # #euclidean distance selection criteria
          # plot_data1$named <- as.factor(ifelse(
          #   (plot_data1$dist > quantile(x = plot_data1$dist, 0.9)
          #    & plot_data1$padj < 0.01),
          #   "named", "not_named"))
          # 
          # plot_data$named[
          #   which(plot_data$Gene.stable.ID %in% 
          #           plot_data1$Gene.stable.ID[plot_data1$named == "named"]
          #         )] <- "named"
        }
        
        #ordering the points based on colour group so that the selected ones 
        #show on top
        plot_data <- plot_data[order(plot_data$coloured, decreasing = T),]
        
        #formatting plot title here instead of in the already bulky ggplot code
        {
          title <- names(merged_results_list[[i]][[j]])[k]
          title <- gsub(pattern = "blood", 
                        replacement = "Blood", 
                        x = title, ignore.case = T)
          title <- gsub(pattern = "spleen", 
                        replacement = "Spleen", 
                        x = title, ignore.case = T)
          title <- gsub(pattern = "lymphnode", 
                        replacement = "LN", 
                        x = title, ignore.case = T)
          title <- gsub(pattern = "peyerspatches", 
                        replacement = "PP", 
                        x = title, ignore.case = F)
          title <- gsub(pattern = "intraepitheliallayer", 
                        replacement = "IEL", 
                        x = title, ignore.case = F)
          title <- gsub(pattern = "laminaproprialymphocyte", 
                        replacement = "LPL", 
                        x = title, ignore.case = F)
          title <- gsub(pattern = "kidney", 
                        replacement = "Kidney", 
                        x = title, ignore.case = F)
          title <- gsub(pattern = "lung", 
                        replacement = "Lung", 
                        x = title, ignore.case = F)
          title <- gsub(pattern = "liver", 
                        replacement = "Liver", 
                        x = title, ignore.case = F)
          title <- gsub(pattern = "pancreas", 
                        replacement = "Pancreas", 
                        x = title, ignore.case = F)
          
          title <- paste0("Comparing ", gsub(pattern = "_", 
                                             replacement = ", ", 
                                             x = gsub(pattern = "__Vs__", 
                                                      replacement = "  vs  ", 
                                                      x = title, 
                                                      ignore.case = T), 
                                             ignore.case = T))
        }
        
        #zeroing the log2 fold changes of "Left" and "Right" genes
        plot_data$log2FoldChange[plot_data$Facet == "Left"] <- #-25
          (ceiling(max(abs(
            plot_data$log2FoldChange[plot_data$Facet == "Both"]))) + 5) * -1
        plot_data$log2FoldChange[plot_data$Facet == "Right"] <- #25
          (ceiling(max(abs(
            plot_data$log2FoldChange[plot_data$Facet == "Both"]))) + 5)
        #plot_data <- plot_data[complete.cases(plot_data),]
        #plot_data <- plot_data[!is.finite(plot_data$log2FoldChange),]
        
        scatter_plot <- ggplot(data = plot_data, 
                               aes_(x = plot_data$log2FoldChange, 
                                    #x = plot_data$RealLog2FC, 
                                   y = -log10(plot_data$padj), 
                                   colour = plot_data$coloured,
                                   #color = plot_data$dist, 
                                   text = plot_data$MGI.symbol, 
                                   label = as.character(
                                     unlist(lapply(str_split(
                                       plot_data$`MGI symbol`, '; '),
                                       `[[`, 1))),
                                   log2FC = plot_data$log2FoldChange,
                                   #log2FC = plot_data$RealLog2FC,
                                   sig = plot_data$padj
                                   ),
                               ) + 
          #geom_point(shape = 19, size = 0.5) + 
          #scatter_plot <- scatter_plot + 
          #geom_hline(yintercept = 0, colour = "#4d4d4d", size = 0.25) + 
          geom_hline(yintercept = 0, colour = "#4d4d4d", size = 0.25) + 
          geom_hline(yintercept = -log10(0.01), 
                     colour = "#4d4d4d", size = 0.25) + 
          geom_vline(xintercept = 0, colour = "#4d4d4d", size = 0.25) + 
          geom_vline(xintercept = 2, colour = "#4d4d4d", size = 0.25) + 
          geom_vline(xintercept = -2, colour = "#4d4d4d", size = 0.25) + 
          geom_point(shape = 19, size = 0.5) + 
          #geom_point(shape = 19, size = 1) + 
          # # stat_dens2d_filter_g(geom = "debug", 
          # #                    h = c(MASS::bandwidth.nrd(
          # #                      plot_data$log2FoldChange), 
          # #                      mean(-log10(plot_data$padj))),
          # #                    n = 200, 
          # #                    keep.fraction = 0.00005, 
          # #                    keep.sparse = T) + 
          # geom_text_repel(aes(label = ifelse(
          #   plot_data$named == "named", #& plot_data$coloured == "coloured",
          #   #as.character(str_extract(plot_data$MGI.symbol, '\\w*')), '')),
          #   as.character(unlist(
          #     lapply(str_split(plot_data$MGI.symbol, '; '), `[[`, 1))),
          #   ''), segment.square = F, segment.inflect = T
          #   ),
          #   #the above picks up the first alias, therefore bring prefered
          #   #aliases to the beginning of the row in the output_table file
          #   #as.character(plot_data$MGI.symbol), '')),
          #   #hjust = 1, vjust = 1.5,
          #   size = 2, segment.size = 0.25, segment.alpha = 0.25,
          #   point.padding = 0.1, #xlim = c(-30, 30), ylim = c(-5, 100),
          #   #force = 5, force_pull = 5,
          #   direction = "both", show.legend = F, 
          #   nudge_y = 0, nudge_x = 0) +
          #expand_limits(x = c(-30, 30), y = c(-10, 100)) + 
          #scale_x_continuous(breaks = seq(-30,30,10)) + 
          #scale_y_continuous(breaks = seq(-20,100,20)) + 
          scale_color_manual(
            labels = c("Log2 |FC|\u2265 2\nand adj.p.value\n\u2264 0.01", 
                       "Other genes"),
            values = c("#fd7239", "darkgrey")) +
          labs(colour = "Genes", 
               #labs(colour = "Distance from centroid", 
               y = "-log10 of Adjusted p.value", 
               #y = bquote('-log'~[10]~' of Adjusted p.value'),
               x = "DESeq log2 Beta", 
               #title = title
               title = ""
               #caption = title
               ) + 
          #scale_fill_discrete(name = "Colour", labels = c("", "1% by quartile")) + 
          #scale_colour_viridis_c(option = "plasma") + 
          theme_light(base_size = 7) + 
          #theme_light(base_size = 9) + 
          theme(plot.caption = element_text(hjust = 0, face= "italic"), 
                legend.key.width = unit(0.15,"line")
                )
          #theme_light(base_size = 15)
          #theme_cowplot(font_size = 10)
        
        #customising x and y axis breaks
        # xbreaks <- sort(c(
        #   na.omit(ggplot_build(scatter_plot)$layout$panel_params[[1]]$x$breaks),
        #   -2, 2))
        #xbreaks <- sort(c(0,-2, 2, 5, -5, 10, -10, 15, -15, 20, -20))
        # xbreaks <- sort(c(0,-2, 2, 5, -5, 10, -10, 15, -15, 
        #                   max(plot_data$log2FoldChange), 
        #                   min(plot_data$log2FoldChange)))
        xbreaks <- 
          sort(c(-2, 2,
                 seq(0, -max(abs(
                   plot_data$log2FoldChange[plot_data$Facet == "Both"])), -5),
                 seq(0, max(abs(
                   plot_data$log2FoldChange[plot_data$Facet == "Both"])), 5)))
      
        ybreaks <- sort(c(
          na.omit(ggplot_build(scatter_plot)$layout$panel_params[[1]]$y$breaks),
          -log10(0.01)))
        # ybreaks <- sort(c(0, -log10(0.01)))
      
        ylabels <- as.character(ybreaks)
        ylabels[ylabels == -log10(0.01)] <- "\nadj.p\n0.01\n\n\n"
      
        scatter_plot <- scatter_plot +
          scale_x_continuous(breaks = xbreaks) +
          scale_y_continuous(breaks = ybreaks, labels = ylabels) + 
          expand_limits(x = c(-ceiling(max(abs(plot_data$log2FoldChange))), 
                              ceiling(max(abs(plot_data$log2FoldChange))))
                        ) + 
          # expand_limits(y = c(-5, max(
          #   na.omit(
          #     ggplot_build(scatter_plot)$layout$panel_params[[1]]$y$breaks))
          #   ))+ 
          theme(panel.grid.minor.x = element_blank(), 
                #panel.grid.major.x = element_blank(), 
                panel.grid.minor.y = element_blank(), 
                #panel.grid.major.y = element_blank(),
                #axis.line = element_line(colour = 'black')
                panel.grid.major = element_line(colour = "#4d4d4d"),
                axis.ticks = element_line(colour = "#4d4d4d")
                )
        
        #adding floating text annotations on the plot area
        label_LHS <- gsub(pattern = ", ", 
                          replacement = "\n", 
                          x = gsub(pattern = "Comparing ", 
                                   replacement = "", 
                                   x = unlist(strsplit(x = title, 
                                                       split = "  vs  "))[2], 
                                   ignore.case = T), 
                          ignore.case = T)
        label_LHS <- paste0("Expressed\nhigher in,\n", label_LHS)
        label_RHS <- gsub(pattern = ", ", 
                          replacement = "\n", 
                          x = gsub(pattern = "Comparing ", 
                                   replacement = "", 
                                   x = unlist(strsplit(x = title, 
                                                       split = "  vs  "))[1], 
                                   ignore.case = T), 
                          ignore.case = T)
        label_RHS <- paste0("Expressed\nhigher in,\n", label_RHS)
        
        # scatter_plot <- scatter_plot + 
        #   coord_cartesian(clip = "off") + 
        #   labs(title = "\n\n") #+ 
        #   # annotate(geom = "text", 
        #   #          x = -ceiling(max(abs(plot_data$log2FoldChange))), 
        #   #          y = ceiling(max(-log10(plot_data$padj))) + 
        #   #            (0.1 * ceiling(max(-log10(plot_data$padj)))), 
        #   #          label = label_LHS, 
        #   #          size = 2, 
        #   #          hjust = 0, 
        #   #          vjust = 0) + 
        #   # annotate(geom = "text", 
        #   #          x = ceiling(max(abs(plot_data$log2FoldChange))), 
        #   #          y = ceiling(max(-log10(plot_data$padj))) + 
        #   #            (0.1 * ceiling(max(-log10(plot_data$padj)))), 
        #   #          label = label_RHS, 
        #   #          size = 2, 
        #   #          hjust = 1, 
        #   #          vjust = 0)
        
        #removing all grid lines while retaining axis labels
        scatter_plot <- scatter_plot + ggExtra::removeGrid()
        #creating hlines and vlines to mark the relevant axis lines
        ## moved to before the plotting of geom_point so that they appear below the dots
        # scatter_plot <- scatter_plot + 
        #   geom_hline(yintercept = 0, colour = "#4d4d4d", size = 0.1) + 
        #   geom_hline(yintercept = -log10(0.01), colour = "#4d4d4d", size = 0.1) + 
        #   geom_vline(xintercept = 0, colour = "#4d4d4d", size = 0.1) + 
        #   geom_vline(xintercept = 2, colour = "#4d4d4d", size = 0.1) + 
        #   geom_vline(xintercept = -2, colour = "#4d4d4d", size = 0.1)
        
        # scatter_plot <- scatter_plot +
        #   geom_text_repel(aes(label = ifelse(
        #     plot_data$named == "named", #& plot_data$coloured == "coloured",
        #     #as.character(str_extract(plot_data$MGI.symbol, '\\w*')), '')),
        #     as.character(unlist(
        #       lapply(str_split(plot_data$MGI.symbol, '; '), `[[`, 1))),
        #     ''), segment.square = F, segment.inflect = T
        #     ),
        #     #the above picks up the first alias, therefore bring prefered
        #     #aliases to the beginning of the row in the output_table file
        #     #as.character(plot_data$MGI.symbol), '')),
        #     #hjust = 1, vjust = 1.5,
        #     size = 1.85, segment.size = 0.25, segment.alpha = 0.25,
        #     #point.padding = 0.1, #xlim = c(-30, 30), ylim = c(-5, 100),
        #     #force = 5, force_pull = 5,
        #     direction = "both", show.legend = F,
        #     nudge_y = 0, nudge_x = 0)
        
        #creating facets
        #scatter_plot <- scatter_plot + facet_grid(cols = vars(Facet))
        
        
        #saving the plots
        saved_plots[[i]][[j]][[length(saved_plots[[i]][[j]])+1]] <- 
          (scatter_plot)
        names(saved_plots[[i]][[j]])[length(saved_plots[[i]][[j]])] <- 
          names(merged_results_list[[i]][[j]])[k]
        
        
        #saving generated plot via ggsave
        ggsave(filename = paste0(names(merged_results_list[[i]][[j]])[k],
                                 ".png"),
               plot = scatter_plot,
               device = "png",
               path = path,
               #scale = 1,
               #width = 9.75,
               #height = 6.75,
               width = 100,
               height = 55,
               units = "mm",
               dpi = 600,
               limitsize = T)

        ggsave(filename = paste0(names(merged_results_list[[i]][[j]])[k],
                                 ".pdf"),
               plot = scatter_plot,
               device = "pdf",
               path = path,
               #scale = 1,
               #width = 9.75,
               #height = 6.75,
               width = 100,
               height = 55,
               units = "mm",
               dpi = 600,
               limitsize = T)
        
        # #saving generated plot via the png device
        # png(filename = paste0(path, 
        #                       names(merged_results_list[[i]][[j]])[k], 
        #                       ".png"), 
        #     width = 900, 
        #     height = 552, 
        #     units = "px", 
        #     bg = "transparent")
        # print(scatter_plot)
        # dev.off()
        
      }
      
    }
  }
  rm(i,j,k)
  
}

#running the above plot code, but for blood treg vs all remaining treg
## get DESeq_object first
{
  DESeq_processed <- readRDS(file = "DESeq_Object.rds")
  results <- results(object = DESeq_processed, 
                     contrast = list("tissueBlood.cellTypeT", 
                                    resultsNames(DESeq_processed)[12:20]),
                     listValues = 
                       c(1, -1/length(resultsNames(DESeq_processed)[12:20])),
                     cooksCutoff = F,
                     independentFiltering = F,
                     test = "Wald",
                     parallel = T)
  results <- as.data.frame(results)
  results <- results[rownames(results) %in% rownames(filtered_tcounts),]
  
  Actuallog2FC <- mean_norm_exp[,"Blood_T", drop = F] - 
    rowMeans(mean_norm_exp[,c(4,6,8,10,12,14,16,18,20)], na.rm = T)
  
  results <- merge(results, Actuallog2FC, by = "row.names")
  rownames(results) <- results$Row.names
  results$Row.names <- NULL
  colnames(results)[ncol(results)] <- "Actuallog2FC"
  #colnames(results)
  
  plot_data <- results
  
  plot_data$Facet <- "Both"
  plot_data$Facet[!is.finite(plot_data$Actuallog2FC) & 
                    plot_data$log2FoldChange > 0] <- "Right"
  plot_data$Facet[!is.finite(plot_data$Actuallog2FC) & 
                  plot_data$log2FoldChange < 0] <- "Left"
  plot_data$Facet <- factor(x = plot_data$Facet, 
                            levels = c("Left", "Both", "Right"))
  
  if(exists("filtered_tcounts")){
    plot_data <- subset(plot_data, rownames(plot_data) %in% 
                          rownames(filtered_tcounts))
  }
  plot_data$`Gene stable ID` <- gsub(pattern = "\\.[0-9]+$", 
                                     replacement = "", 
                                     x = rownames(plot_data))
  plot_data <- merge(x = plot_data, y = output_table[,1:2], 
                     by = "Gene stable ID", 
                     all.x = T, all.y = F, 
                     sort = F)
  
  plot_data <- plot_data[order(plot_data$log2FoldChange, decreasing = T),]
  
  #simple theshold based selection criteria
  plot_data$coloured <- "not_coloured"
  plot_data$coloured[which(plot_data$padj <= 0.01 & 
                             abs(plot_data$log2FoldChange) >= 2)] <- "coloured"
  
  #creating a new column for labelling genes of interest
  plot_data$named <- "not_named"
  #creating a pattern for genes to be labelles
  curr_genes_of_int <- NULL
  
  #ordering the points based on colour group so that the selected ones 
  #show on top
  plot_data <- plot_data[order(plot_data$coloured, decreasing = T),]
  
  #zeroing the log2 fold changes of "Left" and "Right" genes
  plot_data$log2FoldChange[plot_data$Facet == "Left"] <- #-25
    (ceiling(max(abs(
      plot_data$log2FoldChange[plot_data$Facet == "Both"]))) + 2) * -1
  plot_data$log2FoldChange[plot_data$Facet == "Right"] <- #25
    (ceiling(max(abs(
      plot_data$log2FoldChange[plot_data$Facet == "Both"]))) + 2)
  
  scatter_plot <- ggplot(data = plot_data, 
                         aes_(x = plot_data$log2FoldChange, 
                             y = -log10(plot_data$padj), 
                             colour = plot_data$coloured,
                             text = plot_data$MGI.symbol, 
                             label = as.character(
                               unlist(lapply(str_split(
                                 plot_data$`MGI symbol`, '; '),
                                 `[[`, 1))),
                             log2FC = plot_data$log2FoldChange,
                             sig = plot_data$padj
                             ),
                         ) + 
    geom_hline(yintercept = 0, colour = "#4d4d4d", size = 0.25) + 
    geom_hline(yintercept = -log10(0.01), 
               colour = "#4d4d4d", size = 0.25) + 
    geom_vline(xintercept = 0, colour = "#4d4d4d", size = 0.25) + 
    geom_vline(xintercept = 2, colour = "#4d4d4d", size = 0.25) + 
    geom_vline(xintercept = -2, colour = "#4d4d4d", size = 0.25) + 
    geom_point(shape = 19, size = 0.5) + 
    scale_color_manual(
      labels = c("Log2 |FC|\u2265 2\nand adj.p.value\n\u2264 0.01", 
                 "Other genes"),
      values = c("#fd7239", "darkgrey")) +
    labs(colour = "Genes", 
         y = "-log10 of Adjusted p.value", 
         x = "DESeq log2 Beta", 
         title = ""
         ) + 
    theme_light(base_size = 7) + 
    theme(plot.caption = element_text(hjust = 0, face= "italic"), 
          legend.key.width = unit(0.15,"line")
          )
  
  #customising x and y axis breaks
  xbreaks <- 
    sort(c(-2, 2,
           seq(0, -max(abs(
             plot_data$log2FoldChange[plot_data$Facet == "Both"])), -5),
           seq(0, max(abs(
             plot_data$log2FoldChange[plot_data$Facet == "Both"])), 5)))

  ybreaks <- sort(c(
    na.omit(ggplot_build(scatter_plot)$layout$panel_params[[1]]$y$breaks),
    -log10(0.01)))

  ylabels <- as.character(ybreaks)
  ylabels[ylabels == -log10(0.01)] <- "\nadj.p\n0.01\n\n\n"

  scatter_plot <- scatter_plot +
    scale_x_continuous(breaks = xbreaks) +
    scale_y_continuous(breaks = ybreaks, labels = ylabels) + 
    expand_limits(x = c(-ceiling(max(abs(plot_data$log2FoldChange))), 
                        ceiling(max(abs(plot_data$log2FoldChange))))
                  ) + 
    theme(panel.grid.minor.x = element_blank(), 
          panel.grid.minor.y = element_blank(), 
          panel.grid.major = element_line(colour = "#4d4d4d"),
          axis.ticks = element_line(colour = "#4d4d4d")
          )
  
  #removing all grid lines while retaining axis labels
  scatter_plot <- scatter_plot + ggExtra::removeGrid()
  
  #saving generated plot via ggsave
  ggsave(filename = "Blood_Treg_vs_All_Treg.png",
         plot = scatter_plot,
         device = "png",
         path = getwd(),
         #scale = 1,
         #width = 9.75,
         #height = 6.75,
         width = 100,
         height = 55,
         units = "mm",
         dpi = 600,
         limitsize = T)

  ggsave(filename = "Blood_Treg_vs_All_Treg.pdf",
         plot = scatter_plot,
         device = "pdf",
         path = getwd(),
         #scale = 1,
         #width = 9.75,
         #height = 6.75,
         width = 100,
         height = 55,
         units = "mm",
         dpi = 600,
         limitsize = T)
  
}


#creating a list of composite volcano plots via ggplot2 scatter plots using 
#Actuallog2FC
{
  saved_plots <- list()
  
  for(i in seq(1, length(merged_results_list))){
    #print(paste0("i: ", i))
    
    saved_plots[[length(saved_plots)+1]] <- list()
    names(saved_plots)[length(saved_plots)] <- names(merged_results_list)[i]
    
    writeLines(paste0("\nProcessing results from the model: ", 
                      names(merged_results_list)[i]))
    
    for(j in seq(1, length(merged_results_list[[i]]))){
      #print(paste0("j: ", j))
      
      saved_plots[[i]][[length(saved_plots[[i]])+1]] <- list()
      names(saved_plots[[i]])[length(saved_plots[[i]])] <- 
        names(merged_results_list[[i]])[j]
      
      #print(names(merged_results_list[[i]])[j])
      
      #constructing the path and creating directories
      path <- paste0(loc_plots, 
                     "Volcano Plots/", 
                     names(merged_results_list)[i], "/", 
                     names(merged_results_list[[i]])[j], "/")
      dir.create(path = path, recursive = T)
      
      writeLines(paste0("Exporting plots for: ", 
                        names(merged_results_list[[i]])[j]))
      
      for(k in seq(1, length(merged_results_list[[i]][[j]]))){
        #print(paste0("k: ", k))
        
        print(paste0("Exporting plot: ", 
                     names(merged_results_list[[i]][[j]])[k]))
        
        plot_data <- as.data.frame(
          merged_results_list[[i]][[j]][[k]])[
            complete.cases(merged_results_list[[i]][[j]][[k]]),]
        
        #filter the genes if filtered_tcounts exists
        if(exists("filtered_tcounts")){
          plot_data <- subset(plot_data, rownames(plot_data) %in% 
                                rownames(filtered_tcounts))
        }
        
        plot_data$shape <- "real"
        plot_data[abs(plot_data$Actuallog2FC) == Inf,"shape"] <- "beta"
        
        plot_data[plot_data$shape == "beta", "Actuallog2FC"] <- 
          plot_data[plot_data$shape == "beta", "log2FoldChange"]
        
        plot_data$Gene.stable.ID <- gsub(pattern = "\\.[0-9]+$", 
                                         replacement = "", 
                                         x = rownames(plot_data))
        
        plot_data <- merge(x = plot_data, y = output_table[,1:2], 
                           by = "Gene.stable.ID", 
                           all.x = T, all.y = F, 
                           sort = F)
        
        plot_data <- plot_data[order(plot_data$Actuallog2FC, decreasing = T),]
        
        #simple theshold based selection criteria
        plot_data$coloured <- "not_coloured"
        plot_data$coloured[which(
          plot_data$padj <= 0.01 & 
            abs(plot_data$Actuallog2FC) >= 2)] <- "coloured"
        
        #creating a new column for labelling genes of interest
        plot_data$named <- "not_named"
        plot_data$named[grepl(pattern = genes_of_int, 
                              x = plot_data$MGI.symbol, 
                              ignore.case = T)] <- "named"
        
        #calculating euclidean distance for finding significant genes
        {
          # #coordinates <- cbind(plot_data$log2FoldChange, plot_data$padj)
          # centroid <- c(mean(plot_data$Actuallog2FC), 
          #               mean(-log10(plot_data$padj)))
          # #weighting adj.p.value at half due to its higher dynamic range in 
          # #-log10
          # weights <- c(1, 0.5)
          # #subsetting relevant (adj.p < 0.05) data for coloured euclidean dist
          # plot_data1 <- subset(plot_data, plot_data$padj <= 0.01 & 
          #                        abs(plot_data$Actuallog2FC) >=5)
          # #plot_data2 <- subset(plot_data, plot_data$padj > 0.05)
          # coordinates <- cbind(plot_data1$Actuallog2FC, 
          #                      -log10(plot_data1$padj))
          # 
          # plot_data1$dist  <- foreach(l=1:nrow(coordinates),
          #                            .combine = 'c',
          #                            .inorder = TRUE
          #                            ) %dopar% {
          #                              sqrt(sum(((coordinates[l,]*weights) - 
          #                                          centroid)^2))
          #                            }
          # 
          # #euclidean distance selection criteria
          # plot_data1$named <- as.factor(ifelse(
          #   (plot_data1$dist > quantile(x = plot_data1$dist, 0.9, na.rm = T)
          #    & plot_data1$padj < 0.01),
          #   "named", "not_named"))
          # 
          # plot_data$named[
          #   which(plot_data$Gene.stable.ID %in% 
          #           plot_data1$Gene.stable.ID[plot_data1$named == "named"]
          #         )] <- "named"
        }
        
        #ordering the points based on colour group so that the selected ones 
        #show on top
        plot_data <- plot_data[order(plot_data$coloured, decreasing = T),]
        
        #formatting plot title here instead of in the already bulky ggplot code
        {
          title <- names(merged_results_list[[i]][[j]])[k]
          title <- gsub(pattern = "blood", 
                        replacement = "Blood", 
                        x = title, ignore.case = T)
          title <- gsub(pattern = "spleen", 
                        replacement = "Spleen", 
                        x = title, ignore.case = T)
          title <- gsub(pattern = "lymphnode", 
                        replacement = "Lymph node", 
                        x = title, ignore.case = T)
          title <- gsub(pattern = "peyerspatches", 
                        replacement = "Peyer's patches", 
                        x = title, ignore.case = F)
          title <- gsub(pattern = "intraepitheliallayer", 
                        replacement = "Intraepithelial layer", 
                        x = title, ignore.case = F)
          title <- gsub(pattern = "laminaproprialymphocyte", 
                        replacement = "Lamina propria lymphocyte", 
                        x = title, ignore.case = F)
          title <- gsub(pattern = "kidney", 
                        replacement = "Kidney", 
                        x = title, ignore.case = F)
          title <- gsub(pattern = "lung", 
                        replacement = "Lung", 
                        x = title, ignore.case = F)
          title <- gsub(pattern = "liver", 
                        replacement = "Liver", 
                        x = title, ignore.case = F)
          title <- gsub(pattern = "pancreas", 
                        replacement = "Pancreas", 
                        x = title, ignore.case = F)
          
          title <- paste0("Comparing ", gsub(pattern = "_", 
                                             replacement = ", ", 
                                             x = gsub(pattern = "__Vs__", 
                                                      replacement = "  vs  ", 
                                                      x = title, 
                                                      ignore.case = T), 
                                             ignore.case = T))
        }
        
        
        scatter_plot <- ggplot(data = plot_data, 
                               aes_(x = plot_data$Actuallog2FC, 
                                    #x = plot_data$Actuallog2FC, 
                                   y = -log10(plot_data$padj), 
                                   colour = plot_data$coloured,
                                   shape = plot_data$shape, 
                                   #color = plot_data$dist, 
                                   text = plot_data$MGI.symbol, 
                                   # label = as.character(
                                   #   unlist(lapply(str_split(
                                   #     plot_data$MGI.symbol, '; '), 
                                   #     `[[`, 1))),
                                   log2FC = plot_data$Actuallog2FC,
                                   #log2FC = plot_data$Actuallog2FC,
                                   sig = plot_data$padj
                                   ),
                               ) + 
          #geom_point(shape = 19, size = 0.5) + 
          geom_point(size = 1.5, stroke = 0) + 
          # #geom_point() + 
          # # stat_dens2d_filter_g(geom = "debug", 
          # #                    h = c(MASS::bandwidth.nrd(
          # #                      plot_data$log2FoldChange), 
          # #                      mean(-log10(plot_data$padj))),
          # #                    n = 200, 
          # #                    keep.fraction = 0.00005, 
          # #                    keep.sparse = T) + 
          # geom_text_repel(aes(label = ifelse(
          #   plot_data$named == "named", #& plot_data$coloured == "coloured",
          #   #as.character(str_extract(plot_data$MGI.symbol, '\\w*')), '')),
          #   as.character(unlist(
          #     lapply(str_split(plot_data$MGI.symbol, '; '), `[[`, 1))),
          #   ''), segment.square = F, segment.inflect = T
          #   ),
          #   #the above picks up the first alias, therefore bring prefered
          #   #aliases to the beginning of the row in the output_table file
          #   #as.character(plot_data$MGI.symbol), '')),
          #   #hjust = 1, vjust = 1.5,
          #   size = 3, segment.size = 0.25, segment.alpha = 0.25,
          #   point.padding = 0.1, #xlim = c(-30, 30), ylim = c(-5, 100),
          #   force = 5, force_pull = 5,
          #   direction = "both", show.legend = F, nudge_y = 0,
          #   nudge_x = 0) +
          #expand_limits(x = c(-30, 30), y = c(-10, 100)) + 
          # expand_limits(x = c(-ceiling(max(abs(plot_data$Actuallog2FC))), 
          #                     ceiling(max(abs(plot_data$Actuallog2FC))))
          #               ) + 
          #scale_x_continuous(breaks = seq(-30,30,10)) + 
          #scale_y_continuous(breaks = seq(-20,100,20)) + 
          scale_color_manual(
            labels = c("Log2 |FC|\u2265 2 and\nadj.p.value\u2264 0.01", 
                       "Other genes"),
            values = c("#fd7239", "darkgrey")) +
          labs(colour = "Genes", 
               #labs(colour = "Distance from centroid", 
               y = "-Log10 of Adjusted P.value", 
               x = "Log2 Fold Change", 
               #title = title
               caption = title
               ) + 
          scale_shape_manual(labels = c("DESeq2 Estimated\nBeta", 
                                        "Actual Log2\nFold Change"), 
                             values = c(17, 19)#, 
                             #sizes = c(1, 0.5)
                             ) + 
          #scale_fill_discrete(name = "Colour", labels = c("", "1% by quartile")) + 
          #scale_colour_viridis_c(option = "plasma") + 
          theme_light(base_size = 12) +
          theme(plot.caption = element_text(hjust = 0, face= "italic"))
          #theme_light(base_size = 15)
        
        #customising x and y axis breaks
        # xbreaks <- sort(c(
        #   na.omit(ggplot_build(scatter_plot)$layout$panel_params[[1]]$x$breaks),
        #   -2, 2))
        xbreaks <- sort(c(0,-2, 2))
      
        # ybreaks <- sort(c(
        #   na.omit(ggplot_build(scatter_plot)$layout$panel_params[[1]]$y$breaks),
        #   -log10(0.01)))
        ybreaks <- sort(c(0, -log10(0.01)))
      
        ylabels <- as.character(ybreaks)
        ylabels[ylabels == -log10(0.01)] <- "\nadj.p\n0.01\n\n\n"
      
        scatter_plot <- scatter_plot +
          scale_x_continuous(breaks = xbreaks) +
          scale_y_continuous(breaks = ybreaks, labels = ylabels) + 
          expand_limits(x = c(-ceiling(max(abs(plot_data$Actuallog2FC))), 
                              ceiling(max(abs(plot_data$Actuallog2FC))))
                        ) + 
          expand_limits(y = c(-5, max(
            na.omit(
              ggplot_build(scatter_plot)$layout$panel_params[[1]]$y$breaks))
            ))+ 
          theme(panel.grid.minor.x = element_blank(), 
                #panel.grid.major.x = element_blank(), 
                panel.grid.minor.y = element_blank(), 
                #panel.grid.major.y = element_blank(),
                #axis.line = element_line(colour = 'black')
                panel.grid.major = element_line(colour = "#4d4d4d"),
                axis.ticks = element_line(colour = "#4d4d4d")
                )
        
        #adding floating text annotations on the plot area
        label_LHS <- gsub(pattern = ", ", 
                          replacement = "\n", 
                          x = gsub(pattern = "Comparing ", 
                                   replacement = "", 
                                   x = unlist(strsplit(x = title, 
                                                       split = "  vs  "))[2], 
                                   ignore.case = T), 
                          ignore.case = T)
        label_LHS <- paste0("Expressed higher in,\n", label_LHS)
        label_RHS <- gsub(pattern = ", ", 
                          replacement = "\n", 
                          x = gsub(pattern = "Comparing ", 
                                   replacement = "", 
                                   x = unlist(strsplit(x = title, 
                                                       split = "  vs  "))[1], 
                                   ignore.case = T), 
                          ignore.case = T)
        label_RHS <- paste0("Expressed higher in,\n", label_RHS)
        
        scatter_plot <- scatter_plot + 
          annotate(geom = "text", 
                   x = -ceiling(max(abs(plot_data$Actuallog2FC))), 
                   y = ceiling(max(-log10(plot_data$padj))), 
                   label = label_LHS, 
                   size = 3, 
                   hjust = 0, 
                   vjust = 1) + 
          annotate(geom = "text", 
                   x = ceiling(max(abs(plot_data$Actuallog2FC))), 
                   y = ceiling(max(-log10(plot_data$padj))), 
                   label = label_RHS, 
                   size = 3, 
                   hjust = 1, 
                   vjust = 1)
        
        #saving generated plot via ggsave
        ggsave(filename = paste0(names(merged_results_list[[i]][[j]])[k],
                                 ".png"),
               plot = scatter_plot,
               device = "png",
               path = path,
               #scale = 1,
               #width = 9.75,
               #height = 6.75,
               width = 6.5,
               height = 4.5,
               units = "in",
               #dpi = 500,
               limitsize = T)

        ggsave(filename = paste0(names(merged_results_list[[i]][[j]])[k],
                                 ".svg"),
               plot = scatter_plot,
               device = "svg",
               path = path,
               #scale = 1,
               #width = 9.75,
               #height = 6.75,
               width = 6.5,
               height = 4.5,
               units = "in",
               #dpi = 500,
               limitsize = T)
        
        # #saving generated plot via the png device
        # png(filename = paste0(path, 
        #                       names(merged_results_list[[i]][[j]])[k], 
        #                       ".png"), 
        #     width = 900, 
        #     height = 552, 
        #     units = "px", 
        #     bg = "transparent")
        # print(scatter_plot)
        # dev.off()
        
      }
      
    }
  }
  rm(i,j,k)
  
}

#creating a list of volcano plots via ggplot2 scatter plots using only 
#Actuallog2FC
{
  saved_plots <- list()
  
  for(i in seq(1, length(merged_results_list))){
    #print(paste0("i: ", i))
    
    saved_plots[[length(saved_plots)+1]] <- list()
    names(saved_plots)[length(saved_plots)] <- names(merged_results_list)[i]
    
    writeLines(paste0("\nProcessing results from the model: ", 
                      names(merged_results_list)[i]))
    
    for(j in seq(1, length(merged_results_list[[i]]))){
      #print(paste0("j: ", j))
      
      saved_plots[[i]][[length(saved_plots[[i]])+1]] <- list()
      names(saved_plots[[i]])[length(saved_plots[[i]])] <- 
        names(merged_results_list[[i]])[j]
      
      #print(names(merged_results_list[[i]])[j])
      
      #constructing the path and creating directories
      path <- paste0(loc_plots, 
                     "Volcano Plots/", 
                     names(merged_results_list)[i], "/", 
                     names(merged_results_list[[i]])[j], "/")
      dir.create(path = path, recursive = T)
      
      writeLines(paste0("Exporting plots for: ", 
                        names(merged_results_list[[i]])[j]))
      
      for(k in seq(1, length(merged_results_list[[i]][[j]]))){
        #print(paste0("k: ", k))
        
        print(paste0("Exporting plot: ", 
                     names(merged_results_list[[i]][[j]])[k]))
        
        plot_data <- as.data.frame(
          merged_results_list[[i]][[j]][[k]])[
            complete.cases(merged_results_list[[i]][[j]][[k]]),]
        
        #filter the genes if filtered_tcounts exists
        if(exists("filtered_tcounts")){
          plot_data <- subset(plot_data, rownames(plot_data) %in% 
                                rownames(filtered_tcounts))
        }
        
        plot_data$Gene.stable.ID <- gsub(pattern = "\\.[0-9]+$", 
                                         replacement = "", 
                                         x = rownames(plot_data))
        
        plot_data <- merge(x = plot_data, y = output_table[,1:2], 
                           by = "Gene.stable.ID", 
                           all.x = T, all.y = F, 
                           sort = F)
        
        plot_data <- plot_data[order(plot_data$Actuallog2FC, decreasing = T),]
        
        #simple theshold based selection criteria
        plot_data$coloured <- "not_coloured"
        plot_data$coloured[which(
          plot_data$padj <= 0.01 & 
            abs(plot_data$Actuallog2FC) >= 2)] <- "coloured"
        
        #creating a new column for labelling genes of interest
        plot_data$named <- "not_named"
        plot_data$named[grepl(pattern = genes_of_int, 
                              x = plot_data$MGI.symbol, 
                              ignore.case = T)] <- "named"
        
        #calculating euclidean distance for finding significant genes
        {
          #coordinates <- cbind(plot_data$log2FoldChange, plot_data$padj)
          centroid <- c(mean(plot_data$RealLog2FC), 
                        mean(-log10(plot_data$padj)))
          #weighting adj.p.value at half due to its higher dynamic range in 
          #-log10
          weights <- c(1, 0.5)
          #subsetting relevant (adj.p < 0.05) data for coloured euclidean dist
          plot_data1 <- subset(plot_data, plot_data$padj <= 0.01 & 
                                 abs(plot_data$RealLog2FC) >=2)
          #plot_data2 <- subset(plot_data, plot_data$padj > 0.05)
          coordinates <- cbind(plot_data1$RealLog2FC, 
                               -log10(plot_data1$padj))
  
          plot_data1$dist  <- foreach(l=1:nrow(coordinates),
                                     .combine = 'c',
                                     .inorder = TRUE
                                     ) %dopar% {
                                       sqrt(sum(((coordinates[l,]*weights) - 
                                                   centroid)^2))
                                     }
          
          #euclidean distance selection criteria
          plot_data1$named <- as.factor(ifelse(
            (plot_data1$dist > quantile(x = plot_data1$dist, 0.9, na.rm = T)
             & plot_data1$padj < 0.01),
            "named", "not_named"))
          
          plot_data$named[
            which(plot_data$Gene.stable.ID %in% 
                    plot_data1$Gene.stable.ID[plot_data1$named == "named"]
                  )] <- "named"
        }
        
        #ordering the points based on colour group so that the selected ones 
        #show on top
        plot_data <- plot_data[order(plot_data$coloured, decreasing = T),]
        
        #formatting plot title here instead of in the already bulky ggplot code
        {
          title <- names(merged_results_list[[i]][[j]])[k]
          title <- gsub(pattern = "blood", 
                        replacement = "Blood", 
                        x = title, ignore.case = T)
          title <- gsub(pattern = "spleen", 
                        replacement = "Spleen", 
                        x = title, ignore.case = T)
          title <- gsub(pattern = "lymphnode", 
                        replacement = "Lymph node", 
                        x = title, ignore.case = T)
          title <- gsub(pattern = "peyerspatches", 
                        replacement = "Peyer's patches", 
                        x = title, ignore.case = F)
          title <- gsub(pattern = "intraepitheliallayer", 
                        replacement = "Intraepithelial layer", 
                        x = title, ignore.case = F)
          title <- gsub(pattern = "laminaproprialymphocyte", 
                        replacement = "Lamina propria lymphocyte", 
                        x = title, ignore.case = F)
          title <- gsub(pattern = "kidney", 
                        replacement = "Kidney", 
                        x = title, ignore.case = F)
          title <- gsub(pattern = "lung", 
                        replacement = "Lung", 
                        x = title, ignore.case = F)
          title <- gsub(pattern = "liver", 
                        replacement = "Liver", 
                        x = title, ignore.case = F)
          title <- gsub(pattern = "pancreas", 
                        replacement = "Pancreas", 
                        x = title, ignore.case = F)
          
          title <- paste0("Comparing ", gsub(pattern = "_", 
                                             replacement = ", ", 
                                             x = gsub(pattern = "__Vs__", 
                                                      replacement = "  vs  ", 
                                                      x = title, 
                                                      ignore.case = T), 
                                             ignore.case = T))
        }
          
        
        scatter_plot <- ggplot(data = plot_data, 
                               aes_(x = plot_data$Actuallog2FC, 
                                    #x = plot_data$Actuallog2FC, 
                                   y = -log10(plot_data$padj), 
                                   colour = plot_data$coloured,
                                   #color = plot_data$dist, 
                                   text = plot_data$MGI.symbol, 
                                   # label = as.character(
                                   #   unlist(lapply(str_split(
                                   #     plot_data$MGI.symbol, '; '), 
                                   #     `[[`, 1))),
                                   log2FC = plot_data$Actuallog2FC,
                                   #log2FC = plot_data$Actuallog2FC,
                                   sig = plot_data$padj
                                   ),
                               ) + 
          geom_point(shape = 19, size = 0.75) + 
          # stat_dens2d_filter_g(geom = "debug", 
          #                    h = c(MASS::bandwidth.nrd(
          #                      plot_data$log2FoldChange), 
          #                      mean(-log10(plot_data$padj))),
          #                    n = 200, 
          #                    keep.fraction = 0.00005, 
          #                    keep.sparse = T) + 
          geom_text_repel(aes(label = ifelse(
            plot_data$named == "named", #& plot_data$coloured == "coloured",
            #as.character(str_extract(plot_data$MGI.symbol, '\\w*')), '')),
            as.character(unlist(
              lapply(str_split(plot_data$MGI.symbol, '; '), `[[`, 1))),
            ''), segment.square = F, segment.inflect = T
            ),
            #the above picks up the first alias, therefore bring prefered
            #aliases to the beginning of the row in the output_table file
            #as.character(plot_data$MGI.symbol), '')),
            #hjust = 1, vjust = 1.5,
            size = 3, segment.size = 0.25, segment.alpha = 0.25,
            point.padding = 0.1, #xlim = c(-30, 30), ylim = c(-5, 100),
            force = 5, force_pull = 5,
            direction = "both", show.legend = F, nudge_y = 0,
            nudge_x = 0) +
          #expand_limits(x = c(-30, 30), y = c(-10, 100)) + 
          #scale_x_continuous(breaks = seq(-30,30,10)) + 
          #scale_y_continuous(breaks = seq(-20,100,20)) + 
          scale_color_manual(
            labels = c("Log2 |FC|\u2265 2 and\nadj.p.value\u2264 0.01", 
                       "Other genes"),
            values = c("#fd7239", "darkgrey")) +
          labs(colour = "Genes", 
               #labs(colour = "Distance from centroid", 
               y = "-Log10 of Adjusted P.value", 
               x = "Log2 Fold Change", 
               #title = title
               caption = title
               ) + 
          #scale_fill_discrete(name = "Colour", labels = c("", "1% by quartile")) + 
          #scale_colour_viridis_c(option = "plasma") + 
          theme_light(base_size = 12)+
          theme(plot.caption = element_text(hjust = 0, face= "italic"))
          #theme_light(base_size = 15)
          #theme_cowplot(font_size = 10)
        
        #customising x and y axis breaks
        # xbreaks <- sort(c(
        #   na.omit(ggplot_build(scatter_plot)$layout$panel_params[[1]]$x$breaks),
        #   -2, 2))
        xbreaks <- sort(c(0,-2, 2))
      
        # ybreaks <- sort(c(
        #   na.omit(ggplot_build(scatter_plot)$layout$panel_params[[1]]$y$breaks),
        #   -log10(0.01)))
        ybreaks <- sort(c(0, -log10(0.01)))
      
        ylabels <- as.character(ybreaks)
        ylabels[ylabels == -log10(0.01)] <- "\nadj.p\n0.01\n\n\n"
      
        scatter_plot <- scatter_plot +
          scale_x_continuous(breaks = xbreaks) +
          scale_y_continuous(breaks = ybreaks, labels = ylabels) + 
          expand_limits(x = c(-ceiling(max(abs(plot_data$Actuallog2FC))), 
                              ceiling(max(abs(plot_data$Actuallog2FC))))
                        ) + 
          expand_limits(y = c(-5, max(
            na.omit(
              ggplot_build(scatter_plot)$layout$panel_params[[1]]$y$breaks))
            ))+ 
          theme(panel.grid.minor.x = element_blank(), 
                #panel.grid.major.x = element_blank(), 
                panel.grid.minor.y = element_blank(), 
                #panel.grid.major.y = element_blank(),
                #axis.line = element_line(colour = 'black')
                panel.grid.major = element_line(colour = "#4d4d4d"),
                axis.ticks = element_line(colour = "#4d4d4d")
                )
        
        #adding floating text annotations on the plot area
        label_LHS <- gsub(pattern = ", ", 
                          replacement = "\n", 
                          x = gsub(pattern = "Comparing ", 
                                   replacement = "", 
                                   x = unlist(strsplit(x = title, 
                                                       split = "  vs  "))[2], 
                                   ignore.case = T), 
                          ignore.case = T)
        label_LHS <- paste0("Expressed higher in,\n", label_LHS)
        label_RHS <- gsub(pattern = ", ", 
                          replacement = "\n", 
                          x = gsub(pattern = "Comparing ", 
                                   replacement = "", 
                                   x = unlist(strsplit(x = title, 
                                                       split = "  vs  "))[1], 
                                   ignore.case = T), 
                          ignore.case = T)
        label_RHS <- paste0("Expressed higher in,\n", label_RHS)
        
        scatter_plot <- scatter_plot + 
          annotate(geom = "text", 
                   x = -ceiling(max(abs(plot_data$Actuallog2FC))), 
                   y = ceiling(max(-log10(plot_data$padj))), 
                   label = label_LHS, 
                   size = 3, 
                   hjust = 0, 
                   vjust = 1) + 
          annotate(geom = "text", 
                   x = ceiling(max(abs(plot_data$Actuallog2FC))), 
                   y = ceiling(max(-log10(plot_data$padj))), 
                   label = label_RHS, 
                   size = 3, 
                   hjust = 1, 
                   vjust = 1)
        
          saved_plots[[i]][[j]][[length(saved_plots[[i]][[j]])+1]] <- 
          (scatter_plot)
        names(saved_plots[[i]][[j]])[length(saved_plots[[i]][[j]])] <- 
          names(merged_results_list[[i]][[j]])[k]
        
        
        #saving generated plot via ggsave
        ggsave(filename = paste0(names(merged_results_list[[i]][[j]])[k],
                                 ".png"),
               plot = scatter_plot,
               device = "png",
               path = path,
               #scale = 1,
               #width = 9.75,
               #height = 6.75,
               width = 6.5,
               height = 4.5,
               units = "in",
               #dpi = 500,
               limitsize = T)

        ggsave(filename = paste0(names(merged_results_list[[i]][[j]])[k],
                                 ".svg"),
               plot = scatter_plot,
               device = "svg",
               path = path,
               #scale = 1,
               #width = 9.75,
               #height = 6.75,
               width = 6.5,
               height = 4.5,
               units = "in",
               #dpi = 500,
               limitsize = T)
        
        # #saving generated plot via the png device
        # png(filename = paste0(path, 
        #                       names(merged_results_list[[i]][[j]])[k], 
        #                       ".png"), 
        #     width = 900, 
        #     height = 552, 
        #     units = "px", 
        #     bg = "transparent")
        # print(scatter_plot)
        # dev.off()
        
      }
      
    }
  }
  rm(i,j,k)
  
}


#print tree structure of the saved_plots
#xtree <- FromListSimple(saved_plots, nodeName = "Root")
#xtree

#creating a list of volcano plots via ggplot2 scatter plots and selecting genes
# of interest using euclidean distance
{
  saved_plots <- list()
  
  for(i in seq(1, length(merged_results_list))){
    #print(paste0("i: ", i))
    
    saved_plots[[length(saved_plots)+1]] <- list()
    names(saved_plots)[length(saved_plots)] <- names(merged_results_list)[i]
    
    writeLines(paste0("\nProcessing results from the model: ", 
                      names(merged_results_list)[i]))
    
    for(j in seq(1, length(merged_results_list[[i]]))){
      #print(paste0("j: ", j))
      
      saved_plots[[i]][[length(saved_plots[[i]])+1]] <- list()
      names(saved_plots[[i]])[length(saved_plots[[i]])] <- 
        names(merged_results_list[[i]])[j]
      
      #print(names(merged_results_list[[i]])[j])
      
      #constructing the path and creating directories
      path <- paste0(loc_plots, 
                     "Volcano Plots/", 
                     names(merged_results_list)[i], "/", 
                     names(merged_results_list[[i]])[j], "/")
      dir.create(path = path, recursive = T)
      
      writeLines(paste0("Exporting plots for: ", 
                        names(merged_results_list[[i]])[j]))
      
      for(k in seq(1, length(merged_results_list[[i]][[j]]))){
        #print(paste0("k: ", k))
        
        print(paste0("Exporting plot: ", 
                     names(merged_results_list[[i]][[j]])[k]))
        
        plot_data <- as.data.frame(
          merged_results_list[[i]][[j]][[k]])[
            complete.cases(merged_results_list[[i]][[j]][[k]]),]
        
        plot_data$Gene.stable.ID <- gsub(pattern = "\\.[0-9]+$", 
                                         replacement = "", 
                                         x = rownames(plot_data))
        
        plot_data <- merge(x = plot_data, y = output_table[,1:2], 
                           by = "Gene.stable.ID", 
                           all.x = T, all.y = F, 
                           sort = F)
        
        plot_data <- plot_data[order(plot_data$log2FoldChange, decreasing = T),]
        
        #calculating euclidean distance
        #coordinates <- cbind(plot_data$log2FoldChange, plot_data$padj)
        centroid <- c(mean(plot_data$log2FoldChange), 
                      mean(-log10(plot_data$padj)))
        #weighting adj.p.value at half due to its higher dynamic range in -log10
        weights <- c(1, 0.5)
        #subsetting relevant (adj.p < 0.05) data for coloured euclidean dist
        plot_data1 <- subset(plot_data, plot_data$padj <= 0.05)
        plot_data2 <- subset(plot_data, plot_data$padj > 0.05)
        coordinates <- cbind(plot_data1$log2FoldChange, -log10(plot_data1$padj))

        plot_data1$dist  <- foreach(l=1:nrow(coordinates),
                                   .combine = 'c',
                                   .inorder = TRUE
                                   ) %dopar% {
                                     sqrt(sum(((coordinates[l,]*weights) - centroid)^2))
                                   }
        
        #euclidean distance selection criteria
        plot_data1$coloured <- as.factor(ifelse(
          (plot_data1$dist > quantile(x = plot_data1$dist, 0.97)
           &
           # (-log10(plot_data$padj) > quantile(x = -log10(plot_data$padj),
           #                                      0.99) &
               plot_data1$padj < 0.05
           ) |
             #(plot_data$padj < 0.05)) |
            grepl(pattern = genes_of_int, x = plot_data1$MGI.symbol,
                  ignore.case = T),
          "selected", "not_selected"))
        
        #genes of interest selection criteria
        plot_data2$coloured <- as.factor(ifelse(grepl(pattern = genes_of_int, 
                                                      x = plot_data2$MGI.symbol,
                                                      ignore.case = T), 
                                                "selected", 
                                                "not_selected"))
        
        #ordering the points based on colour group so that the selected ones 
        #show on top
        plot_data1 <- plot_data1[order(plot_data1$coloured),]
        plot_data2 <- plot_data2[order(plot_data2$coloured),]
        
        #formatting plot title here instead of in the already bulky ggplot code
        {
          title <- names(merged_results_list[[i]][[j]])[k]
          title <- gsub(pattern = "blood", 
                        replacement = "Blood", 
                        x = title, ignore.case = T)
          title <- gsub(pattern = "spleen", 
                        replacement = "Spleen", 
                        x = title, ignore.case = T)
          title <- gsub(pattern = "lymphnode", 
                        replacement = "Lymph node", 
                        x = title, ignore.case = T)
          title <- gsub(pattern = "peyerspatches", 
                        replacement = "Peyer's patches", 
                        x = title, ignore.case = F)
          title <- gsub(pattern = "intraepitheliallayer", 
                        replacement = "Intraepithelial layer", 
                        x = title, ignore.case = F)
          title <- gsub(pattern = "laminaproprialymphocyte", 
                        replacement = "Lamina propria lymphocyte", 
                        x = title, ignore.case = F)
          title <- gsub(pattern = "kidney", 
                        replacement = "Kidney", 
                        x = title, ignore.case = F)
          title <- gsub(pattern = "lung", 
                        replacement = "Lung", 
                        x = title, ignore.case = F)
          title <- gsub(pattern = "liver", 
                        replacement = "Liver", 
                        x = title, ignore.case = F)
          title <- gsub(pattern = "pancreas", 
                        replacement = "Pancreas", 
                        x = title, ignore.case = F)
          
          title <- paste0("Comparing ", gsub(pattern = "_", 
                                             replacement = ", ", 
                                             x = gsub(pattern = "__Vs__", 
                                                      replacement = "  vs  ", 
                                                      x = title, 
                                                      ignore.case = T), 
                                             ignore.case = T))
        }
        
        
        scatter_plot <- ggplot(data = plot_data,
                               aes_(text = plot_data$MGI.symbol,
                                    log2FC = plot_data$log2FoldChange,
                                    sig = plot_data$padj)
                               ) + 
          geom_point(data = plot_data2, shape = 19, size = 0.8,
                     aes_(x = plot_data2$log2FoldChange,
                          y = -log10(plot_data2$padj),
                          colour = plot_data2$coloured,
                          text = plot_data2$MGI.symbol,
                          log2FC = plot_data2$log2FoldChange,
                          sig = plot_data2$padj
                          )
                     ) + 
          expand_limits(x = c(-30, 30), y = c(-10, 100)) + 
          scale_x_continuous(breaks = seq(-30,30,10)) + 
          scale_y_continuous(breaks = seq(-20,100,20)) + 
          scale_color_manual(labels = c("Other genes", "Genes of\ninterest"),
                         values = c("grey", "black")) + 
          labs(colour = "Adjusted\np.value > 0.05", 
           y = "-Log10 of Adjusted P.value", 
           x = "Log2 Fold Change", 
           title = title) + 
          #theme_light(base_size = 8) + 
          theme_light(base_size = 10) + 
          
          new_scale_color() + 
          
          geom_point(data = plot_data1, shape = 19, size = 0.8,
                     aes_(x = plot_data1$log2FoldChange,
                          y = -log10(plot_data1$padj),
                          colour = plot_data1$dist,
                          text = plot_data1$MGI.symbol,
                          log2FC = plot_data1$log2FoldChange,
                          sig = plot_data1$padj
                          )
                     ) + 
          labs(colour = "Adjusted\np.value \u2264 0.05;\nDistance from\ncentroid", 
           y = "-Log10 of Adjusted P.value", 
           x = "Log2 Fold Change", 
           title = title) + 
          scale_colour_viridis_c(option = "plasma", direction = -1) + 
          #theme_light(base_size = 8)
          theme_light(base_size = 10)
          
        #ggplotly(scatter_plot)
        
          saved_plots[[i]][[j]][[length(saved_plots[[i]][[j]])+1]] <- 
          (scatter_plot)
        names(saved_plots[[i]][[j]])[length(saved_plots[[i]][[j]])] <- 
          names(merged_results_list[[i]][[j]])[k]
        
        
        # #saving generated plot via ggsave
        # ggsave(filename = paste0(names(merged_results_list[[i]][[j]])[k],
        #                          ".png"),
        #        plot = scatter_plot,
        #        device = "png",
        #        path = path,
        #        #scale = 1,
        #        #width = 9.75,
        #        #height = 6.75,
        #        width = 6.5,
        #        height = 4.5,
        #        units = "in",
        #        #dpi = 500,
        #        limitsize = T)
        # 
        # ggsave(filename = paste0(names(merged_results_list[[i]][[j]])[k], 
        #                          ".svg"),
        #        plot = scatter_plot,
        #        device = "svg",
        #        path = path,
        #        #scale = 1,
        #        #width = 9.75,
        #        #height = 6.75,
        #        width = 6.5,
        #        height = 4.5,
        #        units = "in",
        #        #dpi = 500,
        #        limitsize = T)
        
        # #saving generated plot via the png device
        # png(filename = paste0(path, 
        #                       names(merged_results_list[[i]][[j]])[k], 
        #                       ".png"), 
        #     width = 900, 
        #     height = 552, 
        #     units = "px", 
        #     bg = "transparent")
        # print(scatter_plot)
        # dev.off()
        
      }
      
    }
  }
  rm(i,j,k)
  
}


#saving the plots to nested directories in the same structure as the list of 
#plots
{
  # for(i in seq(1, length(merged_results_list))){
  #   #print(paste0("i: ", i))
  #   
  #   for(j in seq(1, length(merged_results_list[[i]]))){
  #     #print(paste0("j: ", j))
  #     
  #     #constructing the path and creating directories
  #     path <- paste0(loc_plots, 
  #                    "Volcano Plots/", 
  #                    names(merged_results_list)[i], "/", 
  #                    names(merged_results_list[[i]])[j], "/")
  #     dir.create(path = path, recursive = T)
  #     
  #     for(k in seq(1, length(merged_results_list[[i]][[j]]))){
  #       #print(paste0("k: ", k))
  #       
  #       # ggsave(filename = paste0(names(saved_plots[[i]][[j]])[k], ".png"),
  #       #        plot = saved_plots[[i]][[j]][[k]],
  #       #        device = "png",
  #       #        path = path,
  #       #        #scale = 1,
  #       #        #width = 9.75,
  #       #        #height = 6.75,
  #       #        width = 6.5,
  #       #        height = 4.5,
  #       #        units = "in",
  #       #        #dpi = 500,
  #       #        limitsize = T)
  #       
  #       png(filename = paste0(names(saved_plots[[i]][[j]])[k], ".png"), 
  #           width = 900, 
  #           height = 552, 
  #           units = "px", 
  #           bg = "transparent")
  #       
  #       print(saved_plots[[i]][[j]][[k]])
  #       
  #       dev.off()
  #       
  #       # save_plot(filename = paste0(names(saved_plots[[i]][[j]])[k], ".png"),
  #       #       # plot = plot_grid(plotlist = saved_plots[[i]][[j]][[k]],
  #       #       #                  align = "none",
  #       #       #                  ncol = 1),
  #       #       plot = saved_plots[[i]][[j]][[k]],
  #       #       ncol = 1,
  #       #       nrow = 1,
  #       #       #base_height = 6.75,
  #       #       #base_width = 7.75,
  #       #       base_asp = 1.618,
  #       #       dpi = 600,
  #       #       device = "png",
  #       #       path = path,
  #       #       limitsize = T)
  #       
  #     }
  #   }
  # }
  # #rm(i,j,k)
  # 
}



#reading in the differential expression files from the webtool
{
  
  #calculating mean counts based on tissue and cell type
  {
    mean_norm_counts <- data.frame(matrix(nrow = nrow(norm_counts), ncol = 0))
    
    #iterate thorugh condition labels and calculate means for Tregs
    for(i in seq(1, length(condition_labels))){
      
      columns <- norm_counts[, grepl(pattern = paste0(
        condition_labels[i], "_[0-9]{1}$"), 
        x = colnames(norm_counts), 
        ignore.case = T), 
        drop = F]
      
      if(ncol(columns)>1) mean_norm_counts <- cbind(mean_norm_counts, 
                                                      rowMeans(
                                                        as.matrix(columns))
                                                      )
      else mean_norm_counts <- cbind(mean_norm_counts, columns)
      
      colnames(mean_norm_counts)[ncol(mean_norm_counts)] <-
          paste0(gsub(pattern = "_[0-9]{1}$", 
                      replacement = "", 
                      x = condition_labels[i], 
                      ignore.case = T))
    }
    #removing iteration related variables
    rm(i, columns)
    
    #formatting the row names to Gene.stable.ID for merging with other results
    mean_norm_counts$Gene.stable.ID <- gsub(pattern = "\\..*",
                                                replacement = "",
                                                x = rownames(norm_counts))
  }
  
  #renaming mean normalised counts columns
  {
    colnames(mean_norm_counts) <- gsub(pattern = "_NT", 
                                       replacement = "_CD4", 
                                       x = colnames(mean_norm_counts), 
                                       ignore.case = T)
    
    colnames(mean_norm_counts) <- gsub(pattern = "_T", 
                                       replacement = "_Tregs", 
                                       x = colnames(mean_norm_counts), 
                                       ignore.case = T)
    
    colnames(mean_norm_counts) <- gsub(pattern = "IntraEpithelialLayer", 
                                       replacement = "IEL", 
                                       x = colnames(mean_norm_counts), 
                                       ignore.case = T)
    
    colnames(mean_norm_counts) <- gsub(pattern = "LaminaPropriaLymphocyte", 
                                       replacement = "LPL", 
                                       x = colnames(mean_norm_counts), 
                                       ignore.case = T)
    
    colnames(mean_norm_counts) <- gsub(pattern = "LymphNode", 
                                       replacement = "LN", 
                                       x = colnames(mean_norm_counts), 
                                       ignore.case = T)
    
    colnames(mean_norm_counts) <- gsub(pattern = "PeyersPatches", 
                                       replacement = "PP", 
                                       x = colnames(mean_norm_counts), 
                                       ignore.case = T)
  }
  
  files <- list.files(path = loc_diffexprs_sngltrnscrpt, full.names = T)
  
  DiffExprsSnglTrnscrpt <- list()
  
  # foreach(i = 1:length(files)) %do% 
  #   DiffExprsSnglTrnscrpt[[i]] <- read.delim(file = files[i], sep = "\t", 
  #                                            stringsAsFactors = F)
  
  for(i in seq(1, length(files))){
    DiffExprsSnglTrnscrpt[[i]] <- read.delim(file = files[i], sep = "\t", 
                                             stringsAsFactors = F)
  }
  names(DiffExprsSnglTrnscrpt) <- gsub(pattern = "(^.*/|.tab$)", 
                                       replacement = "", 
                                       x = files, 
                                       ignore.case = T)
  
  mean_norm_counts <- subset(mean_norm_counts, 
                             mean_norm_counts$Gene.stable.ID %in% 
                               DiffExprsSnglTrnscrpt[[1]]$Gene.ID)
  
  
  for(i in seq(1, length(DiffExprsSnglTrnscrpt))){
    
    if(grepl(pattern = "Tregs", 
             x = names(DiffExprsSnglTrnscrpt)[i], 
             ignore.case = T)){
      subset_counts <- mean_norm_counts[, grepl(pattern = "Tregs", 
                                                x = colnames(mean_norm_counts), 
                                                ignore.case = T)]
      
      subset_counts <- subset_counts[, !grepl(
        pattern = gsub(pattern = "_.*$", 
                       replacement = "", 
                       x = names(DiffExprsSnglTrnscrpt)[i], 
                       ignore.case = T), 
        x = colnames(subset_counts), 
        ignore.case = T)]
      
      subset_counts$Gene.stable.ID <- mean_norm_counts$Gene.stable.ID
      
    }
    else{
      subset_counts <- mean_norm_counts
      
      subset_counts <- subset_counts[, !grepl(
        pattern = gsub(pattern = "_vs.*$", 
                       replacement = "", 
                       x = names(DiffExprsSnglTrnscrpt)[i], 
                       ignore.case = T), 
        x = colnames(subset_counts), 
        ignore.case = T)]
      
    }
    
    row_marker <- logical(length = nrow(subset_counts))
    
    for(j in seq(1, nrow(subset_counts))){
      if(all(subset_counts[j, -ncol(subset_counts)] < 10)){
        row_marker[j] <- TRUE
      }
    }
    
    subset_counts$row_marker <- row_marker
    subset_counts <- subset_counts[order(subset_counts$Gene.stable.ID, 
                                         decreasing = F),]
    
    DiffExprsSnglTrnscrpt[[i]] <- subset(
      x = DiffExprsSnglTrnscrpt[[i]], 
      #abs(DiffExprsSnglTrnscrpt[[i]]$log2.Fold.Change) > 5 & 
      DiffExprsSnglTrnscrpt[[i]]$Adjusted.p.value < 0.01 & 
        subset_counts$row_marker)
    
  }
  rm(i, j)
  
  
  #writing out files
  for(i in seq(1, length(DiffExprsSnglTrnscrpt))){
    if(nrow(DiffExprsSnglTrnscrpt[[i]]) > 0){
      write.table(x = DiffExprsSnglTrnscrpt[[i]], 
                  file = paste0(loc_tissue_specific_exprs, 
                                names(DiffExprsSnglTrnscrpt)[i]), 
                  sep = "\t", 
                  quote = F, 
                  row.names = F, 
                  col.names = T)
    }
  }
  
}



```


#pairwise DESeq2 analysis
```{r}

#reading in DESeq2 outputs
{
  #read normalised counts from the '#writing out the differential expression 
  # files' code block to get them filtered by min transcripts per cell
  #reading norm counts
  # norm_counts <- read.delim(file = paste0(loc_counts, "normalised_counts.txt"), 
  #                           sep = "\t", 
  #                           stringsAsFactors = F)
  #norm_counts <- norm_counts[rowSums(as.matrix(norm_counts))>0,]
  
  #reading the DESeq2 analysis output table
  merged_results_list <- readRDS(file = paste0(loc_diffexprs, 
                                        "merged_results_list.rds"))
  
  contrast_groups <- list(c("Blood"), 
                          c("Spleen", "Lymph Node", "Peyer's Patches"), 
                          c("Intraepithelial Layer", 
                            "Lamina Propria Lymphocyte"), 
                          c("Kidney", "Lung", "Liver", "Pancreas"),
                          c("cellTypeT"),
                          c("cellTypeNT"))
  
  #ensuring that names in contrast_groups list are all lower letter and do not 
  # contain spaces or punctuation
  for(i in seq(1, length(contrast_groups)))
  {
    contrast_groups[[i]] <- tolower(unique(
      gsub("_.*|[[:space:]]|[[:punct:]]", "", contrast_groups[[i]])
      ))
  }
  rm(i)
  names(contrast_groups) <- c("Blood", 
                              "Lymphoid", 
                              "Gut-Associated", 
                              "Non-Lymphoid", 
                              "Treg", 
                              "Tconv")
  
  # output_table <- read.delim(file = paste0(loc_diffexprs_merged, 
  #                                          "output_table_combined.txt"), 
  #                            sep = "\t", stringsAsFactors = F)
  output_table <- read.delim(file = paste0(loc_diffexprs_merged, 
                                           "output_table.txt"), 
                             sep = "\t", stringsAsFactors = F)
  
  colnames(output_table) <- output_table[3,]
  output_table <- output_table[-(1:3),]
  
  mean_norm_counts <- output_table[, c(1, 44:63)]
  rownames(mean_norm_counts) <- mean_norm_counts$`Gene stable ID`
  #mean_norm_counts$`Gene stable ID` <- output_table$`Gene stable ID`
  #colnames(temp) <- temp[3,]
  #temp <- temp[4:nrow(temp),]
  
}

#calculating pairwise contrasts of each treg tissue with other treg tissues
##selecting only those differentially expressed genes which are differentially 
## expressed across all pairwise contrasts for that treg tissue

#formatting counts matrix with updated/finalised names
{
  counts_matrix_renamed <- counts_matrix
  
  counts_matrix_renamed <- 
    counts_matrix_renamed[rowSums(counts_matrix_renamed) > 0 , ]
  
  colnames(counts_matrix_renamed) <- gsub(pattern = "_NT", 
                                          replacement = "_Tconv", 
                                          x = gsub(pattern = "_T", 
                                                   replacement = "_Treg", 
                                                   x = colnames(
                                                     counts_matrix_renamed), 
                                                   ignore.case = T), 
                                          ignore.case = T)
  
  colnames(counts_matrix_renamed) <- gsub(pattern = "IntraEpithelialLayer", 
                                          replacement = "IEL", 
                                          x = colnames(counts_matrix_renamed), 
                                          ignore.case = T)
  colnames(counts_matrix_renamed) <- gsub(pattern = "LaminaPropriaLymphocyte", 
                                          replacement = "LPL", 
                                          x = colnames(counts_matrix_renamed), 
                                          ignore.case = T)
  colnames(counts_matrix_renamed) <- gsub(pattern = "LymphNode", 
                                          replacement = "LN", 
                                          x = colnames(counts_matrix_renamed), 
                                          ignore.case = T)
  colnames(counts_matrix_renamed) <- gsub(pattern = "PeyersPatches", 
                                          replacement = "PP", 
                                          x = colnames(counts_matrix_renamed), 
                                          ignore.case = T)
  colnames(counts_matrix_renamed) <- gsub(pattern = "_",
                                          replacement = " ",
                                          x = colnames(counts_matrix_renamed),
                                          ignore.case = T)
}

#formatting annotations to match the new naming
{
  annotations_renamed <- as.data.frame(
    t(as.data.frame(strsplit(colnames(counts_matrix_renamed), " "))))
  
  rownames(annotations_renamed) <- colnames(counts_matrix_renamed)
  annotations_renamed <- annotations_renamed[,-3]
  colnames(annotations_renamed) <- c("tissue", "cellType")
  
  annotations_renamed$tissue <- as.factor(annotations_renamed$tissue)
  annotations_renamed$cellType <- as.factor(annotations_renamed$cellType)
}

DESeq_object <- DESeqDataSetFromMatrix(countData = counts_matrix_renamed,
                                         colData = annotations_renamed,
                                         design = ~0+tissue:cellType)
DESeq_object <- DESeq(object = DESeq_object, 
                        test = "Wald", 
                        betaPrior = F, 
                        minReplicatesForReplace = Inf, 
                        parallel = T)

result_names <- resultsNames(DESeq_object)
result_names_Treg <- result_names[grepl(pattern = "Treg", 
                                        x = result_names, 
                                        ignore.case = T)]
result_names_Tconv <- result_names[grepl(pattern = "Tconv", 
                                        x = result_names, 
                                        ignore.case = T)]

contrasts_Treg_self <- list()
for(contrast in result_names_Treg){

  contrasts_Treg_self[[length(contrasts_Treg_self)+1]] <- list()
  names(contrasts_Treg_self)[length(contrasts_Treg_self)] <- contrast
  
  for(contrasted_against in result_names_Treg[result_names_Treg != contrast]){
    
    #print(contrasted_against)
    contrasts_Treg_self[[length(contrasts_Treg_self)]][[
      length(contrasts_Treg_self[[length(contrasts_Treg_self)]])+1]] <- 
      as.data.frame(
        results(object = DESeq_object,
                contrast = list(contrast, contrasted_against),
                listValues = c(1, -1),
                cooksCutoff = F,
                independentFiltering = F,
                test = "Wald",
                parallel = T)
      )
    
    results <- contrasts_Treg_self[[length(contrasts_Treg_self)]][[
      length(contrasts_Treg_self[[length(contrasts_Treg_self)]])]]
    
    results <- subset(results, 
                      gsub(pattern = "\\.[0-9]+$", 
                           replacement = "", 
                           x = rownames(results), 
                           ignore.case = T) %in% 
                        mean_norm_counts$`Gene stable ID`)
    
    numerator <- contrast
    numerator <- gsub(pattern = "\\.cellType", 
                      replacement = " ", 
                      x = gsub(pattern = "tissue", 
                               replacement = "", 
                               x = numerator, 
                               ignore.case = T), 
                      ignore.case = T)
    
    denominator <- contrasted_against
    denominator <- gsub(pattern = "\\.cellType", 
                        replacement = " ", 
                        x = gsub(pattern = "tissue", 
                                 replacement = "", 
                                 x = denominator, 
                                 ignore.case = T), 
                        ignore.case = T)
    
    #mean_norm_counts are already in log2
      if(length(numerator) < 2){
        num_means <- 
          data.matrix(mean_norm_counts[, grepl(pattern = paste(numerator, 
                                                               collapse = "|"), 
                                               x = colnames(mean_norm_counts), 
                                               ignore.case = T)])
      }
      else{
        num_means <- 
          rowMeans(data.matrix(
            mean_norm_counts[, grepl(pattern = paste(numerator, 
                                                     collapse = "|"), 
                                     x = colnames(mean_norm_counts), 
                                     ignore.case = T)]))
      }
      if(length(denominator) < 2){
        denom_means <- 
          data.matrix(mean_norm_counts[, grepl(pattern = paste(denominator, 
                                                               collapse = "|"), 
                                               x = colnames(mean_norm_counts), 
                                               ignore.case = T)])
      }
      else{
        denom_means <- 
          rowMeans(data.matrix(
            mean_norm_counts[, grepl(pattern = paste(denominator, 
                                                     collapse = "|"), 
                                     x = colnames(mean_norm_counts), 
                                     ignore.case = T)]))
      }
      # Actuallog2FC <- 
      #   rowMeans(data.matrix(mean_norm_counts[, grepl(pattern = paste(numerator, 
      #                                                       collapse = "|"), 
      #                                       x = colnames(mean_norm_counts), 
      #                                       ignore.case = T)])) -  
      #   rowMeans(data.matrix(mean_norm_counts[, grepl(pattern = paste(denominator, 
      #                                                       collapse = "|"), 
      #                                       x = colnames(mean_norm_counts), 
      #                                       ignore.case = T)]))
      Actuallog2FC <- as.numeric(num_means) - as.numeric(denom_means)
      Actuallog2FC[!is.finite(Actuallog2FC)] <- NA
      names(Actuallog2FC) <- rownames(mean_norm_counts)
      Actuallog2FC <- Actuallog2FC[match(gsub(pattern = "\\.[0-9]+$", 
                                              replacement = "", 
                                              x = rownames(results), 
                                              ignore.case = T), 
                                         names(Actuallog2FC))]
      results$Actuallog2FC <- Actuallog2FC
    
    
    contrasts_Treg_self[[length(contrasts_Treg_self)]][[
      length(contrasts_Treg_self[[length(contrasts_Treg_self)]])]] <- results
    
    names(contrasts_Treg_self[[length(contrasts_Treg_self)]])[
      length(contrasts_Treg_self[[length(contrasts_Treg_self)]])] <- 
      contrasted_against
  }
  # names(contrasts_Treg_self)[length(contrasts_Treg_self)] <- contrast
}

contrasts_Tconv_self <- list()
for(contrast in result_names_Tconv){
  
  #print(contrast)
  contrasts_Tconv_self[[length(contrasts_Tconv_self)+1]] <- list()
  names(contrasts_Tconv_self)[length(contrasts_Tconv_self)] <- contrast
  
  for(contrasted_against in result_names_Tconv[result_names_Tconv != contrast]){
    
    #print(contrasted_against)
    contrasts_Tconv_self[[length(contrasts_Tconv_self)]][[
      length(contrasts_Tconv_self[[length(contrasts_Tconv_self)]])+1]] <- 
      as.data.frame(
        results(object = DESeq_object,
                contrast = list(contrast, contrasted_against),
                listValues = c(1, -1),
                cooksCutoff = F,
                independentFiltering = F,
                test = "Wald",
                parallel = T)
      )
    
    results <- contrasts_Tconv_self[[length(contrasts_Tconv_self)]][[
      length(contrasts_Tconv_self[[length(contrasts_Tconv_self)]])]]
    
    results <- subset(results, 
                      gsub(pattern = "\\.[0-9]+$", 
                           replacement = "", 
                           x = rownames(results), 
                           ignore.case = T) %in% 
                        mean_norm_counts$`Gene stable ID`)
    
    numerator <- contrast
    numerator <- gsub(pattern = "\\.cellType", 
                      replacement = " ", 
                      x = gsub(pattern = "tissue", 
                               replacement = "", 
                               x = numerator, 
                               ignore.case = T), 
                      ignore.case = T)
    
    denominator <- contrasted_against
    denominator <- gsub(pattern = "\\.cellType", 
                        replacement = " ", 
                        x = gsub(pattern = "tissue", 
                                 replacement = "", 
                                 x = denominator, 
                                 ignore.case = T), 
                        ignore.case = T)
    
    #mean_norm_counts are already in log2
      if(length(numerator) < 2){
        num_means <- 
          data.matrix(mean_norm_counts[, grepl(pattern = paste(numerator, 
                                                               collapse = "|"), 
                                               x = colnames(mean_norm_counts), 
                                               ignore.case = T)])
      }
      else{
        num_means <- 
          rowMeans(data.matrix(
            mean_norm_counts[, grepl(pattern = paste(numerator, 
                                                     collapse = "|"), 
                                     x = colnames(mean_norm_counts), 
                                     ignore.case = T)]))
      }
      if(length(denominator) < 2){
        denom_means <- 
          data.matrix(mean_norm_counts[, grepl(pattern = paste(denominator, 
                                                               collapse = "|"), 
                                               x = colnames(mean_norm_counts), 
                                               ignore.case = T)])
      }
      else{
        denom_means <- 
          rowMeans(data.matrix(
            mean_norm_counts[, grepl(pattern = paste(denominator, 
                                                     collapse = "|"), 
                                     x = colnames(mean_norm_counts), 
                                     ignore.case = T)]))
      }
      # Actuallog2FC <- 
      #   rowMeans(data.matrix(mean_norm_counts[, grepl(pattern = paste(numerator, 
      #                                                       collapse = "|"), 
      #                                       x = colnames(mean_norm_counts), 
      #                                       ignore.case = T)])) -  
      #   rowMeans(data.matrix(mean_norm_counts[, grepl(pattern = paste(denominator, 
      #                                                       collapse = "|"), 
      #                                       x = colnames(mean_norm_counts), 
      #                                       ignore.case = T)]))
      Actuallog2FC <- as.numeric(num_means) - as.numeric(denom_means)
      Actuallog2FC[!is.finite(Actuallog2FC)] <- NA
      names(Actuallog2FC) <- rownames(mean_norm_counts)
      Actuallog2FC <- Actuallog2FC[match(gsub(pattern = "\\.[0-9]+$", 
                                              replacement = "", 
                                              x = rownames(results), 
                                              ignore.case = T), 
                                         names(Actuallog2FC))]
      results$Actuallog2FC <- Actuallog2FC
    
    contrasts_Tconv_self[[length(contrasts_Tconv_self)]][[
      length(contrasts_Tconv_self[[length(contrasts_Tconv_self)]])]] <- results
      
    names(contrasts_Tconv_self[[length(contrasts_Tconv_self)]])[
      length(contrasts_Tconv_self[[length(contrasts_Tconv_self)]])] <- 
      contrasted_against
  }
  # names(contrasts_Tconv_self)[length(contrasts_Tconv_self)] <- contrast
}

#contrasting tissue groups vs remaining one-by-one (group-based equivalent of above two)
{
  # contrasts_TregGroups_paired <- list()
  # results_names_TregGroups <- list(
  #   result_names_Treg[grepl(pattern = "Blood", 
  #                           x = result_names_Treg, 
  #                           ignore.case = T)], 
  #   result_names_Treg[grepl(pattern = "(IEL|LPL)", 
  #                           x = result_names_Treg, 
  #                           ignore.case = T)], 
  #   result_names_Treg[grepl(pattern = "(LN|PP|Spleen)", 
  #                           x = result_names_Treg, 
  #                           ignore.case = T)], 
  #   result_names_Treg[grepl(pattern = "(Kidney|Liver|Lung|Pancreas)", 
  #                           x = result_names_Treg, 
  #                           ignore.case = T)]
  #   )
  # names(results_names_TregGroups) <- c("Blood", 
  #                                      "Gut-Associated", 
  #                                      "Lymphoid", 
  #                                      "Non-Lymphoid")
  # 
  # #for(contrast in results_names_TregGroups){
  # for(i in seq(1, length(results_names_TregGroups))){
  #   
  #   contrast <- results_names_TregGroups[[i]]
  #   contrasted_against <- result_names_Treg[!(result_names_Treg %in% contrast)]
  #   
  #   contrasts_TregGroups_paired[[length(contrasts_TregGroups_paired)+1]] <- 
  #     list()
  #   for(j in contrasted_against){
  #     
  #     contrasts_TregGroups_paired[[length(contrasts_TregGroups_paired)]][[
  #       length(contrasts_TregGroups_paired[[length(
  #         contrasts_TregGroups_paired)]])+1]] <- 
  #       as.data.frame(results(object = DESeq_object,
  #                             contrast = list(contrast, j),
  #                             listValues = c(1/length(contrast), -1),
  #                             cooksCutoff = F,
  #                             independentFiltering = F,
  #                             test = "Wald",
  #                             parallel = T)
  #                     )
  #     
  #     # contrasts_TregGroups_paired[[length(contrasts_TregGroups_paired)+1]][[
  #     #   length(contrasts_TregGroups_paired[[length(
  #     #     contrasts_TregGroups_paired)+1]])+1]] <- 
  #     #   as.data.frame(results(object = DESeq_object,
  #     #                         contrast = list(contrast, j),
  #     #                         listValues = c(1/length(contrast), -1),
  #     #                         cooksCutoff = F,
  #     #                         independentFiltering = F,
  #     #                         test = "Wald",
  #     #                         parallel = T)
  #     #                 )
  #     
  #     results <- contrasts_TregGroups_paired[[
  #       length(contrasts_TregGroups_paired)]][[length(
  #         contrasts_TregGroups_paired[[length(contrasts_TregGroups_paired)]])]]
  # 
  #     results <- subset(results,
  #                       gsub(pattern = "\\.[0-9]+$",
  #                            replacement = "",
  #                            x = rownames(results),
  #                            ignore.case = T) %in%
  #                         mean_norm_counts$`Gene stable ID`)
  #   
  #     numerator <- contrast
  #     numerator <- gsub(pattern = "\\.cellType",
  #                       replacement = " ",
  #                       x = gsub(pattern = "tissue",
  #                                replacement = "",
  #                                x = numerator,
  #                                ignore.case = T),
  #                       ignore.case = T)
  #   
  #     denominator <- j
  #     denominator <- gsub(pattern = "\\.cellType",
  #                         replacement = " ",
  #                         x = gsub(pattern = "tissue",
  #                                  replacement = "",
  #                                  x = denominator,
  #                                  ignore.case = T),
  #                         ignore.case = T)
  #   
  # #mean_norm_counts are already in log2
  #     if(length(numerator) < 2){
  #       num_means <- 
  #         data.matrix(mean_norm_counts[, grepl(pattern = paste(numerator, 
  #                                                              collapse = "|"), 
  #                                              x = colnames(mean_norm_counts), 
  #                                              ignore.case = T)])
  #     }
  #     else{
  #       num_means <- 
  #         rowMeans(data.matrix(
  #           mean_norm_counts[, grepl(pattern = paste(numerator, 
  #                                                    collapse = "|"), 
  #                                    x = colnames(mean_norm_counts), 
  #                                    ignore.case = T)]))
  #     }
  #     if(length(denominator) < 2){
  #       denom_means <- 
  #         data.matrix(mean_norm_counts[, grepl(pattern = paste(denominator, 
  #                                                              collapse = "|"), 
  #                                              x = colnames(mean_norm_counts), 
  #                                              ignore.case = T)])
  #     }
  #     else{
  #       denom_means <- 
  #         rowMeans(data.matrix(
  #           mean_norm_counts[, grepl(pattern = paste(denominator, 
  #                                                    collapse = "|"), 
  #                                    x = colnames(mean_norm_counts), 
  #                                    ignore.case = T)]))
  #     }
  #     # Actuallog2FC <- 
  #     #   rowMeans(data.matrix(mean_norm_counts[, grepl(pattern = paste(numerator, 
  #     #                                                       collapse = "|"), 
  #     #                                       x = colnames(mean_norm_counts), 
  #     #                                       ignore.case = T)])) -  
  #     #   rowMeans(data.matrix(mean_norm_counts[, grepl(pattern = paste(denominator, 
  #     #                                                       collapse = "|"), 
  #     #                                       x = colnames(mean_norm_counts), 
  #     #                                       ignore.case = T)]))
# Actuallog2FC <- as.numeric(num_means) - as.numeric(denom_means)
#       Actuallog2FC[!is.finite(Actuallog2FC)] <- NA
#       names(Actuallog2FC) <- rownames(mean_norm_counts)
  #     Actuallog2FC <- Actuallog2FC[match(gsub(pattern = "\\.[0-9]+$", 
  #                                             replacement = "", 
  #                                             x = rownames(results), 
  #                                             ignore.case = T), 
  #                                        names(Actuallog2FC))]
  #     results$Actuallog2FC <- Actuallog2FC
  #   
  #     contrasts_TregGroups_paired[[length(contrasts_TregGroups_paired)]][[
  #       length(contrasts_TregGroups_paired[[
  #         length(contrasts_TregGroups_paired)]])]] <- results
  #   
  #     names(contrasts_TregGroups_paired[[length(
  #       contrasts_TregGroups_paired)]])[length(
  #         contrasts_TregGroups_paired[[length(
  #           contrasts_TregGroups_paired)]])] <- j
  #   }
  #   
  #   names(contrasts_TregGroups_paired)[length(contrasts_TregGroups_paired)] <- 
  #     names(results_names_TregGroups)[i]
  #   
  # }
}

#plotting contrasts_Treg_self and contrasts_TregGroups_paired up and down reg
{
  # appended_list <- append(contrasts_Treg_self, contrasts_TregGroups_paired[-1])
  # 
  # plot_data <- data.frame(Tissues = gsub(pattern = "\\.cellType",
  #                                        replacement = " ",
  #                                        x = gsub(pattern = "tissue",
  #                                                 replacement = "",
  #                                                 x =
  #                                                   names(appended_list),
  #                                                 ignore.case = T),
  #                                        ignore.case = T),
  #                         Higher = NA, 
  #                         Lower = NA, 
  #                         stringsAsFactors = F)
  # 
  # for(i in seq(1, length(appended_list))){
  #   
  #   print(i)
  #   
  #   significs_up <-
  #     rownames(subset(appended_list[[i]][[1]],
  #                     #appended_list[[i]][[1]]$log2FoldChange >= 1 &
  #                     appended_list[[i]][[1]]$log2FoldChange >= 2 &
  #                       appended_list[[i]][[1]]$padj <= 0.01 &
  #                       is.finite(appended_list[[i]][[1]]$Actuallog2FC))
  #              )
  #   
  #   significs_down <-
  #     rownames(subset(appended_list[[i]][[1]],
  #                     #appended_list[[i]][[1]]$log2FoldChange <= -1 &
  #                     appended_list[[i]][[1]]$log2FoldChange <= -2 &
  #                       appended_list[[i]][[1]]$padj <= 0.01 &
  #                       is.finite(appended_list[[i]][[1]]$Actuallog2FC))
  #              )
  # 
  #   for(j in seq(2, length(appended_list[[i]]))){
  #     
  #     #print(j)
  #     
  #     significs_up <-
  #       intersect(significs_up,
  #                 rownames(
  #                   subset(appended_list[[i]][[j]],
  #                          #appended_list[[i]][[j]]$log2FoldChange >= 1 &
  #                          appended_list[[i]][[j]]$log2FoldChange >= 2 &
  #                            appended_list[[i]][[j]]$padj <= 0.01 &
  #                            is.finite(appended_list[[i]][[j]]$Actuallog2FC))
  #              ))
  #     significs_down <-
  #       intersect(significs_down,
  #                 rownames(
  #                   subset(appended_list[[i]][[j]],
  #                          #appended_list[[i]][[j]]$log2FoldChange <= -1 &
  #                          appended_list[[i]][[j]]$log2FoldChange <= -2 &
  #                            appended_list[[i]][[j]]$padj <= 0.01 &
  #                            is.finite(appended_list[[i]][[j]]$Actuallog2FC))
  #              ))
  #   }
  # 
  #   plot_data$Higher[i] <- length(significs_up)
  #   plot_data$Lower[i] <- length(significs_down)
  # 
  # }
  # 
  # #plot_data$Tissues <- as.factor(plot_data$Tissues)
  # #plot_data[plot_data$Counts == 0, 2] <- NA
  # plot_data <- melt(plot_data)
  # #plot_data$Tissues <- as.factor(plot_data$Tissues)
  # colnames(plot_data) <- c("Tissues", "Expression", "Counts")
  # 
  # #removing any bars for zero counts
  # #plot_data$Counts[plot_data$Counts == 0] <- -1
  # 
  # plot_data$Tissues <- factor(x = plot_data$Tissues, 
  #                            levels = (plot_data$Tissues)[
  #                              match(unique(c(
  #                                as.character(
  #                                  plot_data$Tissues[grepl(pattern = "Treg$",
  #                                                         x = plot_data$Tissues,
  #                                                         ignore.case = T)]), 
  #                                as.character(
  #                                  plot_data$Tissues[!grepl(pattern = "Treg$",
  #                                                         x = plot_data$Tissues,
  #                                                         ignore.case = T)])
  #                                )), 
  #                                plot_data$Tissues)
  #                              ]
  #                            )
  # 
  # plot_contrasts_Treg_self <- ggplot(data = plot_data,
  #                                    aes(x = Tissues,
  #                                        y = Counts,
  #                                        #y = value,
  #                                        #colour = Tissues,
  #                                        #colour = Expression,
  #                                        #fill = Tissues)
  #                                        fill = Expression)
  #                                    ) +
  #   geom_bar(stat="identity", position = "stack") +
  #   # geom_text(aes(label = ifelse(value != 0, value, "")), 
  #   #           vjust=-0.3, size=7.5) +
  #   #ylim(0, 30) +
  #   labs(x = "", 
  #        y = "Number of\nSignificant Genes") +
  #   theme_light(base_size = 14.5) +
  #   theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, 
  #                                    size = 11),
  #         panel.grid.minor = element_blank(),
  #         panel.grid.major = element_blank(),
  #         #legend.position = "none"
  #         legend.position = "top"
  #         )
  # 
  # plot_contrasts_Treg_self
  # 
  # ggsave(filename = "plot_contrasts_Tregs.png",
  #        plot = plot_contrasts_Treg_self,
  #        device = "png",
  #        width = 15,
  #        height = 12,
  #        units = "cm",
  #        dpi = 300,
  #        limitsize = T)
  # 
  # ggsave(filename = "plot_contrasts_Tregs.svg",
  #        plot = plot_contrasts_Treg_self,
  #        device = "svg",
  #        width = 14,
  #        height = 10,
  #        units = "cm",
  #        dpi = 300,
  #        limitsize = T)
  # 
  # ggsave(filename = "plot_contrasts_Tregs.pdf",
  #        plot = plot_contrasts_Treg_self,
  #        device = "pdf",
  #        width = 14,
  #        height = 10,
  #        units = "cm",
  #        dpi = 300,
  #        limitsize = T)
  # 
}

#step-wise loss of sig.diff.exprs.genes plot
{
  plot_data_list <- list()
  
  #making the lines for each tissue
  for(i in seq(1, length(contrasts_Treg_self))){
    
    #specifying the data
    selected_set <- NULL     #the largest set which is from the most different tissue
    ordered_max <- NULL      #the order in which sig.diff.exprs.genes reduces
    ordered_index <- NULL    #the order of indices in terms of the intersection
    index <- NULL            #the index curently under consideration
    ordered_labels <- NULL   #column names from available combinations
    
    while(length(ordered_index) < length(contrasts_Treg_self[[i]])){
      
      #resetting variables
      significs <- NULL        #the initial set of sig.diff.exprs.genes
      max <- -1                #guaranteeing that at least one tissue is chosen
      index <- NULL            #the index curently under consideration
      col_count <- 0           #counter for counting columns of the combinations
      
      #getting available leave-1-out combinations
      if(is.null(ordered_index)){
        # available_combn <- combn(x = names(contrasts_Treg_self[[i]]), 
        #                          m = length(contrasts_Treg_self[[i]])-1)
        available_combn <- combn(x = seq(1, length(contrasts_Treg_self[[i]])), 
                                 m = length(contrasts_Treg_self[[i]])-1)
      }
      else{
        available_combn <- 
          combn(x = seq(1, length(contrasts_Treg_self[[i]]))[-ordered_index], 
                m = length(contrasts_Treg_self[[i]])-1-length(ordered_index))
      }
      #formatting the names of missing contrast per column of available_combn
      col_names <- NULL
      for(col in seq(1, ncol(available_combn))){
        
        if(is.null(ordered_index)){
          col_names <- 
            c(col_names, 
              names(contrasts_Treg_self[[i]])[-(available_combn[,col])])
        }
        else{
          col_names <- 
            c(col_names, 
              names(contrasts_Treg_self[[i]])[-c(available_combn[,col],
                                                 ordered_index)])
        }
      }
      colnames(available_combn) <- col_names
      
      
      #if there are no available combination (at the end of step-wise 
      # leave-1-out) then add in the last tissue manually and exit the while 
      # loop
      if(nrow(available_combn) < 1){
        # index <- seq(1, length(contrasts_Treg_self[[i]]))[-ordered_index]
        # # ordered_index <- c(ordered_index, 
        # #                    seq(1, length(
        # #                      contrasts_Treg_self[[i]]))[-ordered_index])
        # ordered_index <- c(ordered_index, index)
        # ordered_labels <- c(ordered_labels, colnames(available_combn))
        # 
        # significs <- rownames(
        #   subset(contrasts_Treg_self[[i]][[index]],
        #          abs(contrasts_Treg_self[[i]][[index]]$log2FoldChange) >= 2 &
        #            contrasts_Treg_self[[i]][[index]]$padj <= 0.01 &
        #            is.finite(contrasts_Treg_self[[i]][[index]]$Actuallog2FC)
        #          ))
        # max <- length(significs)
        # ordered_max <- c(ordered_max, max)
        
        break
      }
      
      #otherwise, continue below
      #calculating and retaining the intersection with the largest sig.diff.exprs.genes
      #for(j in seq(1, ncol(available_combn))){
      for(j in (available_combn[1,])){
        
        col_count <- col_count + 1
        #print(col_count)
        #first intersection
        significs <- rownames(
            subset(contrasts_Treg_self[[i]][[j]],
               abs(contrasts_Treg_self[[i]][[j]]$log2FoldChange) >= 2 &
                 contrasts_Treg_self[[i]][[j]]$padj <= 0.01 &
                 is.finite(contrasts_Treg_self[[i]][[j]]$Actuallog2FC)
               )
            )
        
        for(k in (available_combn[-1,col_count])){
          
          significs <- intersect(significs, rownames(
            subset(contrasts_Treg_self[[i]][[k]],
                   abs(contrasts_Treg_self[[i]][[k]]$log2FoldChange) >= 2 &
                     contrasts_Treg_self[[i]][[k]]$padj <= 0.01 &
                     is.finite(contrasts_Treg_self[[i]][[k]]$Actuallog2FC))
            ))
        }
        
        if(length(significs) > max){
          max <- length(significs)
          under_consid <- significs
          index <- col_count
          }
      }
      
      ordered_max <- c(ordered_max, max)
      ordered_labels <- c(ordered_labels, colnames(available_combn)[index])
      
      #getting the true contrast index from data list
      if(is.null(ordered_index)){
        index <- 
          seq(1, 
              length(contrasts_Treg_self[[i]]))[-c(available_combn[,index])]
      }
      else{
        index <- 
          seq(1, 
              length(contrasts_Treg_self[[i]]))[-c(available_combn[,index], 
                                                   ordered_index)]
      }
      ordered_index <- c(ordered_index, index)
      
    }
    
    plot_data_entry <- ordered_max
    names(plot_data_entry) <- names(contrasts_Treg_self[[i]])[ordered_index]
    
    plot_data_list[[length(plot_data_list)+1]] <- plot_data_entry
    names(plot_data_list)[length(plot_data_list)] <- 
      names(contrasts_Treg_self)[i]

  }
  
  plot_data <- data.frame(Tissue = gsub(pattern = "\\.cellType",
                                       replacement = " ",
                                       x = gsub(pattern = "tissue",
                                                replacement = "",
                                                x =
                                                  names(plot_data_list)[1],
                                                ignore.case = T),
                                       ignore.case = T),
                          Step = seq(1, length(plot_data_list[[1]])), 
                          Excluded = gsub(pattern = "\\.cellType.*$", 
                                          replacement = "", 
                                          x = gsub(pattern = "tissue", 
                                                   replacement = "", 
                                                   x = names(plot_data_list[[1]]), 
                                                   ignore.case = T
                                                   )
                                          ), 
                          Counts = plot_data_list[[1]], 
                          stringsAsFactors = F)
  
  #populating the plot_data dataframe
  for(l in seq(2, length(plot_data_list))){
    plot_data <- rbind(plot_data, 
                       data.frame(Tissue = gsub(pattern = "\\.cellType",
                                       replacement = " ",
                                       x = gsub(pattern = "tissue",
                                                replacement = "",
                                                x =
                                                  names(plot_data_list)[l],
                                                ignore.case = T),
                                       ignore.case = T),
                          Step = seq(1, length(plot_data_list[[l]])), 
                          Excluded = gsub(pattern = "\\.cellType.*$", 
                                          replacement = "", 
                                          x = gsub(pattern = "tissue", 
                                                   replacement = "", 
                                                   x = names(plot_data_list[[l]]), 
                                                   ignore.case = T
                                                   )
                                          ),
                          Counts = plot_data_list[[l]], 
                          stringsAsFactors = F))
  }
  
  plot_data$Tissue <- as.factor(plot_data$Tissue)
  #plot_data[plot_data$Counts == 0, 2] <- NA

  plot_stepwise_Treg <- ggplot(data = plot_data,
                               aes(x = Step,
                                   y = Counts,
                                   colour = Tissue#,
                                   #fill = Tissue
                                   )
                               ) + 
    geom_line(size = 3, alpha = 0.35) + 
    geom_point(size = 4) + 
    ylim(0, 1800) + 
    geom_text(aes(x = Step, 
                  y = 1800, 
                  label = paste0(Excluded, ": ", Counts), 
                  group = Tissue), 
              size = 2.8, 
              vjust = sort((rep(seq(1, 20, 2), 8)))#, hjust = 0.5
              ) + 
    labs(x = "Stepwise Leave-1-out Tissue Removal",
         y = "Number of Significant Genes") +
    theme_light(base_size = 18) +
    theme(#axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
          panel.grid.minor = element_blank(),
          panel.grid.major = element_blank()#,
          #legend.position = "none"
          )
  
  plot_stepwise_Treg

  ggsave(filename = "plot_stepwise_Treg.png",
         plot = plot_stepwise_Treg,
         device = "png",
         width = 23.774,
         height = 14.605,
         units = "cm",
         dpi = 300,
         limitsize = T)

  ggsave(filename = "plot_stepwise_Treg.svg",
         plot = plot_stepwise_Treg,
         device = "svg",
         width = 23.774,
         height = 14.605,
         units = "cm",
         dpi = 300,
         limitsize = T)

  ggsave(filename = "plot_stepwise_Treg.pdf",
         plot = plot_stepwise_Treg,
         device = "pdf",
         width = 23.774,
         height = 14.605,
         units = "cm",
         dpi = 300,
         limitsize = T)
  
  
  plots_list <- list()
  for(i in unique(as.character(plot_data$Tissue))){
    plot_subdata <- plot_data[plot_data$Tissue == i,]
    
    plots_list[[length(plots_list)+1]] <- ggplot(data = plot_subdata, 
                                                 aes(x = Step, 
                                                     y = Counts, 
                                                     colour = Tissue#, 
                                                     #fill = Tissue 
                                                     ) 
                                                 ) +  
      geom_line(size = 2, alpha = 0.35) + 
      geom_point(size = 2) + 
      #ylim(0, 1800) + 
      geom_text(aes(x = Step, 
                    y = Counts+50, 
                    label = paste0(Excluded, ": ", Counts), 
                    group = Tissue), 
                size = 2#, 
                #vjust = sort((rep(seq(1, 20, 2), 8)))#, hjust = 0.5
                ) + 
      labs(x = "Stepwise Leave-1-out Tissue Removal", 
           y = "Number of Significant Genes", 
           title = i) + 
      theme_light(base_size = 8) + 
      theme(#axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
        panel.grid.minor = element_blank(), 
        panel.grid.major = element_blank(), 
        legend.position = "none" 
        )
  }
  
  do.call("grid.arrange", c(plots_list, ncol = floor(sqrt(length(plots_list)))))
  
}

#calculating correlation heatmap for mean log2 expression of tissues
{
  mean_norm_exp_tissues <- mean_norm_counts
  
  #getting a list of genes for which we have all data
  {
    norm_exp_log2 <- data.matrix(log2(norm_counts))
    norm_exp_log2[!is.finite(norm_exp_log2)] <- NA
    norm_exp_log2 <- 
      as.data.frame(norm_exp_log2[complete.cases(norm_exp_log2),])
  }
  
  #filtering mean_norm_exp_tissues based on the norm_exp_log2 genes
  mean_norm_exp_tissues <- subset(mean_norm_exp_tissues, 
                                  rownames(mean_norm_exp_tissues) %in% 
                                    gsub(pattern = "\\.[0-9]+$", 
                                         replacement = "", 
                                         x = rownames(norm_exp_log2), 
                                         ignore.case = T)
                                  )
  mean_norm_exp_tissues <- as.data.frame(mean_norm_exp_tissues)
  
  mean_norm_exp_tissues$`Gene stable ID` <- NULL
  mean_norm_exp_tissues <- data.matrix(mean_norm_exp_tissues)
  
  # mean_norm_exp_Treg <- mean_norm_counts[,grepl(pattern = "Treg", 
  #                                               x = colnames(mean_norm_counts), 
  #                                               ignore.case = T)]
  # mean_norm_exp_Treg$`Gene stable ID` <- NULL
  # mean_norm_exp_Treg <- data.matrix(mean_norm_exp_Treg)

  tissue_groups <- c("^(LN|PP|Spleen) Treg$",
                     "^(IEL|LPL) Treg$",
                     "^(Kidney|Liver|Lung|Pancreas) Treg$",
                     "^(LN|PP|Spleen) Tconv$",
                     "^(IEL|LPL) Tconv$",
                     "^(Kidney|Liver|Lung|Pancreas) Tconv$"
                     )
  names(tissue_groups) <- apply(expand.grid(c("Lymphoid",
                                              "Gut-Associated",
                                              "Non-lymphoid"),
                                            c("Treg",
                                              "Tconv")),
                                            #c("Treg")),
                                1, paste, collapse=" ")

  for(i in seq(1, length(tissue_groups))){

    mean_norm_exp_tissues <-
      cbind(mean_norm_exp_tissues,
            rowMeans(x = data.matrix(
              mean_norm_exp_tissues[, grepl(pattern = tissue_groups[i],
                                            x = colnames(mean_norm_exp_tissues),
                                            ignore.case = T)]),
              na.rm = T)
            )
    colnames(mean_norm_exp_tissues)[ncol(mean_norm_exp_tissues)] <-
      names(tissue_groups)[i]
  }

  corr_mat_pearson <- 
    cor(mean_norm_exp_tissues[,grepl(pattern = "Treg", 
                                     x = colnames(mean_norm_exp_tissues), 
                                     ignore.case = T)],
                          use = "complete.obs",
                          method = "pearson")
  # corr_mat_spearman <- cor(mean_norm_exp_Treg,
  #                         use = "complete.obs",
  #                         method = "spearman")
  # corr_mat_kendall <- cor(mean_norm_exp_Treg,
  #                         use = "complete.obs",
  #                         method = "kendall")

  #constructing rowside and colside colours to identify TregvTconv and Tissue groups
  # colsidecol <- colnames(corr_mat_pearson)
  # colsidecol[grepl(pattern = "Treg",
  #                  x = colsidecol,
  #                  ignore.case = T)] <- "Treg" #"#E66100"
  # colsidecol[grepl(pattern = "Tconv",
  #                  x = colsidecol,
  #                  ignore.case = T)] <- "Tconv" #"#5D3A9B"
  # names(colsidecol) <- colsidecol
  # colsidecol_colours <- unique(colsidecol)
  # names(colsidecol_colours) <- colsidecol_colours
  # colsidecol_colours[1:length(colsidecol_colours)] <- c("#5D3A9B",
  #                                                       "#E66100")
  rowsidecol <- colnames(corr_mat_pearson)
  rowsidecol[grepl(pattern = "Blood",
                   x = rowsidecol,
                   ignore.case = T)] <- "Blood" #"#D81B60"
  rowsidecol[grepl(pattern = "(Spleen|PP|LN|^Lymphoid)",
                   x = rowsidecol,
                   ignore.case = T)] <- "Lymphoid" #"#FFC107"
  rowsidecol[grepl(pattern = "(IEL|LPL|Gut)",
                   x = rowsidecol,
                   ignore.case = T)] <- "Gut-Associated" #"#1E88E5"
  rowsidecol[grepl(pattern = "(Kidney|Lung|Liver|Pancreas|Non)",
                   x = rowsidecol,
                   ignore.case = T)] <- "Non-lymphoid" #"#004D40"
  names(rowsidecol) <- rowsidecol
  rowsidecol_colours <- unique(rowsidecol)
  names(rowsidecol_colours) <- rowsidecol_colours
  rowsidecol_colours[1:length(rowsidecol_colours)] <- c("#D81B60",
                                                        "#1E88E5",
                                                        "#004D40",
                                                        "#FFC107")

  # heatmap_annot <- data.frame(cbind(rowsidecol, colsidecol),
  #                             stringsAsFactors = F)
  heatmap_annot <- data.frame(rowsidecol, stringsAsFactors = F)
  #colnames(heatmap_annot) <- c("Tissue", "Cell Type")
  colnames(heatmap_annot) <- c("Tissue")
  # heatmap_annot_colours <- list(Tissue = rowsidecol_colours,
  #                               `Cell Type` = colsidecol_colours)
  heatmap_annot_colours <- list(Tissue = rowsidecol_colours)

  #col<- colorRampPalette(c("#0064cd", "white", "#ff3232"))(100)
  #col<- colorRampPalette(c("white", "#ff3232"))(100)
  
  aheatmap(x = corr_mat_pearson,
           cellwidth = 11,
           cellheight = 11,
           #color = rev(heat.colors(12)), 
           color = brewer.pal(n = 12, name = "YlOrRd"), 
           #Colv = Rowv, 
           treeheight = 30, 
           #revC = T, 
           #color = rev(hcl.colors(100)),
           #breaks = seq(-1, 1, length.out = 100),
           annRow = heatmap_annot, 
           annCol = heatmap_annot, 
           annColors = heatmap_annot_colours, 
           fontsize = 9,
           #cexRow = 1,
           #cexCol = 0.8,
           main = "Treg Pearson Correlation",
           filename = "Pearson Correlation.pdf")

  # aheatmap(x = corr_mat_spearman, 
  #          cellwidth = 11,
  #          cellheight = 11,
  #          color = rev(heat.colors(12)),
  #          Colv = F,
  #          treeheight = 30,
  #          #color = rev(hcl.colors(100)),
  #          #breaks = seq(-1, 1, length.out = 100),
  #          annRow = heatmap_annot,
  #          annColors = heatmap_annot_colours,
  #          fontsize = 9,
  #          #cexRow = 1,
  #          #cexCol = 0.8,
  #          main = "Spearman Correlation",
  #          filename = "Spearman Correlation.pdf")
  # 
  # aheatmap(x = corr_mat_kendall,
  #          cellwidth = 11,
  #          cellheight = 11,
  #          color = rev(heat.colors(12)),
  #          Colv = F,
  #          treeheight = 30,
  #          #color = rev(hcl.colors(100)),
  #          #breaks = seq(-1, 1, length.out = 100),
  #          annRow = heatmap_annot,
  #          annColors = heatmap_annot_colours,
  #          fontsize = 9,
  #          #cexRow = 1,
  #          #cexCol = 0.8,
  #          main = "Kendall Correlation",
  #          filename = "Kendall Correlation.pdf")
  
  
  #making a heatmap for Tconv pearson by subtracting it from Treg pearson
  corr_mat_pearson_Tconv <- 
    cor(mean_norm_exp_tissues[,grepl(pattern = "Tconv", 
                                     x = colnames(mean_norm_exp_tissues), 
                                     ignore.case = T)],
                          use = "complete.obs",
                          method = "pearson")
  corr_mat_pearson_Tconv <- abs(corr_mat_pearson - corr_mat_pearson_Tconv)
  colnames(corr_mat_pearson_Tconv) <- gsub(pattern = "Treg", 
                                           replacement = "Tconv", 
                                           x = colnames(corr_mat_pearson_Tconv), 
                                           ignore.case = T)
  rownames(corr_mat_pearson_Tconv) <- gsub(pattern = "Treg", 
                                           replacement = "Tconv", 
                                           x = rownames(corr_mat_pearson_Tconv), 
                                           ignore.case = T)
  
  #reordering rows to match the pearson above ##doublecheck when the above corr
  # changes
  # curr order: LN, Lung, Blood, Lymphoid, Spleen, Liver, LPL, IEL, Gut-associated, 
  #              Kidney, Pancreas, PP, Non-lymphoid
  corr_mat_pearson_Tconv <- 
    corr_mat_pearson_Tconv[c(7, 6, 1, 11, 10, 5, 4, 2, 12, 3, 8, 9, 13), 
                           c(7, 6, 1, 11, 10, 5, 4, 2, 12, 3, 8, 9, 13)]
  
  aheatmap(x = corr_mat_pearson_Tconv,
           cellwidth = 11,
           cellheight = 11,
           #color = rev(heat.colors(12)), 
           #color = brewer.pal(n = 2, name = "Blues"), 
           color = brewer.pal(n = 2, name = "Greys"), 
           Rowv = NA, 
           Colv = NA, 
           #treeheight = 30, 
           revC = T, 
           #color = rev(hcl.colors(100)),
           #breaks = seq(-1, 1, length.out = 100),
           #annRow = heatmap_annot, 
           #annCol = heatmap_annot, 
           #annColors = heatmap_annot_colours, 
           fontsize = 9,
           #cexRow = 1,
           #cexCol = 0.8,
           main = "Difference Pearson Correlation",
           filename = "Pearson Correlation Difference.pdf")
}

#alternative implementation
#step-wise loss of sig.diff.exprs.genes plot 
# gives almost the same as the brute force implementation above
{
  # plot_data_list <- list()
  # 
  # #making the lines for each tissue
  # for(i in seq(1, length(contrasts_Treg_self))){
  #   
  #   #specifying the data
  #   significs <- NULL        #the initial set of sig.diff.exprs.genes
  #   selected_set <- NULL     #the largest set which is from the most different tissue
  #   max <- -1                #guaranteeing that at least one tissue is chosen
  #   ordered_max <- NULL      #the order in which sig.diff.exprs.genes reduces
  #   ordered_index <- NULL    #the order of indices in terms of the intersection
  #   index <- NULL            #the index curently under consideration
  #   
  #   #for starting from the smallest
  #   min <- Inf
  #   
  #   #select initial tissue which is the most distinct from the contrasted tissue
  #   for(j in seq(1, length(contrasts_Treg_self[[i]]))){
  #     
  #     significs <- rownames(
  #       subset(contrasts_Treg_self[[i]][[j]],
  #              abs(contrasts_Treg_self[[i]][[j]]$log2FoldChange) >= 2 &
  #                contrasts_Treg_self[[i]][[j]]$padj <= 0.01 &
  #                is.finite(contrasts_Treg_self[[i]][[j]]$Actuallog2FC)
  #              )
  #       )
  #     
  #     #checks if a more distinct tissue contrast has been found or not
  #     if(length(significs) > max){
  #       max <- length(significs)
  #       ordered_max <- max
  #       selected_set <- significs
  #       #index <- j
  #       ordered_index <- j
  #     
  #     # if(length(significs) < min){
  #     #   min <- length(significs)
  #     #   ordered_max <- min
  #     #   selected_set <- significs
  #     #   #index <- j
  #     #   ordered_index <- j
  #     }
  #   }
  #   
  #   #while there are still tissues left to be ordered in a step-wise manner
  #   while(length(ordered_index) < length(contrasts_Treg_self[[i]])){
  #     
  #     #resetting the variables to guarantee at least one selection
  #     significs <- NULL
  #     under_consid <- NULL
  #     max <- -1
  #     index <- NULL
  #     
  #     #for calculating the most lost per step
  #     min <- Inf
  #     
  #     #checking each of the remaining tissues for largest retention of selected_set
  #     for(k in seq(1, length(contrasts_Treg_self[[i]]))[-ordered_index]){
  #       
  #       significs <- intersect(selected_set, rownames(
  #         subset(contrasts_Treg_self[[i]][[k]],
  #                abs(contrasts_Treg_self[[i]][[k]]$log2FoldChange) >= 2 &
  #                  contrasts_Treg_self[[i]][[k]]$padj <= 0.01 &
  #                  is.finite(contrasts_Treg_self[[i]][[k]]$Actuallog2FC))
  #         ))
  #       
  #       #temporarily hold the current best solution
  #       if(length(significs) > max){
  #         max <- length(significs)
  #         under_consid <- significs
  #         index <- k
  #       # if(length(significs) < min){
  #       #   min <- length(significs)
  #       #   under_consid <- significs
  #       #   index <- k
  #       }
  #     }
  #   
  #     #commit the best found solution
  #     selected_set <- under_consid
  #     ordered_index <- c(ordered_index, index)
  #     ordered_max <- c(ordered_max, max)
  #     #ordered_max <- c(ordered_max, min)
  #   }
  #   
  #   plot_data_entry <- ordered_max
  #   names(plot_data_entry) <- names(contrasts_Treg_self[[i]])[ordered_index]
  #   
  #   plot_data_list[[length(plot_data_list)+1]] <- plot_data_entry
  #   names(plot_data_list)[length(plot_data_list)] <- 
  #     names(contrasts_Treg_self)[i]
  # 
  # }
  # 
  # plot_data <- data.frame(Tissue = gsub(pattern = "\\.cellType",
  #                                      replacement = " ",
  #                                      x = gsub(pattern = "tissue",
  #                                               replacement = "",
  #                                               x =
  #                                                 names(plot_data_list)[1],
  #                                               ignore.case = T),
  #                                      ignore.case = T),
  #                         Step = seq(1, length(plot_data_list[[1]])), 
  #                         Excluded = gsub(pattern = "\\.cellType.*$", 
  #                                         replacement = "", 
  #                                         x = gsub(pattern = "tissue", 
  #                                                  replacement = "", 
  #                                                  x = names(plot_data_list[[1]]), 
  #                                                  ignore.case = T
  #                                                  )
  #                                         ), 
  #                         Counts = plot_data_list[[1]], 
  #                         stringsAsFactors = F)
  # 
  # #populating the plot_data dataframe
  # for(l in seq(2, length(plot_data_list))){
  #   plot_data <- rbind(plot_data, 
  #                      data.frame(Tissue = gsub(pattern = "\\.cellType",
  #                                      replacement = " ",
  #                                      x = gsub(pattern = "tissue",
  #                                               replacement = "",
  #                                               x =
  #                                                 names(plot_data_list)[l],
  #                                               ignore.case = T),
  #                                      ignore.case = T),
  #                         Step = seq(1, length(plot_data_list[[l]])), 
  #                         Excluded = gsub(pattern = "\\.cellType.*$", 
  #                                         replacement = "", 
  #                                         x = gsub(pattern = "tissue", 
  #                                                  replacement = "", 
  #                                                  x = names(plot_data_list[[l]]), 
  #                                                  ignore.case = T
  #                                                  )
  #                                         ),
  #                         Counts = plot_data_list[[l]], 
  #                         stringsAsFactors = F))
  # }
  # 
  # plot_data$Tissue <- as.factor(plot_data$Tissue)
  # #plot_data[plot_data$Counts == 0, 2] <- NA
  # 
  # plot_stepwise_Treg <- ggplot(data = plot_data,
  #                              aes(x = Step,
  #                                  y = Counts,
  #                                  colour = Tissue#,
  #                                  #fill = Tissue
  #                                  )
  #                              ) + 
  #   geom_line(aes(linetype = Tissue)) + 
  #   #geom_step() + 
  #   #geom_point(aes(shape = Tissue)) + 
  #   geom_point() + 
  #   #geom_text(aes(label = Counts), vjust=-0.3, size=7.5) +
  #   #ylim(0, 20) +
  #   labs(x = "Step-wise removal",
  #        y = "Number of\nSignificant Genes") +
  #   theme_light(base_size = 16) +
  #   theme(#axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
  #         panel.grid.minor = element_blank(),
  #         panel.grid.major = element_blank()#,
  #         #legend.position = "none"
  #         )
  # 
  # plot_stepwise_Treg
  # 
  # ggsave(filename = "plot_stepwise_Treg.png",
  #        plot = plot_stepwise_Treg,
  #        device = "png",
  #        width = 15,
  #        height = 12,
  #        units = "cm",
  #        dpi = 300,
  #        limitsize = T)
  # 
  # ggsave(filename = "plot_stepwise_Treg.svg",
  #        plot = plot_stepwise_Treg,
  #        device = "svg",
  #        width = 14,
  #        height = 10,
  #        units = "cm",
  #        dpi = 300,
  #        limitsize = T)
  # 
  # ggsave(filename = "plot_stepwise_Treg.pdf",
  #        plot = plot_stepwise_Treg,
  #        device = "pdf",
  #        width = 14,
  #        height = 10,
  #        units = "cm",
  #        dpi = 300,
  #        limitsize = T)
}



#contrasting tissue groups vs other tissue groups (group-based equivalent of above two)
{
  contrasts_TregGroups <- list()
  results_names_TregGroups <- list(
    result_names_Treg[grepl(pattern = "Blood", 
                            x = result_names_Treg, 
                            ignore.case = T)], 
    result_names_Treg[grepl(pattern = "(IEL|LPL)", 
                            x = result_names_Treg, 
                            ignore.case = T)], 
    result_names_Treg[grepl(pattern = "(LN|PP|Spleen)", 
                            x = result_names_Treg, 
                            ignore.case = T)], 
    result_names_Treg[grepl(pattern = "(Kidney|Liver|Lung|Pancreas)", 
                            x = result_names_Treg, 
                            ignore.case = T)]
    )
  names(results_names_TregGroups) <- c("Blood", 
                                       "Gut-Associated", 
                                       "Lymphoid", 
                                       "Non-Lymphoid")
  
  #for(contrast in results_names_TregGroups){
  for(i in seq(1, length(results_names_TregGroups))){
    
    contrast <- results_names_TregGroups[[i]]
    #contrasted_against <- result_names_Treg[!(result_names_Treg %in% contrast)]
    
    contrasts_TregGroups[[length(contrasts_TregGroups)+1]] <- 
      list()
    #for(j in contrasted_against){
    for(j in seq(1, length(results_names_TregGroups))){
      
      if(j == i) next
      
      contrasted_against <- results_names_TregGroups[[j]]
      
      contrasts_TregGroups[[length(contrasts_TregGroups)]][[
        length(contrasts_TregGroups[[length(
          contrasts_TregGroups)]])+1]] <- 
        as.data.frame(results(object = DESeq_object,
                              contrast = list(contrast, contrasted_against),
                              listValues = c(1/length(contrast), 
                                             -(1/length(contrasted_against))),
                              cooksCutoff = F,
                              independentFiltering = F,
                              test = "Wald",
                              parallel = T)
                      )
      
      # contrasts_TregGroups_paired[[length(contrasts_TregGroups_paired)+1]][[
      #   length(contrasts_TregGroups_paired[[length(
      #     contrasts_TregGroups_paired)+1]])+1]] <- 
      #   as.data.frame(results(object = DESeq_object,
      #                         contrast = list(contrast, j),
      #                         listValues = c(1/length(contrast), -1),
      #                         cooksCutoff = F,
      #                         independentFiltering = F,
      #                         test = "Wald",
      #                         parallel = T)
      #                 )
      
      results <- contrasts_TregGroups[[
        length(contrasts_TregGroups)]][[length(
          contrasts_TregGroups[[length(contrasts_TregGroups)]])]]
  
      results <- subset(results,
                        gsub(pattern = "\\.[0-9]+$",
                             replacement = "",
                             x = rownames(results),
                             ignore.case = T) %in%
                          mean_norm_counts$`Gene stable ID`)
    
      numerator <- contrast
      numerator <- gsub(pattern = "\\.cellType",
                        replacement = " ",
                        x = gsub(pattern = "tissue",
                                 replacement = "",
                                 x = numerator,
                                 ignore.case = T),
                        ignore.case = T)
    
      denominator <- contrasted_against
      denominator <- gsub(pattern = "\\.cellType",
                          replacement = " ",
                          x = gsub(pattern = "tissue",
                                   replacement = "",
                                   x = denominator,
                                   ignore.case = T),
                          ignore.case = T)
    
      #mean_norm_counts are already in log2
      if(length(numerator) < 2){
        num_means <- 
          data.matrix(mean_norm_counts[, grepl(pattern = paste(numerator, 
                                                               collapse = "|"), 
                                               x = colnames(mean_norm_counts), 
                                               ignore.case = T)])
      }
      else{
        num_means <- 
          rowMeans(data.matrix(
            mean_norm_counts[, grepl(pattern = paste(numerator, 
                                                     collapse = "|"), 
                                     x = colnames(mean_norm_counts), 
                                     ignore.case = T)]))
      }
      if(length(denominator) < 2){
        denom_means <- 
          data.matrix(mean_norm_counts[, grepl(pattern = paste(denominator, 
                                                               collapse = "|"), 
                                               x = colnames(mean_norm_counts), 
                                               ignore.case = T)])
      }
      else{
        denom_means <- 
          rowMeans(data.matrix(
            mean_norm_counts[, grepl(pattern = paste(denominator, 
                                                     collapse = "|"), 
                                     x = colnames(mean_norm_counts), 
                                     ignore.case = T)]))
      }
      # Actuallog2FC <- 
      #   rowMeans(data.matrix(mean_norm_counts[, grepl(pattern = paste(numerator, 
      #                                                       collapse = "|"), 
      #                                       x = colnames(mean_norm_counts), 
      #                                       ignore.case = T)])) -  
      #   rowMeans(data.matrix(mean_norm_counts[, grepl(pattern = paste(denominator, 
      #                                                       collapse = "|"), 
      #                                       x = colnames(mean_norm_counts), 
      #                                       ignore.case = T)]))
      Actuallog2FC <- as.numeric(num_means) - as.numeric(denom_means)
      Actuallog2FC[!is.finite(Actuallog2FC)] <- NA
      names(Actuallog2FC) <- rownames(mean_norm_counts)
      Actuallog2FC <- Actuallog2FC[match(gsub(pattern = "\\.[0-9]+$", 
                                              replacement = "", 
                                              x = rownames(results), 
                                              ignore.case = T), 
                                         names(Actuallog2FC))]
      results$Actuallog2FC <- Actuallog2FC
    
      contrasts_TregGroups[[length(contrasts_TregGroups)]][[
        length(contrasts_TregGroups[[
          length(contrasts_TregGroups)]])]] <- results
    
      names(contrasts_TregGroups[[length(
        contrasts_TregGroups)]])[length(
          contrasts_TregGroups[[length(
            contrasts_TregGroups)]])] <- names(results_names_TregGroups)[j]
    }
    
    names(contrasts_TregGroups)[length(contrasts_TregGroups)] <- 
      names(results_names_TregGroups)[i]
    
  }
}

#plotting contrasts_Treg_self and contrasts_TregGroups_paired up and down reg
{
  #appended_list <- append(contrasts_Treg_self, contrasts_TregGroups_paired[-1])
  appended_list <- contrasts_TregGroups
  
  plot_data <- data.frame(Tissues = gsub(pattern = "\\.cellType",
                                         replacement = " ",
                                         x = gsub(pattern = "tissue",
                                                  replacement = "",
                                                  x =
                                                    names(appended_list),
                                                  ignore.case = T),
                                         ignore.case = T),
                          Higher = NA, 
                          Lower = NA, 
                          stringsAsFactors = F)
  
  for(i in seq(1, length(appended_list))){
    
    print(i)
    
    significs_up <-
      rownames(subset(appended_list[[i]][[1]],
                      #appended_list[[i]][[1]]$log2FoldChange >= 1 &
                      appended_list[[i]][[1]]$log2FoldChange >= 2 &
                        appended_list[[i]][[1]]$padj <= 0.01 &
                        is.finite(appended_list[[i]][[1]]$Actuallog2FC))
               )
    
    significs_down <-
      rownames(subset(appended_list[[i]][[1]],
                      #appended_list[[i]][[1]]$log2FoldChange <= -1 &
                      appended_list[[i]][[1]]$log2FoldChange <= -2 &
                        appended_list[[i]][[1]]$padj <= 0.01 &
                        is.finite(appended_list[[i]][[1]]$Actuallog2FC))
               )

    for(j in seq(1, length(appended_list[[i]]))){
      
      #print(j)
      
      significs_up <-
        intersect(significs_up,
                  rownames(
                    subset(appended_list[[i]][[j]],
                           #appended_list[[i]][[j]]$log2FoldChange >= 1 &
                           appended_list[[i]][[j]]$log2FoldChange >= 2 &
                             appended_list[[i]][[j]]$padj <= 0.01 &
                             is.finite(appended_list[[i]][[j]]$Actuallog2FC))
               ))
      significs_down <-
        intersect(significs_down,
                  rownames(
                    subset(appended_list[[i]][[j]],
                           #appended_list[[i]][[j]]$log2FoldChange <= -1 &
                           appended_list[[i]][[j]]$log2FoldChange <= -2 &
                             appended_list[[i]][[j]]$padj <= 0.01 &
                             is.finite(appended_list[[i]][[j]]$Actuallog2FC))
               ))
    }
    
    #filtering minimum transcript per cell
    significs_up <- significs_up[significs_up %in% rownames(filtered_tcounts)]
    significs_down <- significs_down[significs_down %in% 
                                       rownames(filtered_tcounts)]
    
    if(i == 1){
      blood_sig_down <- significs_down
      to_write <- output_table[output_table$`Gene stable ID` %in% 
                                     gsub(pattern = "\\.[0-9]+$", 
                                          replacement = "", 
                                          x = blood_sig_down, 
                                          ignore.case = T), 1:3]
      to_write$direction <- "up"
      to_write$direction[to_write[,1] %in% gsub(pattern = "\\.[0-9]+$", 
                                               replacement = "", 
                                               x = significs_down, 
                                               ignore.case = T)] <- "down"
      #temp <- output_table
      write.table(x = to_write, 
                  file = "blood_sig_down.tab", 
                  quote = F, 
                  sep = "\t", 
                  row.names = F)
    }
    else if(i == 2){
      gut_sigs <- union(significs_up, significs_down)
      to_write <- output_table[output_table$`Gene stable ID` %in% 
                                     gsub(pattern = "\\.[0-9]+$", 
                                          replacement = "", 
                                          x = gut_sigs, 
                                          ignore.case = T), 1:3]
      to_write$direction <- "up"
      to_write$direction[to_write[,1] %in% gsub(pattern = "\\.[0-9]+$", 
                                               replacement = "", 
                                               x = significs_down, 
                                               ignore.case = T)] <- "down"
      #temp <- output_table
      write.table(x = to_write, 
                  file = "gut_sigs.tab", 
                  quote = F, 
                  sep = "\t", 
                  row.names = F)
    }
    else if(i == 4){
      non_sigs <- union(significs_up, significs_down)
      to_write <- output_table[output_table$`Gene stable ID` %in% 
                                     gsub(pattern = "\\.[0-9]+$", 
                                          replacement = "", 
                                          x = non_sigs, 
                                          ignore.case = T), 1:3]
      to_write$direction <- "up"
      to_write$direction[to_write[,1] %in% gsub(pattern = "\\.[0-9]+$", 
                                               replacement = "", 
                                               x = significs_down, 
                                               ignore.case = T)] <- "down"
      #temp <- output_table
      write.table(x = to_write, 
                  file = "non-lymph_sigs.tab", 
                  quote = F, 
                  sep = "\t", 
                  row.names = F)
    }

    plot_data$Higher[i] <- length(significs_up)
    plot_data$Lower[i] <- length(significs_down)

  }

  #plot_data$Tissues <- as.factor(plot_data$Tissues)
  #plot_data[plot_data$Counts == 0, 2] <- NA
  plot_data <- melt(plot_data)
  #plot_data$Tissues <- as.factor(plot_data$Tissues)
  colnames(plot_data) <- c("Tissues", "Expression", "Counts")
  
  #removing any bars for zero counts
  #plot_data$Counts[plot_data$Counts == 0] <- -1
  
  plot_data$Tissues <- factor(x = plot_data$Tissues, 
                             levels = (plot_data$Tissues)[
                               match(unique(c(
                                 as.character(
                                   plot_data$Tissues[grepl(pattern = "Treg$",
                                                          x = plot_data$Tissues,
                                                          ignore.case = T)]), 
                                 as.character(
                                   plot_data$Tissues[!grepl(pattern = "Treg$",
                                                          x = plot_data$Tissues,
                                                          ignore.case = T)])
                                 )), 
                                 plot_data$Tissues)
                               ]
                             )

  plot_contrasts_Treg_self <- ggplot(data = plot_data,
                                     aes(x = Tissues,
                                         y = Counts,
                                         #y = value,
                                         #colour = Tissues,
                                         #colour = Expression,
                                         #fill = Tissues)
                                         fill = Expression)
                                     ) +
    geom_bar(stat="identity", position = "stack") +
    # geom_text(aes(label = ifelse(value != 0, value, "")), 
    #           vjust=-0.3, size=7.5) +
    #ylim(0, 30) +
    labs(x = "", 
         y = "Number of\nSignificant Genes") +
    theme_light(base_size = 12) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, 
                                     size = 11),
          panel.grid.minor = element_blank(),
          panel.grid.major = element_blank(),
          #legend.position = "none"
          legend.position = "top"
          )

  plot_contrasts_Treg_self

  ggsave(filename = "plot_contrasts_Tregs_groups.png",
         plot = plot_contrasts_Treg_self,
         device = "png",
         width = 8,
         height = 12,
         units = "cm",
         dpi = 300,
         limitsize = T)

  ggsave(filename = "plot_contrasts_Tregs_groups.svg",
         plot = plot_contrasts_Treg_self,
         device = "svg",
         width = 8,
         height = 12,
         units = "cm",
         dpi = 300,
         limitsize = T)

  ggsave(filename = "plot_contrasts_Tregs_groups.pdf",
         plot = plot_contrasts_Treg_self,
         device = "pdf",
         width = 8,
         height = 12,
         units = "cm",
         dpi = 300,
         limitsize = T)
  
}

#plotting PCA on the blood_sig_down genes
{
  samples_PCA <- data.matrix(mean_norm_exp)
  samples_PCA <- samples_PCA[rownames(samples_PCA) %in% blood_sig_down,]
  samples_PCA <- samples_PCA[, grepl(pattern = "_T$", 
                                     x = colnames(samples_PCA), 
                                     ignore.case = T)]
  
  samples_PCA[!is.finite(samples_PCA)] <- NA
  samples_PCA <- samples_PCA[complete.cases(samples_PCA),]
  samples_PCA <- prcomp(t(samples_PCA[complete.cases(samples_PCA),]))
  
  #plotting variance of PCs
  variance_proportion <- ((samples_PCA$sdev^2) / (sum(samples_PCA$sdev^2)))*100
  dev.new()
  barplot(variance_proportion, cex.names=1,
          xlab=paste("Principal component (PC), 1-",
                     length(samples_PCA$sdev)), 
          ylab="Proportion of variation (%)",
          main="Scree plot", ylim=c(0,100))
  
  ##PCA Plots
  {
    samples_PCA_data <- as.data.frame(samples_PCA$x)
    samples_PCA_data$group <- rownames(samples_PCA_data)
    
    #plot PCA -- data points grouped by tissue groups
    tissue_group <- samples_PCA_data$group
    
    tissue_group <- gsub(pattern = "_(NT|T)", 
                       replacement = "", 
                       x = tissue_group, 
                       ignore.case = T)
    
    for(i in seq_along(tissue_group)){
      if (tolower(tissue_group[i]) == "blood"){
        tissue_group[i] <- "Blood"
      }
      else if (tolower(tissue_group[i]) %in% c("spleen", "lymphnode", 
                                               "peyerspatches")){
        tissue_group[i] <- "Lymphoid"
      }
      else if (tolower(tissue_group[i]) %in% 
          c("intraepitheliallayer", "laminaproprialymphocyte")){
        tissue_group[i] <- paste0("Gut-associated")
      }
      else if (tolower(tissue_group[i]) %in% 
          c("kidney", "lung", "liver", "pancreas")){
        tissue_group[i] <- "Non-lymphoid"
      }
      else if (tolower(tissue_group[i]) == "celltypet"){
        tissue_group[i] <- "Treg"
      }
      else if (tolower(tissue_group[i]) == "celltypent"){
        tissue_group[i] <- "Tconv"
      }
      
    }
    rm(i)
    
    samples_PCA_data$tissue_group <- as.factor(tissue_group)
    
    #only Treg plot
    {
      subset_data <- samples_PCA_data[grepl(pattern = "_T", 
                                            x = rownames(samples_PCA_data), 
                                            ignore.case = F),]
      
      Type <- as.factor(ifelse(grepl(pattern = "_NT", 
                                      x = rownames(subset_data), 
                                      ignore.case = F), "Tconv", "Treg"))
      
      Label <- qdap::multigsub(pattern = 
                                     c("IntraEpithelialLayer", 
                                       "LaminaPropriaLymphocyte", 
                                       "PeyersPatches", "LymphNode"), 
                                   replacement = c("IEL", "LPL", "PP",
                                                   "LN"), 
                                   order.pattern = F, 
                                   text.var = gsub(pattern = "_.*", 
                                                   replacement = "", 
                                                   x = rownames(
                                                     subset_data), 
                                                   ignore.case = T)
                                     )
      
      #manually assigning class colours
      {
        tissue_group_col <- tissue_group
        names(tissue_group_col) <- tissue_group_col
        
        tissue_group_col[grepl(pattern = "Blood",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#D81B60"
        tissue_group_col[grepl(pattern = "(Spleen|PP|LN|^Lymphoid)",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#FFC107"
        tissue_group_col[grepl(pattern = "(IEL|LPL|Gut)",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#1E88E5"
        tissue_group_col[grepl(pattern = "(Kidney|Lung|Liver|Pancreas|Non)",
                               x = tissue_group_col,
                               ignore.case = T)] <- "#004D40"
      }
      
      
      PCA_plot2 <- ggplot(subset_data, 
                          aes(#Type = Type,
                              #Label = Label, 
                              x = PC1, y = PC2, 
                              #color=tissue_group)
                              color = gsub(pattern = ", ",
                                          replacement = ", \n",
                                          x = tissue_group)
                              )
                       ) + 
        geom_point(aes_(shape = as.factor(ifelse(grepl(pattern = "_NT", 
                                      x = rownames(subset_data), 
                                      ignore.case = F), "Tconv", "Treg"))), 
                   size = 4) + 
        geom_text_repel(aes_(label=
                               qdap::multigsub(pattern =
                                                 c("IntraEpithelialLayer",
                                                   "LaminaPropriaLymphocyte",
                                                   "PeyersPatches",
                                                   "LymphNode"),
                                               replacement = c("IEL", "LPL",
                                                               "PP", "LN"),
                                               order.pattern = F,
                                               text.var = gsub(pattern = "_.*",
                                                               replacement = "",
                                                               x = rownames(
                                                                 subset_data),
                                                               ignore.case = T)
                                                 )
                             ),
                        size = 7, segment.alpha = 0.4, #for offline images
                        #size = 5, segment.alpha = 0.4, #for webtool rds files
                        force = 1, direction = "both", show.legend = F) +
        scale_shape_manual(values = c(20, 4)) + 
        scale_colour_manual(values = tissue_group_col) + 
        labs(colour = "Tissue", 
             shape = "Cell Types",
             title = "Treg samples"
             ) + 
        #theme_light(base_size = 13.5) + #for offline images
        theme_light(base_size = 18) + #for webtool rds files
        theme(legend.key.size = unit(0.15, 'lines'), 
              plot.caption = element_text(hjust = 0, face= "italic"),
              panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank()
              ) + 
        scale_x_continuous(limits = c(min(samples_PCA_data$PC1), 
                                      max(samples_PCA_data$PC1))) + 
        scale_y_continuous(limits = c(min(samples_PCA_data$PC2), 
                                      max(samples_PCA_data$PC2))) + 
        guides(shape = F)
      
      #PCA_plot2
      
      # #for ggplotly in the webtool -- modify the weird brackets in the legend
      # myplot <- ggplotly(PCA_plot2, tooltip = c("Label", "Type", "PC1", "PC2")) %>% 
      #   toWebGL()
      # 
      # for (i in 1:length(myplot$x$data)){
      #   if (!is.null(myplot$x$data[[i]]$name)){
      #     myplot$x$data[[i]]$name =  gsub("\\(|\\)","",str_split(myplot$x$data[[i]]$name,",")[[1]][2])
      #   }
      # }
      # 
      # myplot
      
    }
    
    #multiplot(PCA_plot1, PCA_plot2, PCA_plot3, cols = 2)
    
    # PCA_plot1 + PCA_plot2 + PCA_plot3 + 
    #   plot_layout(ncol = 2, nrow = 2, byrow = T)
    gc()
    
  }
  
}

#perform t-SNE on mean log2 transformed data
{
  #reading in DESeq2 outputs
  {
    contrast_groups <- list(c("Blood"), 
                            c("Spleen", "Lymph Node", "Peyer's Patches"), 
                            c("Intraepithelial Layer", 
                              "Lamina Propria Lymphocyte"), 
                            c("Kidney", "Lung", "Liver", "Pancreas"),
                            c("cellTypeT"),
                            c("cellTypeNT"))
    
    #ensuring that names in contrast_groups list are all lower letter and do not 
    # contain spaces or punctuation
    for(i in seq(1, length(contrast_groups)))
    {
      contrast_groups[[i]] <- tolower(unique(
        gsub("_.*|[[:space:]]|[[:punct:]]", "", contrast_groups[[i]])
        ))
    }
    rm(i)
    names(contrast_groups) <- c("Blood", 
                                "Lymphoid", 
                                "Gut-Associated", 
                                "Non-Lymphoid", 
                                "Treg", 
                                "Tconv")
  }
  
  data_tsne <- data.matrix(mean_norm_exp)
  data_tsne <- data_tsne[rownames(data_tsne) %in% blood_sig_down,]
  data_tsne <- data_tsne[, grepl(pattern = "_T$", 
                                     x = colnames(data_tsne), 
                                     ignore.case = T)]
  
  data_tsne[!is.finite(data_tsne)] <- NA
  data_tsne <- data_tsne[complete.cases(data_tsne),]
  data_tsne <- t(data_tsne)
  
  tissue_group <- rownames(data_tsne)
  tissue_group <- gsub(#pattern = " (Treg|Tconv)$", 
                       pattern = "_T$", 
                       replacement = "", 
                       x = tissue_group, 
                       ignore.case = T)
    
  for(i in seq_along(tissue_group)){
    if (tolower(tissue_group[i]) == "blood"){
      tissue_group[i] <- "Blood"
    }
    else if (tolower(tissue_group[i]) %in% c("spleen", "ln", "pp", 
                                             "lymphnode", "peyerspatches")){
      tissue_group[i] <- "Lymphoid tissue"
    }
    else if (tolower(tissue_group[i]) %in% c("iel", "lpl", 
                                             "intraepitheliallayer", 
                                             "laminaproprialymphocyte")){
      tissue_group[i] <- paste0("Gut-associated")
    }
    else if (tolower(tissue_group[i]) %in% 
        c("kidney", "lung", "liver", "pancreas")){
      tissue_group[i] <- "Non-lymphoid"
    }
    
  }
  
  data_tsne <- as.data.frame(data_tsne)
  
  tsne_results <- Rtsne(X = data_tsne, 
                        dims = 2, 
                        initial_dims = 50, 
                        pca = T, 
                        partial_pca = F, 
                        pca_center = T, 
                        pca_scale = F, 
                        normalize = T, 
                        check_duplicates = F, 
                        verbose = T, 
                        perplexity = 3, 
                        theta = 0, 
                        max_iter = 2000, 
                        num_threads = 9)
  
  #only Treg plot
  {
    subset_data <- data_tsne[grepl(pattern = "_T", 
                                          x = rownames(data_tsne), 
                                          ignore.case = F),]
    
    #manually assigning class colours
    {
      tissue_group_col <- tissue_group
      names(tissue_group_col) <- tissue_group_col
      
      tissue_group_col[grepl(pattern = "Blood",
                             x = tissue_group_col,
                             ignore.case = T)] <- "#D81B60"
      tissue_group_col[grepl(pattern = "(Spleen|PP|LN|^Lymphoid)",
                             x = tissue_group_col,
                             ignore.case = T)] <- "#FFC107"
      tissue_group_col[grepl(pattern = "(IEL|LPL|Gut)",
                             x = tissue_group_col,
                             ignore.case = T)] <- "#1E88E5"
      tissue_group_col[grepl(pattern = "(Kidney|Lung|Liver|Pancreas|Non)",
                             x = tissue_group_col,
                             ignore.case = T)] <- "#004D40"
    }
    
    tSNE_plot2 <- ggplot(as.data.frame(subset_data), 
                        aes_(x=tsne_results$Y[which(
                          rownames(data_tsne) %in% rownames(subset_data)),1], 
                           y=tsne_results$Y[which(
                             rownames(data_tsne) %in% rownames(subset_data)),2], 
                            color= gsub(pattern = ", ",
                                        replacement = ", \n",
                                        x = tissue_group[which(
                                          rownames(data_tsne) %in% 
                                            rownames(subset_data))]))
                     ) + 
      geom_point(aes_(shape = as.factor(ifelse(grepl(pattern = "_NT", 
                                    x = rownames(subset_data), 
                                    ignore.case = F), "Tconv", "Treg"))),
                 size = 4) + 
      geom_text_repel(aes_(label= 
                             qdap::multigsub(pattern = 
                                               c("IntraEpithelialLayer", 
                                                 "LaminaPropriaLymphocyte", 
                                                 "PeyersPatches", "LymphNode"), 
                                             replacement = c("IEL", "LPL", "PP",
                                                             "LN"), 
                                             order.pattern = F, 
                                             text.var = gsub(pattern = "_.*", 
                                                             replacement = "", 
                                                             x = rownames(
                                                               subset_data), 
                                                             ignore.case = T)
                                               )
                           ), 
                      #size = 3.25, segment.alpha = 0.4,
                      size = 7, segment.alpha = 0.4,
                      force = 1, direction = "both", show.legend = F) + 
      scale_shape_manual(values = c(20, 4)) + 
      scale_colour_manual(values = tissue_group_col) + 
      labs(x = "Dimension 1", 
           y = "Dimension 2", 
           colour = "Tissue Groups",
           title = "Treg samples"
           ) + 
      #theme_light(base_size = 13) + 
      theme_light(base_size = 18) + 
      theme(legend.key.size = unit(0.15, 'lines'), 
            plot.caption = element_text(hjust = 0, face= "italic"),
            panel.grid.major = element_blank(), panel.grid.minor = element_blank()
            ) + 
      scale_x_continuous(limits = c(min(tsne_results$Y[,1]), 
                                    max(tsne_results$Y[,1]))) + 
      scale_y_continuous(limits = c(min(tsne_results$Y[,2]), 
                                    max(tsne_results$Y[,2]))) + 
      guides(shape = F)
    
    #tSNE_plot2
  }
  
}





#plotting contrasts_Treg_self #no longer needed
{
  # plot_data <- data.frame(Tissues = gsub(pattern = "\\.cellType",
  #                                        replacement = " ",
  #                                        x = gsub(pattern = "tissue",
  #                                                 replacement = "",
  #                                                 x =
  #                                                   names(contrasts_Treg_self),
  #                                                 ignore.case = T),
  #                                        ignore.case = T),
  #                         Counts = NA,
  #                         stringsAsFactors = F)
  # 
  # for(i in seq(1, length(contrasts_Treg_self))){
  # 
  #   significs <-
  #     rownames(subset(contrasts_Treg_self[[i]][[1]],
  #                     contrasts_Treg_self[[i]][[1]]$log2FoldChange >= 2 &
  #                       contrasts_Treg_self[[i]][[1]]$padj <= 0.01 &
  #                       is.finite(contrasts_Treg_self[[i]][[1]]$Actuallog2FC))
  #              )
  # 
  #   for(j in seq(2, length(contrasts_Treg_self[[i]]))){
  # 
  #     significs <-
  #       intersect(significs,
  #                 rownames(
  #                   subset(contrasts_Treg_self[[i]][[j]],
  #                          contrasts_Treg_self[[i]][[j]]$log2FoldChange >= 2 &
  #                            contrasts_Treg_self[[i]][[j]]$padj <= 0.01 &
  #                            is.finite(contrasts_Treg_self[[i]][[j]]$Actuallog2FC))
  #              ))
  #   }
  # 
  #   plot_data$Counts[i] <- length(significs)
  # 
  # }
  # 
  # plot_data$Tissues <- as.factor(plot_data$Tissues)
  # #plot_data[plot_data$Counts == 0, 2] <- NA
  # 
  # plot_contrasts_Treg_self <- ggplot(data = plot_data,
  #                                    aes(x = Tissues,
  #                                        y = Counts,
  #                                        colour = Tissues,
  #                                        fill = Tissues)
  #                                    ) +
  #   geom_bar(stat="identity") +
  #   geom_text(aes(label = Counts), vjust=-0.3, size=7.5) +
  #   ylim(0, 20) +
  #   labs(y = "Number of Significant\nGenes Expressed Higher") +
  #   theme_light(base_size = 16) +
  #   theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
  #         panel.grid.minor = element_blank(),
  #         panel.grid.major = element_blank(),
  #         legend.position = "none")
  # 
  # plot_contrasts_Treg_self
  # 
  # ggsave(filename = "plot_contrasts_Treg_self.png",
  #        plot = plot_contrasts_Treg_self,
  #        device = "png",
  #        width = 15,
  #        height = 12,
  #        units = "cm",
  #        dpi = 300,
  #        limitsize = T)
  # 
  # ggsave(filename = "plot_contrasts_Treg_self.svg",
  #        plot = plot_contrasts_Treg_self,
  #        device = "svg",
  #        width = 14,
  #        height = 10,
  #        units = "cm",
  #        dpi = 300,
  #        limitsize = T)
  # 
  # ggsave(filename = "plot_contrasts_Treg_self.pdf",
  #        plot = plot_contrasts_Treg_self,
  #        device = "pdf",
  #        width = 14,
  #        height = 10,
  #        units = "cm",
  #        dpi = 300,
  #        limitsize = T)
  
}

#plotting contrasts_Tconv_self #no longer needed
{
  # plot_data <- data.frame(Tissues = gsub(pattern = "\\.cellType", 
  #                                        replacement = " ", 
  #                                        x = gsub(pattern = "tissue", 
  #                                                 replacement = "", 
  #                                                 x = 
  #                                                   names(contrasts_Tconv_self), 
  #                                                 ignore.case = T), 
  #                                        ignore.case = T), 
  #                         Counts = NA, 
  #                         stringsAsFactors = F)
  # 
  # for(i in seq(1, length(contrasts_Tconv_self))){
  #   
  #   significs <- 
  #     rownames(subset(contrasts_Tconv_self[[i]][[1]], 
  #                     contrasts_Tconv_self[[i]][[1]]$log2FoldChange > 0 & 
  #                       contrasts_Tconv_self[[i]][[1]]$padj <= 0.01 & 
  #                       is.finite(contrasts_Tconv_self[[i]][[1]]$Actuallog2FC))
  #              )
  #   
  #   for(j in seq(2, length(contrasts_Tconv_self[[i]]))){
  #     
  #     significs <- 
  #       intersect(significs, 
  #                 rownames(
  #                   subset(contrasts_Tconv_self[[i]][[j]], 
  #                          contrasts_Tconv_self[[i]][[j]]$log2FoldChange > 0 & 
  #                            contrasts_Tconv_self[[i]][[j]]$padj <= 0.01 & 
  #                            is.finite(contrasts_Tconv_self[[i]][[j]]$Actuallog2FC))
  #              ))
  #   }
  #   
  #   plot_data$Counts[i] <- length(significs)
  #   
  # }
  # 
  # plot_data$Tissues <- as.factor(plot_data$Tissues)
  # #plot_data[plot_data$Counts == 0, 2] <- NA
  # 
  # # ggplot(data = plot_data, 
  # #        aes(x = Tissues, 
  # #            y = Counts, 
  # #            colour = Tissues, 
  # #            fill = Tissues)
  # #        ) + 
  # #   geom_bar(stat="identity") + 
  # #   geom_text(aes(label = Counts), vjust=-0.3, size=7.5) + 
  # #   ylim(0, 20) + 
  # #   theme_light(base_size = 16) + 
  # #   theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
  # #         panel.grid.minor = element_blank())
  # 
  # plot_contrasts_Tconv_self <- ggplot(data = plot_data, 
  #                                    aes(x = Tissues, 
  #                                        y = Counts, 
  #                                        colour = Tissues, 
  #                                        fill = Tissues)
  #                                    ) + 
  #   geom_bar(stat="identity") + 
  #   geom_text(aes(label = Counts), vjust=-0.3, size=7.5) + 
  #   ylim(0, 20) + 
  #   labs(y = "Number of Significant\nGenes Expressed Higher") + 
  #   theme_light(base_size = 16) + 
  #   theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
  #         panel.grid.minor = element_blank(), 
  #         panel.grid.major = element_blank(), 
  #         legend.position = "none")
  # 
  # plot_contrasts_Tconv_self
  # 
  # ggsave(filename = "plot_contrasts_Tconv_self.png", 
  #        plot = plot_contrasts_Tconv_self, 
  #        device = "png", 
  #        width = 15, 
  #        height = 12, 
  #        units = "cm", 
  #        dpi = 300, 
  #        limitsize = T)
  # 
  # ggsave(filename = "plot_contrasts_Tconv_self.svg", 
  #        plot = plot_contrasts_Tconv_self, 
  #        device = "svg", 
  #        width = 14, 
  #        height = 10, 
  #        units = "cm", 
  #        dpi = 300, 
  #        limitsize = T)
  # 
  # ggsave(filename = "plot_contrasts_Tconv_self.pdf", 
  #        plot = plot_contrasts_Tconv_self, 
  #        device = "pdf", 
  #        width = 14, 
  #        height = 10, 
  #        units = "cm", 
  #        dpi = 300, 
  #        limitsize = T)
  
}

#plotting intersection of the two above #no longer needed
{
  # plot_data <- data.frame(Tissues = gsub(pattern = "\\..*", 
  #                                        replacement = " ", 
  #                                        x = gsub(pattern = "tissue", 
  #                                                 replacement = "", 
  #                                                 x = 
  #                                                   names(contrasts_Tconv_self), 
  #                                                 ignore.case = T), 
  #                                        ignore.case = T), 
  #                         Counts = NA, 
  #                         stringsAsFactors = F)
  # 
  # significs <- NULL
  # significsTreg <- NULL
  # significsTconv <- NULL
  # 
  # for(i in seq(1, length(contrasts_Treg_self))){
  #   
  #   significsTreg <- 
  #     rownames(subset(contrasts_Treg_self[[i]][[1]], 
  #                     contrasts_Treg_self[[i]][[1]]$log2FoldChange > 0 & 
  #                       contrasts_Treg_self[[i]][[1]]$padj <= 0.01 & 
  #                       is.finite(contrasts_Treg_self[[i]][[1]]$Actuallog2FC))
  #              )
  #   
  #   for(j in seq(2, length(contrasts_Treg_self[[i]]))){
  #     
  #     significsTreg <- 
  #       intersect(significsTreg, 
  #                 rownames(
  #                   subset(contrasts_Treg_self[[i]][[j]], 
  #                          contrasts_Treg_self[[i]][[j]]$log2FoldChange > 0 & 
  #                            contrasts_Treg_self[[i]][[j]]$padj <= 0.01 & 
  #                            is.finite(contrasts_Treg_self[[i]][[j]]$Actuallog2FC))
  #              ))
  #   }
  #   
  #   significsTconv <- 
  #     rownames(subset(contrasts_Tconv_self[[i]][[1]], 
  #                     contrasts_Tconv_self[[i]][[1]]$log2FoldChange > 0 & 
  #                       contrasts_Tconv_self[[i]][[1]]$padj <= 0.01 & 
  #                       is.finite(contrasts_Tconv_self[[i]][[1]]$Actuallog2FC))
  #              )
  #   
  #   for(j in seq(2, length(contrasts_Tconv_self[[i]]))){
  #     
  #     significsTconv <- 
  #       intersect(significsTconv, 
  #                 rownames(
  #                   subset(contrasts_Tconv_self[[i]][[j]], 
  #                          contrasts_Tconv_self[[i]][[j]]$log2FoldChange > 0 & 
  #                            contrasts_Tconv_self[[i]][[j]]$padj <= 0.01 & 
  #                            is.finite(contrasts_Tconv_self[[i]][[j]]$Actuallog2FC))
  #              ))
  #   }
  #   
  #   plot_data$Counts[i] <- length(intersect(significsTconv, significsTreg))
  #   
  # }
  # 
  # plot_data$Tissues <- as.factor(plot_data$Tissues)
  # 
  # plot_contrasts_Treg_pooled <- ggplot(data = plot_data, 
  #                                    aes(x = Tissues, 
  #                                        y = Counts, 
  #                                        colour = Tissues, 
  #                                        fill = Tissues)
  #                                    ) + 
  #   geom_bar(stat="identity") + 
  #   geom_text(aes(label = Counts), vjust=-0.3, size=7.5) + 
  #   ylim(0, 20) + 
  #   labs(y = "Number of Significant\nGenes Expressed Higher") + 
  #   theme_light(base_size = 16) + 
  #   theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), 
  #         panel.grid.minor = element_blank(), 
  #         panel.grid.major = element_blank(), 
  #         legend.position = "none")
  # 
  # plot_contrasts_Treg_pooled
  # 
  # ggsave(filename = "plot_contrasts_Treg_pooled.png", 
  #        plot = plot_contrasts_Treg_pooled, 
  #        device = "png", 
  #        width = 15, 
  #        height = 12, 
  #        units = "cm", 
  #        dpi = 300, 
  #        limitsize = T)
  # 
  # ggsave(filename = "plot_contrasts_Treg_pooled.svg", 
  #        plot = plot_contrasts_Treg_pooled, 
  #        device = "svg", 
  #        width = 14, 
  #        height = 10, 
  #        units = "cm", 
  #        dpi = 300, 
  #        limitsize = T)
  # 
  # ggsave(filename = "plot_contrasts_Treg_pooled.pdf", 
  #        plot = plot_contrasts_Treg_pooled, 
  #        device = "pdf", 
  #        width = 14, 
  #        height = 10, 
  #        units = "cm", 
  #        dpi = 300, 
  #        limitsize = T)
  
}


#calculating and plotting the pairwise correlation between the 
# contrasts_Treg_self and contrasts_Tconv_self ## 1D colour strip
{
  # 
  # pearson_corr <- cor(x = contrasts_Treg_self[[1]][[1]]$log2FoldChange, 
  #                     y = contrasts_Tconv_self[[1]][[1]]$log2FoldChange, 
  #                     use = "complete.obs", 
  #                     method = "pearson")
  # spearman_corr <- cor(x = contrasts_Treg_self[[1]][[1]]$log2FoldChange, 
  #                      y = contrasts_Tconv_self[[1]][[1]]$log2FoldChange, 
  #                      use = "complete.obs", 
  #                      method = "spearman")
  # kendall_corr <- cor(x = contrasts_Treg_self[[1]][[1]]$log2FoldChange, 
  #                      y = contrasts_Tconv_self[[1]][[1]]$log2FoldChange, 
  #                      use = "complete.obs", 
  #                      method = "kendall")
  # 
  # name <- paste0(gsub(pattern = "(tissue|\\..*$)", 
  #                     replacement = "", 
  #                     x = names(contrasts_Tconv_self)[1], 
  #                     ignore.case = T), 
  #                " vs ", 
  #                gsub(pattern = "(tissue|\\..*$)", 
  #                     replacement = "", 
  #                     x = names(contrasts_Tconv_self[[1]])[1], 
  #                     ignore.case = T))
  # names(pearson_corr) <- name
  # names(spearman_corr) <- name
  # names(kendall_corr) <- name
  # 
  # for(i in seq(1, length(contrasts_Treg_self)-1)){
  #   
  #   for( j in seq(ifelse(i == 1, 2, i), length(contrasts_Treg_self[[i]]))){
  #     
  #     pearson_corr <- 
  #       append(pearson_corr, 
  #              cor(x = contrasts_Treg_self[[i]][[j]]$log2FoldChange, 
  #                  y = contrasts_Tconv_self[[i]][[j]]$log2FoldChange, 
  #                  use = "complete.obs", 
  #                  method = "pearson"))
  #     spearman_corr <- 
  #       append(spearman_corr, 
  #              cor(x = contrasts_Treg_self[[i]][[j]]$log2FoldChange, 
  #                  y = contrasts_Tconv_self[[i]][[j]]$log2FoldChange, 
  #                  use = "complete.obs", 
  #                  method = "spearman"))
  #     kendall_corr <- 
  #       append(kendall_corr, 
  #              cor(x = contrasts_Treg_self[[i]][[j]]$log2FoldChange, 
  #                  y = contrasts_Tconv_self[[i]][[j]]$log2FoldChange, 
  #                  use = "complete.obs", 
  #                  method = "spearman"))
  #     
  #     name <- paste0(gsub(pattern = "(tissue|\\..*$)", 
  #                         replacement = "", 
  #                         x = names(contrasts_Tconv_self)[i], 
  #                         ignore.case = T), 
  #                    " vs ", 
  #                    gsub(pattern = "(tissue|\\..*$)", 
  #                         replacement = "", 
  #                         x = names(contrasts_Tconv_self[[i]])[j], 
  #                         ignore.case = T))
  #     names(pearson_corr)[length(pearson_corr)] <- name
  #     names(spearman_corr)[length(spearman_corr)] <- name
  #     names(kendall_corr)[length(kendall_corr)] <- name
  #   }
  # }
  # 
  # corr_matrix <- as.data.frame(cbind(kendall_corr, spearman_corr, pearson_corr))
  # #corr_matrix <- data.frame(Spearman = spearman_corr)
  # corr_matrix$comparison <- rownames(corr_matrix)
  # colnames(corr_matrix)[1:3] <- c("Kendall", "Spearman", "Pearson")
  # corr_matrix <- melt(corr_matrix)
  # 
  # row_plot <- ggplot(data = corr_matrix, 
  #                    aes(x = comparison, 
  #                        y = variable, 
  #                        fill = value)) + 
  #   geom_tile() + 
  #   scale_fill_gradientn(limits = c(-1,1), 
  #                        colours=c("#0064cd", "white", "#ff3232")) + 
  #   labs(x = "Correlation of Treg contrasts to Tconv contrasts", 
  #        y = "") + 
  #   scale_x_discrete(position = "top") + 
  #   theme_light() + 
  #   theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 0), 
  #         legend.justification = "top"#, 
  #         #plot.margin = unit(c(0,2,0,0), "cm"), 
  #         #legend.position = c(1.06,1.3)
  #         )
  # 
  # row_plot
  # 
}

#calculating correlation heatmap for log2 expression of samples
{
  # norm_exp_log2 <- data.matrix(log2(norm_counts))
  # norm_exp_log2[!is.finite(norm_exp_tissues_log2)] <- NA
  # norm_exp_log2 <- norm_exp_tissues_log2[complete.cases(norm_exp_tissues_log2),]
  # #mean_norm_exp_tissues$`Gene stable ID` <- NULL
  # #mean_norm_exp_tissues <- data.matrix(mean_norm_exp_tissues)
  # 
  # #formatting the colnames
  # {
  #   colnames(norm_exp_log2) <- gsub(pattern = "_T_", 
  #                                   replacement = "_Treg_", 
  #                                   x = colnames(norm_exp_log2), 
  #                                   ignore.case = T)
  #   colnames(norm_exp_log2) <- gsub(pattern = "_NT_", 
  #                                   replacement = "_Tconv_", 
  #                                   x = colnames(norm_exp_log2), 
  #                                   ignore.case = T)
  #   colnames(norm_exp_log2) <- gsub(pattern = "IntraEpithelialLayer", 
  #                                   replacement = "IEL", 
  #                                   x = colnames(norm_exp_log2), 
  #                                   ignore.case = T)
  #   colnames(norm_exp_log2) <- gsub(pattern = "LaminaPropriaLymphocyte", 
  #                                   replacement = "LPL", 
  #                                   x = colnames(norm_exp_log2), 
  #                                   ignore.case = T)
  #   colnames(norm_exp_log2) <- gsub(pattern = "LymphNode", 
  #                                   replacement = "LN", 
  #                                   x = colnames(norm_exp_log2), 
  #                                   ignore.case = T)
  #   colnames(norm_exp_log2) <- gsub(pattern = "PeyersPatches", 
  #                                   replacement = "PP", 
  #                                   x = colnames(norm_exp_log2), 
  #                                   ignore.case = T)
  # }
  # 
  # norm_exp_log2_Treg <- norm_exp_log2[, grepl(pattern = "Treg", 
  #                                             x = colnames(norm_exp_log2), 
  #                                             ignore.case = T)]
  # 
  # corr_mat_pearson <- cor(norm_exp_log2_Treg, 
  #                         use = "complete.obs", 
  #                         method = "pearson")
  # corr_mat_spearman <- cor(norm_exp_log2_Treg, 
  #                         use = "complete.obs", 
  #                         method = "spearman")
  # corr_mat_kendall <- cor(norm_exp_log2_Treg, 
  #                         use = "complete.obs", 
  #                         method = "kendall")
  # 
  # #constructing rowside and colside colours to identify TregvTconv and Tissue groups
  # colsidecol <- colnames(corr_mat_pearson)
  # colsidecol[grepl(pattern = "Treg", 
  #                  x = colsidecol, 
  #                  ignore.case = T)] <- "Treg" #"#E66100"
  # colsidecol[grepl(pattern = "Tconv", 
  #                  x = colsidecol, 
  #                  ignore.case = T)] <- "Tconv" #"#5D3A9B"
  # names(colsidecol) <- colsidecol
  # colsidecol_colours <- unique(colsidecol)
  # names(colsidecol_colours) <- colsidecol_colours
  # # colsidecol_colours[1:length(colsidecol_colours)] <- c("#5D3A9B", 
  # #                                                       "#E66100")
  # colsidecol_colours[1:length(colsidecol_colours)] <- c("#E66100")
  # 
  # rowsidecol <- colnames(corr_mat_pearson)
  # rowsidecol[grepl(pattern = "Blood", 
  #                  x = rowsidecol, 
  #                  ignore.case = T)] <- "Blood" #"#D81B60"
  # rowsidecol[grepl(pattern = "(Spleen|PP|LN|^Lymphoid)", 
  #                  x = rowsidecol, 
  #                  ignore.case = T)] <- "Lymphoid" #"#FFC107"
  # rowsidecol[grepl(pattern = "(IEL|LPL|Gut)", 
  #                  x = rowsidecol, 
  #                  ignore.case = T)] <- "Gut-Associated" #"#1E88E5"
  # rowsidecol[grepl(pattern = "(Kidney|Lung|Liver|Pancreas|Non)", 
  #                  x = rowsidecol, 
  #                  ignore.case = T)] <- "Non-lymphoid" #"#004D40"
  # names(rowsidecol) <- rowsidecol
  # rowsidecol_colours <- unique(rowsidecol)
  # names(rowsidecol_colours) <- rowsidecol_colours
  # rowsidecol_colours[1:length(rowsidecol_colours)] <- c("#D81B60", 
  #                                                       "#1E88E5", 
  #                                                       "#004D40", 
  #                                                       "#FFC107")
  # 
  # heatmap_annot <- data.frame(cbind(rowsidecol, colsidecol), 
  #                             stringsAsFactors = F)
  # colnames(heatmap_annot) <- c("Tissue", "Cell Type")
  # heatmap_annot_colours <- list(Tissue = rowsidecol_colours, 
  #                               `Cell Type` = colsidecol_colours)
  # 
  # #col<- colorRampPalette(c("#0064cd", "white", "#ff3232"))(100)
  # #col<- colorRampPalette(c("white", "#ff3232"))(100)
  # 
  # aheatmap(x = corr_mat_pearson, 
  #          cellwidth = 11, 
  #          cellheight = 11, 
  #          color = rev(heat.colors(12)), 
  #          Colv = F, 
  #          treeheight = 30, 
  #          #color = rev(hcl.colors(100)), 
  #          #breaks = seq(-1, 1, length.out = 100), 
  #          annRow = heatmap_annot, 
  #          annColors = heatmap_annot_colours, 
  #          fontsize = 9, 
  #          #cexRow = 1, 
  #          #cexCol = 0.8, 
  #          main = "Pearson Correlation", 
  #          filename = "Pearson Correlation.pdf")
  # 
  # aheatmap(x = corr_mat_spearman, 
  #          cellwidth = 11, 
  #          cellheight = 11, 
  #          color = rev(heat.colors(12)), 
  #          Colv = F, 
  #          treeheight = 30, 
  #          #color = rev(hcl.colors(100)), 
  #          #breaks = seq(-1, 1, length.out = 100), 
  #          annRow = heatmap_annot, 
  #          annColors = heatmap_annot_colours, 
  #          fontsize = 9, 
  #          #cexRow = 1, 
  #          #cexCol = 0.8, 
  #          main = "Spearman Correlation", 
  #          filename = "Spearman Correlation.pdf")
  # 
  # aheatmap(x = corr_mat_kendall, 
  #          cellwidth = 11, 
  #          cellheight = 11, 
  #          color = rev(heat.colors(12)), 
  #          Colv = F, 
  #          treeheight = 30, 
  #          #color = rev(hcl.colors(100)), 
  #          #breaks = seq(-1, 1, length.out = 100), 
  #          annRow = heatmap_annot, 
  #          annColors = heatmap_annot_colours, 
  #          fontsize = 9, 
  #          #cexRow = 1, 
  #          #cexCol = 0.8, 
  #          main = "Kendall Correlation", 
  #          filename = "Kendall Correlation.pdf")
  
}


```


#log10 scaled scatter plots of normalised counts of high variance genes
```{r}

#reading in DESeq2 outputs
{
  #reading norm counts
  norm_counts <- read.delim(file = paste0(loc_counts, "normalised_counts.txt"),
                            sep = "\t",
                            stringsAsFactors = F)
  #norm_counts <- norm_counts[rowSums(as.matrix(norm_counts))>0,]
  
  #reading the DESeq2 analysis output table
  merged_results_list <- readRDS(file = paste0(loc_diffexprs, 
                                        "merged_results_list.rds"))
  
  contrast_groups <- list(c("Blood"), 
                          c("Spleen", "Lymph Node"), 
                          c("Intraepithelial Layer", 
                            "Lamina Propria Lymphocyte", "Peyer's Patches"), 
                          c("Kidney", "Lung", "Liver", "Pancreas"),
                          c("cellTypeT"),
                          c("cellTypeNT"))
  
  #ensuring that names in contrast_groups list are all lower letter and do not 
  # contain spaces or punctuation
  for(i in seq(1, length(contrast_groups)))
  {
    contrast_groups[[i]] <- tolower(unique(
      gsub("_.*|[[:space:]]|[[:punct:]]", "", contrast_groups[[i]])
      ))
  }
  rm(i)
  
  output_table <- read.delim(file = paste0(loc_diffexprs_merged, 
                                           "output_table.txt"), 
                             sep = "\t", stringsAsFactors = F)
  
  colnames(output_table) <- output_table[3,]
  output_table <- output_table[-(1:3),]
  
}

#creating a list of scatter plots for the multiplot script grouped by tissue
{
  data_groups <- gsub(pattern = "_[0-9]$", replacement = "", 
                             x = colnames(norm_counts), 
                             ignore.case = T)
  data_groups <- unique(data_groups[duplicated(x = data_groups)])
  
  scatter_list <- list()
  
  for(i in seq(1, length(data_groups))){
    plot_data <- norm_counts[, grepl(pattern = data_groups[i], 
                                     x = colnames(norm_counts), 
                                     ignore.case = T)]
    plot_data <- plot_data[rowSums(plot_data)>0,]
    rownames(plot_data) <- gsub(pattern = "\\.[0-9]+$", 
                                replacement = "", 
                                x = rownames(plot_data), 
                                ignore.case = T)
    plot_data <- log10(plot_data)
    plot_data[plot_data<0] <- 0
    
    combinations <- combn(x = ncol(plot_data), m = 2)
    plots <- list()
    
    for(j in seq(ncol(combinations))){
      data_x <- plot_data[, combinations[1,j]]
      data_y <- plot_data[, combinations[2,j]]
      
      scatter_plot <- ggplot(data = plot_data,
                             aes_(x = data_x,
                                  y = data_y)
                             ) + 
        geom_point() + 
        expand_limits(x = 0, y = 0) + 
        labs(x = colnames(plot_data)[combinations[1,j]], 
             y = colnames(plot_data)[combinations[2,j]], 
             title = "Log10 scaled, normalised counts") + 
        coord_fixed(ratio = 1, xlim = NULL, ylim = NULL, expand = T, 
                    clip = "on") + 
        scale_x_continuous(breaks = seq(floor(min(data_x)), 
                                        ceiling(max(data_x)), 
                                                1)
                           ) + 
        scale_y_continuous(breaks = seq(floor(min(data_y)), 
                                        ceiling(max(data_y)),
                                                1)
                           ) + 
        geom_abline(slope = 1, intercept = 0, colour = "grey") + 
        theme_light() + 
        theme(panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              plot.margin = unit(c(0.1,1,0.5,0.1), "cm")
              )
      
      plots[[length(plots)+1]] <- scatter_plot
      names(plots)[length(plots)] <- paste0(
        colnames(plot_data)[combinations[1,j]], 
        "__vs__", 
        colnames(plot_data)[combinations[2,j]])
    }
    
    scatter_list[[length(scatter_list)+1]] <- plots
    names(scatter_list)[length(scatter_list)] <- data_groups[i]
  }
  
}

#exporting/writing images
{
  #constructing the path and creating directories
  path <- paste0(loc_plots, 
                 "Scatter Plots/")
  dir.create(path = path, recursive = T)
  
  for(i in seq(1, length(scatter_list))){
    
    # ggsave(filename = paste0(names(scatter_list)[i], ".png"),
    #        plot = multiplot(plotlist = scatter_list[[i]],
    #                         cols = ceiling(sqrt(length(scatter_list[[i]])))
    #                         ),
    #        device = "png",
    #        path = path,
    #        #scale = 1,
    #        width = 6.75,
    #        height = 5.75,
    #        #units = "mm",
    #        dpi = 320,
    #        limitsize = T)
    
    # ggsave(filename = paste0(names(scatter_list)[i], ".png"),
    #        plot = wrap_plots(scatter_list[[i]],
    #                          nrow = ceiling(sqrt(length(scatter_list[[i]]))),
    #                          ncol = ceiling(length(scatter_list[[i]]) /
    #                                           ceiling(sqrt(length(
    #                                             scatter_list[[i]])))),
    #                          byrow = T),
    #        device = "png",
    #        path = path,
    #        #scale = 1,
    #        #width = 7.75,
    #        #height = 6.75,
    #        #units = "mm",
    #        dpi = 320,
    #        limitsize = T)
    
    # ggsave(filename = paste0(names(scatter_list)[i], ".svg"), 
    #        plot = wrap_plots(scatter_list[[i]], 
    #                          ncol = ceiling(sqrt(length(scatter_list[[i]]))), 
    #                          nrow = ceiling(length(scatter_list[[i]]) / 
    #                                           ceiling(sqrt(length(
    #                                             scatter_list[[i]])))), 
    #                          byrow = T), 
    #        device = "svg", 
    #        path = path, 
    #        #scale = 1, 
    #        width = 7.75, 
    #        height = 6.75, 
    #        #units = "mm", 
    #        dpi = 320, 
    #        limitsize = T)
    
    save_plot(filename = paste0(names(scatter_list)[i], ".png"), 
              plot = plot_grid(plotlist = scatter_list[[i]], 
                               align = "none", 
                               ncol = 2), 
              ncol = 2, 
              nrow = length(scatter_list[[i]])/2, 
              #base_height = 6.75, 
              #base_width = 7.75,
              #base_asp = , 
              device = "png",
              path = path,
              limitsize = T)
    
    save_plot(filename = paste0(names(scatter_list)[i], ".svg"), 
              plot = plot_grid(plotlist = scatter_list[[i]], 
                               align = "none", 
                               ncol = 2), 
              ncol = 2, 
              nrow = length(scatter_list[[i]])/2, 
              #base_height = 6.75, 
              #base_width = 7.75,
              #base_asp = , 
              device = "svg",
              path = path,
              limitsize = T)
    
  }
  rm(i)
  
}


#creating a list of scatter plots of high variance genes to write via above code
{
  data_groups <- gsub(pattern = "_[0-9]$", replacement = "", 
                             x = colnames(norm_counts), 
                             ignore.case = T)
  data_groups <- unique(data_groups[duplicated(x = data_groups)])
  
  scatter_list <- list()
  
  for(i in seq(1, length(data_groups))){
    plot_data <- norm_counts[, grepl(pattern = data_groups[i], 
                                     x = colnames(norm_counts), 
                                     ignore.case = T)]
    #plot_data <- plot_data[rowSums(plot_data)>0,]
    
    #using CoV as a filter to select high variability genes with atleast 6 reads
    plot_data <- plot_data[
      rowSums(as.matrix(plot_data)) > 0 & 
        (rowSds(as.matrix(plot_data))/rowMeans(as.matrix(plot_data))
         > quantile(x = rowSds(as.matrix(plot_data))/
                      rowMeans(as.matrix(plot_data)), 
                    prob = 0.75, 
                    na.rm = T))
      ,]
    
    rownames(plot_data) <- gsub(pattern = "\\.[0-9]+$", 
                                replacement = "", 
                                x = rownames(plot_data), 
                                ignore.case = T)
    plot_data <- log10(plot_data)
    plot_data[plot_data<0] <- 0
    
    combinations <- combn(x = ncol(plot_data), m = 2)
    plots <- list()
    
    for(j in seq(ncol(combinations))){
      data_x <- plot_data[, combinations[1,j]]
      data_y <- plot_data[, combinations[2,j]]
      
      scatter_plot <- ggplot(data = plot_data,
                             aes_(x = data_x,
                                  y = data_y)
                             ) + 
        geom_point() + 
        expand_limits(x = 0, y = 0) + 
        labs(x = colnames(plot_data)[combinations[1,j]], 
             y = colnames(plot_data)[combinations[2,j]], 
             title = "Log10 scaled, normalised counts") + 
        coord_fixed(ratio = 1, xlim = NULL, ylim = NULL, expand = T, 
                    clip = "on") + 
        scale_x_continuous(breaks = seq(floor(min(data_x)), 
                                        ceiling(max(data_x)), 
                                                1)
                           ) + 
        scale_y_continuous(breaks = seq(floor(min(data_y)), 
                                        ceiling(max(data_y)),
                                                1)
                           ) + 
        geom_abline(slope = 1, intercept = 0, colour = "grey") + 
        theme_light() + 
        theme(panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              plot.margin = unit(c(0.1,1,0.5,0.1), "cm")
              )
      
      plots[[length(plots)+1]] <- scatter_plot
      names(plots)[length(plots)] <- paste0(
        colnames(plot_data)[combinations[1,j]], 
        "__vs__", 
        colnames(plot_data)[combinations[2,j]])
    }
    
    scatter_list[[length(scatter_list)+1]] <- plots
    names(scatter_list)[length(scatter_list)] <- data_groups[i]
  }
  rm(i, j)
  
}

#creating a list of scatter plots of normalised gene counts with high variance 
# genes highlighted and then write via above 'exporting/writing images' code
{
  data_groups <- gsub(pattern = "_[0-9]$", replacement = "", 
                             x = colnames(norm_counts), 
                             ignore.case = T)
  data_groups <- unique(data_groups[duplicated(x = data_groups)])
  
  scatter_list <- list()
  
  for(i in seq(1, length(data_groups))){
    plot_data <- norm_counts[, grepl(pattern = data_groups[i], 
                                     x = colnames(norm_counts), 
                                     ignore.case = T)]
    plot_data <- plot_data[rowSums(plot_data)>0,]
    
    #using CoV as a filter to select high variability genes with atleast n reads
    # plot_data <- plot_data[
    #   rowSums(as.matrix(plot_data)) > 10 & 
    #     (rowSds(as.matrix(plot_data))/rowMeans(as.matrix(plot_data))
    #      > quantile(x = rowSds(as.matrix(plot_data))/
    #                   rowMeans(as.matrix(plot_data)), 
    #                 prob = 0.75, 
    #                 na.rm = T))
    #   ,]
    
    plot_data$Group <- ifelse(rowSums(as.matrix(plot_data)) > 10 & 
                                (rowSds(as.matrix(plot_data))/
                                   rowMeans(as.matrix(plot_data))
                                 > quantile(x = rowSds(as.matrix(plot_data))/
                                              rowMeans(as.matrix(plot_data)), 
                                            prob = 0.75, 
                                            na.rm = T)), "High CoV", "Low CoV")
    
    rownames(plot_data) <- gsub(pattern = "\\.[0-9]+$", 
                                replacement = "", 
                                x = rownames(plot_data), 
                                ignore.case = T)
    plot_data[,1:(ncol(plot_data)-1)] <- 
      log10(plot_data[,1:(ncol(plot_data)-1)])
    repl <- (plot_data[,1:(ncol(plot_data)-1)])
    repl[repl<0] <- 0
    plot_data[,1:(ncol(plot_data)-1)] <- repl
    
    plot_data <- plot_data[order(plot_data$Group, decreasing = T),]
    
    combinations <- combn(x = ncol(plot_data)-1, m = 2)
    plots <- list()
    
    for(j in seq(ncol(combinations))){
      data_x <- plot_data[, combinations[1,j]]
      data_y <- plot_data[, combinations[2,j]]
      
      #scatter plot
      scatter_plot <- ggplot(data = plot_data,
                             aes_(x = data_x,
                                  y = data_y)
                             ) + 
        geom_point(aes_(colour = factor(plot_data$Group))) + 
        expand_limits(x = 0, y = 0) + 
        labs(x = colnames(plot_data)[combinations[1,j]], 
             y = colnames(plot_data)[combinations[2,j]], 
             title = "Log10 scaled, normalised counts",
             colour = "Variation") + 
        coord_fixed(ratio = 1, xlim = NULL, ylim = NULL, expand = T, 
                    clip = "on") + 
        scale_x_continuous(breaks = seq(floor(min(data_x)), 
                                        ceiling(max(data_x)), 
                                                1)
                           ) + 
        scale_y_continuous(breaks = seq(floor(min(data_y)), 
                                        ceiling(max(data_y)),
                                                1)
                           ) + 
        scale_colour_manual(values = c("black", "grey")) + 
        geom_abline(slope = 1, intercept = 0, colour = "darkgrey") + 
        theme_light() + 
        theme(panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              plot.margin = unit(c(0.1,1,0.5,0.1), "cm")
              )
      
      #plots list
      
      plots[[length(plots)+1]] <- scatter_plot
      names(plots)[length(plots)] <- paste0(
        colnames(plot_data)[combinations[1,j]], 
        "__vs__", 
        colnames(plot_data)[combinations[2,j]])
      
      #hex density plot
      hex_plot <- ggplot(data = plot_data,
                         aes_(x = data_x,
                              y = data_y)
                         ) + 
        geom_hex(bins = 30)  + 
        expand_limits(x = 0, y = 0) + 
        labs(x = colnames(plot_data)[combinations[1,j]], 
             y = colnames(plot_data)[combinations[2,j]], 
             title = "Density Plot", 
             fill = "Density") + 
        coord_fixed(ratio = 1, xlim = NULL, ylim = NULL, expand = T, 
                    clip = "on") + 
        scale_fill_distiller(palette = "Spectral") + 
        scale_x_continuous(breaks = seq(floor(min(data_x)), 
                                        ceiling(max(data_x)), 
                                                1)
                           ) + 
        scale_y_continuous(breaks = seq(floor(min(data_y)), 
                                        ceiling(max(data_y)),
                                                1)
                           ) + 
        theme_light() + 
        theme(panel.grid.major = element_blank(), 
              panel.grid.minor = element_blank(),
              plot.margin = unit(c(0.1,1,0.5,0.1), "cm")
              )
      
      #hex_plot
      
      plots[[length(plots)+1]] <- hex_plot
      names(plots)[length(plots)] <- paste0("Density_", 
        colnames(plot_data)[combinations[1,j]], 
        "__vs__", 
        colnames(plot_data)[combinations[2,j]])
      
    }
    
    scatter_list[[length(scatter_list)+1]] <- plots
    names(scatter_list)[length(scatter_list)] <- data_groups[i]
  }
  rm(i, j)
  
}



#comparing immunoglobulin expression across samples
{
  # temp <- output_table[grepl(pattern = "immunoglobulin", x = output_table$MGI.description, ignore.case = T),]$Gene.stable.ID
  # 
  # temp <- subset(norm_counts, gsub(pattern = "\\.[0-9]+$", replacement = "", x = rownames(norm_counts), ignore.case = T) %in% temp)
  # 
  # temp <- temp[,!grepl(pattern = "_NT", x = colnames(temp), ignore.case = T)]
  # 
  # temp$CoV <- rowSds(as.matrix(temp))/rowMeans(as.matrix(temp))
  # summary(temp$CoV)
  # 
  # temp2 <- temp[temp$CoV > 1 & temp$Means > 5,]
  # temp2$SD <- rowSds(as.matrix(temp2[,1:30]))
  # temp2$Gene.stable.ID <- gsub(pattern = "\\.[0-9]+$", replacement = "", x = rownames(temp2), ignore.case = T)
  # 
  # temp2 <- merge(temp2, output_table[,1:3], by = "Gene.stable.ID", all.x = T, all.y = F)
  # 
  # write.table(temp2, "immunoglobulin_exprs.txt", quote = F, sep = "\t", col.names = T, row.names = F)
}

```


#plots to assess inflection point of filtering the genes -- ended up using the 
#d2UIK method
```{r}

groups <- colnames(norm_counts)[grepl(pattern = "_T_[0-9]+$", 
                                      x = colnames(norm_counts), 
                                      ignore.case = T)]
groups <- unique(gsub(pattern = "_[0-9]+$", 
                      replacement = "", 
                      x = groups, 
                      ignore.case = T))

xrange <- c(-5, 6)
yrange <- c(0, 28)


# png(filename = "logfc_vs_countspcell.png",
#     units = "px", width = 899, height = 552)
# png(filename = "logfc_vs_countspcell_shifted.png",
#     units = "px", width = 899, height = 552)
# png(filename = "cmean_logfc_vs_countspcell.png",
#     units = "px", width = 899, height = 552)
# png(filename = "cmean_logfc_vs_countspcell_shifted.png",
#     units = "px", width = 899, height = 552)
# png(filename = "log2fc_vs_log10countspcell.png", 
#     units = "px", width = 899, height = 552)
# png(filename = "cmeanlog2fc_vs_log10countspcell.png",
#     units = "px", width = 899, height = 552)
# png(filename = "cmeanlog2fc_vs_log10countspcell_v2.png",
#     units = "px", width = 899, height = 552)

plot(x = xrange, 
     y = yrange, 
     type = "n", 
     xlab = "Log10 least counts per cell per gene", 
     #ylab = "Log2 fold change between smallest and largest counts", 
     ylab = "Cumulative mean Log2 fold change between smallest and largest counts",
     cex.lab = 1.25, cex.axis = 1.25)

mean_counts_pcell <- 0 
suggested_counts_pcell <- 0

colours <- rainbow(length(groups))

for(i in seq(1, length(groups))){
  
  subset_data <- as.data.frame(norm_counts[, grepl(pattern = groups[i], 
                                                   x = colnames(norm_counts), 
                                                   ignore.case = T)])
  subset_data[subset_data == 0] <- NA
  subset_data <- subset_data[complete.cases(subset_data),]/2000
  subset_data$rowMins <- rowMin(as.matrix(subset_data))
  #subset_data[, 4][subset_data[, 4] == 0] <- NA
  #subset_data <- subset_data/2000
  subset_data$log2FC <- 
    log2(rowMaxs(as.matrix(subset_data[1<3])) / 
           rowMin(as.matrix(subset_data[,1:3])))
  
  subset_data <- subset_data[order(subset_data$log2FC, decreasing = T),]
  subset_data <- subset_data[order(subset_data$rowMins, decreasing = F),]
  
  subset_data$log2FC <- cumsum(subset_data$log2FC)/seq_along(subset_data$log2FC)
  
  lines(x = log10(subset_data$rowMins),
        y = subset_data$log2FC+((i-1)*2),
        type = "l", 
        #type = "b",
        col = colours[i])
  
  mean_counts_pcell <- mean_counts_pcell + 
    (mean(as.matrix(subset_data[,1:3]))/length(groups))
  
  # cnvrg_th <- ((0.1 * (max(subset_data$log2FC) - min(subset_data$log2FC))) / 
  #                100) + min(subset_data$log2FC)
  # 
  # print(cnvrg_th)
  
  # print(inflection::d2uik(x = subset_data$rowMeans, y = subset_data$log2FC))
  suggested_counts_pcell <- suggested_counts_pcell + 
    (inflection::d2uik(x = subset_data$rowMins, y = subset_data$log2FC)/
       length(groups))

  # print(suggested_counts_pcell)
  
}

#value of suggested_counts_pcell is: 0.004377984

legend(x = xrange[2], 
       y = yrange[2], 
       legend = rev(groups), 
       col = rev(colours), 
       xjust = 1, 
       cex = 1.25, 
       lty = 1)

abline(v = log10(mean_counts_pcell), col = "black", lty = 3)
text(x = log10(mean_counts_pcell),
     y = yrange[2],
     labels = paste0("\nMean counts per cell:\n", 
                     signif(x = mean_counts_pcell, digits = 4)),
     pos = 4, 
     cex = 1)

abline(v = log10(suggested_counts_pcell), col = "black", lty = 5)
text(x = log10(suggested_counts_pcell),
     # y = yrange[2]-5,
     y = yrange[2]-2,
     labels = paste0("\nMean elbow in\ncounts per cell:\n",
                     signif(x = suggested_counts_pcell, digits = 4)),
     pos = 4,
     cex = 1)

# dev.off()

#checking number of selected genes to get an idea of things
# for(i in seq(1, length(groups))){
#   
#   subset_data <- as.data.frame(norm_counts[, grepl(pattern = groups[i], 
#                                                    x = colnames(norm_counts), 
#                                                    ignore.case = T)])
#   subset_data[subset_data == 0] <- NA
#   subset_data <- subset_data[complete.cases(subset_data),]/2000
#   subset_data$rowMins <- rowMin(as.matrix(subset_data))
#   
#   print(paste0("Numer of selected genes: ",
#                dim(subset(subset_data,
#                           subset_data$rowMins > suggested_counts_pcell))[1]
#                ))
# }

#regression and confidence interval
{
  # x <- log10(subset_data$rowMins)
  # y <- subset_data$log2FC
  # 
  # lm.out <- lm(y ~ x)
  # newx = seq(min(x),max(x),by = 0.05)
  # conf_interval <- predict(lm.out, newdata=data.frame(x=newx), interval="confidence",
  #                          level = 0.95)
  # plot(x, y, xlab="x", ylab="y", main="Regression")
  # abline(lm.out, col="lightblue")
  # matlines(newx, conf_interval[,2:3], col = "blue", lty=2)

}

```


#custom deseq2 comparisons
```{r}

#reading in DESeq2 outputs
{
  #reading norm counts
  norm_counts <- read.delim(file = paste0(loc_counts, "normalised_counts.txt"),
                            sep = "\t",
                            stringsAsFactors = F)
  #norm_counts <- norm_counts[rowSums(as.matrix(norm_counts))>0,]
  
  #reading the DESeq2 analysis output table
  merged_results_list <- readRDS(file = paste0(loc_diffexprs, 
                                        "merged_results_list.rds"))
  
  contrast_groups <- list(c("Blood"), 
                          c("Spleen", "Lymph Node", "Peyer's Patches"), 
                          c("Intraepithelial Layer", 
                            "Lamina Propria Lymphocyte"), 
                          c("Kidney", "Lung", "Liver", "Pancreas"),
                          c("cellTypeT"),
                          c("cellTypeNT"))
  
  #ensuring that names in contrast_groups list are all lower letter and do not 
  # contain spaces or punctuation
  for(i in seq(1, length(contrast_groups)))
  {
    contrast_groups[[i]] <- tolower(unique(
      gsub("_.*|[[:space:]]|[[:punct:]]", "", contrast_groups[[i]])
      ))
  }
  rm(i)
  
  output_table <- read.delim(file = paste0(loc_diffexprs_merged, 
                                           "output_table.txt"), 
                             sep = "\t", stringsAsFactors = F)
  
  colnames(output_table) <- output_table[3,]
  output_table <- output_table[-(1:3),]
  
}

#additional genes of interest that need to be highlighted
genes_of_int <- c("CD44", "CD62L", "CD69", "CD103", "ICOS", "KLRG1", "NK1.1", 
                  "NRP1", "ST2", "PD1", "NK1")
genes_of_int <- paste0("\\b(", paste(genes_of_int, collapse = "|"), ")\\b")
genes_of_int <- gsub(pattern = "\\.", replacement = "\\\\.", x = genes_of_int)

#comparison between Liver_T_3 vs mean(Liver_T_1, Liver_T_2)
{
  
  annotations_liver <- data.frame(Outlier = c("No", "No", "Yes"), 
                                  row.names = rownames(
                                    annotations)[grep(
                                      pattern = "Liver_T", 
                                      x = rownames(annotations), 
                                      ignore.case = T)]
                                  )
  
  #initialising DESeq object
  print("initialising DESeq object")
  DESeq_object_liver <- DESeqDataSetFromMatrix(
    countData = counts_matrix[,grepl(pattern = "Liver_T", 
                                     x = colnames(counts_matrix), 
                                     ignore.case = T)],
    colData = annotations_liver, 
    design = ~ 0)
  
  #adding meta_data to the dataset
  print("adding meta_data to the dataset")
  mcols(DESeq_object_liver) <- DataFrame(mcols(DESeq_object_liver),  meta_data)
  
  #assigning formula to the design matrix
  design(DESeq_object_liver) <- formula(~Outlier)
  
  DESeq_object_liver <- DESeq(object = DESeq_object_liver,
                        test = "Wald",
                        betaPrior = F, 
                        minReplicatesForReplace = Inf, 
                        parallel = T)
  
  resultsNames(DESeq_object_liver)
  
  res_liver <- results(object = DESeq_object_liver, 
                       name = "Outlier_Yes_vs_No", 
                       cooksCutoff = F, 
                       independentFiltering = F, 
                       test = "Wald", 
                       parallel = T)
  
  res_liver <- as.data.frame(res_liver)
  res_liver <- res_liver[res_liver$baseMean>0,]
  
  #plotting volcano plot
  {
    plot_data <- res_liver
    
    plot_data$Gene.stable.ID <- gsub(pattern = "\\.[0-9]+$", 
                                     replacement = "", 
                                     x = rownames(plot_data))
    
    plot_data <- merge(x = plot_data, y = output_table[,1:2], 
                           by = "Gene.stable.ID", 
                           all.x = T, all.y = F, 
                           sort = F)
    
    plot_data <- plot_data[order(plot_data$log2FoldChange, decreasing = T),]
    
    plot_data$coloured <- as.factor(ifelse(
      ((plot_data$log2FoldChange <
         quantile(x = (plot_data$log2FoldChange), 0.01) |
         plot_data$log2FoldChange >
         quantile(x = (plot_data$log2FoldChange), 0.99))
      & (-log10(plot_data$padj) > quantile(x = -log10(plot_data$padj),
                                           0.99) &
           plot_data$padj < 0.05) | 
        grepl(pattern = genes_of_int, x = plot_data$MGI.symbol, ignore.case = T)
        # tolower(plot_data$MGI.symbol) %in% tolower(
        #   c("CD44", "CD26L", "CD69", "CD103", "ICOS", "KLRG1", "NK1.1", 
        #     "NRP1", "ST2", "PD1", "NK1"))
      ),
      "selected", "not_selected"))
    
    # plot_data$coloured <- as.factor(ifelse(
    #   (plot_data$log2FoldChange>1 | plot_data$log2FoldChange< -1) & 
    #     (plot_data$padj<0.05),
    #   "selected", "not_selected"
    # ))
    
    scatter_plot <- ggplot(data = plot_data, 
                           aes_(x = plot_data$log2FoldChange, 
                               y = -log10(plot_data$padj), 
                               color = plot_data$coloured, 
                               text = plot_data$MGI.symbol,
                               log2FC = plot_data$log2FoldChange,
                               sig = plot_data$padj
                               ),
                           ) + 
      geom_point() + 
      geom_text_repel(aes_(label = ifelse(
        plot_data$coloured == "selected",
        as.character(str_extract(plot_data$MGI.symbol, '\\w*')), '')),
        #the above picks up the first alias, therefore bring prefered aliases to
        #the beginning of the row in the output_table file
        #as.character(plot_data$MGI.symbol), '')),
        hjust = 1, vjust = 1.5, size = 4, segment.alpha = 0.25, 
        force = 1, direction = "both", show.legend = F, nudge_y = 5, 
        nudge_x = -1) + 
      expand_limits(x = 0, y = 0) + 
      scale_color_manual(labels = c("Other genes", "Genes of\ninterest"), 
                         values = c("grey", "darkviolet")) + 
      labs(colour = "Genes", 
           y = "-Log10 of Adjusted P.value", 
           x = "Log2 Fold Change", 
           title = "Liver_T_3 vs Liver_T_1 and Liver_T_2") + 
      #scale_fill_discrete(breaks = ("Genes of\ninterest", "Other genes")) + 
      theme_light()
    
    scatter_plot
    
  }
  
  #plotting the 125 significantly differentiall expressed genes ((log2FC >1 or 
  #< -1) and p.adj<0.05) from Liver_T_3 against Liver_NT_1 to see if they 
  #align at slope = 1
  diff_exprs_genes <- plot_data$Gene.stable.ID[
    which((plot_data$log2FoldChange>1 | plot_data$log2FoldChange< -1) &
            plot_data$padj<0.05)]
  
  diff_Exprs_liv <- plot_data[
    which((plot_data$log2FoldChange>1 | plot_data$log2FoldChange< -1) &
            plot_data$padj<0.05), ]
  
  diff_Exprs_liv <- merge(x = diff_Exprs_liv, y = output_table[,c(1,3)], 
                      by = "Gene.stable.ID", all.x = T, all.y = F)
  
  # diff_exprs_genes <- plot_data$Gene.stable.ID[
  #   which((plot_data$log2FoldChange > mean(plot_data$log2FoldChange)
  #          + sd(plot_data$log2FoldChange) | plot_data$log2FoldChange <
  #            mean(plot_data$log2FoldChange) - sd(plot_data$log2FoldChange)) &
  #           plot_data$padj<0.05)]
  
  #same as above but with Liver_T_1 with Liver_T_3
  {
    plot_norm_data <- norm_counts[which(
      gsub(pattern = "\\.[0-9]+$", 
           replacement = "", 
           x = rownames(norm_counts), 
           ignore.case = T) %in% 
        diff_exprs_genes)
      , grepl(pattern = "Liver_T_1|Liver_T_3", 
              x = colnames(norm_counts), 
              ignore.case = T)]
    
    plot_norm_data <- log10(plot_norm_data)
    plot_norm_data[plot_norm_data < 0] <- 0
    
    scatter_norm_plot_1 <- ggplot(data = plot_norm_data, 
                           aes_(x = plot_norm_data$Liver_T_3, 
                               y = plot_norm_data$Liver_T_1)
                           ) + 
      geom_point() + 
      expand_limits(x = 0, y = 0) + 
      labs(y = "Liver_T_1", 
           x = "Liver_T_3", 
           title = "Log10 scaled, normalised counts") + 
      coord_fixed(ratio = 1, xlim = NULL, ylim = NULL, expand = T, 
                  clip = "on") + 
      scale_x_continuous(breaks = seq(floor(min(plot_norm_data$Liver_T_3)), 
                                      ceiling(max(plot_norm_data$Liver_T_3)), 
                                              1)
                         ) + 
      scale_y_continuous(breaks = seq(floor(min(plot_norm_data$Liver_T_1)), 
                                      ceiling(max(plot_norm_data$Liver_T_1)),
                                              1)
                         ) + 
      geom_abline(slope = 1, intercept = 0, colour = "grey") + 
      theme_light() + 
      theme(panel.grid.major = element_blank(), 
            panel.grid.minor = element_blank()
            )
    
    #scatter_norm_plot_1
    
  }
  
  #same as above but with Liver_T_2 with Liver_T_3
  {
    plot_norm_data <- norm_counts[which(
      gsub(pattern = "\\.[0-9]+$", 
           replacement = "", 
           x = rownames(norm_counts), 
           ignore.case = T) %in% 
        diff_exprs_genes)
      , grepl(pattern = "Liver_T_2|Liver_T_3", 
              x = colnames(norm_counts), 
              ignore.case = T)]
    
    plot_norm_data <- log10(plot_norm_data)
    plot_norm_data[plot_norm_data < 0] <- 0
    
    scatter_norm_plot_2 <- ggplot(data = plot_norm_data, 
                           aes_(x = plot_norm_data$Liver_T_3, 
                               y = plot_norm_data$Liver_T_2)
                           ) + 
      geom_point() + 
      expand_limits(x = 0, y = 0) + 
      labs(y = "Liver_T_2", 
           x = "Liver_T_3", 
           title = "Log10 scaled, normalised counts") + 
      coord_fixed(ratio = 1, xlim = NULL, ylim = NULL, expand = T, 
                  clip = "on") + 
      scale_x_continuous(breaks = seq(floor(min(plot_norm_data$Liver_T_3)), 
                                      ceiling(max(plot_norm_data$Liver_T_3)), 
                                              1)
                         ) + 
      scale_y_continuous(breaks = seq(floor(min(plot_norm_data$Liver_T_2)), 
                                      ceiling(max(plot_norm_data$Liver_T_2)),
                                              1)
                         ) + 
      geom_abline(slope = 1, intercept = 0, colour = "grey") + 
      theme_light() + 
      theme(panel.grid.major = element_blank(), 
            panel.grid.minor = element_blank()
            )
    
    #scatter_norm_plot_2
    
  }
  
  #same as above but with Liver_T_1 with Liver_T_2
  {
    plot_norm_data <- norm_counts[which(
      gsub(pattern = "\\.[0-9]+$", 
           replacement = "", 
           x = rownames(norm_counts), 
           ignore.case = T) %in% 
        diff_exprs_genes)
      , grepl(pattern = "Liver_T_1|Liver_T_2", 
              x = colnames(norm_counts), 
              ignore.case = T)]
    
    plot_norm_data <- log10(plot_norm_data)
    plot_norm_data[plot_norm_data < 0] <- 0
    
    scatter_norm_plot_3 <- ggplot(data = plot_norm_data, 
                           aes_(x = plot_norm_data$Liver_T_2, 
                               y = plot_norm_data$Liver_T_1)
                           ) + 
      geom_point() + 
      expand_limits(x = 0, y = 0) + 
      labs(y = "Liver_T_1", 
           x = "Liver_T_2", 
           title = "Log10 scaled, normalised counts") + 
      coord_fixed(ratio = 1, xlim = NULL, ylim = NULL, expand = T, 
                  clip = "on") + 
      scale_x_continuous(breaks = seq(floor(min(plot_norm_data$Liver_T_2)), 
                                      ceiling(max(plot_norm_data$Liver_T_2)), 
                                              1)
                         ) + 
      scale_y_continuous(breaks = seq(floor(min(plot_norm_data$Liver_T_1)), 
                                      ceiling(max(plot_norm_data$Liver_T_1)),
                                              1)
                         ) + 
      # labs(caption = 
      #        "Liver_T_1/2/3: Liver regulatory T-cell samples 1,2 or 3") + 
      geom_abline(slope = 1, intercept = 0, colour = "grey") + 
      theme_light() + 
      theme(panel.grid.major = element_blank(), 
            panel.grid.minor = element_blank()
            )
    
    #scatter_norm_plot_3
    
  }
  
  #same as above but with Liver_T_3 with FANTOM5 juvenile liver
  {
    FANTOM <- read.delim(file = paste0(loc_FANTOM, 
                                       "E-MTAB-3579-query-results.tpms.tsv"), 
                         sep = "\t", stringsAsFactors = F, skip = 4)
    
    FANTOM <- cbind(FANTOM[,1:2], 
                    FANTOM[,grepl(pattern = "^juvenile.*liver$", 
                                  x = colnames(FANTOM), 
                                  ignore.case = T), drop = F])
    
    TPM_counts <- counts_matrix/meta_data$Length
    TPM_counts <- t(t(TPM_counts)*1e6/colSums(TPM_counts))
    
    plot_norm_data <- as.data.frame(TPM_counts[which(
      gsub(pattern = "\\.[0-9]+$", 
           replacement = "", 
           x = rownames(TPM_counts), 
           ignore.case = T) %in% 
        diff_exprs_genes)
      , grepl(pattern = "Liver_T_3", 
              x = colnames(TPM_counts), 
              ignore.case = T), drop = F])
    
    plot_norm_data$Gene.ID <- gsub(pattern = "\\.[0-9]+$", 
                                          replacement = "", 
                                          x = rownames(plot_norm_data), 
                                          ignore.case = T)
    
    plot_norm_data <- merge(x = plot_norm_data, y = FANTOM, 
                            by = "Gene.ID", all.x = T, all.y = F)
    rownames(plot_norm_data) <- plot_norm_data$Gene.ID
    plot_norm_data$Gene.Name <- NULL
    plot_norm_data$Gene.ID <- NULL
    
    plot_norm_data[is.na(plot_norm_data)] <- 0
    #plot_norm_data <- log10(plot_norm_data)
    plot_norm_data[plot_norm_data < 0] <- 0
    #plot_norm_data[plot_norm_data == -Inf | plot_norm_data == Inf] <- 0
    
    scatter_norm_plot_4 <- ggplot(data = plot_norm_data, 
                           aes_(x = plot_norm_data$juvenile..liver, 
                               y = plot_norm_data$Liver_T_3)
                           ) + 
      geom_point() + 
      expand_limits(x = 0, y = 0) + 
      labs(y = "Liver_T_3", 
           x = "juvenile..liver", 
           title = "Log10 scaled TPM", 
           subtitle = "(Transcripts per million)") + 
      # coord_fixed(ratio = 1, xlim = NULL, ylim = NULL, expand = T, 
      #             clip = "on") + 
      scale_x_continuous(breaks = seq(floor(min(plot_norm_data$juvenile..liver)), 
                                      ceiling(max(plot_norm_data$juvenile..liver)), 
                                              1)
                         ) + 
      scale_y_continuous(breaks = seq(floor(min(plot_norm_data$Liver_T_3)), 
                                      ceiling(max(plot_norm_data$Liver_T_3)),
                                              1)
                         ) + 
      # labs(caption = 
      #        "Liver_T_1/2/3: Liver regulatory T-cell samples 1,2 or 3") + 
      geom_abline(slope = 1, intercept = 0, colour = "grey") + 
      theme_light() + 
      theme(panel.grid.major = element_blank(), 
            panel.grid.minor = element_blank()
            )
    
    #scatter_norm_plot_4
    
  }
  
  
  multiplot(scatter_norm_plot_1, scatter_norm_plot_3, scatter_norm_plot_2, 
            cols = 2)
  
}


#comparison between PeyersPatches_T_1 vs mean(PeyersPatches_T_2, 
#PeyersPatches_T_3)
{
  
  annotations_peyers <- data.frame(Outlier = c("Yes", "No", "No"), 
                                  row.names = rownames(
                                    annotations)[grep(
                                      pattern = "PeyersPatches_T", 
                                      x = rownames(annotations), 
                                      ignore.case = T)]
                                  )
  
  #initialising DESeq object
  print("initialising DESeq object")
  DESeq_object_peyers <- DESeqDataSetFromMatrix(
    countData = counts_matrix[,grepl(pattern = "PeyersPatches_T", 
                                     x = colnames(counts_matrix), 
                                     ignore.case = T)],
    colData = annotations_peyers, 
    design = ~ 0)
  
  #adding meta_data to the dataset
  print("adding meta_data to the dataset")
  mcols(DESeq_object_peyers) <- DataFrame(mcols(DESeq_object_peyers),  meta_data)
  
  #assigning formula to the design matrix
  design(DESeq_object_peyers) <- formula(~Outlier)
  
  DESeq_object_peyers <- DESeq(object = DESeq_object_peyers,
                        test = "Wald",
                        betaPrior = F, 
                        minReplicatesForReplace = Inf, 
                        parallel = T)
  
  resultsNames(DESeq_object_peyers)
  
  res_peyers <- results(object = DESeq_object_peyers, 
                       name = "Outlier_Yes_vs_No", 
                       cooksCutoff = F, 
                       independentFiltering = F, 
                       test = "Wald", 
                       parallel = T)
  
  res_peyers <- as.data.frame(res_peyers)
  res_peyers <- res_peyers[res_peyers$baseMean>0,]
  
  #plotting volcano plot
  {
    plot_data <- res_peyers
    
    plot_data$Gene.stable.ID <- gsub(pattern = "\\.[0-9]+$", 
                                     replacement = "", 
                                     x = rownames(plot_data))
    
    plot_data <- merge(x = plot_data, y = output_table[,1:2], 
                           by = "Gene.stable.ID", 
                           all.x = T, all.y = F, 
                           sort = F)
    
    plot_data <- plot_data[order(plot_data$log2FoldChange, decreasing = T),]
    
    plot_data$coloured <- as.factor(ifelse(
      ((plot_data$log2FoldChange <
         quantile(x = (plot_data$log2FoldChange), 0.01) |
         plot_data$log2FoldChange >
         quantile(x = (plot_data$log2FoldChange), 0.99))
      & (-log10(plot_data$padj) > quantile(x = -log10(plot_data$padj),
                                           0.99) &
           plot_data$padj < 0.05) | 
        grepl(pattern = genes_of_int, x = plot_data$MGI.symbol, ignore.case = T)
        # tolower(plot_data$MGI.symbol) %in% tolower(
        #   c("CD44", "CD26L", "CD69", "CD103", "ICOS", "KLRG1", "NK1.1", 
        #     "NRP1", "ST2", "PD1", "NK1"))
      ),
      "selected", "not_selected"))
    
    # plot_data$coloured <- as.factor(ifelse(
    #   (plot_data$log2FoldChange>1 | plot_data$log2FoldChange< -1) & 
    #     (plot_data$padj<0.05),
    #   "selected", "not_selected"
    # ))
    
    scatter_plot <- ggplot(data = plot_data, 
                           aes_(x = plot_data$log2FoldChange, 
                               y = -log10(plot_data$padj), 
                               color = plot_data$coloured, 
                               text = plot_data$MGI.symbol,
                               log2FC = plot_data$log2FoldChange,
                               sig = plot_data$padj
                               ),
                           ) + 
      geom_point() + 
      geom_text_repel(aes_(label = ifelse(
        plot_data$coloured == "selected",
        as.character(str_extract(plot_data$MGI.symbol, '\\w*')), '')),
        hjust = 1, vjust = 1.5, size = 4, segment.alpha = 0.25, 
        force = 0.1, direction = "both", show.legend = F) + 
      expand_limits(x = 0, y = 0) + 
      scale_color_manual(values = c("grey", "darkviolet")) + 
      labs(colour = "Contrasts", 
           y = "-Log10 of Adjusted P.value", 
           x = "Log2 Fold Change", 
           title = paste0("PeyersPatches_T_1 vs PeyersPatches_T_2 and ", 
                          "PeyersPatches_T_3")) + 
      theme_light()
    
    scatter_plot
    
  }
  
  #plotting the 125 significantly differentiall expressed genes ((log2FC >1 or 
  #< -1) and p.adj<0.05) from PeyersPatches_T_1 against PeyersPatches_NT_1 to 
  #see if they align at slope = 1
  diff_exprs_genes <- plot_data$Gene.stable.ID[
      which((plot_data$log2FoldChange > mean(plot_data$log2FoldChange) 
             + sd(plot_data$log2FoldChange) | plot_data$log2FoldChange < 
               mean(plot_data$log2FoldChange) - sd(plot_data$log2FoldChange)) & 
              plot_data$padj<0.05)]
  
  diff_Exprs_pp <- plot_data[
    which((plot_data$log2FoldChange>1 | plot_data$log2FoldChange< -1) &
            plot_data$padj<0.05), ]
  
  diff_Exprs_pp <- merge(x = diff_Exprs_pp, y = output_table[,c(1,3)], 
                      by = "Gene.stable.ID", all.x = T, all.y = F)
  
  #same as above but with PeyersPatches_T_3 with PeyersPatches_T_1
  {
    plot_norm_data <- norm_counts[which(
      gsub(pattern = "\\.[0-9]+$", 
           replacement = "", 
           x = rownames(norm_counts), 
           ignore.case = T) %in% 
        diff_exprs_genes)
      , grepl(pattern = "PeyersPatches_T_3|PeyersPatches_T_1", 
              x = colnames(norm_counts), 
              ignore.case = T)]
    
    plot_norm_data <- log10(plot_norm_data)
    plot_norm_data[plot_norm_data < 0] <- 0
    
    scatter_norm_plot_1 <- ggplot(data = plot_norm_data, 
                           aes_(x = plot_norm_data$PeyersPatches_T_1, 
                               y = plot_norm_data$PeyersPatches_T_3)
                           ) + 
      geom_point() + 
      expand_limits(x = 0, y = 0) + 
      labs(y = "PeyersPatches_T_3", 
           x = "PeyersPatches_T_1", 
           title = "Log10 scaled, normalised counts") + 
      coord_fixed(ratio = 1, xlim = NULL, ylim = NULL, expand = T, 
                  clip = "on") + 
      scale_x_continuous(breaks = seq(floor(min(plot_norm_data$PeyersPatches_T_1)), 
                                      ceiling(max(plot_norm_data$PeyersPatches_T_1)), 
                                              1)
                         ) + 
      scale_y_continuous(breaks = seq(floor(min(plot_norm_data$PeyersPatches_T_3)), 
                                      ceiling(max(plot_norm_data$PeyersPatches_T_3)),
                                              1)
                         ) + 
      geom_abline(slope = 1, intercept = 0, colour = "grey") + 
      theme_light() + 
      theme(panel.grid.major = element_blank(), 
            panel.grid.minor = element_blank()
            )
    
    #scatter_norm_plot_1
    
  }
  
  #same as above but with PeyersPatches_T_2 with PeyersPatches_T_1
  {
    plot_norm_data <- norm_counts[which(
      gsub(pattern = "\\.[0-9]+$", 
           replacement = "", 
           x = rownames(norm_counts), 
           ignore.case = T) %in% 
        diff_exprs_genes)
      , grepl(pattern = "PeyersPatches_T_2|PeyersPatches_T_1", 
              x = colnames(norm_counts), 
              ignore.case = T)]
    
    plot_norm_data <- log10(plot_norm_data)
    plot_norm_data[plot_norm_data < 0] <- 0
    
    scatter_norm_plot_2 <- ggplot(data = plot_norm_data, 
                           aes_(x = plot_norm_data$PeyersPatches_T_1, 
                               y = plot_norm_data$PeyersPatches_T_2)
                           ) + 
      geom_point() + 
      expand_limits(x = 0, y = 0) + 
      labs(y = "PeyersPatches_T_2", 
           x = "PeyersPatches_T_1", 
           title = "Log10 scaled, normalised counts") + 
      coord_fixed(ratio = 1, xlim = NULL, ylim = NULL, expand = T, 
                  clip = "on") + 
      scale_x_continuous(breaks = seq(floor(min(plot_norm_data$PeyersPatches_T_1)), 
                                      ceiling(max(plot_norm_data$PeyersPatches_T_1)), 
                                              1)
                         ) + 
      scale_y_continuous(breaks = seq(floor(min(plot_norm_data$PeyersPatches_T_2)), 
                                      ceiling(max(plot_norm_data$PeyersPatches_T_2)),
                                              1)
                         ) + 
      geom_abline(slope = 1, intercept = 0, colour = "grey") + 
      theme_light() + 
      theme(panel.grid.major = element_blank(), 
            panel.grid.minor = element_blank()
            )
    
    #scatter_norm_plot_2
    
  }
  
  #same as above but with PeyersPatches_T_3 with PeyersPatches_T_2
  {
    plot_norm_data <- norm_counts[which(
      gsub(pattern = "\\.[0-9]+$", 
           replacement = "", 
           x = rownames(norm_counts), 
           ignore.case = T) %in% 
        diff_exprs_genes)
      , grepl(pattern = "PeyersPatches_T_3|PeyersPatches_T_2", 
              x = colnames(norm_counts), 
              ignore.case = T)]
    
    plot_norm_data <- log10(plot_norm_data)
    plot_norm_data[plot_norm_data < 0] <- 0
    
    scatter_norm_plot_3 <- ggplot(data = plot_norm_data, 
                           aes_(x = plot_norm_data$PeyersPatches_T_2, 
                               y = plot_norm_data$PeyersPatches_T_3)
                           ) + 
      geom_point() + 
      expand_limits(x = 0, y = 0) + 
      labs(y = "PeyersPatches_T_3", 
           x = "PeyersPatches_T_2", 
           title = "Log10 scaled, normalised counts") + 
      coord_fixed(ratio = 1, xlim = NULL, ylim = NULL, expand = T, 
                  clip = "on") + 
      scale_x_continuous(breaks = seq(floor(min(plot_norm_data$PeyersPatches_T_2)), 
                                      ceiling(max(plot_norm_data$PeyersPatches_T_2)), 
                                              1)
                         ) + 
      scale_y_continuous(breaks = seq(floor(min(plot_norm_data$PeyersPatches_T_3)), 
                                      ceiling(max(plot_norm_data$PeyersPatches_T_3)),
                                              1)
                         ) + 
      # labs(caption = 
      #        "PeyersPatches_T_1/2/3: PeyersPatches regulatory T-cell samples 1,2 or 3") + 
      geom_abline(slope = 1, intercept = 0, colour = "grey") + 
      theme_light() + 
      theme(panel.grid.major = element_blank(), 
            panel.grid.minor = element_blank()
            )
    
    #scatter_norm_plot_3
    
  }

  multiplot(scatter_norm_plot_1, scatter_norm_plot_3, scatter_norm_plot_2, 
            cols = 2)
  
}


#comparison between Pancreas_T_3 vs mean(Pancreas_T_1, Pancreas_T_2)
{
  
  annotations_pancreas <- data.frame(Outlier = c("No", "No", "Yes"), 
                                  row.names = rownames(
                                    annotations)[grep(
                                      pattern = "Pancreas_T", 
                                      x = rownames(annotations), 
                                      ignore.case = T)]
                                  )
  
  #initialising DESeq object
  print("initialising DESeq object")
  DESeq_object_pancreas <- DESeqDataSetFromMatrix(
    countData = counts_matrix[,grepl(pattern = "Pancreas_T", 
                                     x = colnames(counts_matrix), 
                                     ignore.case = T)],
    colData = annotations_pancreas, 
    design = ~ 0)
  
  #adding meta_data to the dataset
  print("adding meta_data to the dataset")
  mcols(DESeq_object_pancreas) <- DataFrame(mcols(DESeq_object_pancreas),  meta_data)
  
  #assigning formula to the design matrix
  design(DESeq_object_pancreas) <- formula(~Outlier)
  
  DESeq_object_pancreas <- DESeq(object = DESeq_object_pancreas,
                        test = "Wald",
                        betaPrior = F, 
                        minReplicatesForReplace = Inf, 
                        parallel = T)
  
  resultsNames(DESeq_object_pancreas)
  
  res_pancreas <- results(object = DESeq_object_pancreas, 
                       name = "Outlier_Yes_vs_No", 
                       cooksCutoff = F, 
                       independentFiltering = F, 
                       test = "Wald", 
                       parallel = T)
  
  res_pancreas <- as.data.frame(res_pancreas)
  res_pancreas <- res_pancreas[res_pancreas$baseMean>0,]
  
  #plotting volcano plot
  {
    plot_data <- res_pancreas
    
    plot_data$Gene.stable.ID <- gsub(pattern = "\\.[0-9]+$", 
                                     replacement = "", 
                                     x = rownames(plot_data))
    
    plot_data <- merge(x = plot_data, y = output_table[,1:2], 
                           by = "Gene.stable.ID", 
                           all.x = T, all.y = F, 
                           sort = F)
    
    plot_data <- plot_data[order(plot_data$log2FoldChange, decreasing = T),]
    
    plot_data$coloured <- as.factor(ifelse(
      ((plot_data$log2FoldChange <
         quantile(x = (plot_data$log2FoldChange), 0.01) |
         plot_data$log2FoldChange >
         quantile(x = (plot_data$log2FoldChange), 0.99))
      & (-log10(plot_data$padj) > quantile(x = -log10(plot_data$padj),
                                           0.99) &
           plot_data$padj < 0.05) | 
        tolower(plot_data$MGI.symbol) %in% tolower(
          c("CD44", "CD26L", "CD69", "CD103", "ICOS", "KLRG1", "NK1.1", 
            "NRP1", "ST2", "PD1", "NK1"))
      ),
      "selected", "not_selected"))
    
    # plot_data$coloured <- as.factor(ifelse(
    #   (plot_data$log2FoldChange>1 | plot_data$log2FoldChange< -1) & 
    #     (plot_data$padj<0.05),
    #   "selected", "not_selected"
    # ))
    
    scatter_plot <- ggplot(data = plot_data, 
                           aes_(x = plot_data$log2FoldChange, 
                               y = -log10(plot_data$padj), 
                               color = plot_data$coloured, 
                               text = plot_data$MGI.symbol,
                               log2FC = plot_data$log2FoldChange,
                               sig = plot_data$padj
                               ),
                           ) + 
      geom_point() + 
      geom_text_repel(aes_(label = ifelse(
        plot_data$coloured == "selected",
        as.character(plot_data$MGI.symbol), '')),
        hjust = 1, vjust = 1.5, size = 4, segment.alpha = 0.25, 
        force = 0.1, direction = "both", show.legend = F) + 
      expand_limits(x = 0, y = 0) + 
      scale_color_manual(values = c("grey", "darkviolet")) + 
      labs(colour = "Contrasts", 
           y = "-Log10 of Adjusted P.value", 
           x = "Log2 Fold Change", 
           title = "Pancreas_T_3 vs Pancreas_T_1 and Pancreas_T_2") + 
      #scale_fill_discrete(name = "Colour", labels = c("", "1% by quartile")) + 
      theme_light()
    
    scatter_plot
    
  }
  
  #plotting the 125 significantly differentiall expressed genes ((log2FC >1 or 
  #< -1) and p.adj<0.05) from Pancreas_T_3 against Pancreas_NT_1 to see if they 
  #align at slope = 1
  diff_exprs_genes <- plot_data$Gene.stable.ID[
      which((plot_data$log2FoldChange > mean(plot_data$log2FoldChange) 
             + sd(plot_data$log2FoldChange) | plot_data$log2FoldChange < 
               mean(plot_data$log2FoldChange) - sd(plot_data$log2FoldChange)) & 
              plot_data$padj<0.05)]
  
  #same as above but with Pancreas_T_1 with Pancreas_T_3
  {
    plot_norm_data <- norm_counts[which(
      gsub(pattern = "\\.[0-9]+$", 
           replacement = "", 
           x = rownames(norm_counts), 
           ignore.case = T) %in% 
        diff_exprs_genes)
      , grepl(pattern = "Pancreas_T_1|Pancreas_T_3", 
              x = colnames(norm_counts), 
              ignore.case = T)]
    
    plot_norm_data <- log10(plot_norm_data)
    plot_norm_data[plot_norm_data < 0] <- 0
    
    scatter_norm_plot_1 <- ggplot(data = plot_norm_data, 
                           aes_(x = plot_norm_data$Pancreas_T_3, 
                               y = plot_norm_data$Pancreas_T_1)
                           ) + 
      geom_point() + 
      expand_limits(x = 0, y = 0) + 
      labs(y = "Pancreas_T_1", 
           x = "Pancreas_T_3", 
           title = "Log10 scaled, normalised counts") + 
      coord_fixed(ratio = 1, xlim = NULL, ylim = NULL, expand = T, 
                  clip = "on") + 
      scale_x_continuous(breaks = seq(floor(min(plot_norm_data$Pancreas_T_3)), 
                                      ceiling(max(plot_norm_data$Pancreas_T_3)), 
                                              1)
                         ) + 
      scale_y_continuous(breaks = seq(floor(min(plot_norm_data$Pancreas_T_1)), 
                                      ceiling(max(plot_norm_data$Pancreas_T_1)),
                                              1)
                         ) + 
      geom_abline(slope = 1, intercept = 0, colour = "grey") + 
      theme_light() + 
      theme(panel.grid.major = element_blank(), 
            panel.grid.minor = element_blank()
            )
    
    #scatter_norm_plot_1
    
  }
  
  #same as above but with Pancreas_T_2 with Pancreas_T_3
  {
    plot_norm_data <- norm_counts[which(
      gsub(pattern = "\\.[0-9]+$", 
           replacement = "", 
           x = rownames(norm_counts), 
           ignore.case = T) %in% 
        diff_exprs_genes)
      , grepl(pattern = "Pancreas_T_2|Pancreas_T_3", 
              x = colnames(norm_counts), 
              ignore.case = T)]
    
    plot_norm_data <- log10(plot_norm_data)
    plot_norm_data[plot_norm_data < 0] <- 0
    
    scatter_norm_plot_2 <- ggplot(data = plot_norm_data, 
                           aes_(x = plot_norm_data$Pancreas_T_3, 
                               y = plot_norm_data$Pancreas_T_2)
                           ) + 
      geom_point() + 
      expand_limits(x = 0, y = 0) + 
      labs(y = "Pancreas_T_2", 
           x = "Pancreas_T_3", 
           title = "Log10 scaled, normalised counts") + 
      coord_fixed(ratio = 1, xlim = NULL, ylim = NULL, expand = T, 
                  clip = "on") + 
      scale_x_continuous(breaks = seq(floor(min(plot_norm_data$Pancreas_T_3)), 
                                      ceiling(max(plot_norm_data$Pancreas_T_3)), 
                                              1)
                         ) + 
      scale_y_continuous(breaks = seq(floor(min(plot_norm_data$Pancreas_T_2)), 
                                      ceiling(max(plot_norm_data$Pancreas_T_2)),
                                              1)
                         ) + 
      geom_abline(slope = 1, intercept = 0, colour = "grey") + 
      theme_light() + 
      theme(panel.grid.major = element_blank(), 
            panel.grid.minor = element_blank()
            )
    
    #scatter_norm_plot_2
    
  }
  
  #same as above but with Pancreas_T_1 with Pancreas_T_2
  {
    plot_norm_data <- norm_counts[which(
      gsub(pattern = "\\.[0-9]+$", 
           replacement = "", 
           x = rownames(norm_counts), 
           ignore.case = T) %in% 
        diff_exprs_genes)
      , grepl(pattern = "Pancreas_T_1|Pancreas_T_2", 
              x = colnames(norm_counts), 
              ignore.case = T)]
    
    plot_norm_data <- log10(plot_norm_data)
    plot_norm_data[plot_norm_data < 0] <- 0
    
    scatter_norm_plot_3 <- ggplot(data = plot_norm_data, 
                           aes_(x = plot_norm_data$Pancreas_T_2, 
                               y = plot_norm_data$Pancreas_T_1)
                           ) + 
      geom_point() + 
      expand_limits(x = 0, y = 0) + 
      labs(y = "Pancreas_T_1", 
           x = "Pancreas_T_2", 
           title = "Log10 scaled, normalised counts") + 
      coord_fixed(ratio = 1, xlim = NULL, ylim = NULL, expand = T, 
                  clip = "on") + 
      scale_x_continuous(breaks = seq(floor(min(plot_norm_data$Pancreas_T_2)), 
                                      ceiling(max(plot_norm_data$Pancreas_T_2)), 
                                              1)
                         ) + 
      scale_y_continuous(breaks = seq(floor(min(plot_norm_data$Pancreas_T_1)), 
                                      ceiling(max(plot_norm_data$Pancreas_T_1)),
                                              1)
                         ) + 
      # labs(caption = 
      #        "Pancreas_T_1/2/3: Pancreas regulatory T-cell samples 1,2 or 3") + 
      geom_abline(slope = 1, intercept = 0, colour = "grey") + 
      theme_light() + 
      theme(panel.grid.major = element_blank(), 
            panel.grid.minor = element_blank()
            )
    
    #scatter_norm_plot_3
    
  }

  
  multiplot(scatter_norm_plot_1, scatter_norm_plot_3, scatter_norm_plot_2, 
            cols = 2)
  
}



#comparison of each t-reg sample with its companion samples in the same tissue
{
  
  #get tissue names
  tissues <- gsub(pattern = "_[0-9]+$", 
                  replacement = "", 
                  x = colnames(norm_counts), 
                  ignore.case = T)
  tissues <- unique(tissues[duplicated(tissues)])
  
  #list to store the results
  results_list <- list()
  
  #perform analysis on each tissue in 'tissues'
  for(i in seq(1, length(tissues))){
    
    writeLines(paste0("\nProcessing tissue: ", tissues[i]))
    
    #subset data
    data_subset <- counts_matrix[,grepl(pattern = tissues[i], 
                                        x = colnames(counts_matrix), 
                                        ignore.case = T)]
    
    if(ncol(data_subset) < 2){
      print("Not enough samples for contrasts. Proceeding to net iteration.")
      next
    }
    
    #comparisons to generate the contrasts
    comparisons <- t(combn(x = ncol(data_subset), 
                           m = ifelse(ncol(data_subset) > 2, 2, 1)))
    
    #list to store contrasts per tissue
    results_tissue <- list()
    
    #going through each available comparison for the current tissue
    for(j in seq(1, nrow(comparisons))){
      
      print(paste0("Current outlier: ", colnames(data_subset)[
        !seq(1, nrow(comparisons)) %in% comparisons[j,]])
        )
      
      #if there are no replicates, then do an estimated analysis, otherwise do
      # proper analysis
      if(ncol(data_subset) == 2){
        print(paste0("No replicates found. Using gene-wise estimates to get ", 
                     "estimated results"))
        
        data_subset <- cbind(data_subset, 
                             data_subset[,
                                         (seq(1, nrow(comparisons)) %in% 
                                            comparisons[j,]), 
                                         drop = F]
                             )
        colnames(data_subset)[ncol(data_subset)] <- "duplicate"
        
        #generate tissue annotations
        outlier <- vector(mode = "character", length = ncol(data_subset))
        outlier[seq(1, nrow(comparisons)) %in% comparisons[j,]] <- "No"
        outlier[outlier != "No"] <- "Yes"
        #rectifying the problem with above lines by reverting the duplicate
        # column to "Yes"
        outlier[3] <- "Yes"
        annotations_tissue <- data.frame(Outlier = outlier, 
                                    row.names = colnames(data_subset)
                                    )
        
        #generate the DESeq object
        DESeq_object <- DESeqDataSetFromMatrix(countData = data_subset, 
                                               colData = annotations_tissue, 
                                               design = ~Outlier)
        
        #because of duplication, error about gene-wise dispersion estimates 
        # being within 2 orders of magnitude. Error description recommends that
        # gene-wise estimates be used as final estimates via the below code
        DESeq_object <- estimateSizeFactors(DESeq_object)
        DESeq_object <- estimateDispersionsGeneEst(DESeq_object)
        dispersions(DESeq_object) <- mcols(DESeq_object)$dispGeneEst
        
        DESeq_object <- nbinomWaldTest(object = DESeq_object, 
                                       betaPrior = F)
        
        #restoring data_subset to its original form to continue analyses
        data_subset <- data_subset[,1:(ncol(data_subset)-1)]
        
      }
      else{
        #generate tissue annotations
        outlier <- vector(mode = "character", length = ncol(data_subset))
        outlier[seq(1, nrow(comparisons)) %in% comparisons[j,]] <- "No"
        outlier[outlier != "No"] <- "Yes"
        annotations_tissue <- data.frame(Outlier = outlier, 
                                    row.names = colnames(data_subset)
                                    )
        
        #generate the DESeq object
        DESeq_object <- DESeqDataSetFromMatrix(countData = data_subset, 
                                               colData = annotations_tissue, 
                                               design = ~Outlier)
        
        #running the analysis
        DESeq_object <- DESeq(object = DESeq_object, 
                              test = "Wald", 
                              betaPrior = F, 
                              minReplicatesForReplace = Inf, 
                              parallel = T)
        
      }
      
      resultsNames(DESeq_object)
      
      res_tissue <- results(object = DESeq_object, 
                            name = "Outlier_Yes_vs_No", 
                            cooksCutoff = F, 
                            independentFiltering = F, 
                            test = "Wald", 
                            parallel = T)
      
      res_tissue <- as.data.frame(res_tissue)
      res_tissue <- res_tissue[res_tissue$baseMean>0,]
      
      results_tissue[[length(results_tissue)+1]] <- res_tissue
      names(results_tissue)[length(results_tissue)] <- paste0(
        colnames(data_subset)[!seq(1, nrow(comparisons)) %in% comparisons[j,]], 
        "__vs__", 
        paste(colnames(
          data_subset)[seq(1, nrow(comparisons)) %in% comparisons[j,]], 
          collapse = "_and_")
      )
      
    }
    
    results_list[[length(results_list)+1]] <- results_tissue
    names(results_list)[length(results_list)] <- unique(
      gsub(pattern = "_.*$", 
           replacement = "", 
           x = colnames(data_subset), 
           ignore.case = T)
      )
    
  }
  rm(i, j)
  
  
  #creating a bar plot of the number of differentially expressed genes
  {
    
    diff_exprs_nums <- vector(mode = "numeric", length = 0)
    
    for(i in seq(1, length(results_list))){
      
      for(j in seq(1, length(results_list[[i]]))){
        
        result_obj <- results_list[[i]][[j]]
        
        diff_exprs_nums <- c(diff_exprs_nums, 
                             nrow(result_obj[result_obj$padj < 0.05 & 
                                               (result_obj$log2FoldChange > 1 | 
                                                  result_obj$log2FoldChange< -1)
                                             ,])
                             )
        
        names(diff_exprs_nums)[length(diff_exprs_nums)] <- gsub(
          pattern = "__vs__.*$", 
          replacement = "", 
          x = names(results_list[[i]][j]), 
          ignore.case = T)
        
      }
    }
    rm(i, j)
    
    plot_data <- data.frame(Samples = names(diff_exprs_nums),
                       No.of.Diff.Exprs = diff_exprs_nums, 
                       #row.names = names(diff_exprs_nums), 
                       stringsAsFactors = F)
    
    #aesthetic formatting of plot_data
    {
      plot_data$Samples <- gsub(pattern = "lymphnode", 
                            replacement = "LN", 
                            x =  plot_data$Samples, 
                            ignore.case = T)
    
      plot_data$Samples <- gsub(pattern = "intraepitheliallayer", 
                              replacement = "IEL", 
                              x =  plot_data$Samples, 
                              ignore.case = T)
      
      plot_data$Samples <- gsub(pattern = "laminaproprialymphocyte", 
                              replacement = "LPL", 
                              x =  plot_data$Samples, 
                              ignore.case = T)
      
      plot_data$Samples <- gsub(pattern = "peyerspatches", 
                              replacement = "PP", 
                              x =  plot_data$Samples, 
                              ignore.case = T)
    }
    
    plot_data$Group <- gsub(pattern = "_.*$", 
                            replacement = "", 
                            x = plot_data$Samples, 
                            ignore.case = T)
    
    # #a numeric vector to introduce gaps in the bars
    # x.seq <- seq(1,39)[!seq(1,39) %in% seq(0,38, 4)]
    
    # col_plot <- ggplot(data = plot_data, 
    #                    aes(x = Samples, 
    #                        y = No.of.Diff.Exprs, 
    #                        fill = Group)
    #                    ) + 
    #   geom_col() + 
    #   labs(x = "", 
    #        y = "Number of differentially expressed genes", 
    #        fill = "Tissue") + 
    #   # scale_x_discrete(breaks = NA) + 
    #   # geom_text(aes(x=c(sum(x.seq[1:2])/2, sum(x.seq[3:4])/2), y=0, 
    #   #               label=c("X","Y")), vjust=1.2, size=8) + 
    #   theme_light() + 
    #   theme(axis.text.x = element_text(angle = 90, hjust = 1), 
    #         panel.grid.major = element_blank(), 
    #         panel.grid.minor = element_blank(),
    #         legend.position = "none"
    #         )
    # 
    # col_plot
    
    col_point_plot <- ggplot(data = plot_data,
                             aes(x = Group,
                                 y = No.of.Diff.Exprs,
                                 colour = Group)
                             ) + 
      geom_point(size = 3) + 
      labs(x = "", 
           y = "Number of differentially expressed genes") + 
      theme_light() + 
      theme(legend.position = "none"
            )
    
    col_point_plot
    
  }
  
}


#comparison of each non-blood sample with blood samples
{

  #excluding CD4 data
  data_subset <- counts_matrix[,!grepl(pattern = "_NT", 
                                       x = colnames(counts_matrix), 
                                       ignore.case = T)]
  
  #removing empty rows
  data_subset <- data_subset[rowSums(data_subset) > 0,]
  
  #creating the initial annotations for the contrast
  annotations_tissue <- data.frame(Outlier = c("No", "No", "No", "Yes"))
  rownames(annotations_tissue) <- c(colnames(data_subset)[1:3], "outlier")
  
  #list to store the results
  results_list <- list()
  
  #iteratively analysing each sample against blood tregs
  for(i in seq(4, ncol(data_subset))){
    
    writeLines(paste0("\nProcessing tissue: ", colnames(data_subset)[i]))
    
    #updating the annotations for the sample in question
    rownames(annotations_tissue)[4] <- colnames(data_subset)[i]
    
    #generate the DESeq object
    DESeq_object <- DESeqDataSetFromMatrix(countData = 
                                             cbind(data_subset[,c(1,2,3,i)]), 
                                           colData = annotations_tissue, 
                                           design = ~Outlier)
    
    #running the analysis
    DESeq_object <- DESeq(object = DESeq_object, 
                          test = "Wald", 
                          betaPrior = F, 
                          minReplicatesForReplace = Inf, 
                          parallel = T)
    
    #resultsNames(DESeq_object)
    res_tissue <- results(object = DESeq_object, 
                          name = "Outlier_Yes_vs_No", 
                          cooksCutoff = F, 
                          independentFiltering = F, 
                          test = "Wald", 
                          parallel = T)
    
    #converting the results to a data frame for storage
    res_tissue <- as.data.frame(res_tissue)
    res_tissue <- res_tissue[res_tissue$baseMean>0,]
    
    results_list[[length(results_list)+1]] <- res_tissue
    names(results_list)[length(results_list)] <- rownames(annotations_tissue)[4]
    
  }
  rm(i)
  
  
  #creating a bar plot of the number of differentially expressed genes
  {
    
    diff_exprs_nums <- vector(mode = "numeric", length = 0)
    
    for(i in seq(1, length(results_list))){
      
      result_obj <- results_list[[i]]
      
      diff_exprs_nums <- c(diff_exprs_nums, 
                           nrow(result_obj[result_obj$padj < 0.05 & 
                                             (result_obj$log2FoldChange > 1 | 
                                                result_obj$log2FoldChange< -1)
                                           ,])
                           )
      
      names(diff_exprs_nums)[length(diff_exprs_nums)] <- gsub(
        pattern = "__vs__.*$", 
        replacement = "", 
        x = names(results_list[i]), 
        ignore.case = T)
      
    }
    rm(i)
    
    plot_data <- data.frame(Samples = names(diff_exprs_nums),
                       No.of.Diff.Exprs = diff_exprs_nums, 
                       #row.names = names(diff_exprs_nums), 
                       stringsAsFactors = F)
    
    #aesthetic formatting of plot_data
    {
      plot_data$Samples <- gsub(pattern = "lymphnode", 
                            replacement = "LN", 
                            x =  plot_data$Samples, 
                            ignore.case = T)
    
      plot_data$Samples <- gsub(pattern = "intraepitheliallayer", 
                              replacement = "IEL", 
                              x =  plot_data$Samples, 
                              ignore.case = T)
      
      plot_data$Samples <- gsub(pattern = "laminaproprialymphocyte", 
                              replacement = "LPL", 
                              x =  plot_data$Samples, 
                              ignore.case = T)
      
      plot_data$Samples <- gsub(pattern = "peyerspatches", 
                              replacement = "PP", 
                              x =  plot_data$Samples, 
                              ignore.case = T)
    }
    
    plot_data$Group <- gsub(pattern = "_.*$", 
                            replacement = "", 
                            x = plot_data$Samples, 
                            ignore.case = T)
    
    # #a numeric vector to introduce gaps in the bars
    # x.seq <- seq(1,39)[!seq(1,39) %in% seq(0,38, 4)]
    
    # col_plot <- ggplot(data = plot_data, 
    #                    aes(x = Samples, 
    #                        y = No.of.Diff.Exprs, 
    #                        fill = Group)
    #                    ) + 
    #   geom_col() + 
    #   labs(x = "", 
    #        y = paste0("Number of differentially expressed genes relative to ", 
    #                   "blood Tregs"), 
    #        fill = "Tissue") + 
    #   scale_fill_manual(values = c("#D89000", "#A3A500", "#39B600", 
    #                                  "#00BF7D", "#00BFC4", "#00B0F6", 
    #                                  "#9590FF", "#E76BF3", "#FF62BC")) + 
    #   # scale_x_discrete(breaks = NA) + 
    #   # geom_text(aes(x=c(sum(x.seq[1:2])/2, sum(x.seq[3:4])/2), y=0, 
    #   #               label=c("X","Y")), vjust=1.2, size=8) + 
    #   theme_light() + 
    #   theme(axis.text.x = element_text(angle = 90, hjust = 1), 
    #         panel.grid.major = element_blank(), 
    #         panel.grid.minor = element_blank(),
    #         legend.position = "none"
    #         )
    # 
    # col_plot
    
    col_point_plot <- ggplot(data = plot_data,
                             aes(x = Group,
                                 y = No.of.Diff.Exprs,
                                 colour = Group)
                             ) + 
      geom_point(size = 3) + 
      labs(x = "", 
           y = paste0("Number of differentially expressed genes relative to ", 
                      "blood Tregs")) + 
      scale_colour_manual(values = c("#D89000", "#A3A500", "#39B600",
                                     "#00BF7D", "#00BFC4", "#00B0F6",
                                     "#9590FF", "#E76BF3", "#FF62BC")) +
      theme_light() + 
      theme(legend.position = "none"
            )
    
    col_point_plot
    
  }
  
}

```


#testing comparisons of different formulae in DESeq object without custom design
#matrix
```{r}

#DESeq2 analysis using unified function
{

  #creating the vector of formulae to use in the unified function
  print("creating the vector of formulae to use in the unified function")
  formulae <- c("~0+tissue",
                "~0+tissue+cellType",
                "~0+cellType",
                "~0+tissue:cellType",
                "~0+tissue+tissue:cellType")

  #initialising DESeq object
  print("initialising DESeq object")
  test_DESeq_object <- DESeqDataSetFromMatrix(countData = counts_matrix,
                                         colData = annotations,
                                         design = ~0+tissue+tissue:cellType)

  #adding meta_data to the dataset
  print("adding meta_data to the dataset")
  mcols(test_DESeq_object) <- DataFrame(mcols(test_DESeq_object),  meta_data)

  #prefiltering low/no count genes
  #print("prefiltering low/no count genes")
  #keep <- rowSums(counts(DESeq_object)) >=10
  #DESeq_object <- DESeq_object[keep,]

  test_DESeq_object <- DESeq(object = test_DESeq_object,
                        test = "Wald",
                        betaPrior = F,
                        minReplicatesForReplace = Inf,
                        parallel = T)

  #pattern <- paste0("(", colnames(annotations)[2], ").+(^", colnames(annotations)[2], ")")
  #grep(pattern = "cellType", x = grep(pattern = "tissue", x = resultsNames(test_DESeq_object), ignore.case = T, invert = F), ignore.case = T, invert = T)
  
  resultsNames(test_DESeq_object)

  colnames(model.matrix(design(test_DESeq_object), colData(test_DESeq_object)))

  allsubstr <- function(x, n) substring(x, 1:(nchar(x) - n + 1), n:nchar(x))
  substrings_table <- function(string) table(unlist(sapply(1:nchar(string), allsubstr, x=string)))

  available_patterns <- colnames(colData(test_DESeq_object))
  available_patterns <- available_patterns[-length(available_patterns)]

  test_list <- list()

  test_list[[1]] <- resultsNames(test_DESeq_object)[intersect(x = grep(pattern = "tissue", x = resultsNames(test_DESeq_object), ignore.case = T, invert = F),
            y = grep(pattern = "cellType", x = resultsNames(test_DESeq_object), ignore.case = T, invert = T))]

  test_list[[2]] <- resultsNames(test_DESeq_object)[intersect(x = grep(pattern = "tissue", x = resultsNames(test_DESeq_object), ignore.case = T, invert = F),
            y = grep(pattern = "cellType", x = resultsNames(test_DESeq_object), ignore.case = T, invert = F))]

  test_list[[3]] <- resultsNames(test_DESeq_object)[intersect(x = grep(pattern = "tissue", x = resultsNames(test_DESeq_object), ignore.case = T, invert = T),
            y = grep(pattern = "cellType", x = resultsNames(test_DESeq_object), ignore.case = T, invert = F))]

  for(i in seq(1, length(test_list)))
  {
    if(!length(test_list[[i]] > 0)) next
    writeLines(paste0("Item ", i, " has a length of: ", length(test_list[[i]])))

  }


  test_res_list <- list()
  test_res_list <- GenerateContrasts(contrast_groups = contrast_groups, available_contrasts = test_list[[2]], results_list = test_res_list, DESeq_object = test_DESeq_object, prefix = "")

  # test_DESeq_object <- estimateSizeFactors(test_DESeq_object)
  # #the below is single threaded and as such takes way too long
  # test_DESeq_object <- estimateDispersions(test_DESeq_object)
  # #the below is the multi-threaded internal function in DESeq2 but this also
  # #automatically fits the model using the formula as the model matrix
  # # test_DESeq_object <- DESeq2:::DESeqParallel(object = test_DESeq_object,
  # #                                             modelMatrix = NULL,
  # #                                             BPPARAM = bpparam(),
  # #                                             quiet = F,
  # #                                             minmu = 0.5,
  # #                                             fitType = c("parametric",
  # #                                                         "local", "mean"),
  # #                                             betaPrior = F,
  # #                                             test = "Wald",
  # #                                             useT = FALSE)
  #
  # #creating the (custom) model matrix to extract contrasts that we need
  # print("creating the (custom) model matrix to extract contrasts that we need")
  # model_matrix <- model.matrix(~0+tissue+tissue:cellType, colData(DESeq_object))
  #
  # test_DESeq_object <- nbinomWaldTest(test_DESeq_object,
  #                                     betaPrior = F,
  #                                     quiet = F,
  #                                     modelMatrix = model_matrix)

  resultsNames(test_DESeq_object)

  test_res <- results(object = test_DESeq_object, test = "Wald",
                contrast = list("tissueBlood", "tissueLiver"),
                listValues = c(1, -1),
                parallel = T,
                cooksCutoff = F,
                independentFiltering = F)
  #32272 non-NA rows with cooks and filtering off, vs 16551 with cooks and filtering on

  writeLines(paste0(dim(subset(test_res, test_res$pvalue<0.05))[1], " genes in comparison"))

  #using no model_matrix for contrast contrast = list("tissueBlood.cellTypeT", "tissueLiver.cellTypeT")
  #1469 for ~0+tissue:cellType
  #1752 for ~0+tissue+tissue:cellType
  #1752 for ~0+tissue+cellType+tissue:cellType ## the addedd +cellType does not add anything

  #using hacked/modified script above
  #1752 for ~0+tissue+cellType+tissue:cellType
  #1752 for ~0+tissue+tissue:cellType
  #1752 for ~0+tissue:cellType

  writeLines(paste0(dim(subset(test_res, test_res$padj<0.05))[1], " genes in comparison"))

  #using no model_matrix for contrast contrast = list("tissueBlood.cellTypeT", "tissueLiver.cellTypeT")
  #238 for ~0+tissue:cellType
  #185 for ~0+tissue+tissue:cellType
  #185 for ~0+tissue+cellType+tissue:cellType ## the addedd +cellType does not add anything

  #using hacked/modified script above
  #1752 for ~0+tissue+cellType+tissue:cellType
  #1752 for ~0+tissue+tissue:cellType
  #1752 for ~0+tissue:cellType


  medians_blood_all <- rowMedians(norm_counts[, grepl(pattern = "Blood", x = colnames(norm_counts), ignore.case = T)])
  medians_liver_all <- rowMedians(norm_counts[, grepl(pattern = "Liver", x = colnames(norm_counts), ignore.case = T)])

  manual_log <- log2((medians_blood_all)/(medians_liver_all))
  names(manual_log) <- row.names(norm_counts)
  manual_log[!is.finite(manual_log)] <- 0

  manual_log <- manual_log[order(rownames(test_res))]

  test_res <- test_res[order(rownames(test_res)),]

  sqrt(sum(((manual_log) - (test_res$log2FoldChange))^2, na.rm = T))

  sd((manual_log) - (test_res$log2FoldChange), na.rm = T)


  names(mcols(test_DESeq_object))[grep("log2 fold change", mcols(mcols(test_DESeq_object))$description)]

}

```



#play area
```{r}

for(i in seq(1, length(merged_results_list))){
  for(j in seq(1, length(merged_results_list[[i]]))){
    for(k in seq(1, length(merged_results_list[[i]][[j]]))){
      # print(dim(merged_results_list[[i]][[j]][[k]]))
      # print(dim(merged_results_list[[i]][[j]][[k]][complete.cases(merged_results_list[[i]][[j]][[k]]),]))
      # print(dim(subset(merged_results_list[[i]][[j]][[k]], is.finite(merged_results_list[[i]][[j]][[k]]$Actuallog2FC))))
      
      print(summary(is.na(merged_results_list[[i]][[j]][[k]]$baseMean)))
      print(summary(is.na(merged_results_list[[i]][[j]][[k]]$log2FoldChange)))
      print(summary(is.na(merged_results_list[[i]][[j]][[k]]$lfcSE)))
      print(summary(is.na(merged_results_list[[i]][[j]][[k]]$stat)))
      print(summary(is.na(merged_results_list[[i]][[j]][[k]]$pvalue)))
      print(summary(is.na(merged_results_list[[i]][[j]][[k]]$padj)))
      print(summary(is.na(merged_results_list[[i]][[j]][[k]]$Actuallog2FC)))
      print("---")
      }
    }
}

# pancreas_norm_counts <- norm_counts[rowSums(norm_counts)>0,grepl(pattern = "pancreas", x = colnames(norm_counts), ignore.case = T)]
# pancreas_norm_counts <- pancreas_norm_counts[rowSums(pancreas_norm_counts)>0,]
# 
# sqrt(sum((pancreas_norm_counts$Pancreas_T_1 - pancreas_norm_counts$Pancreas_T_3)^2, na.rm = T))
# #925322.8
# sd((pancreas_norm_counts$Pancreas_T_1 - pancreas_norm_counts$Pancreas_T_3))
# #6089.453
# 
# sqrt(sum((pancreas_norm_counts$Pancreas_T_2 - pancreas_norm_counts$Pancreas_T_3)^2, na.rm = T))
# #965123.6
# sd((pancreas_norm_counts$Pancreas_T_2 - pancreas_norm_counts$Pancreas_T_3))
# #6351.993
# 
# sqrt(sum((pancreas_norm_counts$Pancreas_T_2 - pancreas_norm_counts$Pancreas_T_1)^2, na.rm = T))
# #91951.52
# sd((pancreas_norm_counts$Pancreas_T_2 - pancreas_norm_counts$Pancreas_T_1))
# #605.4047 
# 
# # nrow(pancreas_norm_counts[(pancreas_norm_counts$Pancreas_T_3>pancreas_norm_counts$Pancreas_T_1),])
# # nrow(pancreas_norm_counts[(pancreas_norm_counts$Pancreas_T_3>pancreas_norm_counts$Pancreas_T_2),])
# # nrow(pancreas_norm_counts[(pancreas_norm_counts$Pancreas_T_3>pancreas_norm_counts$Pancreas_T_2 & pancreas_norm_counts$Pancreas_T_3>pancreas_norm_counts$Pancreas_T_1),])
# 
# 
# liver_norm_counts <- norm_counts[rowSums(norm_counts)>0,grepl(pattern = "liver", x = colnames(norm_counts), ignore.case = T)]
# liver_norm_counts <- liver_norm_counts[rowSums(liver_norm_counts)>0,]
# 
# sqrt(sum((liver_norm_counts$Liver_T_1 - liver_norm_counts$Liver_T_3)^2, na.rm = T))
# #537830.9
# sd((liver_norm_counts$Liver_T_1 - liver_norm_counts$Liver_T_3))
# #3695.151
# 
# sqrt(sum((liver_norm_counts$Liver_T_2 - liver_norm_counts$Liver_T_3)^2, na.rm = T))
# #690812.5
# sd((liver_norm_counts$Liver_T_2 - liver_norm_counts$Liver_T_3))
# #4746.747
# 
# sqrt(sum((liver_norm_counts$Liver_T_2 - liver_norm_counts$Liver_T_1)^2, na.rm = T))
# #171298.1
# sd((liver_norm_counts$Liver_T_2 - liver_norm_counts$Liver_T_1))
# #1176.742
# 
# # nrow(liver_norm_counts[(liver_norm_counts$Liver_T_3>liver_norm_counts$Liver_T_1),])
# # nrow(liver_norm_counts[(liver_norm_counts$Liver_T_3>liver_norm_counts$Liver_T_2),])
# # nrow(liver_norm_counts[(liver_norm_counts$Liver_T_3>liver_norm_counts$Liver_T_2 & liver_norm_counts$Liver_T_3>liver_norm_counts$Liver_T_1),])
# 
# 
# peyers_norm_counts <- norm_counts[rowSums(norm_counts)>0,grepl(pattern = "peyers", x = colnames(norm_counts), ignore.case = T)]
# peyers_norm_counts <- peyers_norm_counts[rowSums(peyers_norm_counts)>0,]
# 
# sqrt(sum((peyers_norm_counts$PeyersPatches_T_1 - peyers_norm_counts$PeyersPatches_T_3)^2, na.rm = T))
# #704768.4
# sd((peyers_norm_counts$PeyersPatches_T_1 - peyers_norm_counts$PeyersPatches_T_3))
# #4592.153
# 
# sqrt(sum((peyers_norm_counts$PeyersPatches_T_2 - peyers_norm_counts$PeyersPatches_T_3)^2, na.rm = T))
# #76522.43
# sd((peyers_norm_counts$PeyersPatches_T_2 - peyers_norm_counts$PeyersPatches_T_3))
# #498.2912
# 
# sqrt(sum((peyers_norm_counts$PeyersPatches_T_2 - peyers_norm_counts$PeyersPatches_T_1)^2, na.rm = T))
# #686506.1
# sd((peyers_norm_counts$PeyersPatches_T_2 - peyers_norm_counts$PeyersPatches_T_1))
# #4472.635
# 
# # nrow(peyers_norm_counts[(peyers_norm_counts$PeyersPatches_T_3>peyers_norm_counts$PeyersPatches_T_1),])
# # nrow(peyers_norm_counts[(peyers_norm_counts$PeyersPatches_T_3>peyers_norm_counts$PeyersPatches_T_2),])
# # nrow(peyers_norm_counts[(peyers_norm_counts$PeyersPatches_T_3>peyers_norm_counts$PeyersPatches_T_2 & peyers_norm_counts$PeyersPatches_T_3>peyers_norm_counts$PeyersPatches_T_1),])
# 
# 
# 
# #old --- individual transcript analysis --- using only norm counts
# #reading norm counts
# # norm_counts <- read.delim(file = paste0(loc_counts, "normalised_counts.txt"), 
# #                           sep = "\t", 
# #                           stringsAsFactors = F)
# # 
# # #calculating mean counts based on tissue and cell type
# # {
# #   mean_norm_counts <- data.frame(matrix(nrow = nrow(norm_counts), ncol = 0))
# #   
# #   #iterate thorugh condition labels and calculate means for Tregs
# #   for(i in seq(1, length(condition_labels))){
# #     
# #     columns <- norm_counts[, grepl(pattern = paste0(
# #       condition_labels[i], "_[0-9]{1}$"), 
# #       x = colnames(norm_counts), 
# #       ignore.case = T), 
# #       drop = F]
# #     
# #     if(ncol(columns)>2) mean_norm_counts <- cbind(mean_norm_counts, 
# #                                                     rowMeans(
# #                                                       as.matrix(columns))
# #                                                     )
# #     else mean_norm_counts <- cbind(mean_norm_counts, columns)
# #     
# #     colnames(mean_norm_counts)[ncol(mean_norm_counts)] <-
# #         paste0(gsub(pattern = "_[0-9]{1}$", 
# #                     replacement = "", 
# #                     x = condition_labels[i], 
# #                     ignore.case = T))
# #   }
# #   #removing iteration related variables
# #   rm(i, columns)
# #   
# #   #removing 0 read rows
# #   mean_norm_counts <- mean_norm_counts[rowSums(mean_norm_counts)>0,]
# #   
# # }
# # 
# # #populating a list of gene ids which are found overtly expressed in each tissue 
# # #only
# # #using SD is not strict enough (>= mean_norm_counts[,j]*10*sd_vector)
# # {
# #   #creating an empty list
# #   uniquely_expressing_genes <- list()
# #   
# #   #row standard deviation vector
# #   sd_vector <- rowSds(as.matrix(mean_norm_counts))
# #   
# #   for(i in seq(1, ncol(mean_norm_counts))){
# #     #dummy vector of TRUEs for initialisation
# #     # selection_vector <- !vector(mode = "logical", 
# #     #                            length = nrow(mean_norm_counts))
# #     
# #     #initialised vector of normalised ocunts greater than 10
# #     
# #     for(j in seq(1, ncol(mean_norm_counts))){
# #       
# #       if(i==j) next
# #       
# #       selection_vector <- selection_vector == 
# #         (mean_norm_counts[,i] >= mean_norm_counts[,j]*10*sd_vector)
# #       
# #     }
# #     
# #     print(all(selection_vector))
# #     
# #     uniquely_expressing_genes[[length(uniquely_expressing_genes)+1]] <- 
# #       gsub(pattern = "\\.[0-9]+$", 
# #            replacement = "", 
# #            x = rownames(mean_norm_counts[selection_vector,]))
# #     
# #     names(uniquely_expressing_genes)[length(uniquely_expressing_genes)] <- 
# #       colnames(mean_norm_counts)[i]
# #   }
# #   
# #   
# # }
# 
# # #constructing new data structure for easier plotting
# # {
# #   plots_list <- list()
# #   
# #   for(i in seq(1, length(merged_results_list))){
# #     #print(paste0("i: ", i))
# #     
# #     plots_list[[length(plots_list)+1]] <- list()
# #     names(plots_list)[length(plots_list)] <- names(merged_results_list)[i]
# #     
# #     for(j in seq(1, length(merged_results_list[[i]]))){
# #       #print(paste0("j: ", j))
# #       
# #       plots_list[[i]][[length(plots_list[[i]])+1]] <- list()
# #       names(plots_list[[i]])[length(plots_list[[i]])] <- 
# #         names(merged_results_list[[i]])[j]
# #       
# #       #print(names(merged_results_list[[i]])[j])
# #       
# #       for(k in seq(1, length(contrast_groups))){
# #         #print(paste0("k: ", k))
# #         
# #         selected_contrasts <- grep(pattern = contrast_groups[[k]], 
# #                                    x = names(merged_results_list[[i]][[j]]), 
# #                                    ignore.case = T)
# #         
# #         #print(selected_contrasts)
# #         
# #         if(!length(selected_contrasts>0)) next
# #         
# #         plot_data <- as.data.frame(matrix(data = 0, nrow = 0, ncol = 4))
# #         
# #         for(l in selected_contrasts){
# #           
# #           #print(paste0("l: ", l))
# #           
# #           selected_data <- as.data.frame(merged_results_list[[i]][[j]][[l]])
# #           selected_data <- selected_data[complete.cases(selected_data),]
# #           selected_data <- selected_data[,c(2,6)]
# #           selected_data$contrast <- names(merged_results_list[[i]][[j]][l])
# #           selected_data$geneID <- gsub(pattern = "\\.[0-9]+$", 
# #                                        replacement = "", 
# #                                        x = rownames(selected_data))
# #           
# #           plot_data <- rbind.data.frame(plot_data, selected_data)
# #         }
# #         colnames(plot_data) <- c("log2FC", "padj", "contrast", "geneID")
# #         plot_data <- merge(x = plot_data, y = output_table[,1:2], 
# #                            by.x = "geneID", by.y = "Gene.stable.ID", 
# #                            all.x = T, all.y = F, 
# #                            sort = F)
# #         
# #         plots_list[[i]][[j]][[length(plots_list[[i]][[j]])+1]] <- plot_data
# #         names(plots_list[[i]][[j]])[length(plots_list[[i]][[j]])] <- 
# #           paste(contrast_groups[[k]], sep = "", collapse = "_")
# #       }
# #       
# #     }
# #   }
# #   
# # }


```






